diff -u -U 2 -r -N -d linux-2.6.0-test11.orig/drivers/net/imq.c linux-2.6.0-test11/drivers/net/imq.c
--- linux-2.6.0-test11.orig/drivers/net/imq.c	2003-12-07 15:23:20.000000000 +0000
+++ linux-2.6.0-test11/drivers/net/imq.c	2003-12-07 15:22:24.000000000 +0000
@@ -38,7 +38,7 @@
 	.owner		= THIS_MODULE,
 	.pf		= PF_INET,
 	.hooknum	= NF_IP_PRE_ROUTING,
-	.priority	= NF_IP_PRI_MANGLE + 1
+	.priority	= NF_IP_PRI_NAT_DST + 1
 };
 
 static struct nf_hook_ops imq_egress_ipv4 = {
@@ -55,7 +55,7 @@
 	.owner		= THIS_MODULE,
 	.pf		= PF_INET6,
 	.hooknum	= NF_IP6_PRE_ROUTING,
-	.priority	= NF_IP6_PRI_MANGLE + 1
+	.priority	= NF_IP6_PRI_NAT_DST + 1
 };
 
 static struct nf_hook_ops imq_egress_ipv6 = {
