diff -urN linux.orig/fs/reiserfs/namei.c linux/fs/reiserfs/namei.c
--- linux.orig/fs/reiserfs/namei.c	Fri Apr 26 21:09:06 2002
+++ linux/fs/reiserfs/namei.c	Fri Apr 26 23:29:03 2002
@@ -306,7 +306,7 @@
       pathrelse (&path);
 	    if (quota_allocated)
 		DQUOT_FREE_BLOCK (dir->i_sb, dir, 1);
-      return -EHASHCOLLISION;
+      return -EBUSY;
     }
     /* adjust offset of directory enrty */
     deh->deh_offset = SET_GENERATION_NUMBER (deh->deh_offset, gen_number);
