diff -urN linux.2.2.21.orig/Makefile linux.2.2.21/Makefile
--- linux.2.2.21.orig/Makefile	Sat Jan 26 13:15:53 2002
+++ linux.2.2.21/Makefile	Sat Jan 26 13:18:26 2002
@@ -500,7 +500,7 @@
 #
 
 scripts/mkdep: scripts/mkdep.c
-	$(HOSTCC) $(HOSTCFLAGS) -o scripts/mkdep scripts/mkdep.c
+	$(HOSTCC) $(HOSTCFLAGS) -I $(HPATH) -o scripts/mkdep scripts/mkdep.c
 
 scripts/split-include: scripts/split-include.c
-	$(HOSTCC) $(HOSTCFLAGS) -o scripts/split-include scripts/split-include.c
+	$(HOSTCC) $(HOSTCFLAGS) -I $(HPATH) -o scripts/split-include scripts/split-include.c
