diff -u --new-file --recursive --exclude-from /usr/src/exclude linux.vanilla/scripts/lxdialog/checklist.c linux.21p2/scripts/lxdialog/checklist.c
--- linux.vanilla/scripts/lxdialog/checklist.c	Sat Dec 29 16:57:25 2001
+++ linux.21p2/scripts/lxdialog/checklist.c	Tue Jan  1 21:23:32 2002
@@ -108,8 +108,8 @@
     print_button (dialog, "Select", y, x, selected == 0);
     print_button (dialog, " Help ", y, x + 14, selected == 1);
 
-    /*    wmove(dialog, y, x+1 + 14*selected);
-    wrefresh (dialog);*/
+    /*    wmove(dialog, y, x+1 + 14*selected);*/
+    wrefresh (dialog);
 }
 
 /*
