diff -urN linux.orig/drivers/ieee1394/linux22compat.h linux/drivers/ieee1394/linux22compat.h
--- linux.orig/drivers/ieee1394/linux22compat.h	Thu Jul 11 14:45:42 2002
+++ linux/drivers/ieee1394/linux22compat.h	Thu Jul 11 14:44:00 2002
@@ -49,9 +49,13 @@
 #include <linux/pci.h>
 #include <asm/io.h>
 
-/* seems to fix PPC build */
-typedef u32 dma_addr_t;
-
+/* seems to fix PPC build
+ * typedef u32 dma_addr_t;
+ *
+ * remove it, our types.h have this definitions
+ * -- Krzysiek Taraszka	<dzimi@pld.org.pl>
+ */
+ 
 #define PCI_DMA_BIDIRECTIONAL	0
 #define PCI_DMA_TODEVICE	1
 #define PCI_DMA_FROMDEVICE	2
