diff -urN linux.orig/arch/ppc/kernel/ppc_ksyms.c linux/arch/ppc/kernel/ppc_ksyms.c
--- linux.orig/arch/ppc/kernel/ppc_ksyms.c	Tue Mar  5 15:14:21 2002
+++ linux/arch/ppc/kernel/ppc_ksyms.c	Tue Mar  5 15:50:13 2002
@@ -14,8 +14,8 @@
 #include <asm/uaccess.h>
 #include <asm/io.h>
 #include <linux/ide.h>
-#include <asm/ide.h>
 #include <linux/ide.h>
+#include <asm/ide.h>
 #include <asm/atomic.h>
 #include <asm/bitops.h>
 #include <asm/checksum.h>
diff -urN linux.orig/drivers/block/ll_rw_blk.c linux/drivers/block/ll_rw_blk.c
--- linux.orig/drivers/block/ll_rw_blk.c	Tue Mar  5 15:14:22 2002
+++ linux/drivers/block/ll_rw_blk.c	Tue Mar  5 15:48:35 2002
@@ -26,6 +26,7 @@
 #include <linux/raid/md.h>
 
 #ifdef CONFIG_POWERMAC
+#include <linux/ide.h>
 #include <asm/ide.h>
 #endif
 
