--- linux/drivers/macintosh/mac_hid.c.ppc	Thu Nov  8 02:21:09 2001
+++ linux/drivers/macintosh/mac_hid.c	Thu Nov  8 02:21:34 2001
@@ -480,13 +480,13 @@
 
 	if (!keyboard_sends_linux_keycodes) {
 #ifdef CONFIG_MAGIC_SYSRQ
-		ppc_md.ppc_kbd_sysrq_xlate   = mac_hid_kbd_sysrq_xlate;
+		ppc_md.sysrq_xlate   = mac_hid_kbd_sysrq_xlate;
 		SYSRQ_KEY                = 0x69;
 #endif
 		memcpy(key_maps, mac_key_maps_save, sizeof(key_maps));
 	} else {
 #ifdef CONFIG_MAGIC_SYSRQ
-		ppc_md.ppc_kbd_sysrq_xlate   = pckbd_sysrq_xlate;
+		ppc_md.sysrq_xlate   = pckbd_sysrq_xlate;
 		SYSRQ_KEY                = 0x54;
 #endif
 	}
