diff -u --new-file --recursive --exclude-from /usr/src/exclude linux.vanilla/drivers/video/Makefile linux.21p2/drivers/video/Makefile
--- linux.vanilla/drivers/video/Makefile	Sun Mar 25 17:37:37 2001
+++ linux.21p2/drivers/video/Makefile	Tue Jan  1 21:25:14 2002
@@ -102,6 +102,10 @@
 
 ifeq ($(CONFIG_FB_ATY),y)
 L_OBJS += atyfb.o
+else
+  ifeq ($(CONFIG_FB_ATY),m)
+  M_OBJS += atyfb.o
+  endif
 endif
 
 ifeq ($(CONFIG_FB_ATY128),y)
