diff -urN linux.orig/include/asm-ppc/atomic.h linux/include/asm-ppc/atomic.h
--- linux.orig/include/asm-ppc/atomic.h	Sun Mar 25 18:31:08 2001
+++ linux/include/asm-ppc/atomic.h	Sun Apr  7 20:38:11 2002
@@ -6,6 +6,9 @@
 #define _ASM_PPC_ATOMIC_H_
 
 #ifdef __SMP__
+
+#include <asm/smp.h>
+
 typedef struct { volatile int counter; } atomic_t;
 #else
 typedef struct { int counter; } atomic_t;
diff -urN linux.orig/include/asm-ppc/hardirq.h linux/include/asm-ppc/hardirq.h
--- linux.orig/include/asm-ppc/hardirq.h	Sun Mar 25 18:31:09 2001
+++ linux/include/asm-ppc/hardirq.h	Sun Apr  7 20:38:49 2002
@@ -1,6 +1,8 @@
 #ifndef __ASM_HARDIRQ_H
 #define __ASM_HARDIRQ_H
 
+#include <linux/tasks.h>
+
 extern unsigned int ppc_local_irq_count[NR_CPUS];
 
 /*
