diff -urN linux.orig/include/linux/skbuff.h linux/include/linux/skbuff.h
--- linux.orig/include/linux/skbuff.h	Wed Jul 31 20:06:04 2002
+++ linux/include/linux/skbuff.h	Sat Aug  3 15:23:56 2002
@@ -83,6 +83,7 @@
 	unsigned char	is_clone,		/* We are a clone				*/
 			cloned, 		/* head may be cloned (check refcnt to be sure). */
   			pkt_type,		/* Packet class					*/
+  			pkt_bridged,		/* Tracker for bridging 			*/
   			ip_summed;		/* Driver fed us an IP checksum			*/
 #ifdef CONFIG_IMQ
 	short		from_imq;		/* Pkt from IMQ */
