diff -u --new-file --recursive --exclude-from /usr/src/exclude linux.vanilla/Documentation/SubmittingDrivers linux.21p1/Documentation/SubmittingDrivers
--- linux.vanilla/Documentation/SubmittingDrivers	Sat Dec 29 16:57:19 2001
+++ linux.21p1/Documentation/SubmittingDrivers	Sat Dec 29 03:19:40 2001
@@ -2,7 +2,7 @@
 ---------------------------------------
 
 This document is intended to explain how to submit device drivers to the
-Linux 2.2 and 2.4test kernel trees. Note that if you are interested in video
+Linux 2.2, 2.4 and 2.5 kernel trees. Note that if you are interested in video
 card drivers you should probably talk to XFree86 (http://wwww.xfree86.org) 
 instead.
 
@@ -10,9 +10,9 @@
 -------------------------
 
 Major and minor numbers for devices are allocated by the Linux assigned name
-and number authority (currently better known as H Peter Anvin). The
-site is http://www.lanana.org/. This also deals with allocating numbers for
-devices that are not going to be submitted to the mainstream kernel.
+and number authority (currently better known as H Peter Anvin). The site is
+http://www.lanana.org/. This also deals with allocating numbers for devices
+that are not going to be submitted to the mainstream kernel.
 
 If you don't use assigned numbers then when you device is submitted it will
 get given an assigned number even if that is different from values you may
@@ -22,21 +22,28 @@
 ------------------------
 
 Linux 2.0:
-	No new drivers are accepted for this kernel tree
+	No new drivers are accepted for this kernel tree. If you have any
+	security or other fixes then please contact David Weinehall
+	<tao@acc.umu.se>
 
 Linux 2.2:
 	If the code area has a general maintainer then please submit it to
 	the maintainer listed in MAINTAINERS in the kernel file. If the
 	maintainer does not respond or you cannot find the appropriate
-	maintainer then please contact Alan Cox <alan@lxorguk.ukuu.org.uk>
+	maintainer then please contact Alan Cox <alan@lxorguk.ukuu.org.uk>.
 
-Linux 2.4test:
+Linux 2.4:
+	If the code area has a general maintainer then please submit it to
+	the maintainer listed in MAINTAINERS in the kernel file. If the
+	maintainer does not respond or you cannot find the appropriate
+	maintainer then please contact Marcelo Tosatti
+	<marcelo@conectiva.com.br>.
+
+Linux 2.5:
 	This kernel tree is under active development. The same rules apply
-	as 2.2 but you may wish to submit your driver via linux-kernel (see
-	resources) and follow that list to track changes in API's. These
-	should no longer be occurring as we are now in a code freeze.
-	The final contact point for Linux 2.4 submissions is 	
-	<torvalds@transmeta.com>.
+	as 2.4 but you may wish to submit your driver via linux-kernel (see
+	resources) and follow that list to track changes in API's. The final
+	contact point for Linux 2.5 submissions is <torvalds@transmeta.com>.
 
 What Criteria Determine Acceptance
 ----------------------------------
diff -u --new-file --recursive --exclude-from /usr/src/exclude linux.vanilla/MAINTAINERS linux.21p1/MAINTAINERS
--- linux.vanilla/MAINTAINERS	Sat Dec 29 16:57:19 2001
+++ linux.21p1/MAINTAINERS	Sat Dec 29 03:12:27 2001
@@ -1000,9 +1000,8 @@
 S:	Maintained
 
 USB SUBSYSTEM
-P:	Johannes Erdfelt
-M	jerdfelt@valinux.com
-M:	johannes@erdfelt.com
+P:	Greg Kroah-Hartman
+M:	greg@kroah.com
 L:	linux-usb-users@lists.sourceforge.net
 L:	linux-usb-devel@lists.sourceforge.net
 W:	http://www.linux-usb.org
