diff -urN linux.orig/drivers/macintosh/macserial.h linux/drivers/macintosh/macserial.h
--- linux.orig/drivers/macintosh/macserial.h	Sun Mar 25 18:31:41 2001
+++ linux/drivers/macintosh/macserial.h	Wed Jul 17 18:22:07 2002
@@ -11,23 +11,6 @@
 
 #define NUM_ZSREGS    16
 
-struct serial_struct {
-	int	type;
-	int	line;
-	int	port;
-	int	irq;
-	int	flags;
-	int	xmit_fifo_size;
-	int	custom_divisor;
-	int	baud_base;
-	unsigned short	close_delay;
-	char	reserved_char[2];
-	int	hub6;
-	unsigned short	closing_wait; /* time to wait before closing */
-	unsigned short	closing_wait2; /* no longer used... */
-	int	reserved[4];
-};
-
 /*
  * For the close wait times, 0 means wait forever for serial port to
  * flush its output.  65535 means don't wait at all.
