diff -urN linux.orig/arch/ppc/kernel/setup.c linux/arch/ppc/kernel/setup.c
--- linux.orig/arch/ppc/kernel/setup.c	Sat Jun  8 15:17:55 2002
+++ linux/arch/ppc/kernel/setup.c	Sat Jun  8 15:16:10 2002
@@ -581,6 +581,9 @@
 	extern unsigned long find_available_memory(void);
 	extern unsigned long *end_of_DRAM;
 
+	/* so udelay does something sensible, assume <= 1000 bogomips */
+	loops_per_jiffy = 500000000 / HZ;
+
 #ifdef CONFIG_XMON
 	{
 		char *p;
