diff -urN linux.orig/drivers/scsi/ips.c linux/drivers/scsi/ips.c
--- linux.orig/drivers/scsi/ips.c	Fri Nov  2 17:39:08 2001
+++ linux/drivers/scsi/ips.c	Tue May 28 15:30:06 2002
@@ -178,9 +178,11 @@
 };
 #endif
 
+/*
 #if !defined(__i386__)
    #error "This driver has only been tested on the x86 platform"
 #endif
+*/
 
 #if LINUX_VERSION_CODE < LinuxVersionCode(2,2,0)
    #error "This driver only works with kernel 2.2.0 and later"
