diff -urN linux.orig/mm/memory.c linux/mm/memory.c
--- linux.orig/mm/memory.c	Wed Aug  7 20:20:39 2002
+++ linux/mm/memory.c	Wed Aug  7 20:12:47 2002
@@ -498,7 +498,7 @@
 			if (vma->vm_start > ptr) {
 				if (!(vma->vm_flags & VM_GROWSDOWN))
 					goto out;
-				if (expand_stack(vma, ptr))
+				if (expand_stack(vma, ptr, NULL))
 					goto out;
 			}
 			err = -EACCES;
diff -urN linux.orig/net/unix/af_unix.c linux/net/unix/af_unix.c
--- linux.orig/net/unix/af_unix.c	Wed Aug  7 20:20:22 2002
+++ linux/net/unix/af_unix.c	Wed Aug  7 20:18:38 2002
@@ -1525,7 +1525,7 @@
 		return;
 	wake_up_interruptible(sk->sleep);
 	if (sk->sndbuf - (int)atomic_read(&sk->wmem_alloc) >= MIN_WRITE_SPACE)
-		sock_wake_async(sk->socket, 2);
+		sock_wake_async(sk->socket, 2, POLL_OUT);
 }
 
 #ifdef CONFIG_PROC_FS
