diff -Naur --exclude '*.orig' --exclude '*.rej' linux-2.6.1-gentoo/arch/i386/kernel/cpu/cpufreq/powernow-k7.c linux-2.6.1-gentoo-test/arch/i386/kernel/cpu/cpufreq/powernow-k7.c
--- linux-2.6.1-gentoo/arch/i386/kernel/cpu/cpufreq/powernow-k7.c	2004-01-09 23:40:12.000000000 +0000
+++ linux-2.6.1-gentoo-test/arch/i386/kernel/cpu/cpufreq/powernow-k7.c	2004-01-09 22:32:23.000000000 +0000
@@ -279,6 +279,8 @@
 	etuple &= 0xf00;
 	etuple |= (c->x86_model<<4)|(c->x86_mask);
 
+	if (etuple == 0x7a0) { etuple = 0x781; }
+
 	for (i=0xC0000; i < 0xffff0 ; i+=16) {
 
 		p = phys_to_virt(i);
