--- linux-2.6.0-test1.orig/drivers/video/vesafb.c       2003-07-13
23:30:36.000000000 -0400
+++ linux/drivers/video/vesafb.c        2003-07-19 20:30:18.000000000 -0400
@@ -227,7 +227,7 @@
 	vesafb_defined.xres = screen_info.lfb_width;
 	vesafb_defined.yres = screen_info.lfb_height;
 	vesafb_fix.line_length = screen_info.lfb_linelength;
-	vesafb_fix.smem_len = screen_info.lfb_size * 65536;
+	vesafb_fix.smem_len = screen_info.lfb_width * screen_info.lfb_height * screen_info.lfb_depth;
 	vesafb_fix.visual   = (vesafb_defined.bits_per_pixel == 8) ?
 		FB_VISUAL_PSEUDOCOLOR : FB_VISUAL_TRUECOLOR;
 
