diff -aur linux.backup/arch/ia64/hp/common/sba_iommu.c linux/arch/ia64/hp/common/sba_iommu.c
--- linux.backup/arch/ia64/hp/common/sba_iommu.c	Sat Aug 16 15:02:36 2003
+++ linux/arch/ia64/hp/common/sba_iommu.c	Sat Aug 16 17:56:19 2003
@@ -1935,10 +1935,10 @@
 }
 
 static struct acpi_driver acpi_sba_ioc_driver = {
-	name:		"IOC IOMMU Driver",
-	ids:		"HWP0001,HWP0004",
-	ops: {
-		add:	acpi_sba_ioc_add,
+	.name		= "IOC IOMMU Driver",
+	.ids		= "HWP0001,HWP0004",
+	.ops		= {
+		.add	= acpi_sba_ioc_add,
 	},
 };
 
diff -aur linux.backup/arch/ia64/kernel/perfmon.c linux/arch/ia64/kernel/perfmon.c
--- linux.backup/arch/ia64/kernel/perfmon.c	Sat Aug 16 15:02:16 2003
+++ linux/arch/ia64/kernel/perfmon.c	Sat Aug 16 15:44:59 2003
@@ -2122,7 +2122,7 @@
 	return 1;
 }
 static struct dentry_operations pfmfs_dentry_operations = {
-	d_delete:	pfmfs_delete_dentry,
+	.d_delete	= pfmfs_delete_dentry,
 };
 
 
diff -aur linux.backup/arch/ia64/sn/io/drivers/ioconfig_bus.c linux/arch/ia64/sn/io/drivers/ioconfig_bus.c
--- linux.backup/arch/ia64/sn/io/drivers/ioconfig_bus.c	Thu Jun 26 23:48:30 2003
+++ linux/arch/ia64/sn/io/drivers/ioconfig_bus.c	Sat Aug 16 17:57:14 2003
@@ -346,9 +346,9 @@
 }
 
 struct file_operations ioconfig_bus_fops = {
-	ioctl:ioconfig_bus_ioctl,
-	open:ioconfig_bus_open,		/* open */
-	release:ioconfig_bus_close	/* release */
+	.ioctl = ioconfig_bus_ioctl,
+	.open = ioconfig_bus_open,		/* open */
+	.release = ioconfig_bus_close	/* release */
 };
 
 
diff -aur linux.backup/arch/ia64/sn/io/sn2/shub.c linux/arch/ia64/sn/io/sn2/shub.c
--- linux.backup/arch/ia64/sn/io/sn2/shub.c	Sat Aug 16 15:02:37 2003
+++ linux/arch/ia64/sn/io/sn2/shub.c	Sat Aug 16 15:45:00 2003
@@ -243,7 +243,7 @@
 }
 
 struct file_operations shub_mon_fops = {
-	        ioctl:          shubstats_ioctl,
+	        .ioctl          = shubstats_ioctl,
 };
 
 /*
diff -aur linux.backup/arch/ia64/sn/kernel/setup.c linux/arch/ia64/sn/kernel/setup.c
--- linux.backup/arch/ia64/sn/kernel/setup.c	Sat Aug 16 15:02:37 2003
+++ linux/arch/ia64/sn/kernel/setup.c	Sat Aug 16 17:57:36 2003
@@ -117,14 +117,14 @@
  * VGA color display.
  */
 struct screen_info sn_screen_info = {
-	orig_x:			 0,
-	orig_y:			 0,
-	orig_video_mode:	 3,
-	orig_video_cols:	80,
-	orig_video_ega_bx:	 3,
-	orig_video_lines:	25,
-	orig_video_isVGA:	 1,
-	orig_video_points:	16
+	.orig_x			= 0,
+	.orig_y			= 0,
+	.orig_video_mode	= 3,
+	.orig_video_cols	= 80,
+	.orig_video_ega_bx	= 3,
+	.orig_video_lines	= 25,
+	.orig_video_isVGA	= 1,
+	.orig_video_points	= 16
 };
 
 /*
