--- linux-2.6.7/net/ipv4/netfilter/ip_tables.c.orig	2004-06-30 20:54:36.000000000 +0200
+++ linux-2.6.7/net/ipv4/netfilter/ip_tables.c	2004-07-01 08:50:43.372358512 +0200
@@ -1472,7 +1472,7 @@
 		int *hotdrop)
 {
 	/* tcp.doff is only 4 bits, ie. max 15 * 4 bytes */
-	char opt[60 - sizeof(struct tcphdr)];
+	u_int8_t opt[60 - sizeof(struct tcphdr)];
 	unsigned int i;
 
 	duprintf("tcp_match: finding option\n");
