--- 25/drivers/video/radeonfb.c~radeonfb-line_length-fix	2003-10-05 09:17:58.000000000 -0700
+++ 25-akpm/drivers/video/radeonfb.c	2003-10-05 09:17:58.000000000 -0700
@@ -2090,7 +2090,7 @@ static int radeonfb_set_par (struct fb_i
 	
 	}
 	/* Update fix */
-        info->fix.line_length = rinfo->pitch*64;
+        info->fix.line_length = mode->xres_virtual*(mode->bits_per_pixel/8);
         info->fix.visual = rinfo->depth == 8 ? FB_VISUAL_PSEUDOCOLOR : FB_VISUAL_DIRECTCOLOR;
 
 #ifdef CONFIG_BOOTX_TEXT
