--- linux-2.6.2-rc3/drivers/char/istallion.c.orig	2003-10-09 10:02:39.000000000 +0200
+++ linux-2.6.2-rc3/drivers/char/istallion.c	2004-01-10 04:18:44.000000000 +0100
@@ -417,7 +417,6 @@
 #ifndef PCI_DEVICE_ID_ECRA
 #define	PCI_DEVICE_ID_ECRA		0x0004
 #endif
-#endif
 
 static struct pci_device_id istallion_pci_tbl[] = {
 	{ PCI_VENDOR_ID_STALLION, PCI_DEVICE_ID_ECRA, PCI_ANY_ID, PCI_ANY_ID, 0, 0, 0 },
@@ -425,6 +424,8 @@
 };
 MODULE_DEVICE_TABLE(pci, istallion_pci_tbl);
 
+#endif /* CONFIG_PCI */
+
 /*****************************************************************************/
 
 /*
