diff -u linux/drivers/media/dvb/dvb-core/dvb_demux.c:1.38 linux/drivers/media/dvb/dvb-core/dvb_demux.c:1.39
--- linux/drivers/media/dvb/dvb-core/dvb_demux.c:1.38	Mon Oct 13 02:34:51 2003
+++ linux/drivers/media/dvb/dvb-core/dvb_demux.c	Thu Oct 16 10:36:58 2003
@@ -107,7 +107,7 @@
 {
 	u32 crc;
 
-	crc = crc32_le(~0, data, length);
+	crc = crc32_be(~0, data, length);
 
 	data[length]   = (crc >> 24) & 0xff;
 	data[length+1] = (crc >> 16) & 0xff;
@@ -118,7 +118,7 @@
 
 static u32 dvb_dmx_crc32 (struct dvb_demux_feed *f, const u8 *src, size_t len)
 {
-	return (f->feed.sec.crc_val = crc32_le (f->feed.sec.crc_val, src, len));
+	return (f->feed.sec.crc_val = crc32_be (f->feed.sec.crc_val, src, len));
 }


diff -u linux/drivers/media/dvb/frontends/tda1004x.c:1.7 linux/drivers/media/dvb/frontends/tda1004x.c:1.8
--- linux/drivers/media/dvb/frontends/tda1004x.c:1.7	Thu Oct  9 01:12:46 2003
+++ linux/drivers/media/dvb/frontends/tda1004x.c	Thu Oct 16 10:40:02 2003
@@ -44,12 +44,12 @@
 #include "dvb_frontend.h"
 #include "dvb_functions.h"
 
-#ifndef CONFIG_TDA1004X_MC_LOCATION
-#define CONFIG_TDA1004X_MC_LOCATION "/etc/dvb/tda1004x.mc"
+#ifndef DVB_TDA1004X_FIRMWARE_FILE
+#define DVB_TDA1004X_FIRMWARE_FILE "/etc/dvb/tda1004x.mc"
 #endif
 
 static int tda1004x_debug = 0;
-static char *tda1004x_firmware = CONFIG_TDA1004X_MC_LOCATION;
+static char *tda1004x_firmware = DVB_TDA1004X_FIRMWARE_FILE;
 
 
 #define TDA10045H_ADDRESS        0x08
diff -u linux/drivers/media/dvb/ttpci/av7110.c:1.90 linux/drivers/media/dvb/ttpci/av7110.c:1.91
--- linux/drivers/media/dvb/ttpci/av7110.c:1.90	Wed Oct 15 12:06:04 2003
+++ linux/drivers/media/dvb/ttpci/av7110.c	Fri Oct 17 19:01:24 2003
@@ -4498,6 +4498,30 @@
 	0x1d, 0x00,
 	0x1e, 0x00,
 
+	0x41, 0x77,
+	0x42, 0x77,
+	0x43, 0x77,
+	0x44, 0x77,
+	0x45, 0x77,
+	0x46, 0x77,
+	0x47, 0x77,
+	0x48, 0x77,
+	0x49, 0x77,
+	0x4a, 0x77,
+	0x4b, 0x77,
+	0x4c, 0x77,
+	0x4d, 0x77,
+	0x4e, 0x77,
+	0x4f, 0x77,
+	0x50, 0x77,
+	0x51, 0x77,
+	0x52, 0x77,
+	0x53, 0x77,
+	0x54, 0x77,
+	0x55, 0x77,
+	0x56, 0x77,
+	0x57, 0xff,
+	
 	0xff
 };
 
diff -u linux/drivers/media/dvb/ttpci/av7110_firm.h:1.18 linux/drivers/media/dvb/ttpci/av7110_firm.h:1.19
--- linux/drivers/media/dvb/ttpci/av7110_firm.h:1.18	Mon Oct 13 19:46:23 2003
+++ linux/drivers/media/dvb/ttpci/av7110_firm.h	Mon Oct 20 17:33:17 2003
@@ -1,7 +1,7 @@
 
 #include <asm/types.h>
 
-u8 Dpram [] __initdata = {
+static u8 Dpram [] = {
 	0xe5, 0x9f, 0xf0, 0x1c, 0xe1, 0xb0, 0xf0, 0x0e, 
 	0xe5, 0x9f, 0xf0, 0x18, 0xe2, 0x5e, 0xf0, 0x04, 
 	0xe2, 0x5e, 0xf0, 0x08, 0xe1, 0xa0, 0x00, 0x00, 
@@ -41,7 +41,7 @@
 };
 
 
-u8 Root [] __initdata = {
+static u8 Root [] = {
 	0xb4, 0x90, 0x49, 0x18, 0x1c, 0x0b, 0x4a, 0x18, 
 	0x1a, 0x50, 0x4f, 0x18, 0x1a, 0x79, 0x10, 0x8f, 
 	0x21, 0x00, 0x2f, 0x00, 0xdd, 0x04, 0xcb, 0x10, 
