--- linux-2.6.0-test10/drivers/video/sis/init301.c.orig	2003-11-24 14:36:57.000000000 +0800
+++ linux-2.6.0-test10/drivers/video/sis/init301.c	2003-11-24 14:37:59.000000000 +0800
@@ -11712,7 +11712,7 @@ SetOEMLCDData(SiS_Private *SiS_Pr, PSIS_
   }
 
   temp = GetOEMLCDPtr(SiS_Pr,HwDeviceExtension, ROMAddr, 1);
-  if(temp = 0xFFFF) return;
+  if(temp == 0xFFFF) return;
 
   index = SiS_Pr->SiS_VBModeIDTable[ModeIdIndex]._VB_LCDHIndex;
   for(i=0x14, j=0; i<=0x17; i++, j++) {
