--- linux-2.6.4-rc1/drivers/macintosh/macserial.c.org	2004-02-27 23:21:29.000000000 +0100
+++ linux-2.6.4-rc1/drivers/macintosh/macserial.c	2004-03-02 21:49:44.533392464 +0100
@@ -1781,6 +1781,7 @@
 {
 	struct mac_serial * info = (struct mac_serial *)tty->driver_data;
 	unsigned char control, status;
+	unsigned int cmd;
 	unsigned long flags;
 
 #ifdef CONFIG_KGDB
@@ -1811,6 +1812,7 @@
 {
 	struct mac_serial * info = (struct mac_serial *)tty->driver_data;
 	unsigned int arg, bits;
+	unsigned int cmd; 
 	unsigned long flags;
 
 #ifdef CONFIG_KGDB
