--- linux-2.6.7/drivers/media/common/saa7146_video.c.orig	2004-06-16 07:19:37.000000000 +0200
+++ linux-2.6.7/drivers/media/common/saa7146_video.c	2004-07-06 23:10:55.768840056 +0200
@@ -1,9 +1,9 @@
 #include <media/saa7146_vv.h>
 
-static int memory = 32;
+static int capmemory = 32;
 
-MODULE_PARM(memory,"i");
-MODULE_PARM_DESC(memory, "maximum memory usage for capture buffers (default: 32Mb)");
+MODULE_PARM(capmemory,"i");
+MODULE_PARM_DESC(capmemory, "maximum memory usage for capture buffers (default: 32Mb)");
 
 #define IS_CAPTURE_ACTIVE(fh) \
 	(((vv->video_status & STATUS_CAPTURE) != 0) && (vv->video_fh == fh))
@@ -1331,9 +1331,9 @@
 
 	*size = fh->video_fmt.sizeimage;
 
-	/* check if we exceed the "memory" parameter */
-	if( (*count * *size) > (memory*1048576) ) {
-		*count = (memory*1048576) / *size;
+	/* check if we exceed the "capmemory" parameter */
+	if( (*count * *size) > (capmemory*1048576) ) {
+		*count = (capmemory*1048576) / *size;
 	}
 	
 	DEB_CAP(("%d buffers, %d bytes each.\n",*count,*size));
