diff -urb linux.orig/drivers/isdn/hardware/eicon/divasmain.c linux/drivers/isdn/hardware/eicon/divasmain.c
--- linux.orig/drivers/isdn/hardware/eicon/divasmain.c  Wed Sep 10 17:29:53 2003
+++ linux/drivers/isdn/hardware/eicon/divasmain.c       Wed Sep 10 17:32:00 2003
@@ -71,8 +71,6 @@
 extern int divasfunc_init(int dbgmask);
 extern void divasfunc_exit(void);

-static devfs_handle_t devfs_handle;
-
 typedef struct _diva_os_thread_dpc {
        struct tasklet_struct divas_task;
        struct work_struct trap_script_task;

