--- linux-2.6.8-rc1/fs/gfs_locking/lock_gulm/utils_verb_flags.c.orig	2004-07-12 10:46:12.000000000 +0200
+++ linux-2.6.8-rc1/fs/gfs_locking/lock_gulm/utils_verb_flags.c	2004-07-12 10:46:36.015314232 +0200
@@ -21,7 +21,7 @@
 #include "gulm_log_msg_bits.h"
 
 static __inline__ int
-strncasecmp (const char *s1, const char *s2, size_t l)
+__strncasecmp (const char *s1, const char *s2, size_t l)
 {
 	char c1 = '\0', c2 = '\0';
 
@@ -231,7 +231,7 @@
 		if (token[0] == '-') {
 			token++;
 			for (i = 0; i < len; i++) {
-				if (strncasecmp
+				if (__strncasecmp
 				    (token, verbose_flags[i].name, tl) == 0) {
 					(*verb) &= ~(verbose_flags[i].val);
 				}
@@ -239,17 +239,17 @@
 		} else if (token[0] == '+') {
 			token++;
 			for (i = 0; i < len; i++) {
-				if (strncasecmp
+				if (__strncasecmp
 				    (token, verbose_flags[i].name, tl) == 0) {
 					(*verb) |= verbose_flags[i].val;
 				}
 			}
 		} else {
-			if (strncasecmp (token, "clear", tl) == 0) {
+			if (__strncasecmp (token, "clear", tl) == 0) {
 				(*verb) = 0;
 			} else {
 				for (i = 0; i < len; i++) {
-					if (strncasecmp
+					if (__strncasecmp
 					    (token, verbose_flags[i].name,
 					     tl) == 0) {
 						(*verb) |= verbose_flags[i].val;
