diff -aur linux.backup/fs/affs/file.c linux/fs/affs/file.c
--- linux.backup/fs/affs/file.c	Mon Jan 27 13:41:42 2003
+++ linux/fs/affs/file.c	Sun Aug 17 00:03:08 2003
@@ -793,8 +793,8 @@
 
 struct address_space_operations affs_aops_ofs = {
 	.readpage = affs_readpage_ofs,
-	//writepage: affs_writepage_ofs,
-	//sync_page: affs_sync_page_ofs,
+	//.writepage = affs_writepage_ofs,
+	//.sync_page = affs_sync_page_ofs,
 	.prepare_write = affs_prepare_write_ofs,
 	.commit_write = affs_commit_write_ofs
 };
diff -aur linux.backup/fs/jffs2/wbuf.c linux/fs/jffs2/wbuf.c
--- linux.backup/fs/jffs2/wbuf.c	Thu Jun 26 23:47:48 2003
+++ linux/fs/jffs2/wbuf.c	Sun Aug 17 00:04:26 2003
@@ -37,8 +37,8 @@
 #define NAND_JFFS2_OOB16_FSDALEN	8
 
 struct nand_oobinfo jffs2_oobinfo = {
-	useecc: 1,
-	eccpos: {JFFS2_OOB_ECCPOS0, JFFS2_OOB_ECCPOS1, JFFS2_OOB_ECCPOS2, JFFS2_OOB_ECCPOS3, JFFS2_OOB_ECCPOS4, JFFS2_OOB_ECCPOS5}
+	.useecc = 1,
+	.eccpos = {JFFS2_OOB_ECCPOS0, JFFS2_OOB_ECCPOS1, JFFS2_OOB_ECCPOS2, JFFS2_OOB_ECCPOS3, JFFS2_OOB_ECCPOS4, JFFS2_OOB_ECCPOS5}
 };
 
 static inline void jffs2_refile_wbuf_blocks(struct jffs2_sb_info *c)
