--- linux-2.6.4-rc1/include/asm-sparc/unistd.h.org	2004-02-27 23:20:57.000000000 +0100
+++ linux-2.6.4-rc1/include/asm-sparc/unistd.h	2004-03-02 13:10:39.229091144 +0100
@@ -461,7 +461,7 @@
 				unsigned long addr, unsigned long len,
 				unsigned long prot, unsigned long flags,
 				unsigned long fd, unsigned long pgoff);
-asmlinkage int sys_ioperm(unsigned long from, unsigned long num, int on);
+asmlinkage long sys_ioperm(unsigned long from, unsigned long num, int on);
 struct sigaction;
 asmlinkage long sys_rt_sigaction(int sig,
 				const struct sigaction __user *act,
