diff -Nru linux-2.6.0-test8/arch/h8300/kernel/time.c linux-2.6.0-test8-h8300/arch/h8300/kernel/time.c
--- linux-2.6.0-test8/arch/h8300/kernel/time.c	2003-10-17 00:07:45.000000000 +0900
+++ linux-2.6.0-test8-h8300/arch/h8300/kernel/time.c	2003-10-19 02:26:08.000000000 +0900
@@ -143,3 +143,8 @@
 }
 
 EXPORT_SYMBOL(do_settimeofday);
+
+unsigned long long sched_clock(void)
+{
+	return (unsigned long long)jiffies * (1000000000 / HZ);
+}
diff -Nru linux-2.6.0-test8/arch/h8300/lib/checksum.c linux-2.6.0-test8-h8300/arch/h8300/lib/checksum.c
--- linux-2.6.0-test8/arch/h8300/lib/checksum.c	2003-10-17 00:07:45.000000000 +0900
+++ linux-2.6.0-test8-h8300/arch/h8300/lib/checksum.c	2003-10-19 02:27:49.000000000 +0900
@@ -32,7 +32,6 @@
    of the assembly has to go. */
 
 #include <net/checksum.h>
-#include <net/module.h>
 
 static inline unsigned short from32to16(unsigned long x)
 {
