--- linux-2.6.7/include/asm-ppc/ppc_asm.h.orig	2004-07-06 16:44:48.000000000 +0200
+++ linux-2.6.7/include/asm-ppc/ppc_asm.h	2004-07-06 16:57:45.784635032 +0200
@@ -73,6 +73,16 @@
 #define REST_16EVR(n,s,base)	REST_8EVR(n,s,base); REST_8EVR(n+8,s,base)
 #define REST_32EVR(n,s,base)	REST_16EVR(n,s,base); REST_16EVR(n+16,s,base)
 
+/*
+ * Once a version of gas that understands the AltiVec instructions
+  * is freely available, we can do this the normal way...  - paulus
+   */
+#define LVX(r,a,b)	.long	(31<<26)+((r)<<21)+((a)<<16)+((b)<<11)+(103<<1)
+#define STVX(r,a,b)	.long	(31<<26)+((r)<<21)+((a)<<16)+((b)<<11)+(231<<1)
+#define MFVSCR(r)	.long	(4<<26)+((r)<<21)+(770<<1)
+#define MTVSCR(r)	.long	(4<<26)+((r)<<11)+(802<<1)
+#define DSSALL		.long	(0x1f<<26)+(0x10<<21)+(0x336<<1)
+
 #ifdef CONFIG_PPC601_SYNC_FIX
 #define SYNC				\
 BEGIN_FTR_SECTION			\
