--- linux-2.6.4/drivers/video/i810/i810_main.h	2003-09-27 18:57:46.000000000 -0700
+++ 25/drivers/video/i810/i810_main.h	2004-03-10 21:52:02.000000000 -0800
@@ -84,7 +84,7 @@ extern void i810fb_init_ringbuffer(struc
 extern void i810fb_load_front     (u32 offset, struct fb_info *info);
 
 /* Conditionals */
-#if defined(__i386__)
+#ifdef CONFIG_X86
 inline void flush_cache(void)
 {
 	asm volatile ("wbinvd":::"memory");
