diff -Nru a/init/Kconfig b/init/Kconfig
--- a/init/Kconfig	Tue Oct 21 16:07:24 2003
+++ b/init/Kconfig	Tue Oct 21 16:07:24 2003
@@ -194,28 +194,28 @@
 	  Disabling this option will cause the kernel to be built without
 	  support for epoll family of system calls.
 
-if X86 || IA64
-
 config CPUSETS
-        bool "cpusets"
-        depends on SMP
+        bool "Cpuset support"
+        depends on SMP && (X86 || IA64) && EXPERIMENTAL
         help
-          This options will let you create and manage sets of cpu where you
-          can run the processes.
+          This options will let you create and manage CPUSET's which
+	  allow partioning a SMP machine into execution areas. Processes
+	  in a CPUSET are not allowed to run on processor's outside
+	  the given group.
   
           Say N if unsure.
 
 config CPUSETS_PROC
         bool "/proc/cpusets support"
-        depends on CPUSETS
+        depends on CPUSETS && PROC_FS
+	default y
         help
           Get some info about the existing cpusets in your system.
-          To use this option, you have to ensure that the "/proc file system
-          support" (CONFIG_PROC_FS) is enabled, too.
 
 config CPUSETS_PROC_CPUINFO
         bool "/proc/cpuinfo uses current cpuset"
         depends on CPUSETS_PROC
+	default y
         help
           With this option enabled, a process reading /proc/cpuinfo will
           only see the CPUs that are in its current cpuset.
@@ -223,11 +223,11 @@
 config CPUSETS_PROC_STAT
         bool "/proc/stat uses current cpuset"
         depends on CPUSETS_PROC
+	default y
         help
           With this option enabled, a process reading /proc/stat will
           only see the CPUs that are in its current cpuset.
 
-endif
 
 source "drivers/block/Kconfig.iosched"
 
