--- linux-2.6.7/fs/gfs_locking/lock_gulm//utils_verb_flags.c.orig	2004-07-07 01:40:49.879527400 +0200
+++ linux-2.6.7/fs/gfs_locking/lock_gulm//utils_verb_flags.c	2004-07-07 01:41:10.846339960 +0200
@@ -20,6 +20,7 @@
 
 #include "gulm_log_msg_bits.h"
 
+#ifndef CONFIG_PPC
 static __inline__ int
 strncasecmp (const char *s1, const char *s2, size_t l)
 {
@@ -40,6 +41,9 @@
 	}
 	return (c1 - c2);
 }
+#else
+extern int strncasecmp (const char *, const char *, int)
+#endif
 
 static int bit_array[16] = { 0, 1, 1, 2, 1, 2, 2, 3, 1, 2, 2, 3, 2, 3, 3, 4 };
 
