--- linux-2.6.5-rc2/arch/sparc64/kernel/binfmt_elf32.c.org	2004-03-22 13:55:59.000000000 +0100
+++ linux-2.6.5-rc2/arch/sparc64/kernel/binfmt_elf32.c	2004-03-22 13:01:24.000000000 +0100
@@ -4,6 +4,7 @@
  * Copyright (C) 1995, 1996, 1997, 1998 David S. Miller	(davem@redhat.com)
  * Copyright (C) 1995, 1996, 1997, 1998 Jakub Jelinek	(jj@ultra.linux.cz)
  */
+#include <asm/pgtable.h>
 
 #define ELF_ARCH		EM_SPARC
 #define ELF_CLASS		ELFCLASS32
