--- linux-2.6.7/drivers/block/cciss.c.orig	2004-07-06 19:11:30.000000000 +0200
+++ linux-2.6.7/drivers/block/cciss.c	2004-07-06 21:53:26.603620600 +0200
@@ -208,7 +208,7 @@
         ctlr_info_t *h = (ctlr_info_t*)data;
         drive_info_struct *drv;
 	unsigned long flags;
-	unsigned int vol_sz, vol_sz_frac;
+	uint64_t vol_sz, vol_sz_frac;
 
         ctlr = h->ctlr;
 
@@ -269,7 +269,7 @@
 		if (drv->raid_level > 5)
 			drv->raid_level = RAID_UNKNOWN;
 		size = sprintf(buffer+len, "cciss/c%dd%d:"
-				"\t%4d.%02dGB\tRAID %s\n",
+				"\t%4llu.%02lluGB\tRAID %s\n",
 				ctlr, i, vol_sz,vol_sz_frac,
 				raid_label[drv->raid_level]);
                 pos += size; len += size;
