diff -u linux-2.6.0-test11/sound/pci/via82xx.c linux-2.6.0-test11-af1/sound/pci/via82xx.c
--- linux-2.6.0-test11/sound/pci/via82xx.c      2003-11-29 19:18:22.000000000 +0100
+++ linux-2.6.0-test11-af1/sound/pci/via82xx.c  2003-11-29 19:38:41.000000000 +0100
@@ -1969,6 +1969,7 @@
 	static struct dxs_whitelist whitelist[] = {
 		{ .vendor = 0x1019, .device = 0x0996, .action = VIA_DXS_48K },
 		{ .vendor = 0x1297, .device = 0xc160, .action = VIA_DXS_ENABLE }, /* Shuttle SK41G */
+		{ .vendor = 0x1695, .device = 0x3005, .action = VIA_DXS_ENABLE }, /* EPoX EP-8K9A  */
 		{ } /* terminator */
 	};
 	struct dxs_whitelist *w;
