--- linux-2.6.7/kernel/signal.c.orig	2004-07-07 16:32:08.000000000 +0200
+++ linux-2.6.7/kernel/signal.c	2004-07-07 18:14:39.978062528 +0200
@@ -1071,7 +1071,7 @@
 
 int __kill_pg_info(int sig, struct siginfo *info, pid_t pgrp)
 {
-	struct task_struct *p;
+	struct task_struct *p = NULL;
 	struct list_head *l;
 	struct pid *pid;
 	int retval, success;
--- linux-2.6.7/sound/oss/ad1889.c.orig	2004-06-16 07:19:12.000000000 +0200
+++ linux-2.6.7/sound/oss/ad1889.c	2004-07-08 14:36:04.048339480 +0200
@@ -851,7 +851,7 @@
 	}
 
 	eid = ad1889_codec_read(ac97, AC97_EXTENDED_ID);
-	if (eid == 0xffffff) {
+	if (eid == 0xffff) {
 		printk(KERN_WARNING DEVNAME ": no codec attached?\n");
 		goto out_free;
 	}
