--- linux-2.6.4-rc1/drivers/net/imq.c.orig	2004-03-03 03:43:30.000000000 +0100
+++ linux-2.6.4-rc1/drivers/net/imq.c	2004-03-03 03:54:28.463396008 +0100
@@ -38,7 +38,7 @@
 	.owner		= THIS_MODULE,
 	.pf		= PF_INET,
 	.hooknum	= NF_IP_PRE_ROUTING,
-	.priority	= NF_IP_PRI_MANGLE + 1
+	.priority	= NF_IP_PRI_NAT_DST + 1
 };
 
 static struct nf_hook_ops imq_egress_ipv4 = {
@@ -55,7 +55,7 @@
 	.owner		= THIS_MODULE,
 	.pf		= PF_INET6,
 	.hooknum	= NF_IP6_PRE_ROUTING,
-	.priority	= NF_IP6_PRI_MANGLE + 1
+	.priority	= NF_IP6_PRI_NAT_DST + 1
 };
 
 static struct nf_hook_ops imq_egress_ipv6 = {
