
--- linux-2.6.2/drivers/video/aty/aty128fb.c~	2004-02-16 14:39:19.000000000 +0000
+++ linux-2.6.2/drivers/video/aty/aty128fb.c	2004-02-16 14:40:25.000000000 +0000
@@ -1952,7 +1952,7 @@
 	bios = aty128_map_ROM(par, pdev);
 #ifdef __i386__
 	if (bios == NULL)
-		bios = aty128_find_mem_vbios(par, pdev);
+		bios = aty128_find_mem_vbios(par);
 #endif
 	if (bios == NULL)
 		printk(KERN_INFO "aty128fb: BIOS not located, guessing timings.\n");
