--- linux-2.6.7/include/linux/swap.h.orig	2004-07-06 22:03:43.000000000 +0200
+++ linux-2.6.7/include/linux/swap.h	2004-07-07 11:30:55.669029256 +0200
@@ -4,7 +4,6 @@
 #include <linux/config.h>
 #include <linux/spinlock.h>
 #include <linux/linkage.h>
-#include <linux/pagemap.h>
 #include <linux/mmzone.h>
 #include <linux/list.h>
 #include <linux/sched.h>
@@ -67,6 +66,13 @@
 } swp_entry_t;
 
 /*
+ * [err] asm-sparc/pgtable.h: attempt to use undefined type (swp_entry_t), compilation failed.
+ * [explain] asm/pgtable.h -> linux/swap.h -> linux/pagemap.h -> linux/mm.h -> asm/pgtable.h
+ * [workaround] swp_entry_t moved outside the cycle.
+ */
+#include <linux/pagemap.h>
+
+/*
  * current->reclaim_state points to one of these when a task is running
  * memory reclaim
  */
