--- linux-2.6.2.org/drivers/scsi/Kconfig	2004-02-05 10:00:16.000000000 +0000
+++ linux-2.6.2/drivers/scsi/Kconfig	2004-02-05 11:59:31.000000000 +0000
@@ -727,7 +727,7 @@
 
 config SCSI_INITIO
 	tristate "Initio 9100U(W) support"
-	depends on PCI && SCSI && BROKEN
+	depends on PCI && SCSI
 	help
 	  This is support for the Initio 91XXU(W) SCSI host adapter.  Please
 	  read the SCSI-HOWTO, available from
