--- linux-2.6.8-rc3/arch/i386/kernel/apm.c.orig	2004-08-09 11:09:31.000000000 +0200
+++ linux-2.6.8-rc3/arch/i386/kernel/apm.c	2004-08-09 16:29:51.085575808 +0200
@@ -2271,10 +2271,12 @@
 	}
 	if ((num_online_cpus() > 1) && !power_off && !smp) {
 		printk(KERN_NOTICE "apm: disabled - APM is not SMP safe.\n");
+		apm_info.disabled = 1;
 		return -ENODEV;
 	}
 	if (PM_IS_ACTIVE()) {
 		printk(KERN_NOTICE "apm: overridden by ACPI.\n");
+		apm_info.disabled = 1;
 		return -ENODEV;
 	}
 	pm_active = 1;
