--- linux-2.6.7/net/bridge/br_sysfs_br.c.orig	2004-06-16 07:19:22.000000000 +0200
+++ linux-2.6.7/net/bridge/br_sysfs_br.c	2004-06-17 13:10:11.000000000 +0200
@@ -305,9 +305,7 @@
  * This is a dummy kset so bridge objects don't cause
  * hotplug events 
  */
-struct subsystem bridge_subsys = { 
-	.kset = { .hotplug_ops = NULL },
-};
+decl_subsys_name(bridge, net_bridge, NULL, NULL);
 
 void br_sysfs_init(void)
 {
