diff -Nur linux-2.6.0-test11.org/fs/jffs/intrep.c linux-2.6.0-test11/fs/jffs/intrep.c
--- linux-2.6.0-test11.org/fs/jffs/intrep.c	2003-11-26 21:44:24.000000000 +0100
+++ linux-2.6.0-test11/fs/jffs/intrep.c	2003-12-13 22:21:25.958506776 +0100
@@ -3431,3 +3431,6 @@
 		up(&fmc->biglock);
 	} /* for (;;) */
 } /* jffs_garbage_collect_thread() */
+
+EXPORT_SYMBOL(set_special_pids);
+
diff -Nur linux-2.6.0-test11.org/net/ipv4/netfilter/ip_pool.c linux-2.6.0-test11/net/ipv4/netfilter/ip_pool.c
--- linux-2.6.0-test11.org/net/ipv4/netfilter/ip_pool.c	2003-12-13 22:22:33.500238872 +0100
+++ linux-2.6.0-test11/net/ipv4/netfilter/ip_pool.c	2003-12-13 21:33:06.000000000 +0100
@@ -330,3 +330,7 @@
 
 module_init(init);
 module_exit(fini);
+
+EXPORT_SYMBOL(ip_pool_match);
+EXPORT_SYMBOL(ip_pool_mod);
+
diff -Nur linux-2.6.0-test11.org/net/ipv4/netfilter/Makefile linux-2.6.0-test11/net/ipv4/netfilter/Makefile
--- linux-2.6.0-test11.org/net/ipv4/netfilter/Makefile	2003-12-13 22:22:33.604223064 +0100
+++ linux-2.6.0-test11/net/ipv4/netfilter/Makefile	2003-12-13 21:33:26.000000000 +0100
@@ -36,7 +36,7 @@
 obj-$(CONFIG_IP_NF_NAT_RTSP) += ip_nat_rtsp.o
 
 obj-$(CONFIG_IP_NF_QUAKE3) += ip_conntrack_quake3.o
-obj-$(CONFIG_IP_NF_PPTP) += ip_conntrack_pptp.o
+obj-$(CONFIG_IP_NF_PPTP) += ip_conntrack_pptp.o ip_conntrack_proto_gre.o
 obj-$(CONFIG_IP_NF_MMS) += ip_conntrack_mms.o
 obj-$(CONFIG_IP_NF_AMANDA) += ip_conntrack_amanda.o
 obj-$(CONFIG_IP_NF_TFTP) += ip_conntrack_tftp.o
