--- linux-2.6.0-test3/drivers/net/wan/sdlamain.orig.c	2003-08-18 21:44:12.000000000 +0530
+++ linux-2.6.0-test3/drivers/net/wan/sdlamain.c	2003-08-18 21:45:07.000000000 +0530
@@ -1029,7 +1029,6 @@
 STATIC irqreturn_t sdla_isr (int irq, void* dev_id, struct pt_regs *regs)
 {
 #define	card	((sdla_t*)dev_id)
-	int handled = 0;
 
 	if(card->hw.type == SDLA_S514) {	/* handle interrrupt on S514 */
                 u32 int_status;
--- linux-2.6.0-test3/drivers/video/tridentfb.orig.c	2003-08-18 22:02:56.000000000 +0530
+++ linux-2.6.0-test3/drivers/video/tridentfb.c	2003-08-18 22:03:36.000000000 +0530
@@ -454,7 +454,7 @@
 static void tridentfb_fillrect(struct fb_info * info, const struct fb_fillrect *fr)
 {
 	int bpp = info->var.bits_per_pixel;
-	int dx,dy,w,h,col;
+	int col;
 	
 	switch (bpp) {
 		case 8: col = fr->color;

--- linux-2.6.0-test3/drivers/char/esp.orig.c	2003-08-18 21:19:20.000000000 +0530
+++ linux-2.6.0-test3/drivers/char/esp.c	2003-08-18 21:20:01.000000000 +0530
@@ -2632,7 +2632,7 @@
 static void __exit espserial_exit(void) 
 {
 	unsigned long flags;
-	int e1, e2;
+	int e1;
 	unsigned int region_start, region_end;
 	struct esp_struct *temp_async;
 	struct esp_pio_buffer *pio_buf;

--- linux-2.6.0-test3/drivers/char/isicom.orig.c	2003-08-18 21:07:10.000000000 
+0530
+++ linux-2.6.0-test3/drivers/char/isicom.c	2003-08-18 21:07:25.000000000 
+0530
@@ -972,7 +972,6 @@
 	struct isi_port * port;
 	struct isi_board * card;
 	unsigned int line, board;
-	unsigned long flags;
 	int error;
 
 #ifdef ISICOM_DEBUG	

