diff -urN linux-2.6.6.orig/drivers/char/lirc/lirc_i2c.c linux-2.6.6/drivers/char/lirc/lirc_i2c.c
--- linux-2.6.6.orig/drivers/char/lirc/lirc_i2c.c	2004-05-12 16:42:35.000000000 +0000
+++ linux-2.6.6/drivers/char/lirc/lirc_i2c.c	2004-05-13 09:11:21.000000000 +0000
@@ -1,4 +1,4 @@
-/*      $Id$      */
+/*      $Id$      */
 
 /*
  * i2c IR lirc plugin for Hauppauge and Pixelview cards - new 2.8.x i2c stack
@@ -134,7 +134,7 @@
 
 	if (rc == -1) {
 		/* should this return -EIO? */	
-		dprintk(DEVICE_NAME ": %s read error\n", ir->c.name);
+		dprintk(DEVICE_NAME ": %s read error\n", ir->client.name);
 		return -ENODATA;
 	}
 
@@ -167,7 +167,7 @@
 	unsigned char codes[2];
 
 	/* poll IR chip */
-	if (3 == i2c_master_recv(&ir->c,keybuf,3)) {
+	if (3 == i2c_master_recv(&ir->client,keybuf,3)) {
 		ir->b[0] = keybuf[0];
 		ir->b[1] = keybuf[1];
 		ir->b[2] = keybuf[2];
@@ -200,7 +200,7 @@
 	unsigned char key;
 	
 	/* poll IR chip */
-	if (1 != i2c_master_recv(&ir->c,&key,1)) {
+	if (1 != i2c_master_recv(&ir->client,&key,1)) {
 		dprintk(DRIVER_NAME ": read error\n");
 		return -1;
 	}
@@ -209,7 +209,9 @@
 
 	/* return it */
 	lirc_buffer_write_1( buf, &key );
-	*key = b;
+/* not sure where b came from
+ 	*key = b;
+*/
 	return 0;
 }
 
@@ -220,7 +222,7 @@
 	unsigned char codes[4];
 	
 	/* poll IR chip */
-	if (1 != i2c_master_recv(&ir->c,&key,1)) {
+	if (1 != i2c_master_recv(&ir->client,&key,1)) {
 		dprintk(DRIVER_NAME ": read error\n");
 		return -ENODATA;
 	}
@@ -245,7 +247,7 @@
 	unsigned char key;
 	
 	/* poll IR chip */
-	if (1 != i2c_master_recv(&ir->c,&key,1)) {
+	if (1 != i2c_master_recv(&ir->client,&key,1)) {
 		dprintk(DRIVER_NAME ": read error\n");
 		return -ENODATA;
 	}
