--- linux-2.6.4-rc2.org/drivers/ide/Kconfig	2004-03-04 06:16:45.000000000 +0000
+++ linux-2.6.4-rc2/drivers/ide/Kconfig	2004-03-08 10:36:10.000000000 +0000
@@ -699,7 +699,7 @@
 
 config PDC202XX_BURST
 	bool "Special UDMA Feature"
-	depends on BLK_DEV_PDC202XX_OLD=y
+	depends on BLK_DEV_PDC202XX_OLD
 	---help---
 	  This option causes the pdc202xx driver to enable UDMA modes on the
 	  PDC202xx even when the PDC202xx BIOS has not done so.
@@ -720,7 +720,7 @@
 # FIXME - probably wants to be one for old and for new
 config PDC202XX_FORCE
 	bool "Enable controller even if disabled by BIOS"
-	depends on BLK_DEV_PDC202XX_NEW=y
+	depends on BLK_DEV_PDC202XX_NEW
 	help
 	  Enable the PDC202xx controller even if it has been disabled in the BIOS setup.
 
