--- linux-2.6.7-rc3/include/asm-alpha/pgalloc.h~	2004-06-09 21:28:45.000000000 +0000
+++ linux-2.6.7-rc3/include/asm-alpha/pgalloc.h	2004-06-09 21:28:45.000000000 +0000
@@ -77,4 +77,8 @@
 
 #define check_pgt_cache()	do { } while (0)
 
+#define arch_add_exec_range(mm, limit)         do { ; } while (0)
+#define arch_flush_exec_range(mm)              do { ; } while (0)
+#define arch_remove_exec_range(mm, limit)      do { ; } while (0)
+
 #endif /* _ALPHA_PGALLOC_H */
