diff -Nur linux-2.6.0-test5.old/drivers/video/aty/aty128fb.c linux-2.6.0-test5/drivers/video/aty/aty128fb.c
--- linux-2.6.0-test5.old/drivers/video/aty/aty128fb.c	Mon Sep  8 19:50:04 2003
+++ linux-2.6.0-test5/drivers/video/aty/aty128fb.c	Fri Sep 12 09:59:55 2003
@@ -2041,9 +2041,9 @@
 #define ATY_MIRROR_CRT_ON	0x00000002
 
 /* out param: u32*	backlight value: 0 to 15 */
-#define FBIO_ATY128_GET_MIRROR	_IOR('@', 1, sizeof(__u32*))
+#define FBIO_ATY128_GET_MIRROR	_IOR('@', 1, long)
 /* in param: u32*	backlight value: 0 to 15 */
-#define FBIO_ATY128_SET_MIRROR	_IOW('@', 2, sizeof(__u32*))
+#define FBIO_ATY128_SET_MIRROR	_IOW('@', 2, long)
 
 static int aty128fb_ioctl(struct inode *inode, struct file *file, u_int cmd,
 			  u_long arg, struct fb_info *info)
diff -Nur linux-2.6.0-test5.old/include/linux/pmu.h linux-2.6.0-test5/include/linux/pmu.h
--- linux-2.6.0-test5.old/include/linux/pmu.h	Mon Sep  8 19:49:52 2003
+++ linux-2.6.0-test5/include/linux/pmu.h	Fri Sep 12 09:27:28 2003
@@ -132,7 +132,7 @@
 /* out param: u32*	can_sleep: 0 or 1 */
 #define PMU_IOC_CAN_SLEEP	_IOR('B', 5, size_t) 
 /* no param */
-#define PMU_IOC_GRAB_BACKLIGHT	_IOR('B', 6, 0) 
+#define PMU_IOC_GRAB_BACKLIGHT	_IOR('B', 6, size_t) 
 
 #ifdef __KERNEL__
 
