--- linux-2.6.3/arch/x86_64/kernel/setup.c.org	2004-02-19 10:17:00.000000000 +0100
+++ linux-2.6.3/arch/x86_64/kernel/setup.c	2004-02-19 10:19:20.158076921 +0100
@@ -589,7 +589,8 @@
 static void __init detect_ht(void)
 {
 	extern	int phys_proc_id[NR_CPUS];
-	
+	int	smp_num_siblings;
+
 	u32 	eax, ebx, ecx, edx;
 	int 	index_lsb, index_msb, tmp;
 	int	initial_apic_id;
--- linux-2.6.3/arch/x86_64/kernel/x8664_ksyms.c.org	2004-02-19 10:17:04.000000000 +0100
+++ linux-2.6.3/arch/x86_64/kernel/x8664_ksyms.c	2004-02-19 10:21:23.993254962 +0100
@@ -194,8 +194,6 @@
 
 EXPORT_SYMBOL(die_chain);
 
-EXPORT_SYMBOL(cpu_sibling_map);
-
 extern void do_softirq_thunk(void);
 EXPORT_SYMBOL_NOVERS(do_softirq_thunk);
 
