--- linux-2.6.8.1/drivers/scsi/Kconfig.org	2004-12-20 10:23:11.000000000 +0000
+++ linux-2.6.8.1/drivers/scsi/Kconfig	2004-12-20 10:28:39.295638928 +0000
@@ -380,7 +380,7 @@
 
 config SCSI_ADVANSYS
 	tristate "AdvanSys SCSI support"
-	depends on (ISA || EISA || PCI) && SCSI && BROKEN
+	depends on (ISA || EISA || PCI) && SCSI
 	help
 	  This is a driver for all SCSI host adapters manufactured by
 	  AdvanSys. It is documented in the kernel source in
