diff -urN linux-2.6.7.orig/drivers/usb/media/Kconfig linux-2.6.7/drivers/usb/media/Kconfig
--- linux-2.6.7.orig/drivers/usb/media/Kconfig	2004-07-11 11:38:13.590875280 +0200
+++ linux-2.6.7/drivers/usb/media/Kconfig	2004-07-11 18:49:41.990240360 +0200
@@ -146,6 +146,10 @@
 	  To compile this driver as a module, choose M here: the
 	  module will be called pwc.
 
+config USB_PWCX
+	tristate "PWCX decompressor module"
+	depends on USB_PWC
+	
 config USB_SE401
 	tristate "USB SE401 Camera support"
 	depends on USB && VIDEO_DEV
diff -urN linux-2.6.7.orig/drivers/usb/media/Makefile linux-2.6.7/drivers/usb/media/Makefile
--- linux-2.6.7.orig/drivers/usb/media/Makefile	2004-07-11 11:38:13.590875280 +0200
+++ linux-2.6.7/drivers/usb/media/Makefile	2004-07-11 18:48:43.055199848 +0200
@@ -3,6 +3,7 @@
 #
 
 pwc-objs	:= pwc-if.o pwc-misc.o pwc-ctrl.o pwc-uncompress.o
+pwcx-objs       := pwcx-glue.o libpwcx.a_
 
 obj-$(CONFIG_USB_DABUSB)	+= dabusb.o
 obj-$(CONFIG_USB_DSBR)		+= dsbr100.o
@@ -10,6 +11,7 @@
 obj-$(CONFIG_USB_KONICAWC)	+= konicawc.o usbvideo.o
 obj-$(CONFIG_USB_OV511)		+= ov511.o
 obj-$(CONFIG_USB_PWC)		+= pwc.o
+obj-$(CONFIG_USB_PWCX)		+= pwcx.o
 obj-$(CONFIG_USB_SE401)		+= se401.o
 obj-$(CONFIG_USB_STV680)	+= stv680.o
 obj-$(CONFIG_USB_VICAM)		+= vicam.o usbvideo.o
