--- linux-2.6.11.5.orig/drivers/pci/pci.c	2005-03-19 07:34:54.000000000 +0100
+++ linux-2.6.11.5/drivers/pci/pci.c	2005-03-28 10:46:16.482755728 +0200
@@ -268,7 +268,7 @@
 		return -EIO; 
 
 	pci_read_config_word(dev,pm + PCI_PM_PMC,&pmc);
-	if ((pmc & PCI_PM_CAP_VER_MASK) > 2) {
+	if ((pmc & PCI_PM_CAP_VER_MASK) != 2) {
 		printk(KERN_DEBUG
 		       "PCI: %s has unsupported PM cap regs version (%u)\n",
 		       dev->slot_name, pmc & PCI_PM_CAP_VER_MASK);
