diff -ruN 999-old/kernel/power/suspend.c 999-new/kernel/power/suspend.c
--- 999-old/kernel/power/suspend.c	2004-09-28 08:21:43.000000000 +1000
+++ 999-new/kernel/power/suspend.c	2004-09-28 08:21:33.000000000 +1000
@@ -325,10 +325,11 @@
 
 	/* Now check for graphics class devices, so we can keep the display on while suspending */
 	class = class_find("graphics");
-	if (class)
+	if (class) {
 		list_for_each_entry(class_dev, &class->children, node)
 			device_switch_trees(class_dev->dev, suspend_device_tree);
-	class_put(class);
+		class_put(class);
+	}
 	return 0;
 }
