--- linux-2.6.7/fs/gfs_locking/lock_gulm/utils_verb_flags.c.orig	2004-07-06 18:20:32.000000000 +0000
+++ linux-2.6.7/fs/gfs_locking/lock_gulm/utils_verb_flags.c	2004-07-06 18:56:35.685436549 +0000
@@ -20,26 +20,6 @@
 
 #include "gulm_log_msg_bits.h"
 
-static __inline__ int
-strncasecmp (const char *s1, const char *s2, size_t l)
-{
-	char c1 = '\0', c2 = '\0';
-
-	while (*s1 && *s2 && l-- > 0) {
-		c1 = *s1++;
-		c2 = *s2++;
-
-		if (c1 >= 'A' && c1 <= 'Z')
-			c1 += 'a' - 'A';
-
-		if (c2 >= 'A' && c2 <= 'Z')
-			c2 += 'a' - 'A';
-
-		if (c1 != c2)
-			break;
-	}
-	return (c1 - c2);
-}
 
 static int bit_array[16] = { 0, 1, 1, 2, 1, 2, 2, 3, 1, 2, 2, 3, 2, 3, 3, 4 };
 
