--- linux-2.6.6/net/ipv6/ip6_output.c.org	2004-05-10 04:31:59.000000000 +0200
+++ linux-2.6.6/net/ipv6/ip6_output.c	2004-05-11 18:35:06.402777016 +0200
@@ -805,6 +805,7 @@
 	*dst = NULL;
 	return err;
 }
+EXPORT_SYMBOL(ip6_dst_lookup);
 
 int ip6_append_data(struct sock *sk, int getfrag(void *from, char *to, int offset, int len, int odd, struct sk_buff *skb),
 		    void *from, int length, int transhdrlen,
