--- linux-2.6.0-test6/drivers/acpi/asus_acpi.c.wiget	2003-10-02 20:47:04.000000000 +0200
+++ linux-2.6.0-test6/drivers/acpi/asus_acpi.c	2003-10-02 22:54:06.000000000 +0200
@@ -34,7 +34,7 @@
 
 #include <linux/config.h>
 #if defined (CONFIG_MODVERSIONS) && !defined (MODVERSIONS) && defined (MODULE)
-#include <linux/modversions.h>
+#include <config/modversions.h>
 #endif
 
 #include <linux/kernel.h>
