diff -urN linux-2.6.7.orig/drivers/usb/media/pwc-if.c linux-2.6.7/drivers/usb/media/pwc-if.c
--- linux-2.6.7.orig/drivers/usb/media/pwc-if.c	2004-07-11 11:38:13.671862968 +0200
+++ linux-2.6.7/drivers/usb/media/pwc-if.c	2004-07-05 22:03:39.000000000 +0200
@@ -130,7 +130,7 @@
 
 static int pwc_video_open(struct inode *inode, struct file *file);
 static int pwc_video_close(struct inode *inode, struct file *file);
-static ssize_t pwc_video_read(struct file *file, char __user *buf,
+static ssize_t pwc_video_read(struct file *file, char *buf,
 			  size_t count, loff_t *ppos);
 static unsigned int pwc_video_poll(struct file *file, poll_table *wait);
 static int  pwc_video_ioctl(struct inode *inode, struct file *file,
@@ -1140,7 +1140,7 @@
                 device is tricky anyhow.
  */
 
-static ssize_t pwc_video_read(struct file *file, char __user *buf,
+static ssize_t pwc_video_read(struct file *file, char *buf,
 			  size_t count, loff_t *ppos)
 {
 	struct video_device *vdev = file->private_data;
@@ -1801,19 +1801,6 @@
 	else if (vendor_id == 0x06be) {
 		switch(product_id) {
 		case 0x8116:
-			/* Basicly the same as the Sotec Afina Eye */                
-			Info("AME CU-001 USB webcam detected.\n");
-			name = "AME CU-001";
-			type_id = 730;
-			break;
-		default:
-			return -ENODEV;
-			break;
-		}
-	}
-	else if (vendor_id == 0x06be) {
-		switch(product_id) {
-		case 0x8116:
 			/* This is essentially the same cam as the Sotec Afina Eye */
 			Info("AME Co. Afina Eye USB webcam detected.\n");
 			name = "AME Co. Afina Eye";
diff -urN linux-2.6.7.orig/drivers/usb/media/pwc-uncompress.h linux-2.6.7/drivers/usb/media/pwc-uncompress.h
--- linux-2.6.7.orig/drivers/usb/media/pwc-uncompress.h	2004-07-11 11:38:13.706857648 +0200
+++ linux-2.6.7/drivers/usb/media/pwc-uncompress.h	2004-07-05 22:03:39.000000000 +0200
@@ -48,10 +48,9 @@
 	int  type;		/* type of camera (645, 680, etc) */
 	int  table_size;	/* memory needed */
 
-	void (* init)(int type, int release, void *buffer, void *table);	/* Initialization routine; should be called after each set_video_mode */
-	void (* exit)(void);	/* Cleanup routine */
-	void (* decompress)(struct pwc_coord *image, struct pwc_coord *view,
-			    struct pwc_coord *offset,
+	asmlinkage void (* init)(int type, int release, void *buffer, void *table);	/* Initialization routine; should be called after each set_video_mode */
+	asmlinkage void (* exit)(void);	/* Cleanup routine */
+	asmlinkage void (* decompress)(struct pwc_coord *image, struct pwc_coord *view, struct pwc_coord *offset,
                             void *src, void *dst, int flags,
 	                    void *table, int bandlength);
 	void (* lock)(void);	/* make sure module cannot be unloaded */
