--- linux-2.6/drivers/char/ppdev.c.orig	2003-09-12 12:02:43.000000000 -0700
+++ linux-2.6/drivers/char/ppdev.c	2003-12-22 14:08:13.000000000 -0800
@@ -67,6 +67,7 @@
 #include <asm/uaccess.h>
 #include <linux/ppdev.h>
 #include <linux/smp_lock.h>
+#include <linux/device.h>
 
 #define PP_VERSION "ppdev: user-space parallel port driver"
 #define CHRDEV "ppdev"
@@ -782,4 +783,4 @@
 module_exit(ppdev_cleanup);
 
 MODULE_LICENSE("GPL");
-
+MODULE_ALIAS_CHARDEV_MAJOR(PP_MAJOR);
