--- linux-2.6.2-rc3/drivers/media/dvb/frontends/dst.c	2004-01-06 15:29:13.000000000 +0100
+++ linux-m68k-2.6.2-rc3/drivers/media/dvb/frontends/dst.c	2004-01-10 17:10:35.000000000 +0100
@@ -28,8 +28,8 @@
 #include <linux/string.h>
 #include <linux/slab.h>
 #include <linux/vmalloc.h>
+#include <linux/delay.h>
 #include <asm/div64.h>
-#include <asm/delay.h>
 
 #include "dvb_frontend.h"
 #include "dvb_functions.h"
