diff -aur linux.backup/include/asm-arm/proc-armo/processor.h linux/include/asm-arm/proc-armo/processor.h
--- linux.backup/include/asm-arm/proc-armo/processor.h	Thu Oct 31 11:42:20 2002
+++ linux/include/asm-arm/proc-armo/processor.h	Sun Aug 17 00:09:38 2003
@@ -43,7 +43,7 @@
 	uaccess_t	*uaccess;		/* User access functions*/
 
 #define EXTRA_THREAD_STRUCT_INIT		\
-	uaccess:	&uaccess_kernel,
+	.uaccess	= &uaccess_kernel,
 
 #define start_thread(regs,pc,sp)					\
 ({									\
diff -aur linux.backup/include/asm-arm/xor.h linux/include/asm-arm/xor.h
--- linux.backup/include/asm-arm/xor.h	Thu Oct 31 11:42:54 2002
+++ linux/include/asm-arm/xor.h	Sat Aug 16 15:44:59 2003
@@ -125,11 +125,11 @@
 }
 
 static struct xor_block_template xor_block_arm4regs = {
-	name:	"arm4regs",
-	do_2:	xor_arm4regs_2,
-	do_3:	xor_arm4regs_3,
-	do_4:	xor_arm4regs_4,
-	do_5:	xor_arm4regs_5,
+	.name	= "arm4regs",
+	.do_2	= xor_arm4regs_2,
+	.do_3	= xor_arm4regs_3,
+	.do_4	= xor_arm4regs_4,
+	.do_5	= xor_arm4regs_5,
 };
 
 #undef XOR_TRY_TEMPLATES
diff -aur linux.backup/include/asm-arm26/processor.h linux/include/asm-arm26/processor.h
--- linux.backup/include/asm-arm26/processor.h	Mon Jul 21 23:35:02 2003
+++ linux/include/asm-arm26/processor.h	Sat Aug 16 15:44:59 2003
@@ -51,7 +51,7 @@
         uaccess_t       *uaccess;         /* User access functions*/
 
 #define EXTRA_THREAD_STRUCT_INIT                \
-        uaccess:        &uaccess_kernel,
+        .uaccess        = &uaccess_kernel,
 
 // FIXME?!!
 
diff -aur linux.backup/include/asm-arm26/xor.h linux/include/asm-arm26/xor.h
--- linux.backup/include/asm-arm26/xor.h	Thu Jun 26 23:47:49 2003
+++ linux/include/asm-arm26/xor.h	Sat Aug 16 15:44:59 2003
@@ -125,11 +125,11 @@
 }
 
 static struct xor_block_template xor_block_arm4regs = {
-	name:	"arm4regs",
-	do_2:	xor_arm4regs_2,
-	do_3:	xor_arm4regs_3,
-	do_4:	xor_arm4regs_4,
-	do_5:	xor_arm4regs_5,
+	.name	= "arm4regs",
+	.do_2	= xor_arm4regs_2,
+	.do_3	= xor_arm4regs_3,
+	.do_4	= xor_arm4regs_4,
+	.do_5	= xor_arm4regs_5,
 };
 
 #undef XOR_TRY_TEMPLATES
