--- linux-2.6.5-rc3.org/net/ipv6/ip6_output.c	2004-03-30 05:25:31.000000000 +0200
+++ linux-2.6.5-rc3/net/ipv6/ip6_output.c	2004-03-30 21:47:58.000000000 +0200
@@ -791,6 +791,7 @@
 	*dst = NULL;
 	return err;
 }
+EXPORT_SYMBOL(ip6_dst_lookup);
 
 int ip6_append_data(struct sock *sk, int getfrag(void *from, char *to, int offset, int len, int odd, struct sk_buff *skb),
 		    void *from, int length, int transhdrlen,
--- linux-2.6.5-rc3.org/net/irda/irda_device.c	2004-03-30 05:26:13.000000000 +0200
+++ linux-2.6.5-rc3/net/irda/irda_device.c	2004-03-30 21:57:04.000000000 +0200
@@ -232,6 +232,7 @@
 
 	__irda_task_delete(task);
 }
+EXPORT_SYMBOL(irda_task_delete);
 
 /*
  * Function irda_task_kick (task)
