--- linux-2.6.2-rc3/security/selinux/ss/services.c	2004-01-01 20:23:54.000000000 +0100
+++ linux-m68k-2.6.2-rc3/security/selinux/ss/services.c	2004-01-09 22:16:22.000000000 +0100
@@ -16,6 +16,7 @@
 #include <linux/spinlock.h>
 #include <linux/errno.h>
 #include <linux/in.h>
+#include <linux/sched.h>
 #include <asm/semaphore.h>
 #include "flask.h"
 #include "avc.h"
--- linux-2.6.2-rc3/security/selinux/ss/sidtab.c	2004-01-01 20:23:54.000000000 +0100
+++ linux-m68k-2.6.2-rc3/security/selinux/ss/sidtab.c	2004-01-09 22:13:55.000000000 +0100
@@ -7,6 +7,7 @@
 #include <linux/slab.h>
 #include <linux/spinlock.h>
 #include <linux/errno.h>
+#include <linux/sched.h>
 #include "flask.h"
 #include "security.h"
 #include "sidtab.h"
