--- /build/orig/linux-2.6.0-test4-mm6/drivers/acpi/pci_link.c	2003-09-05 19:16:31.000000000 -0700
+++ linux-2.6.0-test4-mm6/drivers/acpi/pci_link.c	2003-09-12 18:55:36.000000000 -0700
@@ -445,6 +445,9 @@
 	/*
 	 * Update penalties to facilitate IRQ balancing.
 	 */
+	if (acpi_fadt.sci_int > 0 && acpi_fadt.sci_int < ACPI_MAX_ISA_IRQ)
+		acpi_irq_penalty[acpi_fadt.sci_int] += 100000;
+
 	list_for_each(node, &acpi_link.entries) {
 
 		link = list_entry(node, struct acpi_pci_link, node);
