--- linux/arch/i386/kernel/pci-irq.c.orig	Mon Feb 25 20:37:00 2002
+++ linux/arch/i386/kernel/pci-irq.c	Tue May 14 11:46:08 2002
@@ -462,6 +462,8 @@
 	  pirq_serverworks_get, pirq_serverworks_set },
 	{ "AMD756 VIPER", PCI_VENDOR_ID_AMD, PCI_DEVICE_ID_AMD_VIPER_740B,
 		pirq_amd756_get, pirq_amd756_set },
+	{ "AMD768", PCI_VENDOR_ID_AMD, PCI_DEVICE_ID_AMD_VIPER_7443,
+		pirq_amd756_get, pirq_amd756_set },
 
 	{ "default", 0, 0, NULL, NULL }
 };
