diff -urN linux/bttv-0.7.87.orig/driver/Makefile bttv-0.7.87/driver/Makefile
--- linux/bttv-0.7.87.orig/driver/Makefile	Tue Dec  4 11:08:37 2001
+++ linux/bttv-0.7.87/driver/Makefile	Thu Jan 10 17:10:47 2002
@@ -3,7 +3,7 @@
 
 # where the kernel sources are located
 RUNNING := /lib/modules/$(shell uname -r)/build
-KDIR := $(shell if test -d $(RUNNING); then echo $(RUNNING); else echo /usr/src/linux; fi)
+KDIR := ../../../linux
 
 # kernel version
 CURRENT	:= $(shell ./uts-release $(KDIR))
@@ -16,9 +16,7 @@
   KCOMPAT := kcompat24.o
 endif
 
-MODULE_DIR := $(shell test -d "/lib/modules/$(CURRENT)/kernel" && \
-	echo "/lib/modules/$(CURRENT)/kernel/drivers/media/video" ||\
-	echo "/lib/modules/$(CURRENT)/misc")
+MODULE_DIR := ../../../linux-installed/lib/modules/$(CURRENT)/misc
 
 EXTRA_CFLAGS=-g -Wmissing-prototypes -Wstrict-prototypes #-DVIDEODAT_HACK
 
@@ -58,7 +56,7 @@
 
 install:
 	strip --strip-debug *.o
-	su -c "set -x; mkdir -p $(MODULE_DIR); cp -v $(M_OBJS) $(MX_OBJS) $(MODULE_DIR); depmod -a"
+	set -x; mkdir -p $(MODULE_DIR); cp -v $(M_OBJS) $(MX_OBJS) $(MODULE_DIR)
 
 clean:
 	-rm -f *.o .*.o.flags *~
