--- linux-2.6.8-rc4/drivers/scsi/dpt_i2o.c.orig	2004-04-21 15:19:04.000000000 +0200
+++ linux-2.6.8-rc4/drivers/scsi/dpt_i2o.c	2004-08-12 19:42:57.351823928 +0200
@@ -76,7 +76,7 @@
 #include <asm/io.h>
 
 #include "scsi.h"
-#include "hosts.h"
+#include <scsi/scsi_host.h>
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(2,5,65))
 #include "sd.h"
 #endif
@@ -569,7 +569,7 @@
 printk (KERN_INFO"pDev=%p\n", pDev);
 adpt_sleep();
 #endif
-		(struct adpt_device*)(cmd->device->hostdata) = pDev;
+		cmd->device->hostdata = pDev;
 	}
 #if 0
 printk (KERN_INFO"pDev->pScsi_dev=%p\n", cmd->device);
@@ -2797,7 +2797,7 @@
 		printk ("%s: scsi_register returned NULL\n",pHba->name);
 		return -1;
 	}
-	(adpt_hba*)(host->hostdata[0]) = pHba;
+	host->hostdata[0] = (unsigned long)pHba;
 	pHba->host = host;
 
 	/*
@@ -3132,7 +3132,7 @@
 					if(pDev->pScsi_dev && !scsi_device_online(pDev->pScsi_dev)) {
 						printk(KERN_WARNING"%s: Setting device (%d,%d,%d) back online\n",
 								pHba->name,bus_no,scsi_id,scsi_lun);
-						pDev->pScsi_dev->online = TRUE;
+						scsi_device_set_state(pDev->pScsi_dev, SDEV_RUNNING);
 					}
 					d = pDev->pI2o_dev;
 					if(le32_to_cpu(d->lct_data.tid) != tid) { // something changed
@@ -3162,7 +3162,7 @@
 			pDev->state = DPTI_DEV_OFFLINE;
 			printk(KERN_WARNING"%s: Device (%d,%d,%d) offline\n",pHba->name,pDev->scsi_channel,pDev->scsi_id,pDev->scsi_lun);
 			if (pDev->pScsi_dev) {
-				pDev->pScsi_dev->online = FALSE;
+				scsi_device_set_state(pDev->pScsi_dev, SDEV_OFFLINE);
 #if (LINUX_VERSION_CODE >= KERNEL_VERSION(2,5,0)) && (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,0))
 				if (atomic_read(&pDev->pScsi_dev->access_count)) {
 #elif (LINUX_VERSION_CODE >= KERNEL_VERSION(2,6,0))
--- linux-2.6.8-rc4/drivers/scsi/dpti.h.orig	2004-04-21 15:19:04.000000000 +0200
+++ linux-2.6.8-rc4/drivers/scsi/dpti.h	2004-08-12 19:24:34.326509264 +0200
@@ -297,7 +297,6 @@
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(2,6,0))
 #define iminor(x)	MINOR(x->i_rdev)
 #endif
-#define scsi_device_online(d) ((d)->online)
 
 #if (LINUX_VERSION_CODE < KERNEL_VERSION(2,4,0))
 #define set_current_state(x)	current->state = x
--- linux-2.6.8-rc4/drivers/scsi/Kconfig.orig	2004-08-13 12:47:02.000000000 +0200
+++ linux-2.6.8-rc4/drivers/scsi/Kconfig	2004-08-13 15:45:44.049526200 +0200
@@ -363,7 +363,7 @@
 # All the I2O code and drivers do not seem to be 64bit safe.
 config SCSI_DPT_I2O
 	tristate "Adaptec I2O RAID support "
-	depends on !64BIT && SCSI && PCI
+	depends on SCSI && PCI
 	help
 	  This driver supports all of Adaptec's I2O based RAID controllers as 
 	  well as the DPT SmartRaid V cards.  This is an Adaptec maintained
