diff -ur linux.vanilla/drivers/char/epca.c linux/drivers/char/epca.c
--- linux.vanilla/drivers/char/epca.c	Mon Mar  5 12:01:51 2001
+++ linux/drivers/char/epca.c	Tue Mar  6 13:37:58 2001
@@ -1574,7 +1574,8 @@
 	cli();
 
 	if ((tty_unregister_driver(&pc_driver)) ||  
-	    (tty_unregister_driver(&pc_callout)))
+	    (tty_unregister_driver(&pc_callout)) ||
+	    (tty_unregister_driver(&pc_info)))
 	{
 		printk(KERN_WARNING "<Error> - DIGI : cleanup_module failed to un-register tty driver\n");
 		restore_flags(flags);
