--- linux-5.4/arch/x86/boot/Makefile.orig	2025-11-07 11:25:25.438194510 +0100
+++ linux-5.4/arch/x86/boot/Makefile	2025-11-07 11:26:24.356663596 +0100
@@ -66,6 +66,7 @@
 # ---------------------------------------------------------------------------
 
 KBUILD_CFLAGS	:= $(REALMODE_CFLAGS) -D_SETUP
+KBUILD_CFLAGS	+= -std=gnu11
 KBUILD_AFLAGS	:= $(KBUILD_CFLAGS) -D__ASSEMBLY__
 GCOV_PROFILE := n
 UBSAN_SANITIZE := n
--- linux-5.4/arch/x86/realmode/rm/Makefile.orig	2019-11-25 01:32:01.000000000 +0100
+++ linux-5.4/arch/x86/realmode/rm/Makefile	2025-11-07 11:27:17.869996903 +0100
@@ -69,7 +69,7 @@
 # ---------------------------------------------------------------------------
 
 KBUILD_CFLAGS	:= $(REALMODE_CFLAGS) -D_SETUP -D_WAKEUP \
-		   -I$(srctree)/arch/x86/boot
+		   -I$(srctree)/arch/x86/boot -std=gnu11
 KBUILD_AFLAGS	:= $(KBUILD_CFLAGS) -D__ASSEMBLY__
 GCOV_PROFILE := n
 UBSAN_SANITIZE := n
