--- linux-6.1/scripts/gcc-plugins/gcc-common.h.orig	2022-12-11 23:15:18.000000000 +0100
+++ linux-6.1/scripts/gcc-plugins/gcc-common.h	2023-05-02 22:21:15.747011409 +0200
@@ -71,8 +71,11 @@
 #include "varasm.h"
 #include "stor-layout.h"
 #include "internal-fn.h"
+#include "gimple.h"
+#include "gimple-iterator.h"
 #include "gimple-expr.h"
 #include "gimple-fold.h"
+#include "gimple-ssa.h"
 #include "context.h"
 #include "tree-ssa-alias.h"
 #include "tree-ssa.h"
@@ -85,11 +88,8 @@
 #include "tree-eh.h"
 #include "stmt.h"
 #include "gimplify.h"
-#include "gimple.h"
 #include "tree-phinodes.h"
 #include "tree-cfg.h"
-#include "gimple-iterator.h"
-#include "gimple-ssa.h"
 #include "ssa-iterators.h"
 
 #include "builtins.h"
