--- linux-5.15/arch/x86/realmode/rm/Makefile~	2021-10-31 21:53:10.000000000 +0100
+++ linux-5.15/arch/x86/realmode/rm/Makefile	2025-11-05 21:00:05.919997261 +0100
@@ -72,7 +72,7 @@
 # ---------------------------------------------------------------------------
 
 KBUILD_CFLAGS	:= $(REALMODE_CFLAGS) -D_SETUP -D_WAKEUP \
-		   -I$(srctree)/arch/x86/boot
+		   -I$(srctree)/arch/x86/boot -std=gnu11
 KBUILD_AFLAGS	:= $(KBUILD_CFLAGS) -D__ASSEMBLY__
 KBUILD_CFLAGS	+= -fno-asynchronous-unwind-tables
 GCOV_PROFILE := n
--- linux-5.15/arch/x86/boot/Makefile~	2025-11-05 21:02:44.000000000 +0100
+++ linux-5.15/arch/x86/boot/Makefile	2025-11-06 11:11:35.376664029 +0100
@@ -70,6 +70,7 @@
 KBUILD_AFLAGS	:= $(KBUILD_CFLAGS) -D__ASSEMBLY__
 KBUILD_CFLAGS	+= $(call cc-option,-fmacro-prefix-map=$(srctree)/=)
 KBUILD_CFLAGS	+= -fno-asynchronous-unwind-tables
+KBUILD_CFLAGS	+= -std=gnu11
 GCOV_PROFILE := n
 UBSAN_SANITIZE := n
 
