--- linux-2.4.2/drivers/i2o/i2o_core.c.orig	Thu Feb 22 18:09:49 2001
+++ linux/drivers/i2o/i2o_core.c	Tue Mar 13 07:41:04 2001
@@ -3113,8 +3113,6 @@
 }
 
 
-#ifdef MODULE
-
 EXPORT_SYMBOL(i2o_controller_chain);
 EXPORT_SYMBOL(i2o_num_controllers);
 EXPORT_SYMBOL(i2o_find_controller);
@@ -3146,6 +3144,8 @@
 EXPORT_SYMBOL(i2o_dump_message);
 
 EXPORT_SYMBOL(i2o_get_class_name);
+
+#ifdef MODULE
 
 MODULE_AUTHOR("Red Hat Software");
 MODULE_DESCRIPTION("I2O Core");
