diff -urN linux.orig/drivers/block/cpqarray.c linux/drivers/block/cpqarray.c
--- linux.orig/drivers/block/cpqarray.c	Fri Nov  2 17:39:06 2001
+++ linux/drivers/block/cpqarray.c	Mon Mar 18 18:46:10 2002
@@ -55,9 +55,9 @@
 #include <linux/genhd.h>
 
 #include "cpqarray.h"
-#include "ida_cmd.h"
+#include "../ide/ida_cmd.h"
 #include "smart1,2.h"
-#include "ida_ioctl.h"
+#include "../ide/ida_ioctl.h"
 
 #define READ_AHEAD	128
 #define NR_CMDS		128 /* This could probably go as high as ~400 */
diff -urN linux.orig/drivers/block/cpqarray.h linux/drivers/block/cpqarray.h
--- linux.orig/drivers/block/cpqarray.h	Sun Mar 25 18:31:24 2001
+++ linux/drivers/block/cpqarray.h	Mon Mar 18 18:45:42 2002
@@ -33,7 +33,7 @@
 #include <linux/timer.h>
 #endif
 
-#include "ida_cmd.h"
+#include "../ide/ida_cmd.h"
 
 #define IO_OK		0
 #define IO_ERROR	1
diff -urN linux.orig/drivers/ide/ida_ioctl.h linux/drivers/ide/ida_ioctl.h
--- linux.orig/drivers/ide/ida_ioctl.h	Mon Mar 18 18:39:49 2002
+++ linux/drivers/ide/ida_ioctl.h	Mon Mar 18 18:46:25 2002
@@ -26,7 +26,7 @@
 #define IDA_IOCTL_H
 
 #include "ida_cmd.h"
-#include "cpqarray.h"
+#include "../block/cpqarray.h"
 
 #define IDAGETDRVINFO		0x27272828
 #define IDAPASSTHRU		0x28282929
