diff -urN linux.orig/arch/ppc/kernel/setup.c linux/arch/ppc/kernel/setup.c
--- linux.orig/arch/ppc/kernel/setup.c	Sat Mar 23 14:25:54 2002
+++ linux/arch/ppc/kernel/setup.c	Sat Mar 23 14:26:34 2002
@@ -685,8 +685,6 @@
 	id->CurAPMvalues   = __le16_to_cpu(id->CurAPMvalues);
 	id->word92         = __le16_to_cpu(id->word92);
 	id->hw_config      = __le16_to_cpu(id->hw_config);
-	for (i = 0; i < 32; i++)
-		id->words94_125[i]  = __le16_to_cpu(id->words94_125[i]);
 	id->last_lun       = __le16_to_cpu(id->last_lun);
 	id->word127        = __le16_to_cpu(id->word127);
 	id->dlf            = __le16_to_cpu(id->dlf);
@@ -696,6 +694,4 @@
 	id->word156        = __le16_to_cpu(id->word156);
 	for (i = 0; i < 4; i++)
 		id->words157_159[i] = __le16_to_cpu(id->words157_159[i]);
-	for (i = 0; i < 96; i++)
-		id->words160_255[i] = __le16_to_cpu(id->words160_255[i]);
 }
diff -urN linux.orig/drivers/block/cpqarray.c linux/drivers/block/cpqarray.c
--- linux.orig/drivers/block/cpqarray.c	Fri Nov  2 17:39:06 2001
+++ linux/drivers/block/cpqarray.c	Mon Mar 18 18:46:10 2002
@@ -55,9 +55,9 @@
 #include <linux/genhd.h>
 
 #include "cpqarray.h"
-#include "ida_cmd.h"
+#include "../ide/ida_cmd.h"
 #include "smart1,2.h"
-#include "ida_ioctl.h"
+#include "../ide/ida_ioctl.h"
 
 #define READ_AHEAD	128
 #define NR_CMDS		128 /* This could probably go as high as ~400 */
diff -urN linux.orig/drivers/block/cpqarray.h linux/drivers/block/cpqarray.h
--- linux.orig/drivers/block/cpqarray.h	Sun Mar 25 18:31:24 2001
+++ linux/drivers/block/cpqarray.h	Mon Mar 18 18:45:42 2002
@@ -33,7 +33,7 @@
 #include <linux/timer.h>
 #endif
 
-#include "ida_cmd.h"
+#include "../ide/ida_cmd.h"
 
 #define IO_OK		0
 #define IO_ERROR	1
diff -urN linux.orig/drivers/ide/ida_ioctl.h linux/drivers/ide/ida_ioctl.h
--- linux.orig/drivers/ide/ida_ioctl.h	Mon Mar 18 18:39:49 2002
+++ linux/drivers/ide/ida_ioctl.h	Mon Mar 18 18:46:25 2002
@@ -26,7 +26,7 @@
 #define IDA_IOCTL_H
 
 #include "ida_cmd.h"
-#include "cpqarray.h"
+#include "../block/cpqarray.h"
 
 #define IDAGETDRVINFO		0x27272828
 #define IDAPASSTHRU		0x28282929
