diff -urN linux.orig/drivers/ide/ide-cd.c linux/drivers/ide/ide-cd.c
--- linux.orig/drivers/ide/ide-cd.c	Mon Mar 18 18:39:49 2002
+++ linux/drivers/ide/ide-cd.c	Mon Mar 18 18:50:03 2002
@@ -3093,5 +3093,4 @@
 #ifdef MODULE
 module_init(ide_cdrom_init);
 module_exit(ide_cdrom_exit);
-MODULE_LICENSE("GPL");
 #endif
diff -urN linux.orig/drivers/ide/ide.c linux/drivers/ide/ide.c
--- linux.orig/drivers/ide/ide.c	Mon Mar 18 18:40:32 2002
+++ linux/drivers/ide/ide.c	Mon Mar 18 18:50:13 2002
@@ -4099,7 +4099,6 @@
 #ifdef MODULE
 char *options = NULL;
 MODULE_PARM(options,"s");
-MODULE_LICENSE("GPL");
 
 static void __init parse_options (char *line)
 {
