diff -urN linux.orig/drivers/block/cpqarray.c linux/drivers/block/cpqarray.c
--- linux.orig/drivers/block/cpqarray.c	Sun Mar 17 20:51:40 2002
+++ linux/drivers/block/cpqarray.c	Sun Mar 17 20:52:08 2002
@@ -55,9 +55,9 @@
 #include <linux/genhd.h>
 
 #include "cpqarray.h"
-#include "ida_cmd.h"
+#include "../ide/ida_cmd.h"
 #include "smart1,2.h"
-#include "ida_ioctl.h"
+#include "../ide/ida_ioctl.h"
 
 #define READ_AHEAD	128
 #define NR_CMDS		128 /* This could probably go as high as ~400 */
diff -urN linux.orig/drivers/block/cpqarray.h linux/drivers/block/cpqarray.h
--- linux.orig/drivers/block/cpqarray.h	Sun Mar 17 20:51:35 2002
+++ linux/drivers/block/cpqarray.h	Sun Mar 17 20:51:51 2002
@@ -33,7 +33,7 @@
 #include <linux/timer.h>
 #endif
 
-#include "../ida_cmd.h"
+#include "../ide/ida_cmd.h"
 
 #define IO_OK		0
 #define IO_ERROR	1
diff -urN linux.orig/drivers/ide/ida_ioctl.h linux/drivers/ide/ida_ioctl.h
--- linux.orig/drivers/ide/ida_ioctl.h	Sun Mar 17 23:42:34 2002
+++ linux/drivers/ide/ida_ioctl.h	Sun Mar 17 20:53:03 2002
@@ -33,7 +33,7 @@
 #include <linux/timer.h>
 #endif
 
-#include "ida_cmd.h"
+#include "../ide/ida_cmd.h"
 
 #define IO_OK		0
 #define IO_ERROR	1
