diff -urN linux.orig/drivers/ide/ide-probe.c linux/drivers/ide/ide-probe.c
--- linux.orig/drivers/ide/ide-probe.c	Sat Apr 20 18:05:49 2002
+++ linux/drivers/ide/ide-probe.c	Sat Apr 20 21:51:21 2002
@@ -491,11 +491,13 @@
 
 	if (hwif->noprobe)
 		return;
+#ifdef CONFIG_BLK_DEV_IDE
 	if (hwif->io_ports[IDE_DATA_OFFSET] == HD_DATA) {
 		extern void probe_cmos_for_drives(ide_hwif_t *);
 
 		probe_cmos_for_drives (hwif);
 	}
+#endif
 
 	if ((hwif->chipset != ide_4drives || !hwif->mate->present) &&
 #if CONFIG_BLK_DEV_PDC4030
