diff -urN linux.orig/drivers/ide/ide-pmac.c linux/drivers/ide/ide-pmac.c
--- linux.orig/drivers/ide/ide-pmac.c	Fri May  3 19:43:16 2002
+++ linux/drivers/ide/ide-pmac.c	Sat May  4 11:05:36 2002
@@ -918,7 +918,7 @@
 	 * because of a pending unit attention. The problem if that if I
 	 * clear the error, the filesystem dies.
 	 */
-	if (used_dma && !ide_spin_wait_hwgroup(drive, &flags)) {
+	if (used_dma && !ide_spin_wait_hwgroup(drive)) {
 		/* Lock HW group */
 		HWGROUP(drive)->busy = 1;
 		pmac_ide_dma_onoff(drive, 1);
@@ -1027,7 +1027,7 @@
 			
 			if (drive->present) {
 				/* Wait for HW group to complete operations */
-				if (ide_spin_wait_hwgroup(drive, &flags)) {
+				if (ide_spin_wait_hwgroup(drive)) {
 					// What can we do here ? Wake drive we had already
 					// put to sleep and return an error ?
 				} else {
