--- linux/net/netsyms.c.orig	Mon Apr 22 16:54:58 2002
+++ linux/net/netsyms.c	Mon Apr 22 16:55:27 2002
@@ -596,7 +596,7 @@
 
 #if defined(CONFIG_IP_NF_MATCH_STEALTH) || defined(CONFIG_IP_NF_MATCH_STEALTH_MODULE)
 extern struct sock *udp_v4_lookup(u32 saddr, u16 sport, u32 daddr, u16 dport, int dif);
-#if !defined (CONFIG_IPV6_MODULE) && !defined (CONFIG_KHTTPD) && !defined (CONFIG_KHTTPD_MODULE)
+#if !defined (CONFIG_IPV6_MODULE) && !defined (CONFIG_KHTTPD) && !defined (CONFIG_KHTTPD_MODULE) && !defined(CONFIG_IP_NF_PERS_MODULE)
 EXPORT_SYMBOL(tcp_v4_lookup_listener);
 #endif
 #endif
