--- jfs-2.4.17-1.0.15-patch.org	Tue Jan  8 16:49:34 2002
+++ jfs-2.4.17-1.0.15-patch	Fri Jan 11 11:01:20 2002
@@ -2,9 +2,9 @@
 --- linux-2.4.7/Documentation/Changes	Thu Dec  6 14:13:50 2001
 +++ jfs-2.4.7-1.0.12/Documentation/Changes	Tue Jan  8 10:49:34 2002
 @@ -54,6 +54,7 @@
- o  util-linux             2.10o                   # fdformat --version
  o  modutils               2.4.2                   # insmod -V
  o  e2fsprogs              1.25                    # tune2fs
+ o  xfsprogs               2.0.0                   # xfs_db -V
 +o  jfsutils               1.0.14                  # fsck.jfs -V
  o  reiserfsprogs          3.x.0j                  # reiserfsck 2>&1|grep reiserfsprogs
  o  pcmcia-cs              3.1.21                  # cardmgr -V
