--- linux-2.6.20/grsecurity/gracl_shm.c~	2007-03-24 04:54:27.000000000 +0100
+++ linux-2.6.20/grsecurity/gracl_shm.c	2007-03-24 04:55:46.332159000 +0100
@@ -6,6 +6,7 @@
 #include <linux/gracl.h>
 #include <linux/grsecurity.h>
 #include <linux/grinternal.h>
+#include <linux/vs_pid.h>
 
 int
 gr_handle_shmat(const pid_t shm_cprid, const pid_t shm_lapid,
--- linux-2.6.20/localversion-grsec~	2007-03-24 04:54:27.000000000 +0100
+++ linux-2.6.20/localversion-grsec	2007-03-24 05:05:10.455414500 +0100
@@ -1 +0,0 @@
--grsec
--- linux-2.6.20/grsecurity/grsec_sock.c~	2007-03-24 05:38:40.000000000 +0100
+++ linux-2.6.20/grsecurity/grsec_sock.c	2007-03-24 05:47:11.347998750 +0100
@@ -7,6 +7,7 @@
 #include <linux/ip.h>
 #include <net/sock.h>
 #include <net/inet_sock.h>
+#include <linux/vs_context.h>
 #include <linux/grsecurity.h>
 #include <linux/grinternal.h>
 #include <linux/gracl.h>
--- linux-2.6.22/arch/i386/mm/boot_ioremap.c~	2007-08-09 23:10:22.000000000 +0200
+++ linux-2.6.22/arch/i386/mm/boot_ioremap.c	2007-08-09 23:26:35.305141000 +0200
@@ -13,6 +13,7 @@
 #include <asm/tlbflush.h>
 #include <linux/init.h>
 #include <linux/stddef.h>
+#include <linux/sched.h>
 
 /*
  * This is only for a caller who is clever enough to page-align
