--- linux-2.6.22.19/net/ipv4/inet_diag.c~	2008-02-28 15:12:43.635402618 +0100
+++ linux-2.6.22.19/net/ipv4/inet_diag.c	2008-03-04 13:30:30.423644777 +0100
@@ -98,8 +98,8 @@
 
 	r->id.idiag_sport = inet->sport;
 	r->id.idiag_dport = inet->dport;
-	r->id.idiag_src[0] = nx_map_sock_lback(skb->sk->sk_nx_info, inet->rcv_saddr);
-	r->id.idiag_dst[0] = nx_map_sock_lback(skb->sk->sk_nx_info, inet->daddr);
+	r->id.idiag_src[0] = nx_map_sock_lback(sk->sk_nx_info, inet->rcv_saddr);
+	r->id.idiag_dst[0] = nx_map_sock_lback(sk->sk_nx_info, inet->daddr);
 
 #if defined(CONFIG_IPV6) || defined (CONFIG_IPV6_MODULE)
 	if (r->idiag_family == AF_INET6) {
@@ -563,8 +563,8 @@
 
 	r->id.idiag_sport = inet->sport;
 	r->id.idiag_dport = ireq->rmt_port;
-	r->id.idiag_src[0] = nx_map_sock_lback(skb->sk->sk_nx_info, ireq->loc_addr);
-	r->id.idiag_dst[0] = nx_map_sock_lback(skb->sk->sk_nx_info, ireq->rmt_addr);
+	r->id.idiag_src[0] = nx_map_sock_lback(sk->sk_nx_info, ireq->loc_addr);
+	r->id.idiag_dst[0] = nx_map_sock_lback(sk->sk_nx_info, ireq->rmt_addr);
 	r->idiag_expires = jiffies_to_msecs(tmo);
 	r->idiag_rqueue = 0;
 	r->idiag_wqueue = 0;
