--- linux-2.6.23/scripts/Makefile.modpost~	2007-12-12 13:30:45.000000000 +0200
+++ linux-2.6.23/scripts/Makefile.modpost	2007-12-12 13:31:55.551062566 +0200
@@ -62,7 +62,7 @@
  $(if $(KBUILD_EXTMOD),-i,-o) $(kernelsymfile)   \
  $(if $(KBUILD_EXTMOD),-I $(modulesymfile))      \
  $(if $(KBUILD_EXTMOD),-o $(modulesymfile))      \
- $(if $(KBUILD_EXTMOD)$(KBUILD_MODPOST_WARN),-w) \
+ $(if $(KBUILD_EXTMOD),$(if $(KBUILD_MODPOST_WARN),-w)) \
  $(if $(cross_build),-c)
 
 quiet_cmd_modpost = MODPOST $(words $(filter-out vmlinux FORCE, $^)) modules
       cmd_modpost = $(modpost) -s
