diff -urN linux-2.5.7.orig/init/do_mounts.c linux-2.5.7/init/do_mounts.c
--- linux-2.5.7.orig/init/do_mounts.c	Mon Mar 18 21:37:12 2002
+++ linux-2.5.7/init/do_mounts.c	Sun Mar 24 12:27:21 2002
@@ -826,7 +826,7 @@
 
 	create_dev("/dev/root", ROOT_DEV, NULL);
 	if (mount_initrd) {
-		if (initrd_load() && kdev_same(ROOT_DEV, mk_kdev(RAMDISK_MAJOR, 0))) {
+		if (initrd_load() && !kdev_same(ROOT_DEV, mk_kdev(RAMDISK_MAJOR, 0))) {
 			handle_initrd();
 			goto out;
 		}
