--- linux-6.3/fs/aufs/Makefile.orig	2023-05-01 20:26:46.511529296 +0200
+++ linux-6.3/fs/aufs/Makefile	2023-05-01 20:28:26.829201887 +0200
@@ -1,10 +1,10 @@
 # SPDX-License-Identifier: GPL-2.0
 
-include ${src}/magic.mk
+include ${srctree}/fs/aufs/magic.mk
 ifeq (${CONFIG_AUFS_FS},m)
-include ${src}/conf.mk
+include ${srctree}/fs/aufs/conf.mk
 endif
--include ${src}/priv_def.mk
+-include ${srctree}/fs/aufs/priv_def.mk
 
 # cf. include/linux/kernel.h
 # enable pr_debug
