--- linux-2.6.22/fs/unionfs/commonfops.c~	2008-05-18 19:35:41.000000000 +0300
+++ linux-2.6.22/fs/unionfs/commonfops.c	2008-05-18 21:53:35.777262423 +0300
@@ -88,7 +88,7 @@
 					    lower_dentry->d_inode);
 	}
 	lower_dir_dentry = lock_parent(lower_dentry);
-	err = vfs_unlink(lower_dir_dentry->d_inode, lower_dentry);
+	err = vfs_unlink(lower_dir_dentry->d_inode, lower_dentry, NULL);
 	unlock_dir(lower_dir_dentry);
 
 out:
--- linux-2.6.22/fs/unionfs/copyup.c~	2008-05-18 19:35:41.000000000 +0300
+++ linux-2.6.22/fs/unionfs/copyup.c	2008-05-18 21:37:37.269847090 +0300
@@ -482,7 +482,7 @@
 	 * quota, or something else happened so let's unlink; we don't
 	 * really care about the return value of vfs_unlink
 	 */
-	vfs_unlink(new_lower_parent_dentry->d_inode, new_lower_dentry);
+	vfs_unlink(new_lower_parent_dentry->d_inode, new_lower_dentry, NULL);
 
 	if (copyup_file) {
 		/* need to close the file */
--- linux-2.6.22/fs/unionfs/dirhelper.c~	2008-05-18 19:35:41.000000000 +0300
+++ linux-2.6.22/fs/unionfs/dirhelper.c	2008-05-18 21:37:12.040388291 +0300
@@ -70,7 +70,7 @@
 				break;
 			}
 			if (lower_dentry->d_inode)
-				err = vfs_unlink(lower_dir, lower_dentry);
+				err = vfs_unlink(lower_dir, lower_dentry, NULL);
 			dput(lower_dentry);
 			if (err)
 				break;
--- linux-2.6.22/fs/unionfs/inode.c~	2008-05-18 19:35:41.000000000 +0300
+++ linux-2.6.22/fs/unionfs/inode.c	2008-05-18 21:27:05.637322201 +0300
@@ -62,7 +62,7 @@
 	lower_dir_dentry = lock_parent_wh(wh_dentry);
 	/* see Documentation/filesystems/unionfs/issues.txt */
 	lockdep_off();
-	err = vfs_unlink(lower_dir_dentry->d_inode, wh_dentry);
+	err = vfs_unlink(lower_dir_dentry->d_inode, wh_dentry, NULL);
 	lockdep_on();
 	unlock_dir(lower_dir_dentry);
 
@@ -341,7 +341,7 @@
 			/* see Documentation/filesystems/unionfs/issues.txt */
 			lockdep_off();
 			err = vfs_unlink(lower_dir_dentry->d_inode,
-					 whiteout_dentry);
+					 whiteout_dentry, NULL);
 			lockdep_on();
 		}
 
@@ -374,7 +374,7 @@
 		/* see Documentation/filesystems/unionfs/issues.txt */
 		lockdep_off();
 		err = vfs_link(lower_old_dentry, lower_dir_dentry->d_inode,
-			       lower_new_dentry);
+			       lower_new_dentry, NULL);
 		lockdep_on();
 	}
 	unlock_dir(lower_dir_dentry);
@@ -403,7 +403,7 @@
 			/* do vfs_link */
 			err = vfs_link(lower_old_dentry,
 				       lower_dir_dentry->d_inode,
-				       lower_new_dentry);
+				       lower_new_dentry, NULL);
 			lockdep_on();
 			unlock_dir(lower_dir_dentry);
 			goto check_link;
@@ -491,7 +491,7 @@
 
 	mode = S_IALLUGO;
 	err = vfs_symlink(lower_parent_dentry->d_inode, lower_dentry,
-			  symname, mode);
+			  symname, mode, NULL);
 	if (!err) {
 		err = PTR_ERR(unionfs_interpose(dentry, parent->i_sb, 0));
 		if (!err) {
@@ -622,7 +622,7 @@
 		}
 
 		err = vfs_mkdir(lower_parent_dentry->d_inode, lower_dentry,
-				mode);
+				mode, NULL);
 
 		unlock_dir(lower_parent_dentry);
 
@@ -725,7 +725,7 @@
 		goto out;
 	}
 
-	err = vfs_mknod(lower_parent_dentry->d_inode, lower_dentry, mode, dev);
+	err = vfs_mknod(lower_parent_dentry->d_inode, lower_dentry, mode, dev, NULL);
 	if (!err) {
 		err = PTR_ERR(unionfs_interpose(dentry, parent->i_sb, 0));
 		if (!err) {
--- linux-2.6.22/fs/unionfs/rename.c~	2008-05-18 19:35:41.000000000 +0300
+++ linux-2.6.22/fs/unionfs/rename.c	2008-05-18 21:52:05.956070003 +0300
@@ -81,7 +81,7 @@
 		err = is_robranch_super(old_dentry->d_sb, bindex);
 		if (!err)
 			err = vfs_unlink(lower_wh_dir_dentry->d_inode,
-					 lower_wh_dentry);
+					 lower_wh_dentry, NULL);
 
 		dput(lower_wh_dentry);
 		unlock_dir(lower_wh_dir_dentry);
@@ -222,7 +222,7 @@
 		err = is_robranch_super(old_dir->i_sb, bindex);
 		if (!err)
 			err = vfs_unlink(unlink_dir_dentry->d_inode,
-					 unlink_dentry);
+					 unlink_dentry, NULL);
 
 		fsstack_copy_attr_times(new_dentry->d_parent->d_inode,
 					unlink_dir_dentry->d_inode);
--- linux-2.6.22/fs/unionfs/sioq.c~	2008-05-18 19:35:41.000000000 +0300
+++ linux-2.6.22/fs/unionfs/sioq.c	2008-05-18 21:54:31.272470465 +0300
@@ -69,7 +69,7 @@
 	struct sioq_args *args = container_of(work, struct sioq_args, work);
 	struct mkdir_args *m = &args->mkdir;
 
-	args->err = vfs_mkdir(m->parent, m->dentry, m->mode);
+	args->err = vfs_mkdir(m->parent, m->dentry, m->mode, NULL);
 	complete(&args->comp);
 }
 
@@ -78,7 +78,7 @@
 	struct sioq_args *args = container_of(work, struct sioq_args, work);
 	struct mknod_args *m = &args->mknod;
 
-	args->err = vfs_mknod(m->parent, m->dentry, m->mode, m->dev);
+	args->err = vfs_mknod(m->parent, m->dentry, m->mode, m->dev, NULL);
 	complete(&args->comp);
 }
 
@@ -87,7 +87,7 @@
 	struct sioq_args *args = container_of(work, struct sioq_args, work);
 	struct symlink_args *s = &args->symlink;
 
-	args->err = vfs_symlink(s->parent, s->dentry, s->symbuf, s->mode);
+	args->err = vfs_symlink(s->parent, s->dentry, s->symbuf, s->mode, NULL);
 	complete(&args->comp);
 }
 
@@ -96,7 +96,7 @@
 	struct sioq_args *args = container_of(work, struct sioq_args, work);
 	struct unlink_args *u = &args->unlink;
 
-	args->err = vfs_unlink(u->parent, u->dentry);
+	args->err = vfs_unlink(u->parent, u->dentry, NULL);
 	complete(&args->comp);
 }
 
--- linux-2.6.22/fs/unionfs/unlink.c~	2008-05-18 19:35:41.000000000 +0300
+++ linux-2.6.22/fs/unionfs/unlink.c	2008-05-18 21:53:00.271209886 +0300
@@ -71,10 +71,10 @@
 			lockdep_off();
 			if (!S_ISDIR(lower_dentry->d_inode->i_mode))
 				err = vfs_unlink(lower_dir_dentry->d_inode,
-								lower_dentry);
+								lower_dentry, NULL);
 			else
 				err = vfs_rmdir(lower_dir_dentry->d_inode,
-								lower_dentry);
+								lower_dentry, NULL);
 			lockdep_on();
 		}
 
@@ -193,7 +193,7 @@
 	if (!err) {
 		/* see Documentation/filesystems/unionfs/issues.txt */
 		lockdep_off();
-		err = vfs_rmdir(lower_dir_dentry->d_inode, lower_dentry);
+		err = vfs_rmdir(lower_dir_dentry->d_inode, lower_dentry, NULL);
 		lockdep_on();
 	}
 	dput(lower_dentry);
