--- linux-2.6.24/arch/powerpc/kernel/ppc_ksyms.c	2008-01-24 23:58:37.000000000 +0100
+++ linux-2.6.24/arch/powerpc/kernel/ppc_ksyms.c	2008-02-11 20:14:01.480974476 +0100
@@ -59,6 +59,7 @@ extern void single_step_exception(struct
 extern int sys_sigreturn(struct pt_regs *regs);
 
 EXPORT_SYMBOL(clear_pages);
+EXPORT_SYMBOL(copy_page);
 EXPORT_SYMBOL(ISA_DMA_THRESHOLD);
 EXPORT_SYMBOL(DMA_MODE_READ);
 EXPORT_SYMBOL(DMA_MODE_WRITE);
