--- include/linux/mc146818rtc.h~	Sun Nov 17 18:32:48 2002
+++ include/linux/mc146818rtc.h	Sun Nov 17 18:36:21 2002
@@ -28,7 +28,7 @@
 })
 
 #ifdef __KERNEL__
-extern spinlock_t rtc_lock;	/* You must hold this lock to use CMOS_* ops */
+extern struct spinlock_t rtc_lock;	/* You must hold this lock to use CMOS_* ops */
 #endif
 
 /**********************************************************************
