diff -urNp -x '*.orig' linux-4.9/fs/aufs/cpup.c linux-4.9/fs/aufs/cpup.c
--- linux-4.9/fs/aufs/cpup.c	2021-02-24 16:25:34.368384227 +0100
+++ linux-4.9/fs/aufs/cpup.c	2021-02-24 16:25:46.172095883 +0100
@@ -24,9 +24,9 @@
 #include <linux/task_work.h>
 #include "aufs.h"
 
-void au_cpup_attr_flags(struct inode *dst, unsigned int iflags)
+void au_cpup_attr_flags(struct inode *dst, unsigned short iflags)
 {
-	const unsigned int mask = S_DEAD | S_SWAPFILE | S_PRIVATE
+	const unsigned short mask = S_DEAD | S_SWAPFILE | S_PRIVATE
 		| S_NOATIME | S_NOCMTIME | S_AUTOMOUNT;
 
 	BUILD_BUG_ON(sizeof(iflags) != sizeof(dst->i_flags));
@@ -159,7 +159,7 @@ void au_dtime_revert(struct au_dtime *dt
 struct au_cpup_reg_attr {
 	int		valid;
 	struct kstat	st;
-	unsigned int	iflags; /* inode->i_flags */
+	unsigned short	iflags; /* inode->i_flags */
 };
 
 static noinline_for_stack
diff -urNp -x '*.orig' linux-4.9/fs/aufs/cpup.h linux-4.9/fs/aufs/cpup.h
--- linux-4.9/fs/aufs/cpup.h	2021-02-24 16:25:34.368384227 +0100
+++ linux-4.9/fs/aufs/cpup.h	2021-02-24 16:25:46.172095883 +0100
@@ -30,7 +30,7 @@ struct inode;
 struct file;
 struct au_pin;
 
-void au_cpup_attr_flags(struct inode *dst, unsigned int iflags);
+void au_cpup_attr_flags(struct inode *dst, unsigned short iflags);
 void au_cpup_attr_timesizes(struct inode *inode);
 void au_cpup_attr_nlink(struct inode *inode, int force);
 void au_cpup_attr_changeable(struct inode *inode);
