--- a/kernel/printk.c	2007-10-02 14:01:04.780023500 +0200
+++ b/kernel/printk.c	2007-10-02 14:09:40.140231500 +0200
@@ -30,10 +30,10 @@
 #include <linux/delay.h>
 #include <linux/smp.h>
 #include <linux/security.h>
+#include <linux/suspend.h>
 #include <linux/bootmem.h>
 #include <linux/memblock.h>
 #include <linux/syscalls.h>
-#include <linux/suspend.h>
 #include <linux/kexec.h>
 #include <linux/kdb.h>
 #include <linux/ratelimit.h>
--- a/include/linux/fs.h	2008-05-07 11:39:56.000000000 +0000
+++ b/include/linux/fs.h	2008-05-07 11:43:49.907309587 +0000
@@ -207,12 +207,15 @@
 #define MS_KERNMOUNT	(1<<22) /* this is a kern_mount call */
 #define MS_I_VERSION	(1<<23) /* Update inode I_version field */
 #define MS_STRICTATIME	(1<<24) /* Always perform atime updates */
-#define MS_FROZEN	(1<<25)	/* Frozen by freeze_filesystems() */
 #define MS_NOSEC	(1<<28)
 #define MS_BORN		(1<<29)
 #define MS_ACTIVE	(1<<30)
 #define MS_NOUSER	(1<<31)
 
+
+
+#define MS_FROZEN	(1<<27) /* Frozen by freeze_filesystems() */
+
 /*
  * Superblock flags that can be altered by MS_REMOUNT
  */
@@ -230,6 +230,8 @@
 
 /* Inode flags - they have nothing to superblock flags now */
 
+#define S_ATOMIC_COPY	8192	/* Pages mapped with this inode need to be
+				  atomically copied (gem) */
 #define S_SYNC		1	/* Writes are synced at once */
 #define S_NOATIME	2	/* Do not update access times */
 #define S_APPEND	4	/* Append-only file */
@@ -240,8 +242,6 @@
 #define S_IMA		1024	/* Inode has an associated IMA struct */
 #define S_AUTOMOUNT	2048	/* Automount/referral quasi-directory */
 #define S_NOSEC		4096	/* no suid or xattr security attributes */
-#define S_ATOMIC_COPY	8192	/* Pages mapped with this inode need to be
-				   atomically copied (gem) */
 
 /*
  * Note that nosuid etc flags are inode-specific: setting some file-system

