--- linux-2.6.24/arch/powerpc/boot/Makefile~	2008-01-24 22:58:37.000000000 +0000
+++ linux-2.6.24/arch/powerpc/boot/Makefile	2008-02-08 13:41:15.000000000 +0000
@@ -35,9 +35,9 @@
 
 BOOTCFLAGS	+= -I$(obj) -I$(srctree)/$(obj)
 
-$(obj)/4xx.o: BOOTCFLAGS += -mcpu=440
-$(obj)/ebony.o: BOOTCFLAGS += -mcpu=440
-$(obj)/treeboot-walnut.o: BOOTCFLAGS += -mcpu=405
+$(obj)/4xx.o: BOOTCFLAGS += -mcpu=common -Wa,-m440
+$(obj)/ebony.o: BOOTCFLAGS += -mcpu=common -Wa,-m440
+$(obj)/treeboot-walnut.o: BOOTCFLAGS += -mcpu=common -Wa,-m405
 
 zlib       := inffast.c inflate.c inftrees.c
 zlibheader := inffast.h inffixed.h inflate.h inftrees.h infutil.h
