--- linux/drivers/sbus/char/zs.h.old	Thu May 31 15:39:41 2001
+++ linux/drivers/sbus/char/zs.h	Thu May 31 15:44:31 2001
@@ -25,23 +25,6 @@
 
 #define NUM_ZSREGS    16
 
-struct serial_struct {
-	int	type;
-	int	line;
-	int	port;
-	int	irq;
-	int	flags;
-	int	xmit_fifo_size;
-	int	custom_divisor;
-	int	baud_base;
-	unsigned short	close_delay;
-	char	reserved_char[2];
-	int	hub6;
-	unsigned short	closing_wait; /* time to wait before closing */
-	unsigned short	closing_wait2; /* no longer used... */
-	int	reserved[4];
-};
-
 /*
  * For the close wait times, 0 means wait forever for serial port to
  * flush its output.  65535 means don't wait at all.
@@ -157,11 +140,6 @@
 
 
 #define SERIAL_MAGIC 0x5301
-
-/*
- * The size of the serial xmit buffer is 1 page, or 4096 bytes
- */
-#define SERIAL_XMIT_SIZE 4096
 
 /*
  * Events are used to schedule things to happen at timer-interrupt
