diff -ur linux-2.2.18-/net/netsyms.c linux-2.2.18/net/netsyms.c
--- linux-2.2.18-/net/netsyms.c	Thu Jun 21 13:32:57 2001
+++ linux-2.2.18/net/netsyms.c	Thu Jun 21 13:33:31 2001
@@ -101,8 +101,9 @@
 #include <net/sock.h>
 #endif
 
-#ifdef CONFIG_SYSCTL
 extern int sysctl_max_syn_backlog;
+
+#ifdef CONFIG_SYSCTL
 extern int sysctl_ip_always_defrag;
 #endif
 
@@ -365,8 +366,9 @@
 EXPORT_SYMBOL(dev_loopback_xmit);
 EXPORT_SYMBOL(tcp_regs);
 
-#ifdef CONFIG_SYSCTL
 EXPORT_SYMBOL(sysctl_max_syn_backlog);
+
+#ifdef CONFIG_SYSCTL
 EXPORT_SYMBOL(sysctl_ip_always_defrag);
 #endif
 #endif
