--- kernel/drivers/char/agp/agpgart_be-i8x0.c~	Sun Jul 14 15:43:16 2002
+++ kernel/drivers/char/agp/agpgart_be-i8x0.c	Sun Jul 14 15:43:16 2002
@@ -65,7 +65,7 @@
 	/* Intel 815 chipsets have a _weird_ APSIZE register with only
 	 * one non-reserved bit, so mask the others out ... */
 	if (agp_bridge.type == INTEL_I815)
-		temp &= (1 << 3)
+		temp &= (1 << 3);
 
 	values = A_SIZE_8(agp_bridge.aperture_sizes);
 
