diff -uNr linux.orig/drivers/scsi/Config.in linux/drivers/scsi/Config.in
--- linux.orig/drivers/scsi/Config.in	Wed Oct 17 10:00:29 2001
+++ linux/drivers/scsi/Config.in	Wed Oct 17 10:25:40 2001
@@ -96,7 +96,8 @@
 if [ "$CONFIG_PCI" = "y" -a "$CONFIG_SCSI_NCR53C7xx" != "y" ]; then
   dep_tristate 'NCR53C8XX SCSI support' CONFIG_SCSI_NCR53C8XX $CONFIG_SCSI
   dep_tristate 'SYM53C8XX SCSI support' CONFIG_SCSI_SYM53C8XX $CONFIG_SCSI
-  if [ "$CONFIG_SCSI_NCR53C8XX" != "n" -o "$CONFIG_SCSI_SYM53C8XX" != "n" ]; then
+  dep_tristate 'SYM53C8XX SCSI ver 2 support' CONFIG_SCSI_SYM53C8XX2 $CONFIG_SCSI
+  if [ "$CONFIG_SCSI_NCR53C8XX" != "n" -o "$CONFIG_SCSI_SYM53C8XX" != "n" -o "$CONFIG_SCSI_SYM53C8XX2" != "n"]; then
     int  '  default tagged command queue depth' CONFIG_SCSI_NCR53C8XX_DEFAULT_TAGS 16
     int  '  maximum number of queued commands' CONFIG_SCSI_NCR53C8XX_MAX_TAGS 32
     int  '  synchronous transfers frequency in MHz' CONFIG_SCSI_NCR53C8XX_SYNC 80
diff -uNr linux.orig/drivers/scsi/Makefile linux/drivers/scsi/Makefile
--- linux.orig/drivers/scsi/Makefile	Wed Oct 17 10:00:29 2001
+++ linux/drivers/scsi/Makefile	Wed Oct 17 10:08:54 2001
@@ -44,11 +44,11 @@
 
 endif
 
-ifeq ($(CONFIG_SCSI_SYM53C8XX),y)
+ifeq ($(CONFIG_SCSI_SYM53C8XX2),y)
   SUB_DIRS += sym53c8xx
-  L_OBJS += sym53c8xx/sym53c8xx.o
+  L_OBJS += sym53c8xx/sym53c8xx2.o
 else
-  ifeq ($(CONFIG_SCSI_SYM53C8XX),m)
+  ifeq ($(CONFIG_SCSI_SYM53C8XX2),m)
   MOD_SUB_DIRS += sym53c8xx
   endif
 endif
@@ -542,6 +542,14 @@
   endif
 endif
 
+ifeq ($(CONFIG_SCSI_SYM53C8XX),y)
+L_OBJS += sym53c8xx.o
+else
+  ifeq ($(CONFIG_SCSI_SYM53C8XX),m)
+  M_OBJS += sym53c8xx.o
+  endif
+endif
+
 ifeq ($(CONFIG_SCSI_PAS16),y)
 L_OBJS += pas16.o
 else
diff -uNr linux.orig/drivers/scsi/sym53c8xx/Makefile linux/drivers/scsi/sym53c8xx/Makefile
--- linux.orig/drivers/scsi/sym53c8xx/Makefile	Wed Oct 17 10:00:29 2001
+++ linux/drivers/scsi/sym53c8xx/Makefile	Wed Oct 17 10:09:46 2001
@@ -1,14 +1,14 @@
 # File: drivers/sym53c8xx/Makefile
 # Makefile for the NCR/SYMBIOS/LSI 53C8XX PCI SCSI controllers driver.
 
-ifeq ($(CONFIG_SCSI_SYM53C8XX),y)
-    O_TARGET := sym53c8xx.o
+ifeq ($(CONFIG_SCSI_SYM53C8XX2),y)
+    O_TARGET := sym53c8xx2.o
     O_OBJS := sym_fw.o sym_glue.o sym_hipd.o sym_malloc.o sym_misc.o sym_nvram.o
 else
-  ifeq ($(CONFIG_SCSI_SYM53C8XX),m)
+  ifeq ($(CONFIG_SCSI_SYM53C8XX2),m)
     MOD_LIST_NAME := SCSI_MODULES
-    M_OBJS := sym53c8xx.o
-    O_TARGET := sym53c8xx.o
+    M_OBJS := sym53c8xx2.o
+    O_TARGET := sym53c8xx2.o
     O_OBJS := sym_fw.o sym_glue.o sym_hipd.o sym_malloc.o sym_misc.o sym_nvram.o
   endif
 endif
