--- linux/net/wanrouter/af_wanpipe.c	Fri Feb 23 17:55:55 2001
+++ linux/net/wanrouter/af_wanpipe.c.sz	Fri Feb 23 17:58:06 2001
@@ -64,9 +64,7 @@
 #include <net/inet_common.h>
 #endif
 
-#ifdef CONFIG_BRIDGE
-#include <net/br.h>
-#endif
+#include <linux/if_bridge.h>
 
 #define SLOW_BACKOFF 0.1*HZ
 #define FAST_BACKOFF 0.01*HZ
@@ -1697,11 +1695,14 @@
 
 		case SIOCGIFBR:
 		case SIOCSIFBR:
-#ifdef CONFIG_BRIDGE		
-			return(br_ioctl(cmd,(void *) arg));
-#else
+#ifdef CONFIG_KMOD
+			if (br_ioctl_hook == NULL)
+				request_module("bride");
+#endif
+			if (br_ioctl_hook != NULL)
+				return br_ioctl_hook(arg);
+			
 			return -ENOPKG;
-#endif						
 			
 #ifdef CONFIG_INET
 		case SIOCADDRT:
