Index: linux-2.6.23/fs/namei.c
===================================================================
RCS file: /ext1/sysadm/transparent/repository/linux-2.6.23/fs/namei.c,v
retrieving revision 1.2
retrieving revision 1.4
diff -u -p -r1.2 -r1.4
--- linux-2.6.23/fs/namei.c	29 Oct 2007 07:01:03 -0000	1.2
+++ linux-2.6.23/fs/namei.c	29 Oct 2007 07:05:53 -0000	1.4
@@ -2893,6 +2893,7 @@
 EXPORT_SYMBOL(vfs_unlink);
 EXPORT_SYMBOL(dentry_unhash);
 EXPORT_SYMBOL(generic_readlink);
+EXPORT_SYMBOL(deny_write_access);
 
 /* to be mentioned only in INIT_TASK */
 struct fs_struct init_fs = {
Index: linux-2.6.27/fs/ecryptfs/inode.c
===================================================================
retrieving revision 1.1
retrieving revision 1.2
diff -u -p -r1.1 -r1.2
--- linux-2.6.27/fs/ecryptfs/inode.c	19 Dec 2008 03:05:27 -0000	1.1
+++ linux-2.6.27/fs/ecryptfs/inode.c	19 Dec 2008 19:52:26 -0000	1.2
@@ -430,9 +430,6 @@ out_lock:
 	unlock_dir(lower_dir_dentry);
 	dput(lower_new_dentry);
 	dput(lower_old_dentry);
-	d_drop(lower_old_dentry);
-	d_drop(new_dentry);
-	d_drop(old_dentry);
 	return rc;
 }
 
@@ -444,7 +441,10 @@ static int ecryptfs_unlink(struct inode 
 	struct dentry *lower_dir_dentry;
 
 	lower_dir_dentry = lock_parent(lower_dentry);
+	dget(lower_dentry);
+	atomic_inc(&lower_dentry->d_inode->i_count);
 	rc = vfs_unlink(lower_dir_inode, lower_dentry);
+	dput(lower_dentry);
 	if (rc) {
 		printk(KERN_ERR "Error in vfs_unlink; rc = [%d]\n", rc);
 		goto out_unlock;
@@ -455,6 +455,7 @@ static int ecryptfs_unlink(struct inode 
 	dentry->d_inode->i_ctime = dir->i_ctime;
 	d_drop(dentry);
 out_unlock:
+	iput(lower_dentry->d_inode);
 	unlock_dir(lower_dir_dentry);
 	return rc;
 }
@@ -538,8 +539,12 @@ static int ecryptfs_rmdir(struct inode *
 	fsstack_copy_attr_times(dir, lower_dir_dentry->d_inode);
 	dir->i_nlink = lower_dir_dentry->d_inode->i_nlink;
 	unlock_dir(lower_dir_dentry);
-	if (!rc)
+	if (!rc) {
+		struct inode *inode = dentry->d_inode;
+		inode->i_nlink = ecryptfs_inode_to_lower(inode)->i_nlink;
+		inode->i_ctime = dir->i_ctime;
 		d_drop(dentry);
+	}
 	dput(dentry);
 	return rc;
 }
Index: linux-2.6.19/fs/super.c
===================================================================
RCS file: /proj/linux-2.6/repository/linux-2.6.19/fs/super.c,v
retrieving revision 1.1
retrieving revision 1.2
diff -u -p -r1.1 -r1.2
--- linux-2.6.19/fs/super.c	20 Feb 2008 05:23:09 -0000	1.1
+++ linux-2.6.19/fs/super.c	20 Feb 2008 05:24:15 -0000	1.2
@@ -250,6 +250,7 @@ int fsync_super(struct super_block *sb)
 	__fsync_super(sb);
 	return sync_blockdev(sb->s_bdev);
 }
+EXPORT_SYMBOL(fsync_super);
 
 /**
  *	generic_shutdown_super	-	common helper for ->kill_sb()
Index: linux-2.6.22/fs/namei.c
===================================================================
RCS file: /ext1/sysadm/transparent/repository/linux-2.6.22/fs/namei.c,v
retrieving revision 1.1
retrieving revision 1.3
diff -u -p -r1.1 -r1.3
--- linux-2.6.22/fs/namei.c	12 Jul 2007 02:55:19 -0000	1.1
+++ linux-2.6.22/fs/namei.c	12 Jul 2007 03:00:50 -0000	1.3
@@ -1280,7 +1280,7 @@ out:
 	return err;
 }
 
-static struct dentry *__lookup_hash(struct qstr *name,
+struct dentry *__lookup_hash(struct qstr *name,
 		struct dentry *base, struct nameidata *nd)
 {
 	struct dentry *dentry;
@@ -2766,6 +2766,7 @@ EXPORT_SYMBOL(follow_up);
 EXPORT_SYMBOL(get_write_access); /* binfmt_aout */
 EXPORT_SYMBOL(getname);
 EXPORT_SYMBOL(lock_rename);
+EXPORT_SYMBOL(__lookup_hash);
 EXPORT_SYMBOL(lookup_one_len);
 EXPORT_SYMBOL(page_follow_link_light);
 EXPORT_SYMBOL(page_put_link);
Index: linux-2.6.22/include/linux/namei.h
===================================================================
RCS file: /ext1/sysadm/transparent/repository/linux-2.6.22/include/linux/namei.h,v
retrieving revision 1.1
retrieving revision 1.2
diff -u -p -r1.1 -r1.2
--- linux-2.6.22/include/linux/namei.h	12 Jul 2007 02:55:19 -0000	1.1
+++ linux-2.6.22/include/linux/namei.h	12 Jul 2007 02:55:20 -0000	1.2
@@ -81,6 +81,7 @@ extern struct file *lookup_instantiate_f
 extern struct file *nameidata_to_filp(struct nameidata *nd, int flags);
 extern void release_open_intent(struct nameidata *);
 
+struct dentry * __lookup_hash(struct qstr *name, struct dentry * base, struct nameidata *nd);
 extern struct dentry *lookup_one_len(const char *, struct dentry *, int);
 extern struct dentry *lookup_one_len_kern(const char *, struct dentry *, int);
 
Index: fs/file_table.c
===================================================================
RCS file: /ext1/sysadm/transparent/repository/linux-2.6.23-rc6/fs/file_table.c,v
retrieving revision 1.1
retrieving revision 1.2
diff -u -p -r1.1 -r1.2
--- linux/fs/file_table.c	12 Sep 2007 16:24:04 -0000	1.1
+++ linux/fs/file_table.c	12 Sep 2007 16:24:38 -0000	1.2
@@ -243,6 +243,7 @@ void put_filp(struct file *file)
 		file_free(file);
 	}
 }
+EXPORT_SYMBOL(put_filp);
 
 void file_move(struct file *file, struct list_head *list)
 {
