--- linux-2.6.4/net/ipv4/netfilter/ipt_osf.c.orig	2004-04-16 16:31:10.000000000 +0200
+++ linux-2.6.4/net/ipv4/netfilter/ipt_osf.c	2004-04-16 17:40:26.174459608 +0200
@@ -845,8 +845,8 @@
 	
 	remove_proc_entry("sys/net/ipv4/osf", NULL);
 	ipt_unregister_match(&osf_match);
-	if (nts && nts->socket)
-		sock_release(nts->socket);
+	if (nts && nts->sk_socket)
+		sock_release(nts->sk_socket);
 
 	list_for_each_safe(ent, n, &finger_list)
 	{
