--- linux-2.6.8-rc3/net/Kconfig.orig	2004-08-03 23:26:45.000000000 +0200
+++ linux-2.6.8-rc3/net/Kconfig	2004-08-07 08:01:30.850514408 +0200
@@ -551,26 +551,6 @@
 
 	  If unsure, say N.
 
-config NET_FASTROUTE
-	bool "Fast switching (read help!)"
-	depends on EXPERIMENTAL
-	---help---
-	  Saying Y here enables direct NIC-to-NIC (NIC = Network Interface
-	  Card) data transfers on the local network, which is fast.
-
-	  IMPORTANT NOTE: This option is NOT COMPATIBLE with "Network packet
-	  filtering" (CONFIG_NETFILTER). Say N here if you say Y there.
-
-	  However, it will work with all options in the "Advanced router"
-	  section (except for "Use TOS value as routing key" and
-	  "Use FWMARK value as routing key").
-
-	  At the moment, few devices support fast switching (tulip is one of
-	  them, a modified 8390 driver can be found at
-	  <ftp://ftp.tux.org/pub/net/ip-routing/fastroute/fastroute-8390.tar.gz>).
-
-	  If unsure, say N.
-
 config NET_HW_FLOWCONTROL
 	bool "Forwarding between high speed interfaces"
 	depends on EXPERIMENTAL
--- linux-2.6.8-rc3/drivers/char/omnibook/ec.c.orig	2004-08-07 15:32:19.080841480 +0200
+++ linux-2.6.8-rc3/drivers/char/omnibook/ec.c	2004-08-07 15:32:21.215516960 +0200
@@ -28,6 +28,8 @@
 #define USE_ACPI_EC
 #endif
 #endif
+#else
+#include <asm-generic/errno.h>
 #endif
 
 #include <asm/io.h>
--- linux-2.6.8-rc3/arch/i386/crypto/Makefile.orig	2004-08-07 14:54:25.000000000 +0200
+++ linux-2.6.8-rc3/arch/i386/crypto/Makefile	2004-08-07 18:41:49.000000000 +0200
@@ -6,4 +6,4 @@
 
 obj-$(CONFIG_CRYPTO_AES_586) += aes-i586.o
 
-aes-i586-y := aes-i586-asm.o aes.o
+aes-i586-objs := aes-i586-asm.o aes.o
--- linux-2.6.8-rc3/arch/i386/power/suspend2.c.orig	2004-08-08 13:26:48.000000000 +0200
+++ linux-2.6.8-rc3/arch/i386/power/suspend2.c	2004-08-08 14:50:21.850891208 +0200
@@ -133,23 +133,11 @@
 static void fix_processor_context(void)
 {
 	int nr = smp_processor_id();
-	struct tss_struct * t = &per_cpu(init_tss,nr);
+	struct tss_struct * t = init_tss + nr;
 
 	set_tss_desc(nr,t);	/* This just modifies memory; should not be neccessary. But... This is neccessary, because 386 hardware has concept of busy tsc or some similar stupidity. */
-#if LINUX_VERSION_CODE > KERNEL_VERSION(2,4,99)
-#if LINUX_VERSION_CODE > KERNEL_VERSION(2,6,7)
-        per_cpu(cpu_gdt_table,nr)[GDT_ENTRY_TSS].b &= 0xfffffdff;
-#else
-        cpu_gdt_table[nr][GDT_ENTRY_TSS].b &= 0xfffffdff;
-#endif
-
+	cpu_gdt_table[nr][GDT_ENTRY_TSS].b &= 0xfffffdff;
 	load_TR_desc();
-#else
-        gdt_table[__TSS(nr)].b &= 0xfffffdff;
-
-	load_TR(nr);		/* This does ltr */
-#endif
-
 	load_LDT(&current->active_mm->context);	/* This does lldt */
 
 	/*
