--- linux.aufs/fs/aufs/cpup.c	1970-01-01 01:00:00.000000000 +0100
+++ linux/fs/aufs/cpup.c	2013-08-23 23:59:39.631583456 +0200
@@ -24,9 +24,9 @@
 #include <linux/mm.h>
 #include "aufs.h"
 
-void au_cpup_attr_flags(struct inode *dst, unsigned int iflags)
+void au_cpup_attr_flags(struct inode *dst, unsigned short iflags)
 {
-	const unsigned int mask = S_DEAD | S_SWAPFILE | S_PRIVATE
+	const unsigned short mask = S_DEAD | S_SWAPFILE | S_PRIVATE
 		| S_NOATIME | S_NOCMTIME | S_AUTOMOUNT;
 
 	BUILD_BUG_ON(sizeof(iflags) != sizeof(dst->i_flags));
@@ -155,7 +155,7 @@ diff -urN /usr/share/empty/fs/aufs/cpup.c linux/fs/aufs/cpup.c
 struct au_cpup_reg_attr {
 	int		valid;
 	struct kstat	st;
-	unsigned int	iflags; /* inode->i_flags */
+	unsigned short	iflags; /* inode->i_flags */
 };
 
 static noinline_for_stack
--- linux.aufs/fs/aufs/cpup.h	1970-01-01 01:00:00.000000000 +0100
+++ linux/fs/aufs/cpup.h	2013-08-23 23:59:39.634916914 +0200
@@ -31,7 +31,7 @@ diff -urN /usr/share/empty/fs/aufs/cpup.h linux/fs/aufs/cpup.h
 struct file;
 struct au_pin;
 
-void au_cpup_attr_flags(struct inode *dst, unsigned int iflags);
+void au_cpup_attr_flags(struct inode *dst, unsigned short iflags);
 void au_cpup_attr_timesizes(struct inode *inode);
 void au_cpup_attr_nlink(struct inode *inode, int force);
 void au_cpup_attr_changeable(struct inode *inode);
