--- linux-2.6.22.14_desktop/scripts/Makefile.modpost~	2007-12-12 13:26:19.000000000 +0200
+++ linux-2.6.22.14_desktop/scripts/Makefile.modpost	2007-12-12 13:26:55.609852824 +0200
@@ -63,7 +63,7 @@
 	$(if $(KBUILD_EXTMOD),-i,-o) $(kernelsymfile) \
 	$(if $(KBUILD_EXTMOD),-I $(modulesymfile)) \
 	$(if $(KBUILD_EXTMOD),-o $(modulesymfile)) \
-	$(if $(KBUILD_EXTMOD)$(KBUILD_MODPOST_WARN),-w)
+	$(if $(KBUILD_EXTMOD),$(if $(KBUILD_MODPOST_WARN),-w))
 
 PHONY += __modpost
 __modpost: $(modules:.ko=.o) FORCE
