--- linux-2.6.8-rc3/net/Kconfig.orig	2004-08-03 23:26:45.000000000 +0200
+++ linux-2.6.8-rc3/net/Kconfig	2004-08-07 08:01:30.850514408 +0200
@@ -551,26 +551,6 @@
 
 	  If unsure, say N.
 
-config NET_FASTROUTE
-	bool "Fast switching (read help!)"
-	depends on EXPERIMENTAL
-	---help---
-	  Saying Y here enables direct NIC-to-NIC (NIC = Network Interface
-	  Card) data transfers on the local network, which is fast.
-
-	  IMPORTANT NOTE: This option is NOT COMPATIBLE with "Network packet
-	  filtering" (CONFIG_NETFILTER). Say N here if you say Y there.
-
-	  However, it will work with all options in the "Advanced router"
-	  section (except for "Use TOS value as routing key" and
-	  "Use FWMARK value as routing key").
-
-	  At the moment, few devices support fast switching (tulip is one of
-	  them, a modified 8390 driver can be found at
-	  <ftp://ftp.tux.org/pub/net/ip-routing/fastroute/fastroute-8390.tar.gz>).
-
-	  If unsure, say N.
-
 config NET_HW_FLOWCONTROL
 	bool "Forwarding between high speed interfaces"
 	depends on EXPERIMENTAL
--- linux-2.6.8-rc3/drivers/char/omnibook/ec.c.orig	2004-08-07 15:32:19.080841480 +0200
+++ linux-2.6.8-rc3/drivers/char/omnibook/ec.c	2004-08-07 15:32:21.215516960 +0200
@@ -28,6 +28,8 @@
 #define USE_ACPI_EC
 #endif
 #endif
+#else
+#include <asm-generic/errno.h>
 #endif
 
 #include <asm/io.h>
