--- ./sound/pcmcia/vx/vxpocket.c~	2003-07-14 05:31:21.000000000 +0200
+++ ./sound/pcmcia/vx/vxpocket.c	2003-07-16 22:06:22.000000000 +0200
@@ -35,6 +35,7 @@
 #include <pcmcia/version.h>
 #include "vxpocket.h"
 #include <sound/initval.h>
+#include <linux/init.h>
 
 /*
  */
--- ./sound/pci/rme9652/hdsp.c~	2003-07-14 05:32:41.000000000 +0200
+++ ./sound/pci/rme9652/hdsp.c	2003-07-16 22:03:09.000000000 +0200
@@ -26,6 +26,7 @@
 #include <linux/interrupt.h>
 #include <linux/slab.h>
 #include <linux/pci.h>
+#include <linux/init.h>
 
 #include <sound/core.h>
 #include <sound/control.h>
--- ./sound/drivers/opl4/opl4_lib.c~	2003-07-17 22:07:25.000000000 +0200
+++ ./sound/drivers/opl4/opl4_lib.c	2003-07-17 22:15:24.000000000 +0200
@@ -21,6 +21,7 @@
 #include <sound/initval.h>
 #include <linux/ioport.h>
 #include <asm/io.h>
+#include <linux/init.h>
 
 MODULE_AUTHOR("Clemens Ladisch <clemens@ladisch.de>");
 MODULE_DESCRIPTION("OPL4 driver");
