--- linux-2.6.4/net/ipv4/netfilter/ipt_string.c.orig	2004-04-16 21:00:42.000000000 +0200
+++ linux-2.6.4/net/ipv4/netfilter/ipt_string.c	2004-04-16 21:06:19.000000000 +0200
@@ -23,6 +23,9 @@
 #include <linux/file.h>
 #include <net/sock.h>
 
+#include <linux/cpumask.h>
+static int smp_num_cpus = num_online_cpus();
+
 #include <linux/netfilter_ipv4/ip_tables.h>
 #include <linux/netfilter_ipv4/ipt_string.h>
 
