--- linux/Makefile.org	Tue Feb  5 11:12:19 2002
+++ linux/Makefile	Wed Feb  6 10:40:35 2002
@@ -562,10 +562,10 @@
 #
 
 scripts/mkdep: scripts/mkdep.c
-	$(HOSTCC) $(HOSTCFLAGS) -o scripts/mkdep scripts/mkdep.c
+	$(HOSTCC) $(HOSTCFLAGS) -I$(HPATH) -o scripts/mkdep scripts/mkdep.c
 
 scripts/split-include: scripts/split-include.c
-	$(HOSTCC) $(HOSTCFLAGS) -o scripts/split-include scripts/split-include.c
+	$(HOSTCC) $(HOSTCFLAGS) -I$(HPATH) -o scripts/split-include scripts/split-include.c
 
 #
 # RPM target
