diff -Nurp linux-2.6.22.9-vs2.3.0.25/kernel/posix-timers.c linux-2.6.22.9-vs2.3.0.25.timer/kernel/posix-timers.c
--- linux-2.6.22.9-vs2.3.0.25/kernel/posix-timers.c	2007-09-29 14:12:16.000000000 +0200
+++ linux-2.6.22.9-vs2.3.0.25.timer/kernel/posix-timers.c	2007-10-01 14:09:31.000000000 +0200
@@ -299,9 +299,11 @@ void do_schedule_next_timer(struct sigin
 int posix_timer_event(struct k_itimer *timr,int si_private)
 {
 	struct vx_info_save vxis;
+	struct vx_info *vxi;
 	int ret;
 
-	enter_vx_info(task_get_vx_info(timr->it_process), &vxis);
+	vxi = task_get_vx_info(timr->it_process);
+	enter_vx_info(vxi, &vxis);
 	memset(&timr->sigq->info, 0, sizeof(siginfo_t));
 	timr->sigq->info.si_sys_private = si_private;
 	/* Send signal to the process that owns this timer.*/
@@ -330,7 +332,7 @@ int posix_timer_event(struct k_itimer *t
 				   timr->it_process);
 out:
 	leave_vx_info(&vxis);
-	put_vx_info(vxis.vxi);
+	put_vx_info(vxi);
 	return ret;
 }
 EXPORT_SYMBOL_GPL(posix_timer_event);
