--- linux-2.6.10-rc3/net/ipv4/netfilter/ipt_XOR.c.orig	2004-12-10 10:54:24.000000000 +0100
+++ linux-2.6.10-rc3/net/ipv4/netfilter/ipt_XOR.c	2004-12-10 21:53:12.578046056 +0100
@@ -40,8 +40,7 @@
 		tcph = (struct tcphdr *) ((*pskb)->data + iph->ihl*4);
 		for (i=0, j=0; i<(ntohs(iph->tot_len) - iph->ihl*4 - tcph->doff*4); ) {
 			for (k=0; k<=info->block_size; k++) {
-				(char) (*pskb)->data[ iph->ihl*4 + tcph->doff*4 + i ] ^= 
-						info->key[j];
+				(*pskb)->data[ iph->ihl*4 + tcph->doff*4 + i ] ^= info->key[j];
 				i++;
 			}
 			j++;
@@ -52,8 +51,7 @@
 		udph = (struct udphdr *) ((*pskb)->data + iph->ihl*4);
 		for (i=0, j=0; i<(ntohs(udph->len)-8); ) {
 			for (k=0; k<=info->block_size; k++) {
-				(char) (*pskb)->data[ iph->ihl*4 + sizeof(struct udphdr) + i ] ^= 
-						info->key[j];
+				(*pskb)->data[ iph->ihl*4 + sizeof(struct udphdr) + i ] ^= info->key[j];
 				i++;
 			}
 			j++;
