diff -urNp -x '*.orig' linux-4.4/Makefile linux-4.4/Makefile
--- linux-4.4/Makefile	2021-02-24 17:41:14.010620493 +0100
+++ linux-4.4/Makefile	2021-02-24 17:41:27.804397388 +0100
@@ -837,6 +837,9 @@ KBUILD_CFLAGS   += $(call cc-option,-Wer
 # Prohibit date/time macros, which would make the build non-deterministic
 KBUILD_CFLAGS   += $(call cc-option,-Werror=date-time)
 
+# enforce correct pointer usage
+KBUILD_CFLAGS   += $(call cc-option,-Werror=incompatible-pointer-types)
+
 # use the deterministic mode of AR if available
 KBUILD_ARFLAGS := $(call ar-option,D)
 
diff -urNp -x '*.orig' linux-4.4/scripts/kconfig/lxdialog/check-lxdialog.sh linux-4.4/scripts/kconfig/lxdialog/check-lxdialog.sh
--- linux-4.4/scripts/kconfig/lxdialog/check-lxdialog.sh	2016-01-11 00:01:32.000000000 +0100
+++ linux-4.4/scripts/kconfig/lxdialog/check-lxdialog.sh	2021-02-24 17:41:27.804397388 +0100
@@ -11,6 +11,12 @@ ldflags()
 			$cc -print-file-name=lib${lib}.${ext} | grep -q /
 			if [ $? -eq 0 ]; then
 				echo "-l${lib}"
+				for libt in tinfow tinfo ; do
+					$cc -print-file-name=lib${libt}.${ext} | grep -q /
+					if [ $? -eq 0 ]; then
+						echo "-l${libt}"
+					fi
+				done
 				exit
 			fi
 		done
diff -urNp -x '*.orig' linux-4.4/scripts/mod/modpost.c linux-4.4/scripts/mod/modpost.c
--- linux-4.4/scripts/mod/modpost.c	2021-02-24 17:41:15.860679986 +0100
+++ linux-4.4/scripts/mod/modpost.c	2021-02-24 17:41:27.804397388 +0100
@@ -19,7 +19,8 @@
 #include <stdbool.h>
 #include <errno.h>
 #include "modpost.h"
-#include "../../include/generated/autoconf.h"
+// PLD architectures don't use CONFIG_SYMBOL_PREFIX
+//#include "../../include/generated/autoconf.h"
 #include "../../include/linux/license.h"
 #include "../../include/linux/export.h"
 
--- linux-4.4/drivers/extcon/extcon-adc-jack.c.org	2016-01-11 00:01:32.000000000 +0100
+++ linux-4.4/drivers/extcon/extcon-adc-jack.c	2021-08-09 08:44:14.895794160 +0200
@@ -105,7 +105,7 @@ static int adc_jack_probe(struct platfor
 		return -EINVAL;
 	}
 
-	data->edev = devm_extcon_dev_allocate(&pdev->dev, pdata->cable_names);
+	data->edev = devm_extcon_dev_allocate(&pdev->dev, (const unsigned int *)pdata->cable_names);
 	if (IS_ERR(data->edev)) {
 		dev_err(&pdev->dev, "failed to allocate extcon device\n");
 		return -ENOMEM;
--- b/include/net/neighbour.h
+++ a/include/net/neighbour.h
@@ -276,5 +276,10 @@
 
 
+static inline bool neigh_key_eq16(const struct neighbour *n, const void *pkey)
+{
+	return *(const u16 *)n->primary_key == *(const u16 *)pkey;
+}
+
 static inline bool neigh_key_eq32(const struct neighbour *n, const void *pkey)
 {
 	return *(const u32 *)n->primary_key == *(const u32 *)pkey;
--- b/drivers/staging/rdma/hfi1/chip.c
+++ a/drivers/staging/rdma/hfi1/chip.c
@@ -7962,7 +7962,6 @@
 
 	if (dd->synth_stats_timer.data)
 		del_timer_sync(&dd->synth_stats_timer);
-	cancel_work_sync(&dd->update_cntr_work);
 	dd->synth_stats_timer.data = 0;
 	ppd = (struct hfi1_pportdata *)(dd + 1);
 	for (i = 0; i < dd->num_pports; i++, ppd++) {
