--- linux-2.6.27/init/Kconfig~	2008-11-20 23:26:34.000000000 +0100
+++ linux-2.6.27/init/Kconfig	2008-12-01 20:37:12.000000000 +0100
@@ -781,8 +781,8 @@
 source "arch/Kconfig"
 
 config PROC_PAGE_MONITOR
- 	default n
-	depends on PROC_FS && MMU && !GRKERNSEC
+ 	default y
+	depends on PROC_FS && MMU
 	bool "Enable /proc page monitoring" if EMBEDDED
  	help
 	  Various /proc files exist to monitor process memory utilization:
@@ -798,9 +798,9 @@
 
 config SLABINFO
 	bool
-	depends on PROC_FS && !GRKERNSEC_PROC_ADD
+	depends on PROC_FS
 	depends on SLAB || SLUB_DEBUG
-	default n
+	default y
 
 config RT_MUTEXES
 	boolean
--- linux-2.6.27/fs/proc/proc_misc.c~	2008-11-20 23:26:30.000000000 +0100
+++ linux-2.6.27/fs/proc/proc_misc.c	2008-12-01 20:35:55.000000000 +0100
@@ -934,8 +934,12 @@
 #endif
 	proc_create("stat", 0, NULL, &proc_stat_operations);
 	proc_create("interrupts", 0, NULL, &proc_interrupts_operations);
-#if defined(CONFIG_SLABINFO) && !defined(CONFIG_GRKERNSEC_PROC_ADD)
+#ifdef CONFIG_SLABINFO
+#ifdef CONFIG_GRKERNSEC_PROC_ADD
+	proc_create("slabinfo",gr_mode,NULL,&proc_slabinfo_operations);
+#else
 	proc_create("slabinfo",S_IWUSR|S_IRUGO,NULL,&proc_slabinfo_operations);
+#endif
 #ifdef CONFIG_DEBUG_SLAB_LEAK
 	proc_create("slab_allocators", 0, NULL, &proc_slabstats_operations);
 #endif
