--- linux-2.6.7-rc2/init/main.c.org	2004-06-04 12:58:31.000000000 +0200
+++ linux-2.6.7-rc2/init/main.c	2004-06-04 13:43:02.000000000 +0200
@@ -103,8 +103,8 @@
 /*
  * Boot command-line arguments
  */
-#define MAX_INIT_ARGS 8
-#define MAX_INIT_ENVS 8
+#define MAX_INIT_ARGS 256
+#define MAX_INIT_ENVS 256
 
 extern void time_init(void);
 /* Default late time init is NULL. archs can override this later. */
