diff -Nurp linux-2.6.22.10-vs2.3.0.29/kernel/vserver/context.c linux-2.6.22.10-vs2.3.0.29.nsproxy/kernel/vserver/context.c
--- linux-2.6.22.10-vs2.3.0.29/kernel/vserver/context.c	2007-10-29 21:23:59.000000000 -0400
+++ linux-2.6.22.10-vs2.3.0.29.nsproxy/kernel/vserver/context.c	2007-12-04 22:09:56.000000000 -0500
@@ -115,8 +115,6 @@ static struct vx_info *__alloc_vx_info(x
 	new->reboot_cmd = 0;
 	new->exit_code = 0;
 
-	new->vx_nsproxy = copy_nsproxy(current->nsproxy);
-
 	vxdprintk(VXD_CBIT(xid, 0),
 		"alloc_vx_info(%d) = %p", xid, new);
 	vxh_alloc_vx_info(new);
