--- linux/include/asm-m68k/processor.h~	Fri Jan 24 09:20:47 2003
+++ linux/include/asm-m68k/processor.h	Fri Jan 24 09:20:47 2003
@@ -27,6 +27,7 @@
  */
 #define EISA_bus 0
 #define MCA_bus 0
+#define MCA_bus__is_a_macro /* for versions in ksyms.c */
 
 /* 
  * if you change this structure, you must change the code and offsets
--- linux/include/linux/mca.h~	Fri Jan 24 09:20:43 2003
+++ linux/include/linux/mca.h	Fri Jan 24 09:20:43 2003
@@ -6,11 +6,14 @@
 #ifndef _LINUX_MCA_H
 #define _LINUX_MCA_H
 
+#include <asm/processor.h>
+#ifndef MCA_bus__is_a_macro
 /* The detection of MCA bus is done in the real mode (using BIOS).
  * The information is exported to the protected code, where this 
  * variable is set to one in case MCA bus was detected.
 */
 extern int  MCA_bus;
+#endif
 
 /* maximal number of MCA slots - actually, some machines have less, but
 they all have sufficient number of POS registers to cover 8. */
