--- linux-6.15/fs/aufs/Makefile~	2025-05-26 10:00:26.000000000 +0200
+++ linux-6.15/fs/aufs/Makefile	2025-05-26 10:01:28.093923904 +0200
@@ -1,10 +1,10 @@
 # SPDX-License-Identifier: GPL-2.0
 
-include ${src}/magic.mk
+include ${srctree}/fs/magic.mk
 ifeq (${CONFIG_AUFS_FS},m)
-include ${src}/conf.mk
+include ${srctree}/fs/conf.mk
 endif
--include ${src}/priv_def.mk
+-include ${srctree}/fs/priv_def.mk
 
 ccflags-y += ${EXTRA_CFLAGS}
 # cf. include/linux/kernel.h
