--- linux/include/asm-sparc64/io.h~	Sun Mar 25 18:31:09 2001
+++ linux/include/asm-sparc64/io.h	Sun Nov 17 18:13:23 2002
@@ -126,6 +126,10 @@
 
 #define inb_p inb
 #define outb_p outb
+#define inw_p inw
+#define outw_p outw
+#define inl_p inl
+#define outl_p outl
 
 extern void outsb(unsigned long addr, const void *src, unsigned long count);
 extern void outsw(unsigned long addr, const void *src, unsigned long count);
