--- linux-2.6.22/arch/powerpc/boot/Makefile~	2008-02-09 17:38:56.000000000 +0200
+++ linux-2.6.22/arch/powerpc/boot/Makefile	2008-02-09 18:29:15.881852252 +0200
@@ -31,8 +31,8 @@
 
 BOOTCFLAGS	+= -I$(obj) -I$(srctree)/$(obj)
 
-$(obj)/44x.o: BOOTCFLAGS += -mcpu=440
-$(obj)/ebony.o: BOOTCFLAGS += -mcpu=440
+$(obj)/44x.o: BOOTCFLAGS += -mcpu=common -Wa,-m440
+$(obj)/ebony.o: BOOTCFLAGS += -mcpu=common -Wa,-m440
 
 zlib       := inffast.c inflate.c inftrees.c
 zlibheader := inffast.h inffixed.h inflate.h inftrees.h infutil.h
