--- linux-7.0/fs/namespace.c~	2026-04-24 21:35:57.000000000 +0200
+++ linux-7.0/fs/namespace.c	2026-04-25 11:43:27.786675705 +0200
@@ -2278,6 +2278,7 @@
 		path_get(p);
 	return res;
 }
+EXPORT_SYMBOL_GPL(collect_paths);
 
 void drop_collected_paths(const struct path *paths, const struct path *prealloc)
 {
@@ -2286,6 +2287,7 @@
 	if (paths != prealloc)
 		kfree(paths);
 }
+EXPORT_SYMBOL_GPL(drop_collected_paths);
 
 static struct mnt_namespace *alloc_mnt_ns(struct user_namespace *, bool);
 
