--- linux-2.6.8-rc3/drivers/char/omnibook/ec.c.orig	2004-08-07 15:32:19.080841480 +0200
+++ linux-2.6.8-rc3/drivers/char/omnibook/ec.c	2004-08-07 15:32:21.215516960 +0200
@@ -28,6 +28,8 @@
 #define USE_ACPI_EC
 #endif
 #endif
+#else
+#include <asm-generic/errno.h>
 #endif
 
 #include <asm/io.h>
--- linux-2.6.8-rc3/arch/i386/crypto/Makefile.orig	2004-08-07 14:54:25.000000000 +0200
+++ linux-2.6.8-rc3/arch/i386/crypto/Makefile	2004-08-07 18:41:49.000000000 +0200
@@ -6,4 +6,4 @@
 
 obj-$(CONFIG_CRYPTO_AES_586) += aes-i586.o
 
-aes-i586-y := aes-i586-asm.o aes.o
+aes-i586-objs := aes-i586-asm.o aes.o

