--- linux/kernel/sysctl.c.org	Tue Nov 21 08:01:25 2000
+++ linux/kernel/sysctl.c	Tue Nov 21 08:00:33 2000
@@ -1173,6 +1173,13 @@
 	return -ENOSYS;
 }
 
+int sysctl_jiffies(ctl_table *table, int *name, int nlen,
+		void *oldval, size_t *oldlenp,
+		void *newval, size_t newlen, void **context)
+{
+	return -ENOSYS;
+}
+
 int proc_dostring(ctl_table *table, int write, struct file *filp,
 		  void *buffer, size_t *lenp)
 {
