diff -urNp -x '*.orig' linux-4.9/drivers/net/usb/rndis_host.c linux-4.9/drivers/net/usb/rndis_host.c
--- linux-4.9/drivers/net/usb/rndis_host.c	2021-02-24 15:44:20.021329751 +0100
+++ linux-4.9/drivers/net/usb/rndis_host.c	2021-02-24 15:44:33.395081159 +0100
@@ -352,8 +352,7 @@ generic_rndis_bind(struct usbnet *dev, s
 		goto fail_and_release;
 	}
 
-	dev->rx_urb_size = dev->hard_mtu + (dev->maxpacket + 1);
-	dev->rx_urb_size &= ~(dev->maxpacket - 1);
+	dev->rx_urb_size = (dev->udev->speed == USB_SPEED_FULL) ? 16384 : 8192;
 	u.init->max_transfer_size = cpu_to_le32(dev->rx_urb_size);
 
 	net->netdev_ops = &rndis_netdev_ops;
