--- linux-2.6.10/Makefile.orig	2005-01-05 00:43:41.000000000 +0100
+++ linux-2.6.10/Makefile	2005-01-05 00:54:23.961980896 +0100
@@ -824,8 +824,13 @@
 	  echo '"$(KERNELRELEASE)" exceeds $(uts_len) characters' >&2; \
 	  exit 1; \
 	fi; \
-	(echo \#define UTS_RELEASE \"$(KERNELRELEASE)\"; \
-	  echo \#define LINUX_VERSION_CODE `expr $(VERSION) \\* 65536 + $(PATCHLEVEL) \\* 256 + $(SUBLEVEL)`; \
+	(echo \#include \<linux/config.h\>; \
+	 echo \#ifndef CONFIG_SMP; \
+	 echo \#define UTS_RELEASE \"$(KERNELRELEASE)\"; \
+	 echo \#else; \
+	 echo \#define UTS_RELEASE \"$(KERNELRELEASE)smp\"; \
+	 echo \#endif; \
+	 echo \#define LINUX_VERSION_CODE `expr $(VERSION) \\* 65536 + $(PATCHLEVEL) \\* 256 + $(SUBLEVEL)`; \
 	 echo '#define KERNEL_VERSION(a,b,c) (((a) << 16) + ((b) << 8) + (c))'; \
 	)
 endef
