#
#	drivers/atm/zatm.c:1608: error: invalid lvalue in assignment
#
--- linux-2.6.10-rc3/drivers/atm/zatm.c.orig	2004-12-03 22:52:57.000000000 +0100
+++ linux-2.6.10-rc3/drivers/atm/zatm.c	2004-12-10 19:00:19.578469344 +0100
@@ -1605,7 +1605,7 @@
 		goto out_disable;
 
 	zatm_dev->pci_dev = pci_dev;
-	ZATM_DEV(dev) = zatm_dev;
+	dev = (struct atm_dev *)zatm_dev;
 	zatm_dev->copper = (int)ent->driver_data;
 	if ((ret = zatm_init(dev)) || (ret = zatm_start(dev)))
 		goto out_release;


#
#	drivers/atm/iphase.c:623: error: invalid storage class for function 'ia_pkt_tx'
#
--- linux-2.6.10-rc3/drivers/atm/iphase.c.orig	2004-12-03 22:54:50.000000000 +0100
+++ linux-2.6.10-rc3/drivers/atm/iphase.c	2004-12-10 19:11:17.505424400 +0100
@@ -615,12 +615,13 @@
    return tmp;
 }    
 
+static int ia_pkt_tx (struct atm_vcc *vcc, struct sk_buff *skb);
+
 static int ia_que_tx (IADEV *iadev) { 
    struct sk_buff *skb;
    int num_desc;
    struct atm_vcc *vcc;
    struct ia_vcc *iavcc;
-   static int ia_pkt_tx (struct atm_vcc *vcc, struct sk_buff *skb);
    num_desc = ia_avail_descs(iadev);
 
    while (num_desc && (skb = skb_dequeue(&iadev->tx_backlog))) {


#
#	drivers/cdrom/sbpcd.c: In function 'get_state_T':
#	drivers/cdrom/sbpcd.c:1168: error: invalid storage class for function 'cmd_out_T'
#	drivers/cdrom/sbpcd.c: In function 'cmd_out_T':
#	drivers/cdrom/sbpcd.c:1317: error: invalid storage class for function 'cc_DriveReset'
#
--- linux-2.6.10-rc3/drivers/cdrom/sbpcd.c.orig	2004-12-03 22:52:56.000000000 +0100
+++ linux-2.6.10-rc3/drivers/cdrom/sbpcd.c	2004-12-10 19:54:14.710629656 +0100
@@ -1161,11 +1161,11 @@
 	return;
 }
 /*==========================================================================*/
+static int cmd_out_T(void);
+
 static int get_state_T(void)
 {
 	int i;
-	
-	static int cmd_out_T(void);
 
 	clr_cmdbuf();
 	current_drive->n_bytes=1;
@@ -1308,13 +1308,14 @@
 	return (i);
 }
 /*==========================================================================*/
+static int cc_DriveReset(void);
+
 static int cmd_out_T(void)
 {
 #undef CMDT_TRIES
 #define CMDT_TRIES 1000
 #define TEST_FALSE_FF 1
-	
-	static int cc_DriveReset(void);
+
 	int i, j, l=0, m, ntries;
 	unsigned long flags;
 
