--- a/kernel/printk.c	2007-10-02 14:01:04.780023500 +0200
+++ b/kernel/printk.c	2007-10-02 14:09:40.140231500 +0200
@@ -30,9 +30,9 @@
 #include <linux/delay.h>
 #include <linux/smp.h>
 #include <linux/security.h>
+#include <linux/suspend.h>
 #include <linux/bootmem.h>
 #include <linux/syscalls.h>
-#include <linux/suspend.h>
 #include <linux/kexec.h>
 
 #include <asm/uaccess.h>
--- e/include/linux/fs.h	2008-05-07 11:39:56.000000000 +0000
+++ e/include/linux/fs.h	2008-05-07 11:43:49.907309587 +0000
@@ -207,10 +207,13 @@
 #define MS_KERNMOUNT	(1<<22) /* this is a kern_mount call */
 #define MS_I_VERSION	(1<<23) /* Update inode I_version field */
 #define MS_STRICTATIME	(1<<24) /* Always perform atime updates */
-#define MS_FROZEN	(1<<25)	/* Frozen by freeze_filesystems() */
 #define MS_ACTIVE	(1<<30)
 #define MS_NOUSER	(1<<31)
 
+
+
+#define MS_FROZEN	(1<<29) /* Frozen by freeze_filesystems() */
+
 /*
  * Superblock flags that can be altered by MS_REMOUNT
  */

--- e/include/linux/fs.h.orig	2009-07-20 14:35:32.000000000 +0200
+++ e/include/linux/fs.h	2009-07-20 14:37:27.086666353 +0200
@@ -230,6 +230,8 @@
 
 /* Inode flags - they have nothing to superblock flags now */
 
+#define S_ATOMIC_COPY	2048	/* Pages mapped with this inode need to be
+				  atomically copied (gem) */
 #define S_SYNC		1	/* Writes are synced at once */
 #define S_NOATIME	2	/* Do not update access times */
 #define S_APPEND	4	/* Append-only file */
@@ -240,8 +242,6 @@
 #define S_NOCMTIME	128	/* Do not update file c/mtime */
 #define S_SWAPFILE	256	/* Do not truncate: swapon got its bmaps */
 #define S_PRIVATE	512	/* Inode is fs-internal */
-#define S_ATOMIC_COPY	1024	/* Pages mapped with this inode need to be
-				   atomically copied (gem) */
 
 /*
  * Note that nosuid etc flags are inode-specific: setting some file-system

