--- a/init/main.c	2007-10-02 14:01:04.772023000 +0200
+++ b/init/main.c	2007-10-02 14:07:34.520380750 +0200
@@ -51,12 +51,12 @@
 #include <linux/key.h>
 #include <linux/unwind.h>
 #include <linux/buffer_head.h>
+#include <linux/dyn_pageflags.h>
 #include <linux/debug_locks.h>
 #include <linux/lockdep.h>
 #include <linux/pid_namespace.h>
 #include <linux/device.h>
 #include <linux/kthread.h>
-#include <linux/dyn_pageflags.h>
 #include <linux/sched.h>
 
 #include <asm/io.h>
--- a/kernel/printk.c	2007-10-02 14:01:04.780023500 +0200
+++ b/kernel/printk.c	2007-10-02 14:09:40.140231500 +0200
@@ -28,10 +28,10 @@
 #include <linux/delay.h>
 #include <linux/smp.h>
 #include <linux/security.h>
+#include <linux/suspend.h>
 #include <linux/bootmem.h>
 #include <linux/syscalls.h>
 #include <linux/jiffies.h>
-#include <linux/suspend.h>
 
 #include <asm/uaccess.h>
 
--- a/kernel/timer.c.org	2008-01-20 11:46:53.931178999 +0100
+++ b/kernel/timer.c	2008-01-20 11:47:09.758637482 +0100
@@ -32,13 +32,13 @@
 #include <linux/time.h>
 #include <linux/jiffies.h>
 #include <linux/posix-timers.h>
+#include <linux/notifier.h>
+#include <linux/suspend.h>
 #include <linux/cpu.h>
 #include <linux/syscalls.h>
 #include <linux/delay.h>
 #include <linux/tick.h>
 #include <linux/kallsyms.h>
-#include <linux/notifier.h>
-#include <linux/suspend.h>
 
 #include <asm/uaccess.h>
 #include <asm/unistd.h>

--- a/include/linux/fs.h.org	2008-01-20 11:56:05.845456959 +0100
+++ b/include/linux/fs.h	2008-01-20 11:56:23.109653995 +0100
@@ -119,6 +119,9 @@
 #define MS_VERBOSE	32768	/* War is peace. Verbosity is silence.
 				   MS_VERBOSE is deprecated. */
 #define MS_SILENT	32768
+
+#define MS_FROZEN	(1<<23)	/* Frozen by freeze_filesystems() */
+
 #define MS_POSIXACL	(1<<16)	/* VFS does not apply the umask */
 #define MS_UNBINDABLE	(1<<17)	/* change to unbindable */
 #define MS_PRIVATE	(1<<18)	/* change to private */
@@ -126,7 +129,6 @@
 #define MS_SHARED	(1<<20)	/* change to shared */
 #define MS_RELATIME	(1<<21)	/* Update atime relative to mtime/ctime. */
 #define MS_KERNMOUNT	(1<<22) /* this is a kern_mount call */
-#define MS_FROZEN	(1<<23)	/* Frozen by freeze_filesystems() */
 #define MS_ACTIVE	(1<<30)
 #define MS_NOUSER	(1<<31)
 
