From jsimmons@infradead.org Fri Apr 11 01:25:27 2003
Date: Fri, 11 Apr 2003 00:12:25 +0100 (BST)
From: James Simmons <jsimmons@infradead.org>

diff -Nru linux-2.5.67/include/linux/pci_ids.h linux-2.5-jw/include/linux/pci_ids.h
--- linux-2.5.67/include/linux/pci_ids.h        2003-04-07 13:31:06.000000000 -0400
+++ linux-2.5-jw/include/linux/pci_ids.h        2003-04-10 18:17:11.000000000 -0400
@@ -287,6 +287,7 @@
  #define PCI_DEVICE_ID_ATI_RADEON_Ie    0x4965
  #define PCI_DEVICE_ID_ATI_RADEON_If    0x4966
  #define PCI_DEVICE_ID_ATI_RADEON_Ig    0x4967
+#define PCI_DEVICE_ID_ATI_RADEON_QM    0x514d
  /* Radeon R300 (9700) */
  #define PCI_DEVICE_ID_ATI_RADEON_ND    0x4e44
  #define PCI_DEVICE_ID_ATI_RADEON_NE    0x4e45


