--- linux-2.6.8.1/fs/binfmt_elf.c.org	2004-10-08 14:53:25.000000000 +0200
+++ linux-2.6.8.1/fs/binfmt_elf.c	2004-10-08 14:58:55.000000000 +0200
@@ -733,7 +733,7 @@
   	struct exec interp_ex;
 	char passed_fileno[6];
 	struct files_struct *files;
-	int have_pt_gnu_stack, executable_stack = EXSTACK_DEFAULT;
+	int executable_stack = EXSTACK_DEFAULT;
 	unsigned long def_flags = 0;
 	unsigned long task_size = TASK_SIZE;
 
