--- linux/drivers/scsi/sym53c8xx_comm.h~	Thu Feb 26 09:20:57 2004
+++ linux/drivers/scsi/sym53c8xx_comm.h	Thu Feb 26 09:20:57 2004
@@ -283,7 +283,7 @@
 	base = pdev->base_address[index];
 #if BITS_PER_LONG > 32
 	if ((base & 0x7) == 0x4)
-		*base |= (((u_long)pdev->base_address[++index]) << 32);
+		base |= (((u_long)pdev->base_address[++index]) << 32);
 #endif
 #endif
 	return (base & ~0x7ul);
--- linux/drivers/scsi/sym53c8xx.c.0523	Wed May 23 18:25:06 2001
+++ linux/drivers/scsi/sym53c8xx.c	Wed May 23 18:42:48 2001
@@ -464,7 +464,7 @@
 	base = pdev->base_address[index];
 #if BITS_PER_LONG > 32
 	if ((base & 0x7) == 0x4)
-		*base |= (((u_long)pdev->base_address[++index]) << 32);
+		base |= (((u_long)pdev->base_address[++index]) << 32);
 #endif
 #endif
 	return (base & ~0x7ul);
@@ -681,7 +681,11 @@
 
 #ifdef __sparc__
 #  include <asm/irq.h>
+#if LINUX_VERSION_CODE >= LinuxVersionCode(2,4,0)
 #  define memcpy_to_pci(a, b, c)	memcpy_toio((a), (b), (c))
+#else
+#  define memcpy_to_pci(a, b, c)	memcpy_toio((void *)(a), (b), (c))
+#endif
 #elif defined(__alpha__)
 #  define memcpy_to_pci(a, b, c)	memcpy_toio((a), (b), (c))
 #else	/* others */
------------------------ Cut HERE -------------------------

Thanks for the report.

  Regards,
     Gérard.
