--- linux-2.6.10/Makefile.orig	2005-01-06 23:41:48.000000000 +0100
+++ linux-2.6.10/Makefile	2005-01-06 23:44:16.000385304 +0100
@@ -824,7 +824,7 @@
 	  echo '"$(KERNELRELEASE)" exceeds $(uts_len) characters' >&2; \
 	  exit 1; \
 	fi; \
-	(echo \#define UTS_RELEASE \"$(KERNELRELEASE)\"; \
+	(echo \#define UTS_RELEASE \"$(KERNELRELEASE)$(LOCALVERSION)\"; \
 	  echo \#define LINUX_VERSION_CODE `expr $(VERSION) \\* 65536 + $(PATCHLEVEL) \\* 256 + $(SUBLEVEL)`; \
 	 echo '#define KERNEL_VERSION(a,b,c) (((a) << 16) + ((b) << 8) + (c))'; \
 	)
