--- ./net/mac80211/ieee80211_rate.h.org	2007-10-09 22:31:38.000000000 +0200
+++ ./net/mac80211/ieee80211_rate.h	2007-10-29 12:51:53.613305271 +0100
@@ -15,8 +15,8 @@
 #include <linux/skbuff.h>
 #include <linux/types.h>
 #include <net/mac80211.h>
-#include "ieee80211_i.h"
-#include "sta_info.h"
+#include <net/ieee80211_i.h>
+#include <net/sta_info.h>
 
 #define RATE_CONTROL_NUM_DOWN 20
 #define RATE_CONTROL_NUM_UP   15
--- ./net/mac80211/ieee80211_i.h.org	2007-10-09 22:31:38.000000000 +0200
+++ ./net/mac80211/ieee80211_i.h	2007-10-29 14:53:16.353770862 +0100
@@ -22,8 +22,8 @@
 #include <linux/types.h>
 #include <linux/spinlock.h>
 #include <net/wireless.h>
-#include "ieee80211_key.h"
-#include "sta_info.h"
+#include <net/ieee80211_key.h>
+#include <net/sta_info.h>
 
 /* ieee80211.o internal definitions, etc. These are not included into
  * low-level drivers. */
