--- linux-3.3/fs/proc/base.c~	2012-03-19 21:44:42.000000000 +0100
+++ linux-3.3/fs/proc/base.c	2012-03-21 12:25:28.051092423 +0100
@@ -568,6 +568,8 @@
 				 struct task_struct *task,
 				 int hide_pid_min)
 {
+	if (vx_check(0, VS_WATCH_P))
+		return true;
 	if (pid->hide_pid < hide_pid_min)
 		return true;
 	if (in_group_p(pid->pid_gid))
diff -NurpP --minimal linux-3.14.17/init/Kconfig linux-3.14.17-vs2.3.6.13/init/Kconfig
--- linux-3.14.17/init/Kconfig	2014-08-14 01:38:34.000000000 +0000
+++ linux-3.14.17-vs2.3.6.13/init/Kconfig	2014-08-30 14:37:19.000000000 +0000
@@ -1117,6 +1118,7 @@ config IPC_NS
 
 config USER_NS
 	bool "User namespace"
+	depends on VSERVER_DISABLED
 	default n
 	help
 	  This allows containers, i.e. vservers, to use user namespaces
