diff -Nur linux.old/include/linux/tty.h linux/include/linux/tty.h
--- linux.old/include/linux/tty.h	Thu Oct 19 18:10:10 2000
+++ linux/include/linux/tty.h	Thu Oct 19 19:57:44 2000
@@ -277,7 +277,6 @@
 	int alt_speed;		/* For magic substitution of 38400 bps */
 	struct wait_queue *write_wait;
 	struct wait_queue *read_wait;
-	struct wait_queue *poll_wait;
 	struct tq_struct tq_hangup;
 	void *disc_data;
 	void *driver_data;
@@ -306,6 +305,7 @@
 	struct semaphore atomic_read;
 	struct semaphore atomic_write;
 	spinlock_t read_lock;
+	struct wait_queue *poll_wait;
 };
 
 /* tty magic number */
