--- linux-2.6.27/scripts/unifdef.c.org	2009-05-25 09:52:07.548880740 +0200
+++ linux-2.6.27/scripts/unifdef.c	2009-05-25 09:52:37.499458519 +0200
@@ -206,7 +206,7 @@
 static void             error(const char *);
 static int              findsym(const char *);
 static void             flushline(bool);
-static Linetype         getline(void);
+static Linetype         unidef_getline(void);
 static Linetype         ifeval(const char **);
 static void             ignoreoff(void);
 static void             ignoreon(void);
@@ -512,7 +512,7 @@
 
 	for (;;) {
 		linenum++;
-		lineval = getline();
+		lineval = unidef_getline();
 		trans_table[ifstate[depth]][lineval]();
 		debug("process %s -> %s depth %d",
 		    linetype_name[lineval],
@@ -526,7 +526,7 @@
  * help from skipcomment().
  */
 static Linetype
-getline(void)
+unidef_getline(void)
 {
 	const char *cp;
 	int cursym;

