--- linux-2.6.18/drivers/char/drm/drm_pciids.h.via	2006-10-09 18:12:45.000000000 +0200
+++ linux-2.6.18/drivers/char/drm/drm_pciids.h	2006-10-09 18:13:56.000000000 +0200
@@ -227,6 +227,7 @@
 	{0x1106, 0x3122, PCI_ANY_ID, PCI_ANY_ID, 0, 0, 0}, \
 	{0x1106, 0x7205, PCI_ANY_ID, PCI_ANY_ID, 0, 0, 0}, \
 	{0x1106, 0x3108, PCI_ANY_ID, PCI_ANY_ID, 0, 0, 0}, \
+	{0x1106, 0x3344, PCI_ANY_ID, PCI_ANY_ID, 0, 0, 0}, \
 	{0, 0, 0}
 
 #define i810_PCI_IDS \
