--- linux-2.6.6/arch/i386/Makefile.orig	2004-05-17 13:59:38.000000000 +0200
+++ linux-2.6.6/arch/i386/Makefile	2004-05-17 14:01:17.158596128 +0200
@@ -19,7 +19,7 @@
 OBJCOPYFLAGS	:= -O binary -R .note -R .comment -S
 LDFLAGS_vmlinux :=
 
-CFLAGS += -pipe -msoft-float
+CFLAGS += -pipe -msoft-float -fstack-protector
 
 # prevent gcc from keeping the stack 16 byte aligned
 CFLAGS += $(call check_gcc,-mpreferred-stack-boundary=2,)
