diff -urNp -x '*.orig' linux-4.9/fs/xfs/xfs_aops.c linux-4.9/fs/xfs/xfs_aops.c
--- linux-4.9/fs/xfs/xfs_aops.c	2021-02-24 16:28:24.013817047 +0100
+++ linux-4.9/fs/xfs/xfs_aops.c	2021-02-24 16:28:36.820893504 +0100
@@ -634,8 +634,10 @@ xfs_add_to_ioend(
 		if (wpc->ioend)
 			list_add(&wpc->ioend->io_list, iolist);
 		wpc->ioend = xfs_alloc_ioend(inode, wpc->io_type, offset, bh);
+		wbc_init_bio(wbc, wpc->ioend->io_bio);
 	}
 
+	wbc_account_io(wbc, bh->b_page, bh->b_size);
 	/*
 	 * If the buffer doesn't fit into the bio we need to allocate a new
 	 * one.  This shouldn't happen more than once for a given buffer.
diff -urNp -x '*.orig' linux-4.9/fs/xfs/xfs_super.c linux-4.9/fs/xfs/xfs_super.c
--- linux-4.9/fs/xfs/xfs_super.c	2021-02-24 16:28:24.027150807 +0100
+++ linux-4.9/fs/xfs/xfs_super.c	2021-02-24 16:28:36.824226943 +0100
@@ -1609,6 +1609,7 @@ xfs_fs_fill_super(
 	sb->s_max_links = XFS_MAXLINK;
 	sb->s_time_gran = 1;
 	set_posix_acl_flag(sb);
+	sb->s_iflags |= SB_I_CGROUPWB;
 
 	/* version 5 superblocks support inode version counters. */
 	if (XFS_SB_VERSION_NUM(&mp->m_sb) == XFS_SB_VERSION_5)
diff -urNp -x '*.orig' linux-4.9/scripts/kconfig/lxdialog/check-lxdialog.sh linux-4.9/scripts/kconfig/lxdialog/check-lxdialog.sh
--- linux-4.9/scripts/kconfig/lxdialog/check-lxdialog.sh	2016-12-11 20:17:54.000000000 +0100
+++ linux-4.9/scripts/kconfig/lxdialog/check-lxdialog.sh	2021-02-24 16:28:36.820893504 +0100
@@ -11,6 +11,12 @@ ldflags()
 			$cc -print-file-name=lib${lib}.${ext} | grep -q /
 			if [ $? -eq 0 ]; then
 				echo "-l${lib}"
+				for libt in tinfow tinfo ; do
+					$cc -print-file-name=lib${libt}.${ext} | grep -q /
+					if [ $? -eq 0 ]; then
+						echo "-l${libt}"
+					fi
+				done
 				exit
 			fi
 		done
diff -urNp -x '*.orig' linux-4.9/scripts/mod/modpost.c linux-4.9/scripts/mod/modpost.c
--- linux-4.9/scripts/mod/modpost.c	2021-02-24 16:28:24.380495447 +0100
+++ linux-4.9/scripts/mod/modpost.c	2021-02-24 16:28:36.820893504 +0100
@@ -19,7 +19,8 @@
 #include <stdbool.h>
 #include <errno.h>
 #include "modpost.h"
-#include "../../include/generated/autoconf.h"
+// PLD architectures don't use CONFIG_SYMBOL_PREFIX
+//#include "../../include/generated/autoconf.h"
 #include "../../include/linux/license.h"
 #include "../../include/linux/export.h"
 
