--- linux-2.6.7-rc2/init/main.c.org	2004-06-04 12:58:31.000000000 +0200
+++ linux-2.6.7-rc2/init/main.c	2004-06-04 13:43:02.000000000 +0200
@@ -103,8 +103,8 @@
 /*
  * Boot command-line arguments
  */
-#define MAX_INIT_ARGS 8
-#define MAX_INIT_ENVS 8
+#define MAX_INIT_ARGS 256
+#define MAX_INIT_ENVS 256
 
 extern void time_init(void);
 /* Default late time init is NULL. archs can override this later. */
diff -Nur --exclude '*.orig' linux-2.6.7-rc3.org/arch/cris/kernel/setup.c linux-2.6.7-rc3/arch/cris/kernel/setup.c
--- linux-2.6.7-rc3.org/arch/cris/kernel/setup.c	2004-06-07 21:14:58.000000000 +0200
+++ linux-2.6.7-rc3/arch/cris/kernel/setup.c	2004-06-08 11:24:42.000000000 +0200
@@ -29,7 +29,7 @@
 extern int root_mountflags;
 extern char _etext, _edata, _end;
 
-#define COMMAND_LINE_SIZE 256
+#define COMMAND_LINE_SIZE 4096
 
 static char command_line[COMMAND_LINE_SIZE] = { 0, };
        char saved_command_line[COMMAND_LINE_SIZE];
diff -Nur --exclude '*.orig' linux-2.6.7-rc3.org/arch/ia64/kernel/setup.c linux-2.6.7-rc3/arch/ia64/kernel/setup.c
--- linux-2.6.7-rc3.org/arch/ia64/kernel/setup.c	2004-06-07 21:13:39.000000000 +0200
+++ linux-2.6.7-rc3/arch/ia64/kernel/setup.c	2004-06-08 11:25:17.000000000 +0200
@@ -88,7 +88,7 @@
 unsigned long ia64_max_iommu_merge_mask = ~0UL;
 EXPORT_SYMBOL(ia64_max_iommu_merge_mask);
 
-#define COMMAND_LINE_SIZE	512
+#define COMMAND_LINE_SIZE	4096
 
 char saved_command_line[COMMAND_LINE_SIZE]; /* used in proc filesystem */
 
diff -Nur --exclude '*.orig' linux-2.6.7-rc3.org/arch/parisc/kernel/setup.c linux-2.6.7-rc3/arch/parisc/kernel/setup.c
--- linux-2.6.7-rc3.org/arch/parisc/kernel/setup.c	2004-06-07 21:15:11.000000000 +0200
+++ linux-2.6.7-rc3/arch/parisc/kernel/setup.c	2004-06-08 11:26:06.000000000 +0200
@@ -45,7 +45,7 @@
 #include <asm/pdc_chassis.h>
 #include <asm/io.h>
 
-#define COMMAND_LINE_SIZE 1024
+#define COMMAND_LINE_SIZE 4096
 char	saved_command_line[COMMAND_LINE_SIZE];
 char	command_line[COMMAND_LINE_SIZE];
 
diff -Nur --exclude '*.orig' linux-2.6.7-rc3.org/arch/sh/kernel/setup.c linux-2.6.7-rc3/arch/sh/kernel/setup.c
--- linux-2.6.7-rc3.org/arch/sh/kernel/setup.c	2004-06-07 21:14:42.000000000 +0200
+++ linux-2.6.7-rc3/arch/sh/kernel/setup.c	2004-06-08 11:26:39.000000000 +0200
@@ -85,7 +85,7 @@
 #define INITRD_SIZE (*(unsigned long *) (PARAM+0x014))
 /* ... */
 #define COMMAND_LINE ((char *) (PARAM+0x100))
-#define COMMAND_LINE_SIZE 256
+#define COMMAND_LINE_SIZE 4096
 
 #define RAMDISK_IMAGE_START_MASK  	0x07FF
 #define RAMDISK_PROMPT_FLAG		0x8000
diff -Nur --exclude '*.orig' linux-2.6.7-rc3.org/include/asm-alpha/system.h linux-2.6.7-rc3/include/asm-alpha/system.h
--- linux-2.6.7-rc3.org/include/asm-alpha/system.h	2004-06-07 21:14:33.000000000 +0200
+++ linux-2.6.7-rc3/include/asm-alpha/system.h	2004-06-08 11:27:50.000000000 +0200
@@ -43,7 +43,7 @@
  */
 #define PARAM			ZERO_PGE
 #define COMMAND_LINE		((char*)(PARAM + 0x0000))
-#define COMMAND_LINE_SIZE	256
+#define COMMAND_LINE_SIZE	4096
 #define INITRD_START		(*(unsigned long *) (PARAM+0x100))
 #define INITRD_SIZE		(*(unsigned long *) (PARAM+0x108))
 
diff -Nur --exclude '*.orig' linux-2.6.7-rc3.org/include/asm-arm/setup.h linux-2.6.7-rc3/include/asm-arm/setup.h
--- linux-2.6.7-rc3.org/include/asm-arm/setup.h	2004-06-07 21:14:12.000000000 +0200
+++ linux-2.6.7-rc3/include/asm-arm/setup.h	2004-06-08 11:28:44.000000000 +0200
@@ -14,7 +14,7 @@
 #ifndef __ASMARM_SETUP_H
 #define __ASMARM_SETUP_H
 
-#define COMMAND_LINE_SIZE 1024
+#define COMMAND_LINE_SIZE 4096
 
 /* The list ends with an ATAG_NONE node. */
 #define ATAG_NONE	0x00000000
diff -Nur --exclude '*.orig' linux-2.6.7-rc3.org/include/asm-arm26/setup.h linux-2.6.7-rc3/include/asm-arm26/setup.h
--- linux-2.6.7-rc3.org/include/asm-arm26/setup.h	2004-06-07 21:14:01.000000000 +0200
+++ linux-2.6.7-rc3/include/asm-arm26/setup.h	2004-06-08 11:29:00.000000000 +0200
@@ -14,7 +14,7 @@
 #ifndef __ASMARM_SETUP_H
 #define __ASMARM_SETUP_H
 
-#define COMMAND_LINE_SIZE 1024
+#define COMMAND_LINE_SIZE 4096
 
 /* The list ends with an ATAG_NONE node. */
 #define ATAG_NONE	0x00000000
diff -Nur --exclude '*.orig' linux-2.6.7-rc3.org/include/asm-i386/setup.h linux-2.6.7-rc3/include/asm-i386/setup.h
--- linux-2.6.7-rc3.org/include/asm-i386/setup.h	2004-06-07 21:14:42.000000000 +0200
+++ linux-2.6.7-rc3/include/asm-i386/setup.h	2004-06-08 11:29:19.000000000 +0200
@@ -17,7 +17,7 @@
 #define MAX_NONPAE_PFN	(1 << 20)
 
 #define PARAM_SIZE 2048
-#define COMMAND_LINE_SIZE 256
+#define COMMAND_LINE_SIZE 4096
 
 #define OLD_CL_MAGIC_ADDR	0x90020
 #define OLD_CL_MAGIC		0xA33F
diff -Nur --exclude '*.orig' linux-2.6.7-rc3.org/include/asm-ppc/machdep.h linux-2.6.7-rc3/include/asm-ppc/machdep.h
--- linux-2.6.7-rc3.org/include/asm-ppc/machdep.h	2004-06-07 21:14:03.000000000 +0200
+++ linux-2.6.7-rc3/include/asm-ppc/machdep.h	2004-06-08 11:29:48.000000000 +0200
@@ -106,7 +106,7 @@
 };
 
 extern struct machdep_calls ppc_md;
-#define COMMAND_LINE_SIZE 512
+#define COMMAND_LINE_SIZE 4096
 extern char cmd_line[COMMAND_LINE_SIZE];
 
 extern void setup_pci_ptrs(void);
diff -Nur --exclude '*.orig' linux-2.6.7-rc3.org/include/asm-ppc64/machdep.h linux-2.6.7-rc3/include/asm-ppc64/machdep.h
--- linux-2.6.7-rc3.org/include/asm-ppc64/machdep.h	2004-06-07 21:14:04.000000000 +0200
+++ linux-2.6.7-rc3/include/asm-ppc64/machdep.h	2004-06-08 11:30:11.000000000 +0200
@@ -112,7 +112,7 @@
 };
 
 extern struct machdep_calls ppc_md;
-#define COMMAND_LINE_SIZE 512
+#define COMMAND_LINE_SIZE 4096
 extern char cmd_line[COMMAND_LINE_SIZE];
 extern char saved_command_line[COMMAND_LINE_SIZE];
 
diff -Nur --exclude '*.orig' linux-2.6.7-rc3.org/include/asm-s390/setup.h linux-2.6.7-rc3/include/asm-s390/setup.h
--- linux-2.6.7-rc3.org/include/asm-s390/setup.h	2004-06-07 21:14:58.000000000 +0200
+++ linux-2.6.7-rc3/include/asm-s390/setup.h	2004-06-08 11:30:38.000000000 +0200
@@ -9,7 +9,7 @@
 #define _ASM_S390_SETUP_H
 
 #define PARMAREA		0x10400
-#define COMMAND_LINE_SIZE 	896
+#define COMMAND_LINE_SIZE 	4096
 #define RAMDISK_ORIGIN		0x800000
 #define RAMDISK_SIZE		0x800000
 
diff -Nur --exclude '*.orig' linux-2.6.7-rc3.org/include/asm-x86_64/bootsetup.h linux-2.6.7-rc3/include/asm-x86_64/bootsetup.h
--- linux-2.6.7-rc3.org/include/asm-x86_64/bootsetup.h	2004-06-07 21:14:56.000000000 +0200
+++ linux-2.6.7-rc3/include/asm-x86_64/bootsetup.h	2004-06-08 11:31:53.000000000 +0200
@@ -30,7 +30,7 @@
 #define EDD_NR     (*(unsigned char *) (PARAM+EDDNR))
 #define EDD_BUF     ((struct edd_info *) (PARAM+EDDBUF))
 #define COMMAND_LINE saved_command_line
-#define COMMAND_LINE_SIZE 256
+#define COMMAND_LINE_SIZE 4096
 
 #define RAMDISK_IMAGE_START_MASK  	0x07FF
 #define RAMDISK_PROMPT_FLAG		0x8000
--- linux-2.6.7/arch/ppc/syslib/prom.c.org	2004-06-16 07:18:57.000000000 +0200
+++ linux-2.6.7/arch/ppc/syslib/prom.c	2004-06-17 13:48:00.000000000 +0200
@@ -85,7 +85,7 @@
 extern void enter_rtas(void *);
 void phys_call_rtas(int, int, int, ...);
 
-extern char cmd_line[512];	/* XXX */
+extern char cmd_line[4096];	/* XXX */
 extern boot_infos_t *boot_infos;
 unsigned long dev_tree_size;
 
