diff -Nur linux-2.6.15.6/drivers/input/joystick/iforce/iforce-serio.c linux-2.6.15.6.iforce/drivers/input/joystick/iforce/iforce-serio.c
--- linux-2.6.15.6/drivers/input/joystick/iforce/iforce-serio.c	2006-03-05 19:07:54.000000000 +0000
+++ linux-2.6.15.6.iforce/drivers/input/joystick/iforce/iforce-serio.c	2006-04-29 23:17:59.000000000 +0000
@@ -175,6 +175,12 @@
 		.id	= SERIO_ANY,
 		.extra	= SERIO_ANY,
 	},
+	{
+		.type	= SERIO_RS232,
+		.proto	= 0x1f, // Trust ForceFeedback Race Master
+		.id	= SERIO_ANY,
+		.extra	= SERIO_ANY,
+	},
 	{ 0 }
 };
 
--- linux-2.6.17/include/asm-ppc/io.h~	2006-06-18 01:49:35.000000000 +0000
+++ linux-2.6.17/include/asm-ppc/io.h	2006-06-22 02:44:19.000000000 +0000
@@ -445,6 +445,10 @@
 #define page_to_phys(page)	(page_to_pfn(page) << PAGE_SHIFT)
 #define page_to_bus(page)	(page_to_phys(page) + PCI_DRAM_OFFSET)
 
+#define isa_virt_to_bus virt_to_phys
+#define isa_page_to_bus page_to_phys
+#define isa_bus_to_virt phys_to_virt
+
 /* Enforce in-order execution of data I/O.
  * No distinction between read/write on PPC; use eieio for all three.
  */
--- linux-2.6.20/drivers/media/common/ir-keymaps.c.orig	2007-02-04 19:44:54.000000000 +0100
+++ linux-2.6.20/drivers/media/common/ir-keymaps.c	2007-02-17 22:22:39.327414970 +0100
@@ -151,10 +151,6 @@
 
 IR_KEYTAB_TYPE ir_codes_pixelview[IR_KEYTAB_SIZE] = {
 
-	[ 0x1e ] = KEY_POWER,       // power
-	[ 0x07 ] = KEY_MEDIA,       // source
-	[ 0x1c ] = KEY_SEARCH,      // scan
-
 /* FIXME: duplicate keycodes?
  *
  * These four keys seem to share the same GPIO as CH+, CH-, <<< and >>>
@@ -175,14 +171,10 @@
  */
 
 	[ 0x03 ] = KEY_TUNER,       // TV/FM
+	[ 0x07 ] = KEY_SEARCH,      // scan
+	[ 0x1c ] = KEY_ZOOM,        // zoom
 
-	[ 0x00 ] = KEY_RECORD,
-	[ 0x08 ] = KEY_STOP,
-	[ 0x11 ] = KEY_PLAY,
-
-	[ 0x1a ] = KEY_PLAYPAUSE,   // freeze
-	[ 0x19 ] = KEY_ZOOM,        // zoom
-	[ 0x0f ] = KEY_TEXT,        // min
+	[ 0x1e ] = KEY_POWER,       // power
 
 	[ 0x01 ] = KEY_1,
 	[ 0x0b ] = KEY_2,
@@ -195,20 +187,22 @@
 	[ 0x12 ] = KEY_9,
 	[ 0x02 ] = KEY_0,
 	[ 0x10 ] = KEY_LAST,        // +100
-	[ 0x13 ] = KEY_LIST,        // recall
+	[ 0x13 ] = KEY_LIST,        // loop
 
-	[ 0x1f ] = KEY_CHANNELUP,   // chn down
-	[ 0x17 ] = KEY_CHANNELDOWN, // chn up
-	[ 0x16 ] = KEY_VOLUMEUP,    // vol down
-	[ 0x14 ] = KEY_VOLUMEDOWN,  // vol up
+	[ 0x1f ] = KEY_VOLUMEUP,    // vol down
+	[ 0x17 ] = KEY_VOLUMEDOWN,  // vol up
+	[ 0x16 ] = KEY_CHANNELUP,   // chn down
+	[ 0x14 ] = KEY_CHANNELDOWN, // chn up
 
-	[ 0x04 ] = KEY_KPMINUS,     // <<<
+	[ 0x00 ] = KEY_MEDIA,       // source
+	[ 0x18 ] = KEY_MUTE,        // mute/unmute
+
+	[ 0x04 ] = KEY_KPMINUS,     // -
 	[ 0x0e ] = KEY_SETUP,       // function
-	[ 0x0c ] = KEY_KPPLUS,      // >>>
+	[ 0x0c ] = KEY_KPPLUS,      // +
 
-	[ 0x0d ] = KEY_GOTO,        // mts
-	[ 0x1d ] = KEY_REFRESH,     // reset
-	[ 0x18 ] = KEY_MUTE         // mute/unmute
+	[ 0x0d ] = KEY_REFRESH,     // reset
+	[ 0x0f ] = KEY_PLAYPAUSE    // freeze
 };
 
 EXPORT_SYMBOL_GPL(ir_codes_pixelview);
--- linux-2.6.22/drivers/sbus/char/cpwatchdog.c~	2007-07-09 01:32:17.000000000 +0200
+++ linux-2.6.22/drivers/sbus/char/cpwatchdog.c	2007-08-12 03:56:02.739993250 +0200
@@ -28,6 +28,7 @@
 #include <asm/ebus.h>
 #include <asm/oplib.h>
 #include <asm/uaccess.h>
+#include <asm/io.h>
 
 #include <asm/watchdog.h>
 
--- linux/include/asm-powerpc/suspend.h	2007-07-09 01:32:17.000000000 +0200
+++ linux/include/asm-powerpc/suspend.h	2007-08-28 23:26:16.629658848 +0200
@@ -6,4 +6,7 @@
 void save_processor_state(void);
 void restore_processor_state(void);
 
+#define suspend2_faulted (0)
+#define clear_suspend2_fault() do { } while(0)
+
 #endif /* __ASM_POWERPC_SUSPEND_H */
