--- linux-2.4.19/drivers/char/pc_keyb.c~	Sat Oct 19 17:06:29 2002
+++ linux-2.4.19/drivers/char/pc_keyb.c	Sat Oct 19 17:06:29 2002
@@ -228,7 +228,14 @@
 #define E0_MSLW	125
 #define E0_MSRW	126
 #define E0_MSTM	127
-
+/*
+ * Some keyboards have Power/Sleep/Wake buttons:
+ * e0 5e for Power, e0 5f for sleep, e0 63 for wake
+*/ 
+#define E0_POWER 124
+#define E0_SLEEP 125
+#define E0_WAKE  126
+ 
 static unsigned char e0_keys[128] = {
   0, 0, 0, 0, 0, 0, 0, 0,			      /* 0x00-0x07 */
   0, 0, 0, 0, 0, 0, 0, 0,			      /* 0x08-0x0f */
@@ -241,8 +248,8 @@
   E0_DO, E0_F17, 0, 0, 0, 0, E0_BREAK, E0_HOME,	      /* 0x40-0x47 */
   E0_UP, E0_PGUP, 0, E0_LEFT, E0_OK, E0_RIGHT, E0_KPMINPLUS, E0_END,/* 0x48-0x4f */
   E0_DOWN, E0_PGDN, E0_INS, E0_DEL, 0, 0, 0, 0,	      /* 0x50-0x57 */
-  0, 0, 0, E0_MSLW, E0_MSRW, E0_MSTM, 0, 0,	      /* 0x58-0x5f */
-  0, 0, 0, 0, 0, 0, 0, 0,			      /* 0x60-0x67 */
+  0, 0, 0, E0_MSLW, E0_MSRW, E0_MSTM, E0_POWER, E0_SLEEP,  /* 0x58-0x5f */
+  0, 0, 0, E0_WAKE, 0, 0, 0, 0,			      /* 0x60-0x67 */
   0, 0, 0, 0, 0, 0, 0, E0_MACRO,		      /* 0x68-0x6f */
   0, 0, 0, 0, 0, 0, 0, 0,			      /* 0x70-0x77 */
   0, 0, 0, 0, 0, 0, 0, 0			      /* 0x78-0x7f */
