diff -urNp -x '*.orig' linux-4.14/scripts/kconfig/lxdialog/check-lxdialog.sh linux-4.14/scripts/kconfig/lxdialog/check-lxdialog.sh
--- linux-4.14/scripts/kconfig/lxdialog/check-lxdialog.sh	2017-11-12 19:46:13.000000000 +0100
+++ linux-4.14/scripts/kconfig/lxdialog/check-lxdialog.sh	2021-02-24 21:43:55.011102640 +0100
@@ -12,6 +12,12 @@ ldflags()
 			$cc -print-file-name=lib${lib}.${ext} | grep -q /
 			if [ $? -eq 0 ]; then
 				echo "-l${lib}"
+				for libt in tinfow tinfo ; do
+					$cc -print-file-name=lib${libt}.${ext} | grep -q /
+					if [ $? -eq 0 ]; then
+						echo "-l${libt}"
+					fi
+				done
 				exit
 			fi
 		done
diff -urNp -x '*.orig' linux-4.14/scripts/mod/modpost.c linux-4.14/scripts/mod/modpost.c
--- linux-4.14/scripts/mod/modpost.c	2021-02-24 21:43:47.051093387 +0100
+++ linux-4.14/scripts/mod/modpost.c	2021-02-24 21:43:55.011102640 +0100
@@ -19,7 +19,8 @@
 #include <stdbool.h>
 #include <errno.h>
 #include "modpost.h"
-#include "../../include/generated/autoconf.h"
+// PLD architectures don't use CONFIG_SYMBOL_PREFIX
+//#include "../../include/generated/autoconf.h"
 #include "../../include/linux/license.h"
 #include "../../include/linux/export.h"
 
diff -urNp -x '*.orig' linux-4.14/security/selinux/include/classmap.h linux-4.14/security/selinux/include/classmap.h
--- linux-4.14/security/selinux/include/classmap.h	2021-02-24 21:43:47.054426725 +0100
+++ linux-4.14/security/selinux/include/classmap.h	2021-02-24 21:43:55.011102640 +0100
@@ -239,9 +239,11 @@ struct security_class_mapping secclass_m
 	  { "access", NULL } },
 	{ "infiniband_endport",
 	  { "manage_subnet", NULL } },
+	{ "xdp_socket",
+	  { COMMON_SOCK_PERMS, NULL } },
 	{ NULL }
   };
 
-#if PF_MAX > 44
+#if PF_MAX > 45
 #error New address family defined, please update secclass_map.
 #endif
