--- linux/drivers/net/Config.in~	Thu Jan 23 15:38:55 2003
+++ linux/drivers/net/Config.in	Thu Jan 23 15:38:55 2003
@@ -169,7 +169,9 @@
   fi
   bool 'Pocket and portable adaptors' CONFIG_NET_POCKET
   if [ "$CONFIG_NET_POCKET" = "y" ]; then
-    bool 'AT-LAN-TEC/RealTek pocket adaptor support' CONFIG_ATP
+    if [ "$CONFIG_X86" = "y" ]; then
+      bool 'AT-LAN-TEC/RealTek pocket adaptor support' CONFIG_ATP
+    fi
     tristate 'D-Link DE600 pocket adaptor support' CONFIG_DE600
     tristate 'D-Link DE620 pocket adaptor support' CONFIG_DE620
   fi
--- drivers/atm/Config.in~	Thu Jan 23 15:42:19 2003
+++ drivers/atm/Config.in	Thu Jan 23 15:42:19 2003
@@ -51,7 +51,7 @@
       if [ "$CONFIG_ATM_FORE200E_PCA" = "y" ]; then
 	bool '   Use default PCA-200E firmware (normally enabled)' CONFIG_ATM_FORE200E_PCA_DEFAULT_FW
         if [ "$CONFIG_ATM_FORE200E_PCA_DEFAULT_FW" = "n" ]; then
-          string '   Pathname of user-supplied binary firmware' CONFIG_ATM_FORE200E_PCA_FW
+          string '   Pathname of user-supplied binary firmware' CONFIG_ATM_FORE200E_PCA_FW ""
         fi
       fi
     fi
