diff -Naur linux/Makefile linux-p/Makefile
--- linux/Makefile	Mon Jan 14 20:31:04 2002
+++ linux-p/Makefile	Mon Jan 14 20:31:53 2002
@@ -511,7 +511,7 @@
 #
 
 scripts/mkdep: scripts/mkdep.c
-	$(HOSTCC) $(HOSTCFLAGS) -o scripts/mkdep scripts/mkdep.c
+	$(HOSTCC) $(HOSTCFLAGS) -I./include/ -o scripts/mkdep scripts/mkdep.c
 
 scripts/split-include: scripts/split-include.c
-	$(HOSTCC) $(HOSTCFLAGS) -o scripts/split-include scripts/split-include.c
+	$(HOSTCC) $(HOSTCFLAGS) -I./include/ -o scripts/split-include scripts/split-include.c
