--- linux-3.3/fs/proc/base.c~	2012-03-19 21:44:42.000000000 +0100
+++ linux-3.3/fs/proc/base.c	2012-03-21 12:25:28.051092423 +0100
@@ -568,6 +568,8 @@
 				 struct task_struct *task,
 				 int hide_pid_min)
 {
+	if (vx_check(0, VS_WATCH_P))
+		return true;
 	if (pid->hide_pid < hide_pid_min)
 		return true;
 	if (in_group_p(pid->pid_gid))
--- linux-4.9.76/include/linux/vs_time.h	1970-01-01 00:00:00.000000000 +0000
+++ linux-4.9.76-vs2.3.9.5/include/linux/vs_time.h	2018-01-10 09:27:12.000000000 +0000
@@ -7,7 +7,7 @@
 
 extern void vx_adjust_timespec(struct timespec *ts);
 extern int vx_settimeofday(const struct timespec *ts);
-extern int vx_settimeofday64(const struct timespec *ts);
+extern int vx_settimeofday64(const struct timespec64 *ts);
 
 #else
 #define	vx_adjust_timespec(t)	do { } while (0)
