--- a/arch/i386/Makefile	2005-03-26 04:28:18.000000000 +0100
+++ b/arch/i386/Makefile	2005-04-07 13:06:01.000000000 +0200
@@ -15,6 +15,13 @@
 #           Added '-march' and '-mpreferred-stack-boundary' support
 #
 
+HAS_BIARCH      := $(call cc-option-yn, -m32)
+ifeq ($(HAS_BIARCH),y)
+AS              := $(AS) --32
+LD              := $(LD) -m elf_i386
+CC              := $(CC) -m32
+endif
+
 LDFLAGS		:= -m elf_i386
 OBJCOPYFLAGS	:= -O binary -R .note -R .comment -S
 LDFLAGS_vmlinux :=
