--- linux/include/asm-alpha/termios.h~	Thu Feb 26 00:47:26 2004
+++ linux/include/asm-alpha/termios.h	Thu Feb 26 00:47:26 2004
@@ -71,6 +71,7 @@
 #define N_SLIP		1
 #define N_MOUSE		2
 #define N_PPP		3
+#define N_STRIP		4
 #define N_AX25		5
 #define N_X25		6	/* X.25 async */
 #define N_6PACK		7
--- linux/drivers/net/sbni.c~	Thu Feb 26 02:28:30 2004
+++ linux/drivers/net/sbni.c	Thu Feb 26 02:28:30 2004
@@ -54,7 +54,9 @@
 
 
 #undef GOODBUS16
-#define CRCASM
+#if defined(__i386__)
+#	define CRCASM
+#endif
 #define KATYUSHA
 
 #include <linux/version.h>
