--- linux/drivers/usb/scanner.c.org	Sat Mar  3 14:07:07 2001
+++ linux/drivers/usb/scanner.c	Sat Mar  3 14:08:38 2001
@@ -303,6 +303,7 @@
 	{ USB_DEVICE(0x04b8, 0x0104) },	/* Perfection 1200U and 1200Photo*/
 	{ USB_DEVICE(0x04b8, 0x0106) },	/* Stylus Scan 2500 */
 	{ USB_DEVICE(0x04b8, 0x0107) },	/* Expression 1600 */
+	{ USB_DEVICE(0x04b8, 0x010a) }, /* perfection 1640SU and 1640SUPhoto */
 	/* Umax */
 	{ USB_DEVICE(0x1606, 0x0010) },	/* Astra 1220U */
 	{ USB_DEVICE(0x1606, 0x0030) },	/* Astra 2000U */
