--- linux-2.4.20.org/drivers/isdn/hisax/avma1_cs.c	Sat Aug  3 00:39:44 2002
+++ linux-2.4.20/drivers/isdn/hisax/avma1_cs.c	Sat Dec 14 08:25:44 2002
@@ -522,9 +522,9 @@
     return 0;
 } /* avma1cs_event */
 
-/*====================================================================*/
+/* ==================================================================== */
 
-static int __init init_avma1_cs(void)
+static int init_avma1_cs(void)
 {
     servinfo_t serv;
     DEBUG(0, "%s\n", version);
@@ -538,7 +538,7 @@
     return 0;
 }
 
-static void __exit exit_avma1_cs(void)
+static int exit_avma1_cs(void)
 {
     DEBUG(0, "avma1_cs: unloading\n");
     unregister_pccard_driver(&dev_info);
