From Fedora (already in linux-2.4 bk); fixes sstfb oops.

--- 1/drivers/video/sstfb.c~	2004-02-07 23:54:31.000000000 +0000
+++ 2/drivers/video/sstfb.c	2004-02-07 23:54:55.000000000 +0000
@@ -1821,6 +1821,8 @@
 	sst_info = (struct sstfb_info*)kmalloc(sizeof(*sst_info), GFP_KERNEL);
 	if (!sst_info)
 		goto fail_kmalloc;
+	memset(sst_info, 0, sizeof(*sst_info));
+	
 	pci_set_drvdata(pdev, sst_info);
 	sst_info->type = id->driver_data;
 	spec = &voodoo_spec[sst_info->type];
