--- linux-2.6.14.2/arch/x86_64/kernel/io_apic.c.orig	2005-11-11 06:33:12.000000000 +0100
+++ linux-2.6.14.2/arch/x86_64/kernel/io_apic.c	2005-11-21 10:16:28.362717500 +0100
@@ -265,18 +266,24 @@ void __init check_ioapic(void) 
 			for (func = 0; func < 8; func++) { 
 				u32 class;
 				u32 vendor;
+				u16 product;
 				u8 type;
 				class = read_pci_config(num,slot,func,
 							PCI_CLASS_REVISION);
+
 				if (class == 0xffffffff)
 					break; 
 
-		       		if ((class >> 16) != PCI_CLASS_BRIDGE_PCI)
+		       		if ((class >> 16) != PCI_CLASS_BRIDGE_PCI && 
+				    (class >> 16) != PCI_CLASS_BRIDGE_HOST)
 					continue; 
 
 				vendor = read_pci_config(num, slot, func, 
 							 PCI_VENDOR_ID);
 				vendor &= 0xffff;
+				
+				product = read_pci_config_16(num, slot, func,
+							  PCI_DEVICE_ID);
 				switch (vendor) { 
 				case PCI_VENDOR_ID_VIA:
 #ifdef CONFIG_GART_IOMMU
@@ -299,8 +306,18 @@ void __init check_ioapic(void) 
 #endif
 					/* RED-PEN skip them on mptables too? */
 					return;
-				} 
 
+				case PCI_VENDOR_ID_ATI:
+					if (product==0x5950 || product==0x5951) {
+						printk(KERN_INFO "ATI board detected - disabling APIC pin 1\n");
+#ifdef CONFIG_ACPI
+						/* This seems to be wrong, too */
+						acpi_skip_timer_override = 1;
+#endif
+						disable_timer_pin_1 = 1;
+					}
+					return;
+				}
 				/* No multi-function device? */
 				type = read_pci_config_byte(num,slot,func,
 							    PCI_HEADER_TYPE);
@@ -1681,6 +1698,8 @@ static inline void check_timer(void)
 			if (nmi_watchdog == NMI_IO_APIC) {
 				setup_nmi();
 			}
+			if (disable_timer_pin_1)
+				clear_IO_APIC_pin(0, pin1);
 			return;
 		}
 		/*
