--- linux-2.6.8.1.orig/drivers/mtd/devices/Makefile	2004-08-14 12:54:46.000000000 +0200
+++ linux-2.6.8.1/drivers/mtd/devices/Makefile	2004-09-25 22:46:42.868745888 +0200
@@ -10,10 +10,10 @@
 # here where previously there was none.  We now have to ensure that
 # doc200[01].o are linked before docprobe.o
 
-obj-$(CONFIG_MTD_DOC2000)	+= doc2000.o
-obj-$(CONFIG_MTD_DOC2001)	+= doc2001.o
-obj-$(CONFIG_MTD_DOC2001PLUS)	+= doc2001plus.o
-obj-$(CONFIG_MTD_DOCPROBE)	+= docprobe.o
+obj-$(CONFIG_MTD_DOC2000)	+= doc2000.o docecc.o
+obj-$(CONFIG_MTD_DOC2001)	+= doc2001.o docecc.o
+obj-$(CONFIG_MTD_DOC2001PLUS)	+= doc2001plus.o docecc.o
+obj-$(CONFIG_MTD_DOCPROBE)	+= docprobe.o docecc.o
 obj-$(CONFIG_MTD_DOCECC)	+= docecc.o
 obj-$(CONFIG_MTD_SLRAM)		+= slram.o
 obj-$(CONFIG_MTD_PHRAM)		+= phram.o
