--- /home/baggins/kernel/linux-2.4.21/drivers/md/dm-bbr.c~	Mon Jun 16 19:37:28 2003
+++ /home/baggins/kernel/linux-2.4.21/drivers/md/dm-bbr.c	Mon Jun 16 22:08:20 2003
@@ -37,6 +37,8 @@
 #include "dm-bbr.h"
 #include "syncio.h"
 
+#define O1_SCHEDULER
+
 static struct sync_io_handle *syncio_handle = NULL;
 static char *no_mem_msg = "could not allocate memory";
 #define BBR_NO_MEM() DMERR("%s: %s", __FUNCTION__, no_mem_msg)
