--- linux-2.6.14.6/drivers/media/video/cx88/cx88-video.c.old	2006-01-08 03:15:24.000000000 +0100
+++ linux-2.6.14.6/drivers/media/video/cx88/cx88-video.c	2006-01-30 17:17:05.000000000 +0100
@@ -1256,11 +1256,10 @@
 {
 	int err;
 
-	if (video_debug > 1)
+	if (video_debug > 1) {
 		cx88_print_ioctl(core->name,cmd);
-	printk( KERN_INFO "CORE IOCTL: 0x%x\n", cmd );
-	cx88_print_ioctl(core->name,cmd);
-	dprintk( 1, "CORE IOCTL: 0x%x\n", cmd );
+		dprintk( 1, "CORE IOCTL: 0x%x\n", cmd );
+	}
 
 	switch (cmd) {
 	/* ---------- tv norms ---------- */
