--- linux-2.4.20/fs/intermezzo/vfs.c.orig	Fri Nov 29 00:53:15 2002
+++ linux-2.4.20/fs/intermezzo/vfs.c	Fri Apr 18 22:16:29 2003
@@ -467,7 +467,7 @@
         struct presto_file_set *fset;
         int error;
 #ifdef  CONFIG_FS_POSIX_ACL
-        int (*set_posix_acl)(struct inode *, int type, posix_acl_t *)=NULL;
+        int (*set_posix_acl)(struct inode *, int type, struct posix_acl *)=NULL;
 #endif
 
         ENTRY;
