Only in linux-2.4.10-pre11-fix/fs/udf: .balloc.c.swp
diff -ru -X /home/jack/.kerndiffexclude linux-2.4.10-pre11/fs/udf/ialloc.c linux-2.4.10-pre11-fix/fs/udf/ialloc.c
--- linux-2.4.10-pre11/fs/udf/ialloc.c	Sat Sep 22 17:28:51 2001
+++ linux-2.4.10-pre11-fix/fs/udf/ialloc.c	Sat Sep 22 17:31:20 2001
@@ -155,7 +155,8 @@
 	unlock_super(sb);
 	if (DQUOT_ALLOC_INODE(sb, inode))
 	{
-		sb->dq_op->drop(inode);
+		DQUOT_DROP(inode);
+		inode->i_flags |= S_NOQUOTA;
 		inode->i_nlink = 0;
 		iput(inode);
 		*err = -EDQUOT;
diff -ru -X /home/jack/.kerndiffexclude linux-2.4.10-pre11/fs/ufs/ialloc.c linux-2.4.10-pre11-fix/fs/ufs/ialloc.c
--- linux-2.4.10-pre11/fs/ufs/ialloc.c	Sat Sep 22 17:41:36 2001
+++ linux-2.4.10-pre11-fix/fs/ufs/ialloc.c	Sat Sep 22 17:42:10 2001
@@ -279,7 +279,8 @@
 	unlock_super (sb);
 
 	if(DQUOT_ALLOC_INODE(sb, inode)) {
-		sb->dq_op->drop(inode);
+		DQUOT_DROP(inode);
+		inode->i_flags |= S_NOQUOTA;
 		inode->i_nlink = 0;
 		iput(inode);
 		*err = -EDQUOT;
@@ -293,6 +294,7 @@
 
 failed:
 	unlock_super (sb);
+	make_bad_inode(inode);
 	iput (inode);
 	UFSD(("EXIT (FAILED)\n"))
 	return NULL;
