--- linux/fs/quota_v1.c.org	Fri Apr 26 11:33:20 2002
+++ linux/fs/quota_v1.c	Fri Apr 26 11:58:33 2002
@@ -234,6 +234,10 @@
 
 EXPORT_NO_SYMBOLS;
 
+MODULE_AUTHOR("");
+MODULE_DESCRIPTION("QUOTA Version 1 support.");
+MODULE_LICENSE("GPL");
+
 module_init(init_v1_quota_format);
 module_exit(exit_v1_quota_format);
 
--- linux/fs/quota_v2.c.org	Fri Apr 26 11:33:20 2002
+++ linux/fs/quota_v2.c	Fri Apr 26 11:58:20 2002
@@ -686,5 +686,9 @@
 
 EXPORT_NO_SYMBOLS;
 
+MODULE_AUTHOR("");
+MODULE_DESCRIPTION("QUOTA Version 2 support.");
+MODULE_LICENSE("GPL");
+
 module_init(init_v2_quota_format);
 module_exit(exit_v2_quota_format);
