--- linux-2.4.22/fs/ext3/super.c~	Mon Sep 22 20:30:44 2003
+++ linux-2.4.22/fs/ext3/super.c	Mon Sep 22 22:14:47 2003
@@ -1868,12 +1868,12 @@
 
 static int __init init_ext3_fs(void)
 {
+	int error = init_ext3_xattr();
 #ifdef CONFIG_QUOTA
 	init_dquot_operations(&ext3_qops);
 	old_sync_dquot = ext3_qops.sync_dquot;
 	ext3_qops.sync_dquot = ext3_sync_dquot;
 #endif
-	int error = init_ext3_xattr();
 	if (error)
 		return error;
 	error = init_ext3_xattr_user();
