--- linux-2.6.3/include/asm-alpha/io.h.orig	2004-02-18 04:59:27.000000000 +0100
+++ linux-2.6.3/include/asm-alpha/io.h	2004-02-25 13:42:10.000000000 +0100
@@ -131,6 +131,15 @@
 #endif /* !__KERNEL__ */
 
 /*
+ * taken from asm-i386
+ * ("ISA I/O bus memory addresses are 1:1 with the physical address")
+ * is it correct?
+ */
+#define isa_virt_to_bus virt_to_phys
+#define isa_page_to_bus page_to_phys
+#define isa_bus_to_virt phys_to_virt
+
+/*
  * There are different chipsets to interface the Alpha CPUs to the world.
  */
 
