--- linux-2.6.11.6/sound/oss/ac97_codec.c.orig	2005-03-26 04:28:15.000000000 +0100
+++ linux-2.6.11.6/sound/oss/ac97_codec.c	2005-04-05 15:01:16.000000000 +0200
@@ -153,6 +153,7 @@
 	{0x43525931, "Cirrus Logic CS4299 rev A", &crystal_digital_ops},
 	{0x43525933, "Cirrus Logic CS4299 rev C", &crystal_digital_ops},
 	{0x43525934, "Cirrus Logic CS4299 rev D", &crystal_digital_ops},
+	{0x43585430, "CXT48",			&default_ops,		AC97_DELUDED_MODEM },
 	{0x43585442, "CXT66",			&default_ops,		AC97_DELUDED_MODEM },
 	{0x44543031, "Diamond Technology DT0893", &default_ops},
 	{0x45838308, "ESS Allegro ES1988",	&null_ops},
