diff -u --recursive --new-file v2.4.13/linux/drivers/usb/usb-ohci.h linux/drivers/usb/usb-ohci.h
--- v2.4.13/linux/drivers/usb/usb-ohci.h	Mon Aug 27 12:41:45 2001
+++ linux/drivers/usb/usb-ohci.h	Wed Oct 24 08:26:12 2001
@@ -373,7 +373,7 @@
 
 typedef struct ohci {
 	struct ohci_hcca *hcca;		/* hcca */
-	u32 hcca_dma;
+	dma_addr_t hcca_dma;
 
 	int irq;
 	int disabled;			/* e.g. got a UE, we're hung */
diff -u --recursive --new-file v2.4.13/linux/drivers/usb/usb-uhci-debug.h linux/drivers/usb/usb-uhci-debug.h
--- v2.4.13/linux/drivers/usb/usb-uhci-debug.h	Wed May 16 10:31:27 2001
+++ linux/drivers/usb/usb-uhci-debug.h	Wed Oct 24 08:26:51 2001
@@ -5,7 +5,7 @@
 		dbg("qh has not QH_TYPE");
 		return;
 	}
-	dbg("QH @ %p/%08X:", qh, qh->dma_addr);
+	dbg("QH @ %p/%08llX:", qh, (u64)qh->dma_addr);
 
 	if (qh->hw.qh.head & UHCI_PTR_TERM)
 		dbg("    Head Terminate");
