--- linux-2.6.14.3/arch/sparc/kernel/sparc_ksyms.c.orig	2004-12-24 22:33:52.000000000 +0100
+++ linux-2.6.14.3/arch/sparc/kernel/sparc_ksyms.c	2005-01-09 08:37:50.199948280 +0100
@@ -216,9 +216,10 @@
 EXPORT_SYMBOL(pci_map_page);
 EXPORT_SYMBOL(pci_unmap_page);
 /* Actually, ioremap/iounmap are not PCI specific. But it is ok for drivers. */
+#endif
+/* or not? not all MTD drivers depend on PCI */
 EXPORT_SYMBOL(ioremap);
 EXPORT_SYMBOL(iounmap);
-#endif
 
 /* in arch/sparc/mm/highmem.c */
 #ifdef CONFIG_HIGHMEM
@@ -255,6 +256,9 @@
 EXPORT_SYMBOL(prom_getint);
 EXPORT_SYMBOL(prom_getintdefault);
 EXPORT_SYMBOL(prom_finddevice);
+EXPORT_SYMBOL(prom_halt);
+EXPORT_SYMBOL(prom_printf);
+EXPORT_SYMBOL(prom_apply_generic_ranges);
 EXPORT_SYMBOL(romvec);
 EXPORT_SYMBOL(__prom_getchild);
 EXPORT_SYMBOL(__prom_getsibling);
@@ -267,6 +271,8 @@
 EXPORT_SYMBOL(memscan);
 EXPORT_SYMBOL(strlen);
 EXPORT_SYMBOL(strnlen);
+EXPORT_SYMBOL(__strlen_user);
+EXPORT_SYMBOL(__strnlen_user);
 EXPORT_SYMBOL(strcpy);
 EXPORT_SYMBOL(strncpy);
 EXPORT_SYMBOL(strcat);
--- linux-2.6.14.3/arch/sparc/kernel/process.c.orig	2005-11-24 22:10:21.000000000 +0000
+++ linux-2.6.14.3/arch/sparc/kernel/process.c	2005-12-11 10:22:00.217937341 +0000
@@ -55,6 +55,7 @@
  * handler when auxio is not present-- unused for now...
  */
 void (*pm_power_off)(void);
+EXPORT_SYMBOL(pm_power_off);
 
 /*
  * sysctl - toggle power-off restriction for serial console 
