--- linux-2.5.73/include/asm-alpha/pci.h	2003-06-22 20:32:28.000000000 +0200
+++ linux-2.5.73.c/include/asm-alpha/pci.h	2003-06-23 23:25:55.000000000 +0200
@@ -197,7 +197,7 @@
 /* Bus number == domain number until we get above 256 busses */
 static inline int pci_name_bus(char *name, struct pci_bus *bus)
 {
-	int domain = pci_domain_nr(bus)
+	int domain = pci_domain_nr(bus);
 	if (domain < 256) {
 		sprintf(name, "%02x", domain);
 	} else {
