--- drivers/ide/hpt366.c.org	Sun Jul 14 22:02:38 2002
+++ drivers/ide/hpt366.c	Sun Jul 14 22:14:22 2002
@@ -664,7 +664,6 @@
 	byte ultra66		= eighty_ninty_three(drive);
 	int  rval;
 
-	config_chipset_for_pio(drive);
 		
 	if ((drive->media != ide_disk) && (speed < XFER_SW_DMA_0))
 		return ((int) ide_dma_off_quietly);
@@ -791,7 +791,6 @@
 		dma_func = ide_dma_off_quietly;
 no_dma_set:
 
-		config_chipset_for_pio(drive);
 	}
 	return HWIF(drive)->dmaproc(dma_func, drive);
 }
