--- linux-2.4.20.orig/net/core/filter.c	2001-05-16 19:32:59.000000000 +0200
+++ linux-2.4.20/net/core/filter.c	2003-07-31 16:50:00.000000000 +0200
@@ -294,10 +294,9 @@ load_b:
 				goto load_b;
 
 			case BPF_LDX|BPF_B|BPF_MSH:
-				k = fentry->k;
-				if(k >= 0 && (unsigned int)k >= len)
+				if(fentry->k >= len)
 					return (0);
-				X = (data[k] & 0xf) << 2;
+				X = (data[fentry->k] & 0xf) << 2;
 				continue;
 
 			case BPF_LD|BPF_IMM:
