--- linux-2.6.19/drivers/net/forcedeth.c.orig   2006-09-20 05:42:06.000000000 +0200
+++ linux-2.6.19/drivers/net/forcedeth.c        2006-11-25 14:19:31.000000000 +0100
@@ -166,7 +166,8 @@
  * Hardware access:
  */
 
-#define DEV_NEED_TIMERIRQ          0x000001  /* set the timer irq flag in the irq mask */
+#define DEV_NEED_TIMERIRQ_ORIG     0x000001  /* set the timer irq flag in the irq mask */
+#define DEV_NEED_TIMERIRQ          0x000000  /* work-around for Wake-On-Lan functionality */
 #define DEV_NEED_LINKTIMER         0x000002  /* poll link settings. Relies on the timer irq */
 #define DEV_HAS_LARGEDESC          0x000004  /* device supports jumbo frames and needs packet format 2 */
 #define DEV_HAS_HIGH_DMA           0x000008  /* device supports 64bit dma */
@@ -4342,7 +4343,7 @@
 			np->msi_flags |= 0x0001;
 	}
 
-	if (id->driver_data & DEV_NEED_TIMERIRQ)
+	if (id->driver_data & DEV_NEED_TIMERIRQ_ORIG)
 		np->irqmask |= NVREG_IRQ_TIMER;
 	if (id->driver_data & DEV_NEED_LINKTIMER) {
 		dprintk(KERN_INFO "%s: link timer on.\n", pci_name(pci_dev));
@@ -4562,7 +4563,7 @@
 MODULE_PARM_DESC(dma_64bit, "High DMA is enabled by setting to 1 and disabled by setting to 0.");
 
 MODULE_AUTHOR("Manfred Spraul <manfred@colorfullife.com>");
-MODULE_DESCRIPTION("Reverse Engineered nForce ethernet driver");
+MODULE_DESCRIPTION("Reverse Engineered nForce ethernet driver (TIMERIRQ DISABLED)");
 MODULE_LICENSE("GPL");
 
 MODULE_DEVICE_TABLE(pci, pci_tbl);
