--- linux-wlan-ng-0.2.1pre21/src/Makefile.orig	2004-02-13 19:42:39.000000000 +0100
+++ linux-wlan-ng-0.2.1pre21/src/Makefile	2004-05-31 09:53:20.232175216 +0200
@@ -44,9 +44,9 @@
 
 include ../config.mk
 
-DIRS=p80211 prism2 shared wlanctl wland nwepgen \
+DIRS=prism2 shared wlanctl wland nwepgen \
 	mkmeta wlancfg
-ALL_DIRS=mkmeta p80211 prism2 shared wlanctl wland \
+ALL_DIRS=mkmeta prism2 shared wlanctl wland \
 	nwepgen wlancfg
 
 ifneq ($(wildcard *.addon),) 
--- linux-wlan-ng-0.2.1pre21/src/prism2/Makefile.orig	2004-02-13 19:47:37.000000000 +0100
+++ linux-wlan-ng-0.2.1pre21/src/prism2/Makefile	2004-05-31 09:53:30.815566296 +0200
@@ -44,7 +44,7 @@
 
 include ../../config.mk
 
-DIRS=driver ridlist download
+DIRS=ridlist download
 
 ifneq ($(wildcard *.addon),) 
 DIRS+=`cat *.addon`
--- linux-wlan-ng-0.2.1-pre26/man/Makefile.orig	2004-08-04 23:09:28.000000000 +0200
+++ linux-wlan-ng-0.2.1-pre26/man/Makefile	2005-07-18 21:20:41.988581896 +0200
@@ -48,8 +48,8 @@
 	@echo "Nothing to do..."
 
 install:
-	mkdir -p $(TARGET_ROOT_ON_HOST)/usr/local/man/man1
-	cp *.1 	$(TARGET_ROOT_ON_HOST)/usr/local/man/man1
+	mkdir -p $(TARGET_ROOT_ON_HOST)/usr/share/man/man1
+	cp *.1 	$(TARGET_ROOT_ON_HOST)/usr/share/man/man1
 
 clean:
 	rm -f core core.* *.o .*.o *.s *.a .depend tmp_make *~ tags
