--- linux-2.4.20-orig/drivers/ide/ide-cd.h	2002-12-10 17:46:28 +0900
+++ linux-2.4.20/drivers/ide/ide-cd.h	2003-05-16 00:59:53 +0900
@@ -437,7 +437,7 @@
 
 	byte     curlba[3];
 	byte     nslots;
-	__u8 short slot_tablelen;
+	__u8     slot_tablelen;
 };
 
 
--- linux-2.4.20-orig/net/core/rtnetlink.c	2002-12-09 16:38:46 +0900
+++ linux-2.4.20/net/core/rtnetlink.c	2003-05-16 01:31:20 +0900
@@ -394,7 +394,7 @@
  * Malformed skbs with wrong lengths of messages are discarded silently.
  */
 
-extern __inline__ int rtnetlink_rcv_skb(struct sk_buff *skb)
+__inline__ int rtnetlink_rcv_skb(struct sk_buff *skb)
 {
 	int err;
 	struct nlmsghdr * nlh;
--- linux-2.4.20/drivers/atm/ambassador.c.orig	Thu Sep 13 22:21:32 2001
+++ linux-2.4.20/drivers/atm/ambassador.c	Wed Jun 11 08:42:37 2003
@@ -290,12 +290,12 @@
 /********** microcode **********/
 
 #ifdef AMB_NEW_MICROCODE
-#define UCODE(x) UCODE1(atmsar12.,x)
+#define UCODE(x) UCODE1(atmsar12,x)
 #else
-#define UCODE(x) UCODE1(atmsar11.,x)
+#define UCODE(x) UCODE1(atmsar11,x)
 #endif
 #define UCODE2(x) #x
-#define UCODE1(x,y) UCODE2(x ## y)
+#define UCODE1(x,y) UCODE2(x.y)
 
 static u32 __initdata ucode_start = 
 #include UCODE(start)
--- linux-2.4.20/include/linux/mtd/nftl.h.org	Fri Jun  6 15:46:54 2003
+++ linux-2.4.20/include/linux/mtd/nftl.h	Wed Jun 11 13:49:12 2003
@@ -97,7 +97,7 @@
 	__u16 lastEUN;                  /* should be suppressed */
 	__u16 numfreeEUNs;
 	__u16 LastFreeEUN; 		/* To speed up finding a free EUN */
-	__u32 long nr_sects;
+	__u32 nr_sects;
 	int head,sect,cyl;
 	__u16 *EUNtable; 		/* [numvunits]: First EUN for each virtual unit  */
 	__u16 *ReplUnitTable; 		/* [numEUNs]: ReplUnitNumber for each */
diff -Nur --exclude='*.orig' --exclude='*.o' linux-2.4.20.old/drivers/net/fealnx.c linux-2.4.20/drivers/net/fealnx.c
--- linux-2.4.20.old/drivers/net/fealnx.c	Mon Jun 16 13:16:08 2003
+++ linux-2.4.20/drivers/net/fealnx.c	Mon Jun 16 13:16:09 2003
@@ -1198,7 +1198,7 @@
 		printk("\n");
 	}
 
-	   + dev->if_port = np->default_port;
+	dev->if_port = np->default_port;
 	/* Reinit. Gross */
 
 	/* Reset the chip's Tx and Rx processes. */
diff -Nur --exclude='*.orig' --exclude='*.o' linux-2.4.20.old/drivers/net/irda/ma600.c linux-2.4.20/drivers/net/irda/ma600.c
--- linux-2.4.20.old/drivers/net/irda/ma600.c	Mon Jun 16 13:16:08 2003
+++ linux-2.4.20/drivers/net/irda/ma600.c	Mon Jun 16 13:16:09 2003
@@ -47,9 +47,9 @@
 	#undef ASSERT(expr, func)
 	#define ASSERT(expr, func) \
 	if(!(expr)) { \
-	        printk( "Assertion failed! %s,%s,%s,line=%d\n",\
-        	#expr,__FILE__,__FUNCTION__,__LINE__); \
-	        ##func}
+	        printk( "Assertion failed! %s,%s,line=%d\n",\
+        	__FILE__,__FUNCTION__,__LINE__); \
+		}
 #endif
 
 /* convert hex value to ascii hex */
diff -Nur --exclude='*.orig' --exclude='*.o' linux-2.4.20.old/drivers/video/controlfb.c linux-2.4.20/drivers/video/controlfb.c
--- linux-2.4.20.old/drivers/video/controlfb.c	Mon Jun 16 13:16:08 2003
+++ linux-2.4.20/drivers/video/controlfb.c	Mon Jun 16 13:16:09 2003
@@ -132,7 +132,7 @@
 };
 
 /* control register access macro */
-#define CNTRL_REG(INFO,REG) (&(((INFO)->control_regs-> ## REG).r))
+#define CNTRL_REG(INFO,REG) (&(((INFO)->control_regs-> REG).r))
 
 
 /******************** Prototypes for exported functions ********************/
diff -Nur --exclude='*.orig' --exclude='*.o' linux-2.4.20.old/include/asm-ppc/unistd.h linux-2.4.20/include/asm-ppc/unistd.h
--- linux-2.4.20.old/include/asm-ppc/unistd.h	Mon Jun 16 13:16:08 2003
+++ linux-2.4.20/include/asm-ppc/unistd.h	Mon Jun 16 13:16:09 2003
@@ -250,7 +250,7 @@
 	       (type) __sc_ret
 
 #define __syscall_clobbers \
-	"r4", "r5", "r6", "r7", "r8", "r9", "r10", "r11", "r12"
+	"r6", "r7", "r8", "r9", "r10", "r11", "r12"
 
 #define _syscall0(type,name)						\
 type name(void)								\
diff -Nur --exclude='*.orig' --exclude='*.o' linux-2.4.20.old/drivers/scsi/aic7xxx/aic79xx_osm.c linux-2.4.20/drivers/scsi/aic7xxx/aic79xx_osm.c
--- linux-2.4.20.old/drivers/scsi/aic7xxx/aic79xx_osm.c	Mon Jun 16 13:16:08 2003
+++ linux-2.4.20/drivers/scsi/aic7xxx/aic79xx_osm.c	Mon Jun 16 13:26:12 2003
@@ -336,31 +336,31 @@
 MODULE_LICENSE("Dual BSD/GPL");
 #endif
 MODULE_PARM(aic79xx, "s");
-MODULE_PARM_DESC(aic79xx, "period delimited, options string.
-	verbose			Enable verbose/diagnostic logging
-	debug			Bitmask of debug values to enable
-	no_reset		Supress initial bus resets
-	extended		Enable extended geometry on all controllers
-	periodic_otag		Send an ordered tagged transaction periodically
-				to prevent tag starvation.  This may be
-				required by some older disk drives/RAID arrays. 
-	reverse_scan		Sort PCI devices highest Bus/Slot to lowest
-	tag_info:<tag_str>	Set per-target tag depth
-	rd_strm:<rd_strm_masks> Set per-target read streaming setting.
-	seltime:<int>		Selection Timeout(0/256ms,1/128ms,2/64ms,3/32ms)
-
-	Sample /etc/modules.conf line:
-		Enable verbose logging
-		Set tag depth on Controller 2/Target 2 to 10 tags
-		Shorten the selection timeout to 128ms from its default of 256
-
-	options aic79xx='\"verbose.tag_info:{{}.{}.{..10}}.seltime:1\"'
-
-	Sample /etc/modules.conf line:
-		Change Read Streaming for Controller's 2 and 3
-
-	options aic79xx='\"rd_strm:{..0xFFF0.0xC0F0}\"'
-");
+MODULE_PARM_DESC(aic79xx, "period delimited, options string."\
+"	verbose			Enable verbose/diagnostic logging"\
+"	debug			Bitmask of debug values to enable"\
+"	no_reset		Supress initial bus resets"\
+"	extended		Enable extended geometry on all controllers"\
+"	periodic_otag		Send an ordered tagged transaction periodically"\
+"				to prevent tag starvation.  This may be"\
+"				required by some older disk drives/RAID arrays. "\
+"	reverse_scan		Sort PCI devices highest Bus/Slot to lowest"\
+"	tag_info:<tag_str>	Set per-target tag depth"\
+"	rd_strm:<rd_strm_masks> Set per-target read streaming setting."\
+"	seltime:<int>		Selection Timeout(0/256ms,1/128ms,2/64ms,3/32ms)"\
+""\
+"	Sample /etc/modules.conf line:"\
+"		Enable verbose logging"\
+"		Set tag depth on Controller 2/Target 2 to 10 tags"\
+"		Shorten the selection timeout to 128ms from its default of 256"\
+""\
+"	options aic79xx='\"verbose.tag_info:{{}.{}.{..10}}.seltime:1\"'"\
+""\
+"	Sample /etc/modules.conf line:"\
+"		Change Read Streaming for Controller's 2 and 3"\
+""\
+"	options aic79xx='\"rd_strm:{..0xFFF0.0xC0F0}\"'"
+);
 #endif
 
 static void ahd_linux_handle_scsi_status(struct ahd_softc *,
diff -Nur --exclude='*.orig' --exclude='*.o' linux-2.4.20.old/drivers/scsi/aic7xxx/aic7xxx_osm.c linux-2.4.20/drivers/scsi/aic7xxx/aic7xxx_osm.c
--- linux-2.4.20.old/drivers/scsi/aic7xxx/aic7xxx_osm.c	Mon Jun 16 13:16:08 2003
+++ linux-2.4.20/drivers/scsi/aic7xxx/aic7xxx_osm.c	Mon Jun 16 13:16:09 2003
@@ -417,26 +417,26 @@
 MODULE_LICENSE("Dual BSD/GPL");
 #endif
 MODULE_PARM(aic7xxx, "s");
-MODULE_PARM_DESC(aic7xxx, "period delimited, options string.
-	verbose			Enable verbose/diagnostic logging
-	no_probe		Disable EISA/VLB controller probing
-	no_reset		Supress initial bus resets
-	extended		Enable extended geometry on all controllers
-	periodic_otag		Send an ordered tagged transaction periodically
-				to prevent tag starvation.  This may be
-				required by some older disk drives/RAID arrays. 
-	reverse_scan		Sort PCI devices highest Bus/Slot to lowest
-	tag_info:<tag_str>	Set per-target tag depth
-	seltime:<int>		Selection Timeout(0/256ms,1/128ms,2/64ms,3/32ms)
-
-	Sample /etc/modules.conf line:
-		Enable verbose logging
-		Disable EISA/VLB probing
-		Set tag depth on Controller 2/Target 2 to 10 tags
-		Shorten the selection timeout to 128ms from its default of 256
-
-	options aic7xxx='\"verbose.no_probe.tag_info:{{}.{}.{..10}}.seltime:1\"'
-");
+MODULE_PARM_DESC(aic7xxx, "period delimited, options string." \
+"	verbose			Enable verbose/diagnostic logging" \
+"	no_probe		Disable EISA/VLB controller probing" \
+"	no_reset		Supress initial bus resets" \
+"	extended		Enable extended geometry on all controllers" \
+"	periodic_otag		Send an ordered tagged transaction periodically" \
+"				to prevent tag starvation.  This may be" \
+"				required by some older disk drives/RAID arrays." \
+"	reverse_scan		Sort PCI devices highest Bus/Slot to lowest" \
+"	tag_info:<tag_str>	Set per-target tag depth" \
+"	seltime:<int>		Selection Timeout(0/256ms,1/128ms,2/64ms,3/32ms)" \
+" " \
+"	Sample /etc/modules.conf line:" \
+"		Enable verbose logging" \
+"		Disable EISA/VLB probing" \
+"		Set tag depth on Controller 2/Target 2 to 10 tags" \
+"		Shorten the selection timeout to 128ms from its default of 256" \
+" " \
+"	options aic7xxx='\"verbose.no_probe.tag_info:{{}.{}.{..10}}.seltime:1\"'"
+);
 #endif
 
 static void ahc_linux_handle_scsi_status(struct ahc_softc *,
diff -Nur --exclude='*.orig' linux-2.4.20.org/drivers/sound/cs46xx.c linux-2.4.20/drivers/sound/cs46xx.c
--- linux-2.4.20.org/drivers/sound/cs46xx.c	Sat Aug  3 00:39:44 2002
+++ linux-2.4.20/drivers/sound/cs46xx.c	Tue Jun 17 06:25:18 2003
@@ -947,8 +947,8 @@
 
 struct InitStruct
 {
-    u32 long off;
-    u32 long val;
+    u32 off;
+    u32 val;
 } InitArray[] = { {0x00000040, 0x3fc0000f},
                   {0x0000004c, 0x04800000},
 
diff -Nur --exclude='*.orig' linux-2.4.20.org/fs/reiserfs/super.c linux-2.4.20/fs/reiserfs/super.c
--- linux-2.4.20.org/fs/reiserfs/super.c	Tue Jun 17 07:17:19 2003
+++ linux-2.4.20/fs/reiserfs/super.c	Tue Jun 17 06:44:51 2003
@@ -944,8 +944,7 @@
     ll_rw_block(READ, 1, &(SB_AP_BITMAP(s)[i].bh)) ;
     wait_on_buffer(SB_AP_BITMAP(s)[i].bh) ;
     if (!buffer_uptodate(SB_AP_BITMAP(s)[i].bh)) {
-      printk("reread_meta_blocks, error reading bitmap block number %d at
-      %ld\n", i, SB_AP_BITMAP(s)[i].bh->b_blocknr) ;
+      printk("reread_meta_blocks, error reading bitmap block number %d at %ld\n", i, SB_AP_BITMAP(s)[i].bh->b_blocknr) ;
       return 1 ;
     }
   }
diff -Nur --exclude='*.orig' linux-2.4.20.org/net/ipv4/netfilter/ip_nat_core.c linux-2.4.20/net/ipv4/netfilter/ip_nat_core.c
--- linux-2.4.20.org/net/ipv4/netfilter/ip_nat_core.c	Tue Jun 17 07:17:18 2003
+++ linux-2.4.20/net/ipv4/netfilter/ip_nat_core.c	Tue Jun 17 07:05:01 2003
@@ -851,8 +851,7 @@
 			if (exp_for_packet(exp, pskb)) {
 				/* FIXME: May be true multiple times in the
 				 * case of UDP!! */
-				DEBUGP("calling nat helper (exp=%p) for
-					packet\n", exp);
+				DEBUGP("calling nat helper (exp=%p) for packet\n", exp);
 				ret = helper->help(ct, exp, info, ctinfo, 
 						   hooknum, pskb);
 				if (ret != NF_ACCEPT) {
diff -Nur --exclude='*.orig' --exclude='*.org' linux-2.4.20.org/drivers/net/tokenring/olympic.c linux-2.4.20/drivers/net/tokenring/olympic.c
--- linux-2.4.20.org/drivers/net/tokenring/olympic.c	2002-11-28 23:53:14.000000000 +0000
+++ linux-2.4.20/drivers/net/tokenring/olympic.c	2003-06-20 12:23:35.000000000 +0000
@@ -655,8 +655,8 @@
 	printk(" stat_ring[7]: %p\n", &(olympic_priv->olympic_rx_status_ring[7])  );
 
 	printk("RXCDA: %x, rx_ring[0]: %p\n",readl(olympic_mmio+RXCDA),&olympic_priv->olympic_rx_ring[0]);
-	printk("Rx_ring_dma_addr = %08x, rx_status_dma_addr =
-%08x\n",olympic_priv->rx_ring_dma_addr,olympic_priv->rx_status_ring_dma_addr) ; 
+	printk("Rx_ring_dma_addr = %08x, rx_status_dma_addr =%08x\n",
+		olympic_priv->rx_ring_dma_addr,olympic_priv->rx_status_ring_dma_addr) ; 
 #endif
 
 	writew((((readw(olympic_mmio+RXENQ)) & 0x8000) ^ 0x8000) | i,olympic_mmio+RXENQ);
diff -Nur --exclude='*.orig' --exclude='*.org' linux-2.4.20.org/drivers/net/wan/sbni.c linux-2.4.20/drivers/net/wan/sbni.c
--- linux-2.4.20.org/drivers/net/wan/sbni.c	2002-11-28 23:53:14.000000000 +0000
+++ linux-2.4.20/drivers/net/wan/sbni.c	2003-06-20 12:47:13.000000000 +0000
@@ -1547,88 +1547,6 @@
 
 /* -------------------------------------------------------------------------- */
 
-#ifdef ASM_CRC
-
-static u32
-calc_crc32( u32  crc,  u8  *p,  u32  len )
-{
-	register u32  _crc __asm ( "ax" );
-	_crc = crc;
-	
-	__asm __volatile (
-		"xorl	%%ebx, %%ebx\n"
-		"movl	%1, %%esi\n" 
-		"movl	%2, %%ecx\n" 
-		"movl	$crc32tab, %%edi\n"
-		"shrl	$2, %%ecx\n"
-		"jz	1f\n"
-
-		".align 4\n"
-	"0:\n"
-		"movb	%%al, %%bl\n"
-		"movl	(%%esi), %%edx\n"
-		"shrl	$8, %%eax\n"
-		"xorb	%%dl, %%bl\n"
-		"shrl	$8, %%edx\n"
-		"xorl	(%%edi,%%ebx,4), %%eax\n"
-
-		"movb	%%al, %%bl\n"
-		"shrl	$8, %%eax\n"
-		"xorb	%%dl, %%bl\n"
-		"shrl	$8, %%edx\n"
-		"xorl	(%%edi,%%ebx,4), %%eax\n"
-
-		"movb	%%al, %%bl\n"
-		"shrl	$8, %%eax\n"
-		"xorb	%%dl, %%bl\n"
-		"movb	%%dh, %%dl\n" 
-		"xorl	(%%edi,%%ebx,4), %%eax\n"
-
-		"movb	%%al, %%bl\n"
-		"shrl	$8, %%eax\n"
-		"xorb	%%dl, %%bl\n"
-		"addl	$4, %%esi\n"
-		"xorl	(%%edi,%%ebx,4), %%eax\n"
-
-		"decl	%%ecx\n"
-		"jnz	0b\n"
-
-	"1:\n"
-		"movl	%2, %%ecx\n"
-		"andl	$3, %%ecx\n"
-		"jz	2f\n"
-
-		"movb	%%al, %%bl\n"
-		"shrl	$8, %%eax\n"
-		"xorb	(%%esi), %%bl\n"
-		"xorl	(%%edi,%%ebx,4), %%eax\n"
-
-		"decl	%%ecx\n"
-		"jz	2f\n"
-
-		"movb	%%al, %%bl\n"
-		"shrl	$8, %%eax\n"
-		"xorb	1(%%esi), %%bl\n"
-		"xorl	(%%edi,%%ebx,4), %%eax\n"
-
-		"decl	%%ecx\n"
-		"jz	2f\n"
-
-		"movb	%%al, %%bl\n"
-		"shrl	$8, %%eax\n"
-		"xorb	2(%%esi), %%bl\n"
-		"xorl	(%%edi,%%ebx,4), %%eax\n"
-	"2:\n"
-		:
-		: "a" (_crc), "g" (p), "g" (len)
-		: "ax", "bx", "cx", "dx", "si", "di"
-	);
-
-	return  _crc;
-}
-
-#else	/* ASM_CRC */
-
 static u32
 calc_crc32( u32  crc,  u8  *p,  u32  len )
 {
@@ -1638,8 +1556,6 @@
 	return  crc;
 }
 
-#endif	/* ASM_CRC */
-
 
 static u32  crc32tab[] __attribute__ ((aligned(8))) = {
 	0xD202EF8D,  0xA505DF1B,  0x3C0C8EA1,  0x4B0BBE37,
diff -Nur --exclude='*.orig' --exclude='*.org' linux-2.4.20.org/drivers/net/wan/sdla_chdlc.c linux-2.4.20/drivers/net/wan/sdla_chdlc.c
--- linux-2.4.20.org/drivers/net/wan/sdla_chdlc.c	2003-06-20 12:41:46.000000000 +0000
+++ linux-2.4.20/drivers/net/wan/sdla_chdlc.c	2003-06-20 12:41:03.000000000 +0000
@@ -591,8 +591,7 @@
 	
 
 		if (chdlc_set_intr_mode(card, APP_INT_ON_TIMER)){
-			printk (KERN_INFO "%s: 
-				Failed to set interrupt triggers!\n",
+			printk (KERN_INFO "%s: Failed to set interrupt triggers!\n",
 				card->devname);
 			return -EIO;	
         	}
--- linux-2.4.20/net/decnet/dn_table.c.org	Fri Dec 21 17:42:05 2001
+++ linux-2.4.20/net/decnet/dn_table.c	Mon Jun 23 13:23:27 2003
@@ -836,8 +836,8 @@
                 return NULL;
 
         if (in_interrupt() && net_ratelimit()) {
-                printk(KERN_DEBUG "DECnet: BUG! Attempt to create routing table 
-from interrupt\n"); 
+                printk(KERN_DEBUG 
+		"DECnet: BUG! Attempt to create routing table from interrupt\n"); 
                 return NULL;
         }
         if ((t = kmalloc(sizeof(struct dn_fib_table), GFP_KERNEL)) == NULL)
--- linux-2.4.20/arch/i386/math-emu/poly.h~	Thu Jul 26 20:37:38 2001
+++ linux-2.4.20/arch/i386/math-emu/poly.h	Thu Jul 10 02:09:29 2003
@@ -75,10 +75,10 @@
 /* Add the 12 byte Xsig x2 to Xsig dest, with no checks for overflow. */
 static inline void add_Xsig_Xsig(Xsig *dest, const Xsig *x2)
 {
-  asm volatile ("movl %1,%%edi; movl %2,%%esi;
-                 movl (%%esi),%%eax; addl %%eax,(%%edi);
-                 movl 4(%%esi),%%eax; adcl %%eax,4(%%edi);
-                 movl 8(%%esi),%%eax; adcl %%eax,8(%%edi);"
+  asm volatile ("movl %1,%%edi; movl %2,%%esi; "
+                 "movl (%%esi),%%eax; addl %%eax,(%%edi); "
+                 "movl 4(%%esi),%%eax; adcl %%eax,4(%%edi); "
+                 "movl 8(%%esi),%%eax; adcl %%eax,8(%%edi); "
                  :"=g" (*dest):"g" (dest), "g" (x2)
                  :"ax","si","di");
 }
@@ -90,16 +90,16 @@
    problem, but keep fingers crossed! */
 static inline void add_two_Xsig(Xsig *dest, const Xsig *x2, long int *exp)
 {
-  asm volatile ("movl %2,%%ecx; movl %3,%%esi;
-                 movl (%%esi),%%eax; addl %%eax,(%%ecx);
-                 movl 4(%%esi),%%eax; adcl %%eax,4(%%ecx);
-                 movl 8(%%esi),%%eax; adcl %%eax,8(%%ecx);
-                 jnc 0f;
-		 rcrl 8(%%ecx); rcrl 4(%%ecx); rcrl (%%ecx)
-                 movl %4,%%ecx; incl (%%ecx)
-                 movl $1,%%eax; jmp 1f;
-                 0: xorl %%eax,%%eax;
-                 1:"
+  asm volatile ("movl %2,%%ecx; movl %3,%%esi; "
+                 "movl (%%esi),%%eax; addl %%eax,(%%ecx); "
+                 "movl 4(%%esi),%%eax; adcl %%eax,4(%%ecx); "
+                 "movl 8(%%esi),%%eax; adcl %%eax,8(%%ecx); "
+                 "jnc 0f; "
+		 "rcrl 8(%%ecx); rcrl 4(%%ecx); rcrl (%%ecx); "
+                 "movl %4,%%ecx; incl (%%ecx); "
+                 "movl $1,%%eax; jmp 1f; "
+                 "0: xorl %%eax,%%eax; "
+                 "1:"
 		:"=g" (*exp), "=g" (*dest)
 		:"g" (dest), "g" (x2), "g" (exp)
 		:"cx","si","ax");
