diff -urN linux-2.4.20.orig/drivers/char/agp/agpgart_be.c linux-2.4.20/drivers/char/agp/agpgart_be.c
--- linux-2.4.20.orig/drivers/char/agp/agpgart_be.c	Sat Jan 18 22:30:47 2003
+++ linux-2.4.20/drivers/char/agp/agpgart_be.c	Sat Jan 18 22:55:46 2003
@@ -4714,6 +4714,12 @@
 		"Via",
 		"Apollo Pro KT266",
 		via_generic_setup },
+        { PCI_DEVICE_ID_VIA_8377_0,
+		PCI_VENDOR_ID_VIA,
+		VIA_APOLLO_KT400,
+		"Via",
+		"Apollo Pro KT400",
+		via_generic_setup },
 	{ 0,
 		PCI_VENDOR_ID_VIA,
 		VIA_GENERIC,
diff -urN linux-2.4.20.orig/drivers/char/drm-4.0/agpsupport.c linux-2.4.20/drivers/char/drm-4.0/agpsupport.c
--- linux-2.4.20.orig/drivers/char/drm-4.0/agpsupport.c	Sat Jan 18 22:30:47 2003
+++ linux-2.4.20/drivers/char/drm-4.0/agpsupport.c	Sat Jan 18 22:55:46 2003
@@ -275,6 +275,8 @@
 			break;
 		case VIA_APOLLO_KT133:	head->chipset = "VIA Apollo KT133"; 
 			break;
+		case VIA_APOLLO_KT400:  head->chipset = "VIA Apollo KT400";
+			break;
 #endif
 
 		case VIA_APOLLO_PRO: 	head->chipset = "VIA Apollo Pro";
diff -urN linux-2.4.20.orig/include/linux/agp_backend.h linux-2.4.20/include/linux/agp_backend.h
--- linux-2.4.20.orig/include/linux/agp_backend.h	Sat Jan 18 22:30:34 2003
+++ linux-2.4.20/include/linux/agp_backend.h	Sat Jan 18 23:04:15 2003
@@ -60,6 +60,7 @@
 	VIA_APOLLO_PRO,
 	VIA_APOLLO_KX133,
 	VIA_APOLLO_KT133,
+	VIA_APOLLO_KT400,
 	SIS_GENERIC,
 	AMD_GENERIC,
 	AMD_IRONGATE,
diff -urN linux-2.4.20.orig/include/linux/pci_ids.h linux-2.4.20/include/linux/pci_ids.h
--- linux-2.4.20.orig/include/linux/pci_ids.h	Sat Jan 18 22:30:34 2003
+++ linux-2.4.20/include/linux/pci_ids.h	Sat Jan 18 23:03:44 2003
@@ -986,6 +986,7 @@
 #define PCI_DEVICE_ID_VIA_8233C_0	0x3109
 #define PCI_DEVICE_ID_VIA_8361		0x3112
 #define PCI_DEVICE_ID_VIA_8233A		0x3147
+#define PCI_DEVICE_ID_VIA_8377_0 	0x3189
 #define PCI_DEVICE_ID_VIA_86C100A	0x6100
 #define PCI_DEVICE_ID_VIA_8231		0x8231
 #define PCI_DEVICE_ID_VIA_8231_4	0x8235
