diff -urN linux-2.4.21.org/kernel/printk.c linux-2.4.21/kernel/printk.c
--- linux-2.4.21.org/kernel/printk.c	2003-08-12 23:34:01.000000000 +0200
+++ linux-2.4.21/kernel/printk.c	2003-08-24 01:13:36.000000000 +0200
@@ -35,10 +35,8 @@
 #define LOG_BUF_LEN	(65536)
 #elif defined(CONFIG_ARCH_S390)
 #define LOG_BUF_LEN	(131072)
-#elif defined(CONFIG_SMP)
-#define LOG_BUF_LEN	(32768)
-#else	
-#define LOG_BUF_LEN	(16384)			/* This must be a power of two */
+#else
+#define LOG_BUF_LEN	(32768)			/* This must be a power of two */
 #endif
 
 #define LOG_BUF_MASK	(LOG_BUF_LEN-1)
