--- linux-2.6.7/arch/ppc/Kconfig.orig	2004-07-09 23:03:07.000000000 +0000
+++ linux-2.6.7/arch/ppc/Kconfig	2004-07-10 09:17:08.000000000 +0000
@@ -1247,6 +1247,8 @@
 
 source "lib/Kconfig"
 
+source "cluster/Kconfig"
+
 source "arch/ppc/oprofile/Kconfig"
 
 menu "Kernel hacking"
--- linux-2.6.7/arch/sparc/Kconfig.orig	2004-06-16 05:18:37.000000000 +0000
+++ linux-2.6.7/arch/sparc/Kconfig	2004-07-10 09:16:34.000000000 +0000
@@ -456,3 +456,4 @@
 
 source "lib/Kconfig"
 
+source "cluster/Kconfig"
--- linux-2.6.7/arch/ia64/Kconfig.orig	2004-07-09 23:03:06.000000000 +0000
+++ linux-2.6.7/arch/ia64/Kconfig	2004-07-10 09:19:56.000000000 +0000
@@ -368,6 +368,8 @@
 
 source "lib/Kconfig"
 
+source "cluster/Kconfig"
+
 source "arch/ia64/hp/sim/Kconfig"
 
 source "arch/ia64/oprofile/Kconfig"
--- linux-2.6.7/arch/x86_64/Kconfig.orig	2004-06-16 05:19:44.000000000 +0000
+++ linux-2.6.7/arch/x86_64/Kconfig	2004-07-10 09:17:45.000000000 +0000
@@ -505,3 +505,4 @@
 
 source "lib/Kconfig"
 
+source "cluster/Kconfig"
