diff -urN linux.orig/linux-wlan-ng-0.1.12/Makefile linux/linux-wlan-ng-0.1.12/Makefile
--- linux.orig/linux-wlan-ng-0.1.12/Makefile	Sat Feb 16 10:46:47 2002
+++ linux/linux-wlan-ng-0.1.12/Makefile	Sat Feb 16 10:46:39 2002
@@ -43,7 +43,7 @@
 # --------------------------------------------------------------------
 
 
-DIRS = src doc man etc
+DIRS = src
 
 help:
 	@echo "Pick one of the following targets:"
diff -urN linux.orig/linux-wlan-ng-0.1.12/src/nwepgen/Makefile linux/linux-wlan-ng-0.1.12/src/nwepgen/Makefile
--- linux.orig/linux-wlan-ng-0.1.12/src/nwepgen/Makefile	Sat Feb 16 10:46:46 2002
+++ linux/linux-wlan-ng-0.1.12/src/nwepgen/Makefile	Sat Feb 16 14:08:42 2002
@@ -91,7 +91,7 @@
 CPPFLAGS= -I../include -I$(LINUX_SRC)/include \
 		-D__LINUX_WLAN__ -D__arm__
 else
-this will probably generate an error.  New arch's need to be added explicitly
+#this will probably generate an error.  New arch's need to be added explicitly
 CPPFLAGS= -I../include -I$(LINUX_SRC)/include \
 		-D__LINUX_WLAN__
 endif
diff -urN linux.orig/linux-wlan-ng-0.1.12/src/Makefile linux/linux-wlan-ng-0.1.12/src/Makefile
--- linux.orig/linux-wlan-ng-0.1.12/src/Makefile	Mon Aug 27 20:49:39 2001
+++ linux/linux-wlan-ng-0.1.12/src/Makefile	Mon Feb 18 21:23:24 2002
@@ -44,11 +44,9 @@
 
 include ../config.mk
 
-DIRS=p80211 prism2 shared wlanctl wland nwepgen \
-	mkmeta wlancfg
-ALL_DIRS=mkmeta p80211 prism2 shared wlanctl wland \
-	nwepgen wlancfg
-
+DIRS=p80211 prism2 shared
+ALL_DIRS=p80211 prism2 shared
+	
 ifneq ($(wildcard *.addon),) 
 DIRS+=`cat *.addon`
 ALL_DIRS+=`cat *.addon`
