--- linux/mm/highmem.c.org	Sat Jul  7 19:46:23 2001
+++ linux/mm/highmem.c	Sat Jul  7 19:51:03 2001
@@ -254,6 +254,13 @@
 
 static __init int init_emergency_pool(void)
 {
+	struct sysinfo i;
+        si_meminfo(&i);
+        si_swapinfo(&i);
+        
+        if (!i.totalhigh)
+        	return;
+
 	spin_lock_irq(&emergency_lock);
 	while (nr_emergency_pages < POOL_SIZE) {
 		struct page * page = alloc_page(GFP_ATOMIC);
