--- Makefile.orig	2005-01-16 19:22:53.000000000 +0100
+++ Makefile	2005-01-16 19:51:34.766170496 +0100
@@ -158,7 +158,6 @@
 	       $(shell cat /dev/null $(localver)) \
 	       $(patsubst "%",%,$(CONFIG_LOCALVERSION)))
 
-KERNELRELEASE=$(VERSION).$(PATCHLEVEL).$(SUBLEVEL)$(EXTRAVERSION)$(LOCALVERSION)
 
 # SUBARCH tells the usermode build what the underlying arch is.  That is set
 # first, and if a usermode build is happening, the "ARCH=um" on the command
@@ -528,6 +527,13 @@
 CFLAGS		+= -g
 endif
 
+__KERNELRELEASE = $(VERSION).$(PATCHLEVEL).$(SUBLEVEL)$(EXTRAVERSION)$(LOCALVERSION)
+ifndef CONFIG_SMP
+KERNELRELEASE   = $(__KERNELRELEASE)
+else
+KERNELRELEASE   = $(__KERNELRELEASE)smp
+endif
+
 include $(srctree)/arch/$(ARCH)/Makefile
 
 # warn about C99 declaration after statement
@@ -829,8 +835,13 @@
 	  echo '"$(KERNELRELEASE)" exceeds $(uts_len) characters' >&2; \
 	  exit 1; \
 	fi; \
-	(echo \#define UTS_RELEASE \"$(KERNELRELEASE)\"; \
-	  echo \#define LINUX_VERSION_CODE `expr $(VERSION) \\* 65536 + $(PATCHLEVEL) \\* 256 + $(SUBLEVEL)`; \
+	(echo '#include <linux/config.h>'; \
+	 echo '#ifndef CONFIG_SMP'; \
+	 echo \#define UTS_RELEASE \"$(__KERNELRELEASE)\"; \
+	 echo '#else'; \
+	 echo \#define UTS_RELEASE \"$(__KERNELRELEASE)smp\"; \
+	 echo '#endif'; \
+	 echo \#define LINUX_VERSION_CODE `expr $(VERSION) \\* 65536 + $(PATCHLEVEL) \\* 256 + $(SUBLEVEL)`; \
 	 echo '#define KERNEL_VERSION(a,b,c) (((a) << 16) + ((b) << 8) + (c))'; \
 	)
 endef
