diff -urN linux-2.4.24.org/fs/intermezzo/vfs.c linux-2.4.24/fs/intermezzo/vfs.c
--- linux-2.4.24.org/fs/intermezzo/vfs.c	2004-01-18 16:55:22.382895163 +0100
+++ linux-2.4.24/fs/intermezzo/vfs.c	2004-01-18 16:56:21.407533002 +0100
@@ -74,7 +74,7 @@
 #ifdef CONFIG_FS_EXT_ATTR
 # include <linux/ext_attr.h>
 
-#ifdef CONFIG_INTERMEZZO_FS_POSIX_ACL
+#ifdef CONFIG_INTERMEZZO_FS_POSIX_ACL && 0
 #  include <linux/posix_acl.h>
 # endif
 #endif
@@ -466,7 +466,7 @@
         struct dentry *dentry;
         struct presto_file_set *fset;
         int error;
-#ifdef CONFIG_INTERMEZZO_FS_POSIX_ACL
+#ifdef CONFIG_INTERMEZZO_FS_POSIX_ACL && 0
         int (*set_posix_acl)(struct inode *, int type, posix_acl_t *)=NULL;
 #endif
 
@@ -507,7 +507,7 @@
                                  (dentry->d_inode->i_mode & ~S_IALLUGO);
                 CDEBUG(D_PIOCTL, "chmod: orig %#o, set %#o, result %#o\n",
                        dentry->d_inode->i_mode, set_mode, iattr->ia_mode);
-#ifdef CONFIG_INTERMEZZO_FS_POSIX_ACL
+#ifdef CONFIG_INTERMEZZO_FS_POSIX_ACL && 0
                 /* ACl code interacts badly with setattr 
                  * since it tries to modify the ACL using 
                  * set_ext_attr which recurses back into presto.  
@@ -535,7 +535,7 @@
                 }
         }
 
-#ifdef CONFIG_INTERMEZZO_FS_POSIX_ACL
+#ifdef CONFIG_INTERMEZZO_FS_POSIX_ACL && 0
         /* restore the inode_operations if we changed them*/
         if (iattr->ia_valid & ATTR_MODE) 
                 dentry->d_inode->i_op->set_posix_acl=set_posix_acl;
@@ -2271,7 +2271,7 @@
 
 #ifdef CONFIG_FS_EXT_ATTR
 
-#ifdef CONFIG_INTERMEZZO_FS_POSIX_ACL
+#ifdef CONFIG_INTERMEZZO_FS_POSIX_ACL && 0
 /* Posix ACL code changes i_mode without using a notify_change (or
  * a mark_inode_dirty!). We need to duplicate this at the reintegrator
  * which is done by this function. This function also takes care of 
@@ -2414,7 +2414,7 @@
                 goto exit;
         }
 
-#ifdef CONFIG_INTERMEZZO_FS_POSIX_ACL
+#ifdef CONFIG_INTERMEZZO_FS_POSIX_ACL && 0
         /* Reset mode if specified*/
         /* XXX: when we do native acl support, move this code out! */
         if (mode != NULL) {
