diff -ur linux-2.4.20/drivers/scsi/Makefile linux-2.4.20-sym53c/drivers/scsi/Makefile
--- linux-2.4.20/drivers/scsi/Makefile	Mon Mar 17 19:09:47 2003
+++ linux-2.4.20-sym53c/drivers/scsi/Makefile	Mon Mar 17 19:16:53 2003
@@ -103,7 +103,7 @@
 ifeq ($(CONFIG_SCSI_SYM53C8XX_2),y)
   obj-$(CONFIG_SCSI_SYM53C8XX_2)	+= sym53c8xx_2/sym53c8xx.o
 endif
-obj-$(CONFIG_SCSI_SYM53C8XX)	+= sym53c8xx.o 
+obj-$(CONFIG_SCSI_SYM53C8XX)	+= sym53c8xx_old.o 
 obj-$(CONFIG_SCSI_NCR53C8XX)	+= ncr53c8xx.o 
 obj-$(CONFIG_SCSI_EATA_DMA)	+= eata_dma.o
 obj-$(CONFIG_SCSI_EATA_PIO)	+= eata_pio.o
