diff -urN linux-2.4.21.org/drivers/md/raid5.c linux-2.4.21/drivers/md/raid5.c
--- linux-2.4.21.org/drivers/md/raid5.c	Tue Aug 12 22:22:46 2003
+++ linux-2.4.21/drivers/md/raid5.c	Tue Aug 12 22:54:34 2003
@@ -1365,6 +1365,7 @@
 	struct disk_info *disk;
 	struct md_list_head *tmp;
 	int start_recovery = 0;
+	int spares = 0;
 
 	MOD_INC_USE_COUNT;
 
@@ -1462,6 +1463,7 @@
 			disk->write_only = 0;
 			disk->spare = 1;
 			disk->used_slot = 1;
+			spares ++;
 		}
 	}
 
@@ -1554,6 +1556,7 @@
 		}
 	}
 	sb->active_disks = conf->working_disks;
+	sb->spare_disks = spares;
 
 	if (sb->active_disks == sb->raid_disks)
 		printk("raid5: raid level %d set md%d active with %d out of %d devices, algorithm %d\n", conf->level, mdidx(mddev), sb->active_disks, sb->raid_disks, conf->algorithm);
