--- linux/fs/quota_v1.c.org	Fri Apr 26 11:33:20 2002
+++ linux/fs/quota_v1.c	Fri Apr 26 11:58:33 2002
@@ -234,6 +234,10 @@
 
 EXPORT_NO_SYMBOLS;
 
+MODULE_AUTHOR("");
+MODULE_DESCRIPTION("QUOTA Version 1 support.");
+MODULE_LICENSE("GPL");
+
 module_init(init_v1_quota_format);
 module_exit(exit_v1_quota_format);
 
--- linux/fs/quota_v2.c.org	Fri Apr 26 11:33:20 2002
+++ linux/fs/quota_v2.c	Fri Apr 26 11:58:20 2002
@@ -686,5 +686,9 @@
 
 EXPORT_NO_SYMBOLS;
 
+MODULE_AUTHOR("");
+MODULE_DESCRIPTION("QUOTA Version 2 support.");
+MODULE_LICENSE("GPL");
+
 module_init(init_v2_quota_format);
 module_exit(exit_v2_quota_format);
diff -urN linux.org/net/unix/af_unix.c linux/net/unix/af_unix.c
--- linux.org/net/unix/af_unix.c	Sun May 19 02:49:28 2002
+++ linux/net/unix/af_unix.c	Sun May 19 02:50:14 2002
@@ -1880,6 +1880,10 @@
 	remove_proc_entry("net/unix", 0);
 }
 
+MODULE_AUTHOR("Alan Cox");
+MODULE_DESCRIPTION("Implementation of BSD Unix domain sockets.");
+MODULE_LICENSE("GPL");
+
 module_init(af_unix_init);
 module_exit(af_unix_exit);
 
--- linux-2.4.20/drivers/block/ps2esdi.c.org	Fri Nov 29 00:53:12 2002
+++ linux-2.4.20/drivers/block/ps2esdi.c	Mon Dec 16 15:32:21 2002
@@ -30,8 +30,6 @@
 #include <linux/config.h>
 #include <linux/major.h>
 
-#ifdef  CONFIG_BLK_DEV_PS2
-
 #define MAJOR_NR PS2ESDI_MAJOR
 
 #include <linux/errno.h>
@@ -1173,5 +1171,3 @@
 	}
 	wake_up(&ps2esdi_int);
 }
-
-#endif
--- linux-2.4.20/drivers/char/synclinkmp.c.org	Fri Nov 29 00:53:12 2002
+++ linux-2.4.20/drivers/char/synclinkmp.c	Mon Dec 16 15:00:41 2002
@@ -87,6 +87,10 @@
 
 #include "linux/synclink.h"
 
+MODULE_AUTHOR("Paul Fulghum <paulkf@microgate.com>");
+MODULE_DESCRIPTION("Device driver for Microgate SyncLink Multiport");
+MODULE_LICENSE("GPL");
+
 static MGSL_PARAMS default_params = {
 	MGSL_MODE_HDLC,			/* unsigned long mode */
 	0,				/* unsigned char loopback; */
--- linux-2.4.20/drivers/net/hamradio/soundmodem/sm.c.org	Wed Apr 18 23:40:05 2001
+++ linux-2.4.20/drivers/net/hamradio/soundmodem/sm.c	Mon Dec 16 15:12:44 2002
@@ -634,6 +634,7 @@
 
 MODULE_AUTHOR("Thomas M. Sailer, sailer@ife.ee.ethz.ch, hb9jnx@hb9w.che.eu");
 MODULE_DESCRIPTION("Soundcard amateur radio modem driver");
+MODULE_LICENSE("GPL2");
 
 /* --------------------------------------------------------------------- */
 
--- linux-2.4.20/drivers/net/wan/8253x/8253xini.c.org	Sat Aug  3 02:39:44 2002
+++ linux-2.4.20/drivers/net/wan/8253x/8253xini.c	Mon Dec 16 15:36:42 2002
@@ -3041,3 +3041,4 @@
 module_exit(auraXX20_cleanup);
 MODULE_DESCRIPTION("Aurora Multiport Multiprotocol Serial Driver");
 MODULE_AUTHOR("Joachim Martillo <martillo@telfordtools.com>");
+MODULE_LICENSE("GPL");
--- linux-2.4.20/drivers/scsi/mca_53c9x.c.org	Thu Oct 25 22:53:51 2001
+++ linux-2.4.20/drivers/scsi/mca_53c9x.c	Mon Dec 16 15:24:43 2002
@@ -51,6 +51,11 @@
 
 #include <asm/pgtable.h>
 
+#include <linux/module.h>
+MODULE_DESCRIPTION("Driver for the SCSI adapter on NCR 35xx Microchannel machines");
+MODULE_AUTHOR("unknown");
+MODULE_LICENSE("unknown");
+
 static int  dma_bytes_sent(struct NCR_ESP *, int);
 static int  dma_can_transfer(struct NCR_ESP *, Scsi_Cmnd *);
 static void dma_dump_state(struct NCR_ESP *);
--- linux-2.4.20/net/ipv4/netfilter/ipfwadm_core.c.org	Mon Feb 25 20:38:14 2002
+++ linux-2.4.20/net/ipv4/netfilter/ipfwadm_core.c	Mon Dec 16 15:27:52 2002
@@ -130,6 +130,10 @@
 #include <linux/stat.h>
 #include <linux/version.h>
 
+MODULE_AUTHOR("");
+MODULE_DESCRIPTION("IP firewalling code");
+MODULE_LICENSE("GPL");
+
 /*
  *	Implement IP packet firewall
  */
--- linux-2.4.20/net/ipv6/netfilter/ip6t_length.c.org	Fri Nov 29 00:53:15 2002
+++ linux-2.4.20/net/ipv6/netfilter/ip6t_length.c	Mon Dec 16 15:31:08 2002
@@ -5,6 +5,8 @@
 #include <linux/netfilter_ipv6/ip6t_length.h>
 #include <linux/netfilter_ipv6/ip6_tables.h>
 
+MODULE_LICENSE("unknown");
+
 static int
 match(const struct sk_buff *skb,
       const struct net_device *in,
