--- linux-2.4.21/drivers/char/agp/agpgart_be.c~	Mon Jun 30 19:05:11 2003
+++ linux-2.4.21/drivers/char/agp/agpgart_be.c	Mon Jun 30 19:36:34 2003
@@ -4738,7 +4738,6 @@
 static int __init uninorth_setup (struct pci_dev *pdev)
 {
 	agp_bridge.masks = uninorth_masks;
-	agp_bridge.num_of_masks = 1;
 	agp_bridge.aperture_sizes = (void *)uninorth_sizes;
 	agp_bridge.size_type = U32_APER_SIZE;
 	agp_bridge.num_aperture_sizes = 4; //7;
