diff -urNp linux-1090/arch/i386/kernel/pci-irq.c linux-1100/arch/i386/kernel/pci-irq.c
--- linux-1090/arch/i386/kernel/pci-irq.c	
+++ linux-1100/arch/i386/kernel/pci-irq.c	
@@ -36,7 +36,7 @@ unsigned int pcibios_irq_mask = 0xfff8;
 
 static int pirq_penalty[16] = {
 	1000000, 1000000, 1000000, 1000, 1000, 0, 1000, 1000,
-	0, 0, 0, 0, 1000, 100000, 100000, 100000
+	0, 0, 0, 0, 2500, 100000, 100000, 100000
 };
 
 struct irq_router {
