--- linux-2.6.19/include/linux/freezer.h~	2006-12-09 00:12:07.000000000 +0100
+++ linux-2.6.19/include/linux/freezer.h	2006-12-09 00:33:27.000000000 +0100
@@ -95,8 +95,10 @@ static inline int thaw_process(struct ta
 static inline void frozen_process(struct task_struct *p) { BUG(); }
 
 static inline void refrigerator(void) {}
+#if 0
 static inline int freeze_processes(void) { BUG(); return 0; }
 static inline void thaw_processes(int which_threads) {}
+#endif
 
 static inline int try_to_freeze(void) { return 0; }
 
