--- linux-2.4.19/drivers/char/drm/i830_dma.c~	Sun Aug 11 15:46:15 2002
+++ linux-2.4.19/drivers/char/drm/i830_dma.c	Sun Aug 11 17:09:29 2002
@@ -316,9 +316,8 @@
 	if(page == 0UL) 
 		return;
 	
-	atomic_dec(&virt_to_page(page)->count);
-	clear_bit(PG_locked, &virt_to_page(page)->flags);
-	wake_up(&virt_to_page(page)->wait);
+	put_page(virt_to_page(page));
+	UnlockPage(virt_to_page(page));
 	free_page(page);
 	return;
 }
--- linux-2.4.19/drivers/char/drm/i810_dma.c~	Sun Aug 11 15:46:15 2002
+++ linux-2.4.19/drivers/char/drm/i810_dma.c	Sun Aug 11 17:08:56 2002
@@ -294,9 +294,8 @@
 	if(page == 0UL)
 		return;
 
-	atomic_dec(&virt_to_page(page)->count);
-	clear_bit(PG_locked, &virt_to_page(page)->flags);
-	wake_up(&virt_to_page(page)->wait);
+	put_page(virt_to_page(page));
+	UnlockPage(virt_to_page(page));
 	free_page(page);
 	return;
 }
