--- linux-2.5.72/arch/alpha/kernel/srmcons.c~	2003-06-17 06:20:22.000000000 +0200
+++ linux-2.5.72/arch/alpha/kernel/srmcons.c	2003-06-22 09:46:55.000000000 +0200
@@ -291,6 +291,7 @@
 		driver->type = TTY_DRIVER_TYPE_SYSTEM;
 		driver->subtype = SYSTEM_TYPE_SYSCONS;
 		driver->init_termios = tty_std_termios;
+		tty_set_operations(driver, &srmcons_ops);
 		err = tty_register_driver(driver);
 		if (err) {
 			put_tty_driver(driver);
