diff -urN linux.orig/linux-wlan-ng-0.1.12/Makefile linux/linux-wlan-ng-0.1.12/Makefile
--- linux.orig/linux-wlan-ng-0.1.12/Makefile	Thu Apr  5 22:38:48 2001
+++ linux/linux-wlan-ng-0.1.12/Makefile	Fri Feb 15 23:22:55 2002
@@ -43,7 +43,7 @@
 # --------------------------------------------------------------------
 
 
-DIRS = src doc man etc
+DIRS = src
 
 help:
 	@echo "Pick one of the following targets:"
