--- linux-2.6.10/include/linux/pci_ids.h.orig	2004-12-24 22:35:50.000000000 +0100
+++ linux-2.6.10/include/linux/pci_ids.h	2005-02-14 18:39:28.601179992 +0100
@@ -548,6 +548,7 @@
 #define PCI_DEVICE_ID_CT_65550		0x00e0
 #define PCI_DEVICE_ID_CT_65554		0x00e4
 #define PCI_DEVICE_ID_CT_65555		0x00e5
+#define PCI_DEVICE_ID_CT_69000		0x00c0
 
 #define PCI_VENDOR_ID_MIRO		0x1031
 #define PCI_DEVICE_ID_MIRO_36050	0x5601
