--- ./include/linux/pci_ids.h.org	2006-09-20 05:42:06.000000000 +0200
+++ ./include/linux/pci_ids.h	2006-11-05 02:05:18.000000000 +0100
@@ -760,6 +760,7 @@
 #define PCI_DEVICE_ID_TI_1510		0xac56
 #define PCI_DEVICE_ID_TI_X620		0xac8d
 #define PCI_DEVICE_ID_TI_X420		0xac8e
+#define PCI_DEVICE_ID_TI_X420_X620	0xac8f
 
 #define PCI_VENDOR_ID_SONY		0x104d
 
--- ./drivers/mmc/sdhci.c.org	2006-09-20 05:42:06.000000000 +0200
+++ ./drivers/mmc/sdhci.c	2006-11-05 02:06:21.000000000 +0100
@@ -61,6 +61,14 @@
 		.driver_data	= SDHCI_QUIRK_FORCE_DMA,
 	},
 
+	{
+		.vendor		= PCI_VENDOR_ID_TI,
+		.device		= PCI_DEVICE_ID_TI_X420_X620,
+		.subvendor	= PCI_ANY_ID,
+		.subdevice	= PCI_ANY_ID,
+		.driver_data	= SDHCI_QUIRK_FORCE_DMA,
+	},
+
 	{	/* Generic SD host controller */
 		PCI_DEVICE_CLASS((PCI_CLASS_SYSTEM_SDHCI << 8), 0xFFFF00)
 	},
