--- linux/kernel/syscall_ksyms.c~	Sun Feb  3 20:25:43 2002
+++ linux/kernel/syscall_ksyms.c	Sun Feb  3 20:32:21 2002
@@ -87,7 +87,6 @@
 EXPORT_SYMBOL(sys_sched_setparam);
 EXPORT_SYMBOL(sys_sched_getscheduler);
 EXPORT_SYMBOL(sys_sched_getparam);
-EXPORT_SYMBOL(sys_sched_yield);
 EXPORT_SYMBOL(sys_sched_get_priority_max);
 EXPORT_SYMBOL(sys_sched_get_priority_min);
 EXPORT_SYMBOL(sys_sched_rr_get_interval);
