diff -Nur --exclude='*.orig' linux-2.4.20.org/drivers/macintosh/adbhid.c linux-2.4.20/drivers/macintosh/adbhid.c
--- linux-2.4.20.org/drivers/macintosh/adbhid.c	Sat Aug  3 00:39:44 2002
+++ linux-2.4.20/drivers/macintosh/adbhid.c	Fri Jun  6 11:05:44 2003
@@ -550,7 +550,7 @@
 	input_register_device(&adbhid[id]->input);
 
-	printk(KERN_INFO "input%d: ADB HID on ID %d:%d.%02x\n",
+	printk(KERN_INFO "input: ADB HID on ID %d:%d.%02x\n",
-	       adbhid[id]->input.number, id, default_id, original_handler_id);
+	       id, default_id, original_handler_id);
 
 	if (default_id == ADB_KEYBOARD) {
 		/* HACK WARNING!! This should go away as soon there is an utility
diff -Nur --exclude='*.orig' linux-2.4.20.org/drivers/macintosh/mac_hid.c linux-2.4.20/drivers/macintosh/mac_hid.c
--- linux-2.4.20.org/drivers/macintosh/mac_hid.c	Sat Aug  3 00:39:44 2002
+++ linux-2.4.20/drivers/macintosh/mac_hid.c	Fri Jun  6 11:04:30 2003
@@ -469,7 +469,7 @@
 
 	input_register_device(&emumousebtn);
 
-	printk(KERN_INFO "input%d: Macintosh mouse button emulation\n", emumousebtn.number);
+	printk(KERN_INFO "input: Macintosh mouse button emulation\n");
 }
 #endif
 
