--- linux-2.4.20/arch/i386/lib/csumcpy.S.org	Tue Mar 11 09:31:08 2003
+++ linux-2.4.20/arch/i386/lib/csumcpy.S	Tue Mar 11 09:45:52 2003
@@ -115,7 +115,7 @@
 	testl	$2, %edi
 	jz	40f
 30:
-    # 2-aligned, but not 4-aligned
+//    # 2-aligned but not 4-aligned
 	cmpl	$3, %ecx
 	jbe	60f
 	movw	(%esi), %bx	# eat 2 bytes
--- linux-2.4.20/arch/i386/lib/csumcpy_sse.inc.org	Tue Mar 11 09:31:08 2003
+++ linux-2.4.20/arch/i386/lib/csumcpy_sse.inc	Tue Mar 11 09:45:15 2003
@@ -130,7 +130,7 @@
 	ret
 
 
-# 16-align edi and get back
+//# 16-align edi and get back
 5500:	cmp	$ITER_SZ, %ecx	# edi is 4-aligned here
 	mov	%ecx, %edx      # edx needed at 20:
 	jb	20b		# not worthy: too short
