diff -ur linux-2.6.10.orig/arch/sparc/kernel/sparc_ksyms.c linux-2.6.10/arch/sparc/kernel/sparc_ksyms.c
--- linux-2.6.10.orig/arch/sparc/kernel/sparc_ksyms.c	2004-12-24 22:33:52.000000000 +0100
+++ linux-2.6.10/arch/sparc/kernel/sparc_ksyms.c	2005-01-09 08:37:50.199948280 +0100
@@ -218,9 +218,10 @@
 EXPORT_SYMBOL(pci_map_page);
 EXPORT_SYMBOL(pci_unmap_page);
 /* Actually, ioremap/iounmap are not PCI specific. But it is ok for drivers. */
+#endif
+/* or not? not all MTD drivers depend on PCI */
 EXPORT_SYMBOL(ioremap);
 EXPORT_SYMBOL(iounmap);
-#endif
 
 /* in arch/sparc/mm/highmem.c */
 #ifdef CONFIG_HIGHMEM
@@ -256,6 +257,9 @@
 EXPORT_SYMBOL(prom_getint);
 EXPORT_SYMBOL(prom_getintdefault);
 EXPORT_SYMBOL(prom_finddevice);
+EXPORT_SYMBOL(prom_halt);
+EXPORT_SYMBOL(prom_printf);
+EXPORT_SYMBOL(prom_apply_generic_ranges);
 EXPORT_SYMBOL(romvec);
 EXPORT_SYMBOL(__prom_getchild);
 EXPORT_SYMBOL(__prom_getsibling);
