diff -uNr linux-2.6.16.orig/sound/pci/hda/hda_intel.c linux-2.6.16/sound/pci/hda/hda_intel.c
--- linux-2.6.16.orig/sound/pci/hda/hda_intel.c	2007-06-24 09:22:14.983683000 +0200
+++ linux-2.6.16/sound/pci/hda/hda_intel.c	2007-06-24 12:13:35.246039069 +0200
@@ -1602,6 +1602,7 @@
 	{ 0x10b9, 0x5461, PCI_ANY_ID, PCI_ANY_ID, 0, 0, AZX_DRIVER_ULI }, /* ULI M5461 */
 	{ 0x10de, 0x026c, PCI_ANY_ID, PCI_ANY_ID, 0, 0, AZX_DRIVER_NVIDIA }, /* NVIDIA 026c */
 	{ 0x10de, 0x0371, PCI_ANY_ID, PCI_ANY_ID, 0, 0, AZX_DRIVER_NVIDIA }, /* NVIDIA 0371 */
+        { 0x10de, 0x03f0, PCI_ANY_ID, PCI_ANY_ID, 0, 0, AZX_DRIVER_NVIDIA }, /* NVIDIA 03f0 */
 	{ 0, }
 };
 MODULE_DEVICE_TABLE(pci, azx_ids);
