--- linux.orig/include/linux/tasks.h	Mon Dec 11 01:49:44 2000
+++ linux/include/linux/tasks.h	Sat Dec 23 04:31:48 2000
@@ -11,10 +11,10 @@
 #define NR_CPUS 1
 #endif
 
-#define NR_TASKS	512	/* On x86 Max about 4000 */
+#define NR_TASKS	4080	/* On x86 Max about 4000 */
 
-#define MAX_TASKS_PER_USER (NR_TASKS/2)
-#define MIN_TASKS_LEFT_FOR_ROOT 4
+#define MAX_TASKS_PER_USER 256
+#define MIN_TASKS_LEFT_FOR_ROOT 64
 
 
 /*
