diff -Nurp linux-2.6.22.9-vs2.3.0.27/net/ipv4/inet_connection_sock.c linux-2.6.22.9-vs2.3.0.27.v6tw/net/ipv4/inet_connection_sock.c
--- linux-2.6.22.9-vs2.3.0.27/net/ipv4/inet_connection_sock.c	2007-10-10 11:13:10.000000000 +0200
+++ linux-2.6.22.9-vs2.3.0.27.v6tw/net/ipv4/inet_connection_sock.c	2007-10-17 13:31:31.000000000 +0200
@@ -39,7 +39,7 @@ int ipv4_rcv_saddr_equal(const struct so
 	__be32	sk1_rcv_saddr = inet_rcv_saddr(sk1),
 		sk2_rcv_saddr = inet_rcv_saddr(sk2);
 
-	if (ipv6_only_sock(sk2))
+	if (inet_v6_ipv6only(sk2))
 		return 0;
 
 	if (sk1_rcv_saddr &&
