diff -urN linux-2.4.20/drivers/char/agp/agpgart_be.c linux-2.4.20+KM266/drivers/char/agp/agpgart_be.c
--- linux-2.4.20/drivers/char/agp/agpgart_be.c	2002-12-21 03:39:12.000000000 +0100
+++ linux-2.4.20+KM266/drivers/char/agp/agpgart_be.c	2003-05-19 21:57:46.000000000 +0200
@@ -4714,6 +4714,12 @@
 		"Via",
 		"Apollo Pro KT400",
 		via_generic_setup },
+        { PCI_DEVICE_ID_VIA_8375,
+		PCI_VENDOR_ID_VIA,
+		VIA_APOLLO_KM266,
+		"Via",
+		"Apollo Pro KM266 / KL266",
+		via_generic_setup },
 	{ 0,
 		PCI_VENDOR_ID_VIA,
 		VIA_GENERIC,
diff -urN linux-2.4.20/drivers/char/drm-4.0/agpsupport.c linux-2.4.20+KM266/drivers/char/drm-4.0/agpsupport.c
--- linux-2.4.20/drivers/char/drm-4.0/agpsupport.c	2002-02-25 20:37:57.000000000 +0100
+++ linux-2.4.20+KM266/drivers/char/drm-4.0/agpsupport.c	2003-05-19 21:57:46.000000000 +0200
@@ -275,6 +275,8 @@
 			break;
 		case VIA_APOLLO_KT400:	head->chipset = "VIA Apollo KT400"; 
 			break;
+		case VIA_APOLLO_KM266:  head->chipset = "VIA Apollo KM266 / KL266";
+			break;
 #endif
 
 		case VIA_APOLLO_PRO: 	head->chipset = "VIA Apollo Pro";
diff -urN linux-2.4.20/drivers/pci/pci.ids linux-2.4.20+KM266/drivers/pci/pci.ids
--- linux-2.4.20/drivers/pci/pci.ids	2002-11-29 00:53:14.000000000 +0100
+++ linux-2.4.20+KM266/drivers/pci/pci.ids	2003-05-19 21:57:46.000000000 +0200
@@ -2741,6 +2741,7 @@
 	3104  USB 2.0
 	3109  VT8233C PCI to ISA Bridge
 	3112  VT8361 [KLE133] Host Bridge
+	3116  VT8375 [KM266/KL266 AGP] Host Bridge
 	3128  VT8753 [P4X266 AGP]
 	3133  VT3133 Host Bridge
 	3147  VT8233A ISA Bridge
diff -urN linux-2.4.20/include/linux/agp_backend.h linux-2.4.20+KM266/include/linux/agp_backend.h
--- linux-2.4.20/include/linux/agp_backend.h	2002-11-29 00:53:15.000000000 +0100
+++ linux-2.4.20+KM266/include/linux/agp_backend.h	2003-05-19 21:57:46.000000000 +0200
@@ -60,6 +60,7 @@
 	VIA_APOLLO_KX133,
 	VIA_APOLLO_KT133,
 	VIA_APOLLO_KT400,
+	VIA_APOLLO_KM266,
 	SIS_GENERIC,
 	AMD_GENERIC,
 	AMD_IRONGATE,
diff -urN linux-2.4.20/include/linux/pci_ids.h linux-2.4.20+KM266/include/linux/pci_ids.h
--- linux-2.4.20/include/linux/pci_ids.h	2002-11-30 05:49:28.000000000 +0100
+++ linux-2.4.20+KM266/include/linux/pci_ids.h	2003-05-19 21:57:46.000000000 +0200
@@ -988,6 +988,7 @@
 #define PCI_DEVICE_ID_VIA_8622		0x3102
 #define PCI_DEVICE_ID_VIA_8233C_0	0x3109
 #define PCI_DEVICE_ID_VIA_8361		0x3112
+#define PCI_DEVICE_ID_VIA_8375		0x3116
 #define PCI_DEVICE_ID_VIA_8233A		0x3147
 #define PCI_DEVICE_ID_VIA_8237_SATA	0x3149
 #define PCI_DEVICE_ID_VIA_P4X333		0x3168
