--- linux-2.6.17/drivers/media/video/cx88/cx88-tvaudio.c.orig	2006-09-15 21:37:08.000000000 +0200
+++ linux-2.6.17/drivers/media/video/cx88/cx88-tvaudio.c	2006-09-15 22:05:19.000000000 +0200
@@ -148,11 +148,10 @@
 
 	if (cx88_boards[core->board].blackbird) {
 		/* sets sound input from external adc */
-		if (core->board == CX88_BOARD_HAUPPAUGE_ROSLYN)
-			cx_clear(AUD_CTL, EN_I2SIN_ENABLE);
-		else
-			cx_set(AUD_CTL, EN_I2SIN_ENABLE);
-
+		cx_clear(AUD_CTL, EN_I2SIN_ENABLE);
+#if 0
+		cx_set(AUD_CTL, EN_I2SIN_ENABLE);
+#endif
 		cx_write(AUD_I2SINPUTCNTL, 4);
 		cx_write(AUD_BAUDRATE, 1);
 		/* 'pass-thru mode': this enables the i2s output to the mpeg encoder */
