--- linux-wlan-ng-0.2.1pre21/src/Makefile.orig	2004-02-13 19:42:39.000000000 +0100
+++ linux-wlan-ng-0.2.1pre21/src/Makefile	2004-05-31 09:53:20.232175216 +0200
@@ -44,9 +44,9 @@
 
 include ../config.mk
 
-DIRS=p80211 prism2 shared wlanctl wland nwepgen \
+DIRS=prism2 shared wlanctl wland nwepgen \
 	mkmeta wlancfg
-ALL_DIRS=mkmeta p80211 prism2 shared wlanctl wland \
+ALL_DIRS=mkmeta prism2 shared wlanctl wland \
 	nwepgen wlancfg
 
 ifneq ($(wildcard *.addon),) 
--- linux-wlan-ng-0.2.1pre21/src/prism2/Makefile.orig	2004-02-13 19:47:37.000000000 +0100
+++ linux-wlan-ng-0.2.1pre21/src/prism2/Makefile	2004-05-31 09:53:30.815566296 +0200
@@ -44,7 +44,7 @@
 
 include ../../config.mk
 
-DIRS=driver ridlist download
+DIRS=ridlist download
 
 ifneq ($(wildcard *.addon),) 
 DIRS+=`cat *.addon`
--- linux-wlan-ng-0.2.1-pre12.orig/man/Makefile	2003-09-24 18:31:45.000000000 +0200
+++ linux-wlan-ng-0.2.1-pre12/man/Makefile	2003-09-24 18:32:10.000000000 +0200
@@ -48,9 +48,9 @@
 	@echo "Nothing to do..."
 
 install:
-	mkdir -p $(TARGET_ROOT_ON_HOST)/usr/local/man/man1
+	mkdir -p $(TARGET_ROOT_ON_HOST)/usr/share/man/man1
 	cp nwepgen.man wlancfg.man wlanctl-ng.man wland.man \
-	$(TARGET_ROOT_ON_HOST)/usr/local/man/man1
+	$(TARGET_ROOT_ON_HOST)/usr/share/man/man1
 
 clean:
 	rm -f core core.* *.o .*.o *.s *.a .depend tmp_make *~ tags
--- linux-wlan-ng-0.2.1pre21/src/p80211/Makefile.orig	2004-05-04 17:15:21.000000000 +0200
+++ linux-wlan-ng-0.2.1pre21/src/p80211/Makefile	2004-05-31 07:52:51.212153304 +0200
@@ -100,10 +100,10 @@
 	rm -f *.o *.ko .*.cmd *.mod.c  *.flags .*.flags
 
 install:
-	mkdir -p $(TARGET_MODDIR)/linux-wlan-ng/net
+	mkdir -p $(TARGET_MODDIR)/kernel/drivers/net/wireless
 ifeq ($(KERN_25), y)
-	cp p80211.ko	$(TARGET_MODDIR)/linux-wlan-ng
+	cp p80211.ko	$(TARGET_MODDIR)/kernel/drivers/net/wireless
 else
-	cp p80211.o	$(TARGET_MODDIR)/linux-wlan-ng
+	cp p80211.o	$(TARGET_MODDIR)/kernel/drivers/net/wireless
 endif
 
--- linux-wlan-ng-0.2.1pre21/src/prism2/driver/Makefile.orig	2004-05-04 17:15:22.000000000 +0200
+++ linux-wlan-ng-0.2.1pre21/src/prism2/driver/Makefile	2004-05-31 07:54:07.998480008 +0200
@@ -104,8 +104,8 @@
 
 install:
 ifeq ($(KERN_25), y)
-	cp $(obj-m:.o=.ko)	$(TARGET_MODDIR)/linux-wlan-ng
+	cp $(obj-m:.o=.ko)	$(TARGET_MODDIR)/kernel/drivers/net/wireless
 else
-	cp $(obj-m)	$(TARGET_MODDIR)/linux-wlan-ng
+	cp $(obj-m)	$(TARGET_MODDIR)/kernel/drivers/net/wireless
 endif
 
