diff -Nur --exclude='*.orig' linux-2.4.20.org/drivers/macintosh/adbhid.c linux-2.4.20/drivers/macintosh/adbhid.c
--- linux-2.4.20.org/drivers/macintosh/adbhid.c	Sat Aug  3 00:39:44 2002
+++ linux-2.4.20/drivers/macintosh/adbhid.c	Fri Jun  6 11:05:44 2003
@@ -550,7 +550,7 @@
 	input_register_device(&adbhid[id]->input);
 
-	printk(KERN_INFO "input%d: ADB HID on ID %d:%d.%02x\n",
+	printk(KERN_INFO "input: ADB HID on ID %d:%d.%02x\n",
-	       adbhid[id]->input.number, id, default_id, original_handler_id);
+	       id, default_id, original_handler_id);
 
 	if (default_id == ADB_KEYBOARD) {
 		/* HACK WARNING!! This should go away as soon there is an utility
diff -Nur --exclude='*.orig' linux-2.4.20.org/drivers/macintosh/mac_hid.c linux-2.4.20/drivers/macintosh/mac_hid.c
--- linux-2.4.20.org/drivers/macintosh/mac_hid.c	Sat Aug  3 00:39:44 2002
+++ linux-2.4.20/drivers/macintosh/mac_hid.c	Fri Jun  6 11:04:30 2003
@@ -469,7 +469,7 @@
 
 	input_register_device(&emumousebtn);
 
-	printk(KERN_INFO "input%d: Macintosh mouse button emulation\n", emumousebtn.number);
+	printk(KERN_INFO "input: Macintosh mouse button emulation\n");
 }
 #endif
 
--- linux-2.4.20/drivers/usb/aiptek.c.org	Fri Nov 29 00:53:14 2002
+++ linux-2.4.20/drivers/usb/aiptek.c	Fri Jun  6 15:16:32 2003
@@ -289,8 +289,8 @@
 
 	input_register_device(&aiptek->dev);
 
-	printk(KERN_INFO "input%d: %s on usb%d:%d.%d\n",
-	       aiptek->dev.number, aiptek->features->name, dev->bus->busnum,
+	printk(KERN_INFO "input: %s on usb%d:%d.%d\n",
+	       aiptek->features->name, dev->bus->busnum,
 	       dev->devnum, ifnum);
 
 	return aiptek;
--- linux-2.4.20/drivers/usb/powermate.c.org	Fri Jun  6 13:45:12 2003
+++ linux-2.4.20/drivers/usb/powermate.c	Fri Jun  6 15:38:15 2003
@@ -306,7 +306,7 @@
 		printk(KERN_INFO "input: %s on %s\n", pm->input.name, pm->input.path);
 	}
 #else
-	printk(KERN_INFO "input: %s on input%d\n", pm->input.name, pm->input.number);
+	printk(KERN_INFO "input: %s on input\n", pm->input.name);
 #endif
 	
 	/* force an update of everything */
