--- linux-2.4.21/drivers/i2c/i2c-velleman.c~	Mon Aug 11 14:45:01 2003
+++ linux-2.4.21/drivers/i2c/i2c-velleman.c	Mon Aug 11 17:14:18 2003
@@ -29,6 +29,7 @@
 #include <linux/i2c.h>
 #include <linux/i2c-algo-bit.h>
 #include <asm/io.h>
+#include <asm/param.h>
 
 /* ----- global defines -----------------------------------------------	*/
 #define DEB(x)		/* should be reasonable open, close &c. 	*/
--- linux-2.4.21/drivers/i2c/scx200_acb.c~	Mon Aug 11 14:45:01 2003
+++ linux-2.4.21/drivers/i2c/scx200_acb.c	Mon Aug 11 17:17:00 2003
@@ -33,6 +33,7 @@
 #include <linux/i2c.h>
 #include <linux/smp_lock.h>
 #include <linux/pci.h>
+#include <linux/slab.h>
 #include <asm/io.h>
 
 #include <linux/scx200.h>
