--- linux-2.4.20/arch/alpha/kernel/alpha_ksyms.c.org	Mon Feb 25 19:37:52 2002
+++ linux-2.4.20/arch/alpha/kernel/alpha_ksyms.c	Fri May 24 19:48:14 2002
@@ -264,3 +264,6 @@
 EXPORT_SYMBOL(irongate_ioremap);
 EXPORT_SYMBOL(irongate_iounmap);
 #endif
+
+EXPORT_SYMBOL(plat_node_data);
+
--- linux-2.4.20/drivers/char/ipmi/ipmi_watchdog.c.org	Sat Jun  7 14:11:42 2003
+++ linux-2.4.20/drivers/char/ipmi/ipmi_watchdog.c	Sat Jun  7 14:12:42 2003
@@ -31,6 +31,7 @@
  *  675 Mass Ave, Cambridge, MA 02139, USA.
  */
 
+#include <asm/types.h>
 #include <linux/config.h>
 #include <linux/module.h>
 #include <linux/ipmi.h>
--- linux-2.4.20/drivers/i2c/i2c-velleman.c.org	Sat Jun  7 13:14:12 2003
+++ linux-2.4.20/drivers/i2c/i2c-velleman.c	Sat Jun  7 14:21:02 2003
@@ -20,6 +20,7 @@
 
 /* $Id$ */
 
+#include <asm/param.h>
 #include <linux/kernel.h>
 #include <linux/ioport.h>
 #include <linux/module.h>
diff -Nur --exclude='*.orig' linux-2.4.20.org/arch/alpha/kernel/alpha_ksyms.c linux-2.4.20/arch/alpha/kernel/alpha_ksyms.c
--- linux-2.4.20.org/arch/alpha/kernel/alpha_ksyms.c	Tue Jun 17 10:17:35 2003
+++ linux-2.4.20/arch/alpha/kernel/alpha_ksyms.c	Mon Jun 16 15:54:59 2003
@@ -269,3 +269,5 @@
 
 EXPORT_SYMBOL(plat_node_data);
 
+EXPORT_SYMBOL(ioremap);
+EXPORT_SYMBOL(iounmap);
diff -Nur --exclude='*.orig' linux-2.4.20.org/drivers/net/wan/farsync.c linux-2.4.20/drivers/net/wan/farsync.c
--- linux-2.4.20.org/drivers/net/wan/farsync.c	Fri Dec 21 18:41:55 2001
+++ linux-2.4.20/drivers/net/wan/farsync.c	Mon Jun 16 16:08:26 2003
@@ -23,6 +23,7 @@
 #include <linux/if_arp.h>
 #include <asm/uaccess.h>
 #include <net/syncppp.h>
+#include <asm/io.h>
 
 #include "farsync.h"
 
diff -Nur --exclude='*.orig' linux-2.4.20.org/drivers/net/wireless/hostap_pci.c linux-2.4.20/drivers/net/wireless/hostap_pci.c
--- linux-2.4.20.org/drivers/net/wireless/hostap_pci.c	Tue Jun 17 10:17:27 2003
+++ linux-2.4.20/drivers/net/wireless/hostap_pci.c	Mon Jun 16 21:13:05 2003
@@ -24,6 +24,8 @@
 #include <linux/ioport.h>
 #include <linux/pci.h>
 
+#include <asm/io.h>
+
 #include "hostap_wlan.h"
 
 
diff -Nur --exclude='*.orig' linux-2.4.20.org/drivers/net/wireless/hostap_plx.c linux-2.4.20/drivers/net/wireless/hostap_plx.c
--- linux-2.4.20.org/drivers/net/wireless/hostap_plx.c	Tue Jun 17 10:17:27 2003
+++ linux-2.4.20/drivers/net/wireless/hostap_plx.c	Mon Jun 16 21:12:37 2003
@@ -27,6 +27,8 @@
 #include <linux/ioport.h>
 #include <linux/pci.h>
 
+#include <asm/io.h>
+
 #include "hostap_wlan.h"
 
 
diff -Nur --exclude='*.orig' linux-2.4.20.org/drivers/sound/forte.c linux-2.4.20/drivers/sound/forte.c
--- linux-2.4.20.org/drivers/sound/forte.c	Fri Nov 29 00:53:14 2002
+++ linux-2.4.20/drivers/sound/forte.c	Mon Jun 16 21:15:00 2003
@@ -54,6 +54,8 @@
 #include <asm/uaccess.h>
 #include <asm/hardirq.h>
 
+#include <asm/io.h>
+
 #define DRIVER_NAME	"forte"
 #define DRIVER_VERSION 	"$Id$"
 #define PFX 		DRIVER_NAME ": "
