--- linux/fs/ext3/file.c~	Sat Jul 28 18:07:14 2001
+++ linux/fs/ext3/file.c	Sat Jul 28 23:20:34 2001
@@ -65,11 +65,7 @@
 }
 
 struct file_operations ext3_file_operations = {
-#ifdef EXT3_NEW_QUOTAS	/* Really -ac */
-	llseek:		generic_file_llseek,	/* BKL held */
-#else
 	llseek:		default_llseek,		/* BKL held */
-#endif
 	read:		generic_file_read,	/* BKL not held.  Don't need */
 	write:		ext3_file_write,	/* BKL not held.  Don't need */
 	ioctl:		ext3_ioctl,		/* BKL held */
--- linux/include/linux/ext3_fs.h~	Sat Jul 28 18:24:50 2001
+++ linux/include/linux/ext3_fs.h	Sat Jul 28 23:21:14 2001
@@ -712,9 +712,9 @@
 /*
  * Detect whether we're running old (Linus) quotas or new (Alan) quotas
  */
-#ifdef DQINFOOFF
+/* #ifdef DQINFOOFF */
 #define EXT3_NEW_QUOTAS
-#endif
+/* #endif */
 
 #endif	/* __KERNEL__ */
 
