--- linux/drivers/net/comx-hw-comx.c~	Thu Feb 26 02:20:49 2004
+++ linux/drivers/net/comx-hw-comx.c	Thu Feb 26 02:20:49 2004
@@ -48,7 +48,7 @@
 
 #include <asm/uaccess.h>
 #include <asm/io.h>
-#include <asm/delay.h>
+#include <linux/delay.h>
 #include <linux/config.h>
 #include <linux/module.h>
 #include <linux/version.h>
--- linux/drivers/net/c101.c~	Thu Feb 26 02:22:52 2004
+++ linux/drivers/net/c101.c	Thu Feb 26 02:22:52 2004
@@ -26,10 +26,10 @@
 #include <linux/types.h>
 #include <linux/string.h>
 #include <linux/errno.h>
+#include <linux/delay.h>
 #include <linux/init.h>
 #include <linux/netdevice.h>
 #include <asm/io.h>
-#include <asm/delay.h>
 
 #include "syncppp.h"
 #include <linux/hdlc.h>
--- linux/drivers/scsi/seagate.c~	Thu Feb 26 18:31:32 2004
+++ linux/drivers/scsi/seagate.c	Thu Feb 26 18:31:32 2004
@@ -88,6 +88,7 @@
 #include <linux/sched.h>
 #include <linux/string.h>
 #include <linux/proc_fs.h>
+#include <linux/delay.h>
 #include <linux/init.h>
 
 #include <linux/blk.h>
@@ -99,7 +100,6 @@
 #include <asm/uaccess.h>
 #include "sd.h"
 #include <scsi/scsi_ioctl.h>
-#include <asm/delay.h>
 
 #ifdef DEBUG
 #define DPRINTK( when, msg... ) do { if ( (DEBUG & (when)) == (when) ) printk( msg ); } while (0)
