diff -urN linux-2.5.74-bk1/drivers/ieee1394/sbp2.c linux-2.5.74-bk2/drivers/ieee1394/sbp2.c
--- linux-2.5.74-bk1/drivers/ieee1394/sbp2.c	2003-07-02 13:41:44.000000000 -0700
+++ linux-2.5.74-bk2/drivers/ieee1394/sbp2.c	2003-07-04 05:07:11.000000000 -0700
@@ -56,6 +56,8 @@
 #include <linux/smp_lock.h>
 #include <linux/init.h>
 #include <linux/version.h>
+#include <linux/pci.h>
+
 #include <asm/current.h>
 #include <asm/uaccess.h>
 #include <asm/io.h>
