diff -urNp linux-660/arch/i386/kernel/edd.c linux-661/arch/i386/kernel/edd.c
--- linux-660/arch/i386/kernel/edd.c	
+++ linux-661/arch/i386/kernel/edd.c	
@@ -11,7 +11,7 @@
  * fn41 - Check Extensions Present and
  * fn48 - Get Device Parametes with EDD extensions
  * made in setup.S, copied to safe structures in setup.c,
- * and presents it in driverfs.
+ * and presents it in /proc.
  *
  * This program is free software; you can redistribute it and/or modify
  * it under the terms of the GNU General Public License v2.0 as published by
@@ -26,7 +26,6 @@
 
 /*
  * TODO:
- * - Convert to using /proc instead of driverfs
  * - move edd.[ch] to better locations if/when one is decided
  * - keep current with 2.5 EDD code changes
  */
@@ -47,7 +46,7 @@ MODULE_AUTHOR("Matt Domsch <Matt_Domsch@
 MODULE_DESCRIPTION("proc interface to BIOS EDD information");
 MODULE_LICENSE("GPL");
 
-#define EDD_VERSION "0.08 2003-Jan-05"
+#define EDD_VERSION "0.09 2003-Jan-22"
 #define EDD_DEVICE_NAME_SIZE 16
 #define REPORT_URL "http://domsch.com/linux/edd30/results.html"
 
@@ -601,7 +600,7 @@ edd_device_register(struct edd_device *e
 }
 
 /**
- * edd_init() - creates driverfs tree of EDD data
+ * edd_init() - creates /proc tree of EDD data
  *
  * This assumes that eddnr and edd were
  * assigned in setup.c already.
diff -urNp linux-660/arch/i386/kernel/setup.c linux-661/arch/i386/kernel/setup.c
--- linux-660/arch/i386/kernel/setup.c	
+++ linux-661/arch/i386/kernel/setup.c	
@@ -698,7 +698,7 @@ static int __init copy_e820_map(struct e
 
 #if defined(CONFIG_EDD) || defined(CONFIG_EDD_MODULE)
 unsigned char eddnr;
-struct edd_info edd[EDDNR];
+struct edd_info edd[EDDMAXNR];
 /**
  * copy_edd() - Copy the BIOS EDD information
  *              from empty_zero_page into a safe place.
diff -urNp linux-660/include/asm-i386/edd.h linux-661/include/asm-i386/edd.h
--- linux-660/include/asm-i386/edd.h	
+++ linux-661/include/asm-i386/edd.h	
@@ -165,7 +165,7 @@ struct edd_info {
 	struct edd_device_params params;
 } __attribute__ ((packed));
 
-extern struct edd_info edd[EDDNR];
+extern struct edd_info edd[EDDMAXNR];
 extern unsigned char eddnr;
 #endif				/*!__ASSEMBLY__ */
 
