--- linux/include/linux/agp_backend.h.chmou	Mon Dec 11 07:21:13 2000
+++ linux/include/linux/agp_backend.h	Mon Dec 11 07:21:44 2000
@@ -45,6 +45,7 @@
 	INTEL_BX,
 	INTEL_GX,
 	INTEL_I810,
+	INTEL_I815,
 	INTEL_I840,
 	VIA_GENERIC,
 	VIA_VP3,
--- linux/drivers/char/agp/agp_backend.h.chmou	Mon Dec 11 07:21:10 2000
+++ linux/drivers/char/agp/agp_backend.h	Mon Dec 11 07:21:44 2000
@@ -45,6 +45,7 @@
 	INTEL_BX,
 	INTEL_GX,
 	INTEL_I810,
+	INTEL_I815,
 	INTEL_I840,
 	VIA_GENERIC,
 	VIA_VP3,
--- linux/drivers/char/agp/agpgart_be.c.chmou	Mon Dec 11 07:21:10 2000
+++ linux/drivers/char/agp/agpgart_be.c	Mon Dec 11 07:21:44 2000
@@ -2026,6 +2026,13 @@
 		"Intel",
 		"440GX",
 		intel_generic_setup },
+	/* could we add support for PCI_DEVICE_ID_INTEL_815_1 too ? */
+	{ PCI_DEVICE_ID_INTEL_815_0,
+		PCI_VENDOR_ID_INTEL,
+		INTEL_I815,
+		"Intel",
+		"i815",
+		intel_generic_setup },
 	{ PCI_DEVICE_ID_INTEL_840_0,
 		PCI_VENDOR_ID_INTEL,
 		INTEL_I840,
--- linux/drivers/char/Config.in.chmou	Mon Dec 11 07:21:14 2000
+++ linux/drivers/char/Config.in	Mon Dec 11 07:21:44 2000
@@ -131,8 +131,8 @@
 if [ "$CONFIG_EXPERIMENTAL" = "y" ]; then
    tristate '/dev/agpgart (AGP Support) (EXPERIMENTAL)' CONFIG_AGP n
    if [ "$CONFIG_AGP" != "n" ]; then
-      bool '   Intel 440LX/BX/GX support' CONFIG_AGP_INTEL
-      bool '   Intel I810/I810 DC100/I810e support' CONFIG_AGP_I810
+      bool '   Intel 440LX/BX/GX and I815/820 support' CONFIG_AGP_INTEL
+      bool '   Intel I810/I815 (on-board video) support' CONFIG_AGP_I810
       bool '   VIA VP3/MVP3/Apollo Pro support' CONFIG_AGP_VIA
       bool '   AMD Irongate support' CONFIG_AGP_AMD
       bool '   Generic SiS support' CONFIG_AGP_SIS
