--- linux-2.4.20/arch/ppc/Makefile.org	Fri Nov 29 00:53:11 2002
+++ linux-2.4.20/arch/ppc/Makefile	Thu Feb 20 12:54:13 2003
@@ -25,7 +25,7 @@
 CPPFLAGS	:= $(CPPFLAGS) -I$(TOPDIR)/arch/$(ARCH)
 AFLAGS		:= $(AFLAGS) -I$(TOPDIR)/arch/$(ARCH)
 CFLAGS		:= $(CFLAGS) -I$(TOPDIR)/arch/$(ARCH) -fsigned-char \
-		-msoft-float -pipe -ffixed-r2 -Wno-uninitialized \
+		-pipe -ffixed-r2 -Wno-uninitialized \
 		-mmultiple -mstring
 HOSTCFLAGS	+= -I$(TOPDIR)/arch/$(ARCH)
 CPP		= $(CC) -E $(CFLAGS)
