--- linux-2.6.19/drivers/media/video/cx88/cx88-tvaudio.c~	2006-11-29 22:57:37.000000000 +0100
+++ linux-2.6.19/drivers/media/video/cx88/cx88-tvaudio.c	2006-12-12 13:00:23.000000000 +0100
@@ -150,6 +150,7 @@ static void set_audio_finish(struct cx88
 		case CX88_BOARD_KWORLD_HARDWARE_MPEG_TV_XPERT:
 		case CX88_BOARD_PIXELVIEW_PLAYTV_P7000:
 		case CX88_BOARD_ASUS_PVR_416:
+		case CX88_BOARD_LEADTEK_PVR2000:
 			cx_clear(AUD_CTL, EN_I2SIN_ENABLE);
 			break;
 		default:
