--- linux-2.6.11.5/fs/iriver/inode.c~	2005-03-25 12:39:37.000732080 +0100
+++ linux-2.6.11.5/fs/iriver/inode.c	2005-03-25 13:12:05.700484688 +0100
@@ -135,7 +135,9 @@
 //Used in RAMFS
 static struct backing_dev_info iriver_backing_dev_info = {
 	.ra_pages	= 0,	/* No readahead */
-	.memory_backed	= 1,	/* Does not contribute to dirty memory */
+	.capabilities	= BDI_CAP_NO_ACCT_DIRTY | BDI_CAP_NO_WRITEBACK |
+			  BDI_CAP_MAP_DIRECT | BDI_CAP_MAP_COPY |
+			  BDI_CAP_READ_MAP | BDI_CAP_WRITE_MAP | BDI_CAP_EXEC_MAP,
 };
 
 static inline void cleanup(struct super_block *sb) {
