 # fixes CAN-2003-0464
--- linux-2.4.20/net/sunrpc/svcsock.c.orig	Fri Nov 29 00:53:16 2002
+++ linux-2.4.20/net/sunrpc/svcsock.c	Tue Sep  9 15:42:19 2003
@@ -1229,7 +1229,8 @@
 		return error;
 
 	if (sin != NULL) {
-		sock->sk->reuse = 1; /* allow address reuse */
+		if (type == SOCK_STREAM)
+			sock->sk->reuse = 1; /* allow address reuse */
 		error = sock->ops->bind(sock, (struct sockaddr *) sin,
 						sizeof(*sin));
 		if (error < 0)
