--- linux-2.4.19/net/netsyms.c.org	Tue Nov 26 14:57:52 2002
+++ linux-2.4.19/net/netsyms.c	Wed Nov 27 12:52:03 2002
@@ -603,11 +603,6 @@
 EXPORT_SYMBOL(grsec_enable_randsrc);
 EXPORT_SYMBOL(grsec_enable_randrpc);
 
-#ifdef CONFIG_UNIX_MODULE
-EXPORT_SYMBOL(gr_check_create);
-EXPORT_SYMBOL(gr_handle_create);
-#endif
-
 #if defined(CONFIG_NET_RADIO) || defined(CONFIG_NET_PCMCIA_RADIO)
 /* Don't include the whole header mess for a single function */
 extern void wireless_send_event(struct net_device *dev, unsigned int cmd, union iwreq_data *wrqu, char *extra);
