Fix possible information leak through e1000 driver (CAN-2004-0535)

--- linux-2.4.26/drivers/net/e1000/e1000_ethtool.c.orig	Wed Apr 14 15:05:30 2004
+++ linux-2.4.26/drivers/net/e1000/e1000_ethtool.c	Fri Jun 25 13:46:51 2004
@@ -1421,6 +1421,9 @@
 
 		if(copy_from_user(&regs, addr, sizeof(regs)))
 			return -EFAULT;
+		memset(regs_buff, 0, sizeof(regs_buff));
+		if (regs.len > E1000_REGS_LEN)
+			regs.len = E1000_REGS_LEN;
 		e1000_ethtool_gregs(adapter, &regs, regs_buff);
 		if(copy_to_user(addr, &regs, sizeof(regs)))
 			return -EFAULT;
