--- linux-2.4.21/crypto/ciphers/des/des_locl.h.orig	2003-08-16 23:32:05.000000000 +0200
+++ linux-2.4.21/crypto/ciphers/des/des_locl.h	2003-08-16 23:28:11.000000000 +0200
@@ -98,11 +98,11 @@
 #endif
 
 /* Unroll the inner loop, this sometimes helps, sometimes hinders.
- * Very mucy CPU dependant */
+ * Very mucy CPU dependant *
 #ifndef DES_UNROLL
 #define DES_UNROLL
 #endif
-
+*/
 /* These default values were supplied by
  * Peter Gutman <pgut001@cs.auckland.ac.nz>
  * They are only used if nothing else has been defined */
@@ -116,7 +116,7 @@
 #if defined( sun )		/* Newer Sparc's */
   #define DES_PTR
   #define DES_RISC1
-  #define DES_UNROLL
+  /*#define DES_UNROLL*/
 #elif defined( __ultrix )	/* Older MIPS */
   #define DES_PTR
   #define DES_RISC2
