--- Linux/fs/partitions/check.c	Wed Jul  4 14:00:15 2001
+++ linux/fs/partitions/check.c	Wed Jul  4 14:02:32 2001
@@ -164,6 +164,15 @@
  			sprintf(buf, "%s/c%dd%dp%d", maj, ctlr, disk, part);
  		return buf;
  	}
+	if (hd->major == ATARAID_MAJOR) {
+ 		int disk = minor >> hd->minor_shift;
+ 		int part = minor & (( 1 << hd->minor_shift) - 1);
+ 		if (part == 0)
+ 			sprintf(buf, "%s/d%d", maj, disk);
+ 		else
+ 			sprintf(buf, "%s/d%dp%d", maj, disk, part);
+ 		return buf;
+ 	}
 	if (part)
 		sprintf(buf, "%s%c%d", maj, unit, part);
 	else
