--- linux-2.5.0or/net/core/dev.c	Thu Nov  8 00:39:36 2001
+++ linux/net/core/dev.c	Tue Dec 11 21:30:49 2001
@@ -1688,7 +1688,7 @@
 	int size;

 	if (stats)
-		size = sprintf(buffer, "%6s:%8lu %7lu %4lu %4lu %4lu %5lu %10lu %9lu %8lu %7lu %4lu %4lu %4lu %5lu %7lu %10lu\n",
+		size = sprintf(buffer, "%6s:%14llu %7lu %4lu %4lu %4lu %5lu %10lu %9lu %14llu %7lu %4lu %4lu %4lu %5lu %7lu %10lu\n",
  		   dev->name,
 		   stats->rx_bytes,
 		   stats->rx_packets, stats->rx_errors,
--- linux-2.5.0or/include/linux/netdevice.h	Thu Nov 22 21:47:09 2001
+++ linux/include/linux/netdevice.h	Tue Dec 11 21:24:54 2001
@@ -97,8 +97,9 @@
 {
 	unsigned long	rx_packets;		/* total packets received	*/
 	unsigned long	tx_packets;		/* total packets transmitted	*/
-	unsigned long	rx_bytes;		/* total bytes received 	*/
-	unsigned long	tx_bytes;		/* total bytes transmitted	*/
+        /* rx and tx counters fixed up to 64-bit */
+        long long       rx_bytes;               /* total bytes received         */
+        long long       tx_bytes;               /* total bytes transmitted      */
 	unsigned long	rx_errors;		/* bad packets received		*/
 	unsigned long	tx_errors;		/* packet transmit problems	*/
 	unsigned long	rx_dropped;		/* no space in linux buffers	*/

