diff -urNp linux-1230/drivers/md/lvm-snap.c linux-1250/drivers/md/lvm-snap.c
--- linux-1230/drivers/md/lvm-snap.c	
+++ linux-1250/drivers/md/lvm-snap.c	
@@ -113,6 +113,8 @@ lvm_find_exception_table(kdev_t org_dev,
 	lv_block_exception_t * ret;
 	int i = 0;
 
+	if (!hash_table)
+		BUG();
 	hash_table = &hash_table[hashfn(org_dev, org_start, mask, chunk_size)];
 	ret = NULL;
 	for (next = hash_table->next; next != hash_table; next = next->next)
