--- linux-2.6.16.14/include/linux/sched.h~	2006-05-05 16:35:38.000000000 +0200
+++ linux-2.6.16.14/include/linux/sched.h	2006-05-05 17:33:02.000000000 +0200
@@ -1486,8 +1486,10 @@
 static inline void frozen_process(struct task_struct *p) { BUG(); }
 
 static inline void refrigerator(void) {}
+#ifndef freeze_processes
 static inline int freeze_processes(void) { BUG(); return 0; }
 static inline void thaw_processes(int which_threads) {}
+#endif
 
 static inline int try_to_freeze(void) { return 0; }
 
--- linux-2.6.16.14/include/linux/suspend.h~	2006-05-05 16:35:34.000000000 +0200
+++ linux-2.6.16.14/include/linux/suspend.h	2006-05-05 17:33:35.000000000 +0200
@@ -59,8 +59,10 @@
 	printk("Warning: fake suspend called\n");
 	return -EPERM;
 }
+#ifndef freeze_processes
 static inline int freeze_processes(void) { return 0; }
 static inline void thaw_processes(int which_threads) { }
+#endif
 #endif /* CONFIG_PM */
 
 extern char resume2_file[256];
