--- linux-2.6.28/fs/proc/Kconfig~       2008-11-20 23:26:34.000000000 +0100
+++ linux-2.6.28/fs/proc/Kconfig        2008-12-01 20:37:12.000000000 +0100
@@ -59,8 +59,8 @@
 	  limited in memory.
 
 config PROC_PAGE_MONITOR
- 	default n
-	depends on PROC_FS && MMU && !GRKERNSEC
+ 	default y
+	depends on PROC_FS && MMU
 	bool "Enable /proc page monitoring" if EMBEDDED
  	help
 	  Various /proc files exist to monitor process memory utilization:
--- linux-2.6.28/mm/slab.c~	2009-01-18 02:10:12.395711069 +0100
+++ linux-2.6.28/mm/slab.c	2009-01-18 02:18:05.632401077 +0100
@@ -4496,8 +4496,15 @@
 
 static int __init slab_proc_init(void)
 {
-#if !defined(CONFIG_GRKERNSEC_PROC_ADD)
-	proc_create("slabinfo",S_IWUSR|S_IRUGO,NULL,&proc_slabinfo_operations);
+	int gr_mode = S_IWUSR|S_IRUGO;
+#ifdef CONFIG_GRKERNSEC_PROC_USER
+	gr_mode = S_IRUSR;
+#elif defined(CONFIG_GRKERNSEC_PROC_USERGROUP)
+	gr_mode = S_IRUSR | S_IRGRP;
+#endif
+
+#ifdef CONFIG_GRKERNSEC_PROC_ADD
+	proc_create("slabinfo",gr_mode,NULL,&proc_slabinfo_operations);
 #ifdef CONFIG_DEBUG_SLAB_LEAK
 	proc_create("slab_allocators", 0, NULL, &proc_slabstats_operations);
 #endif
