diff -Nur linux-2.4.20.org/include/linux/netfilter_ipv4/ip_conntrack_tuple.h linux-2.4.20/include/linux/netfilter_ipv4/ip_conntrack_tuple.h
--- linux-2.4.20.org/include/linux/netfilter_ipv4/ip_conntrack_tuple.h	2003-09-22 11:57:14.000000000 +0000
+++ linux-2.4.20/include/linux/netfilter_ipv4/ip_conntrack_tuple.h	2003-09-22 11:32:04.000000000 +0000
@@ -74,8 +74,8 @@
  * really care about is the  source/destination unions */
 #define IP_CT_TUPLE_BLANK(tuple) 				\
 	do {							\
-		tuple->src.u.all = 0;				\
-		tuple->dst.u.all = 0;				\
+		(tuple)->src.u.all = 0;				\
+		(tuple)->dst.u.all = 0;				\
 	} while (0)
 
 enum ip_conntrack_dir
diff -Nur linux-2.4.20.org/net/ipv6/netfilter/Config.in linux-2.4.20/net/ipv6/netfilter/Config.in
--- linux-2.4.20.org/net/ipv6/netfilter/Config.in	2003-09-22 11:57:14.000000000 +0000
+++ linux-2.4.20/net/ipv6/netfilter/Config.in	2003-09-22 11:59:43.000000000 +0000
@@ -92,7 +92,7 @@
   if [ "$CONFIG_IP6_NF_MANGLE" != "n" ]; then
 #    dep_tristate '    TOS target support' CONFIG_IP6_NF_TARGET_TOS $CONFIG_IP_NF_MANGLE
     dep_tristate '    MARK target support' CONFIG_IP6_NF_TARGET_MARK $CONFIG_IP6_NF_MANGLE
-    dep_mbool '    ROUTE target support' CONFIG_IP6_NF_TARGET_ROUTE $CONFIG_IP6_NF_MANGLE
+    dep_tristate '    ROUTE target support' CONFIG_IP6_NF_TARGET_ROUTE $CONFIG_IP6_NF_MANGLE
     dep_tristate '    IMQ target support' CONFIG_IP6_NF_TARGET_IMQ $CONFIG_IP6_NF_MANGLE
   fi
   #dep_tristate '  LOG target support' CONFIG_IP6_NF_TARGET_LOG $CONFIG_IP6_NF_IPTABLES
