--- linux-2.4.20-orig/drivers/ide/ide-cd.h	2002-12-10 17:46:28 +0900
+++ linux-2.4.20/drivers/ide/ide-cd.h	2003-05-16 00:59:53 +0900
@@ -437,7 +437,7 @@
 
 	byte     curlba[3];
 	byte     nslots;
-	__u8 short slot_tablelen;
+	__u8     slot_tablelen;
 };
 
 
--- linux-2.4.20-orig/net/core/rtnetlink.c	2002-12-09 16:38:46 +0900
+++ linux-2.4.20/net/core/rtnetlink.c	2003-05-16 01:31:20 +0900
@@ -394,7 +394,7 @@
  * Malformed skbs with wrong lengths of messages are discarded silently.
  */
 
-extern __inline__ int rtnetlink_rcv_skb(struct sk_buff *skb)
+__inline__ int rtnetlink_rcv_skb(struct sk_buff *skb)
 {
 	int err;
 	struct nlmsghdr * nlh;
