--- linux-2.6.8-rc1/net/ipv6/netfilter/ip6t_REJECT.c.orig	2004-07-12 07:43:10.738370632 +0200
+++ linux-2.6.8-rc1/net/ipv6/netfilter/ip6t_REJECT.c	2004-07-12 07:42:55.497687568 +0200
@@ -22,6 +22,7 @@
 #include <net/ipv6.h>
 #include <net/tcp.h>
 #include <net/icmp.h>
+#include <net/ip6_checksum.h>
 #include <net/ip6_fib.h>
 #include <net/ip6_route.h>
 #include <net/flow.h>
--- linux-2.6.8.1.orig/drivers/mtd/devices/Makefile	2004-08-14 12:54:46.000000000 +0200
+++ linux-2.6.8.1/drivers/mtd/devices/Makefile	2004-09-25 22:46:42.868745888 +0200
@@ -10,10 +10,10 @@
 # here where previously there was none.  We now have to ensure that
 # doc200[01].o are linked before docprobe.o
 
-obj-$(CONFIG_MTD_DOC2000)	+= doc2000.o
-obj-$(CONFIG_MTD_DOC2001)	+= doc2001.o
-obj-$(CONFIG_MTD_DOC2001PLUS)	+= doc2001plus.o
-obj-$(CONFIG_MTD_DOCPROBE)	+= docprobe.o
+obj-$(CONFIG_MTD_DOC2000)	+= doc2000.o docecc.o
+obj-$(CONFIG_MTD_DOC2001)	+= doc2001.o docecc.o
+obj-$(CONFIG_MTD_DOC2001PLUS)	+= doc2001plus.o docecc.o
+obj-$(CONFIG_MTD_DOCPROBE)	+= docprobe.o docecc.o
 obj-$(CONFIG_MTD_DOCECC)	+= docecc.o
 obj-$(CONFIG_MTD_SLRAM)		+= slram.o
 obj-$(CONFIG_MTD_PHRAM)		+= phram.o
