--- linux-2.6.7/arch/sparc/kernel/sparc_ksyms.c.orig	2004-07-10 09:36:21.000000000 +0000
+++ linux-2.6.7/arch/sparc/kernel/sparc_ksyms.c	2004-07-10 13:48:28.000000000 +0000
@@ -215,9 +215,10 @@
 EXPORT_SYMBOL(pci_dma_sync_sg_for_cpu);
 EXPORT_SYMBOL(pci_dma_sync_sg_for_device);
 /* Actually, ioremap/iounmap are not PCI specific. But it is ok for drivers. */
+#endif
+/* or not? not all MTD drivers depend on PCI */
 EXPORT_SYMBOL(ioremap);
 EXPORT_SYMBOL(iounmap);
-#endif
 
 /* in arch/sparc/mm/highmem.c */
 #ifdef CONFIG_HIGHMEM
@@ -253,9 +254,13 @@
 EXPORT_SYMBOL(prom_getint);
 EXPORT_SYMBOL(prom_getintdefault);
 EXPORT_SYMBOL(prom_finddevice);
+EXPORT_SYMBOL(prom_halt);
+EXPORT_SYMBOL(prom_printf);
+EXPORT_SYMBOL(prom_apply_generic_ranges);
 EXPORT_SYMBOL(romvec);
 EXPORT_SYMBOL(__prom_getchild);
 EXPORT_SYMBOL(__prom_getsibling);
+EXPORT_SYMBOL(sun_do_break);
 
 /* sparc library symbols */
 EXPORT_SYMBOL(bcopy);
