--- linux/drivers/net/pcnet32.c.org	Wed Apr 18 12:47:35 2001
+++ linux/drivers/net/pcnet32.c	Wed Apr 18 12:53:33 2001
@@ -626,7 +626,7 @@
 	ltint = 1;
     }
     
-    dev = init_etherdev(NULL, 0);
+    dev = alloc_etherdev(0);
     if(dev==NULL)
 	return -ENOMEM;
 
@@ -788,8 +788,8 @@
     lp->next = pcnet32_dev;
     pcnet32_dev = dev;
 
-    /* Fill in the generic fields of the device structure. */
-    ether_setup(dev);
+    /* Register the device structure. */
+    register_netdev(dev);
     return 0;
 }
 
