--- linux.orig/net/Makefile	Fri Nov 30 14:57:51 2001
+++ linux/net/Makefile	Fri Nov 30 15:04:31 2001
@@ -10,7 +10,7 @@
 MOD_SUB_DIRS := ipv4
 ALL_SUB_DIRS := 802 ax25 bridge core ethernet ipv4 ipv6 ipx unix appletalk \
 		netrom rose lapb x25 wanrouter netlink sched packet sunrpc \
-		econet irda 802_1Q #decnet
+		econet atm irda 802_1Q #decnet
 SUB_DIRS     := core ethernet sched
 MOD_LIST_NAME := NET_MISC_MODULES
 
