--- linux-2.6.9-rc2/include/asm-alpha/io.h.orig	2004-09-28 15:28:32.376274856 +0200
+++ linux-2.6.9-rc2/include/asm-alpha/io.h	2004-09-28 15:34:56.009953672 +0200
@@ -5,6 +5,7 @@
 
 #include <linux/config.h>
 #include <linux/kernel.h>
+#include <linux/mm.h>
 #include <asm/compiler.h>
 #include <asm/system.h>
 #include <asm/pgtable.h>
@@ -127,6 +128,15 @@
 	return (long)address <= 0 ? NULL : virt;
 }
 
+#define isa_virt_to_bus virt_to_bus
+#define isa_bus_to_virt bus_to_virt
+static inline unsigned long isa_page_to_bus(struct page *page)
+{
+	unsigned long phys = page_to_phys(page);
+	unsigned long bus = phys + __direct_map_base;
+	return phys <= __direct_map_size ? bus : 0;
+}
+
 /*
  * There are different chipsets to interface the Alpha CPUs to the world.
  */
