--- linux-2.4.19/drivers/usb/hid-input.c.org	Sun Nov 17 09:16:51 2002
+++ linux-2.4.19/drivers/usb/hid-input.c	Sun Nov 17 09:17:28 2002
@@ -271,7 +271,7 @@
 	set_bit(usage->type, input->evbit);
 
 	while (usage->code <= max && test_and_set_bit(usage->code, bit)) {
-		usage->code = find_next_zero_bit(bit, usage->code, max + 1);
+		usage->code = find_next_zero_bit(bit, max + 1, usage->code);
 	}
 
 	if (usage->code > max) return;
