--- drivers/net/irda/nsc-ircc.c.old	2003-08-13 01:01:56.000000000 +0200
+++ drivers/net/irda/nsc-ircc.c	2003-08-13 01:03:25.000000000 +0200
@@ -704,11 +704,12 @@
 		   driver_name, version);
 
 	/* Should be 0x2? */
+#if 0
 	if (0x20 != (version & 0xf0)) {
 		ERROR("%s, Wrong chip version %02x\n", driver_name, version);
 		return -1;
 	}
-
+#endif
 	/* Switch to advanced mode */
 	switch_bank(iobase, BANK2);
 	outb(ECR1_EXT_SL, iobase+ECR1);
