diff -urN l.org/arch/i386/kernel/acpi/sleep.c l/arch/i386/kernel/acpi/sleep.c
--- l.org/arch/i386/kernel/acpi/sleep.c	2004-07-26 23:41:38.467403872 +0200
+++ l/arch/i386/kernel/acpi/sleep.c	2004-07-26 23:42:27.698919544 +0200
@@ -78,8 +78,7 @@
 		return;
 	}
 #ifdef CONFIG_X86_PAE
-	printk(KERN_ERR "ACPI: S3 and PAE do not like each other for now, S3 disabled.\n");
-	return;
+	printk(KERN_ERR "ACPI: S3 and PAE do not like each other for some machines, be warned.\n");
 #endif
 	acpi_wakeup_address = (unsigned long)alloc_bootmem_low(PAGE_SIZE);
 	if (!acpi_wakeup_address)
