--- a/drivers/char/vt_ioctl.c	2005-03-30 11:51:25.000000000 +0200
+++ b/drivers/char/vt_ioctl.c	2005-03-30 12:41:31.000000000 +0200
@@ -367,7 +367,14 @@
 	void __user *up = (void __user *)arg;
 	int i, perm;
 	
-	console = vc->vc_num;
+	acquire_console_sem();
+	if (!vc) {
+		printk(KERN_ERR "vt_ioctl: argh, driver_data is NULL !\n");
+		release_console_sem();
+		return -EINVAL;
+	}
+ 	console = vc->vc_num;
+	release_console_sem();
 
 	if (!vc_cons_allocated(console)) 	/* impossible? */
 		return -ENOIOCTLCMD;
