--- linux-2.4.20/arch/alpha/kernel/alpha_ksyms.c.org	Mon Feb 25 19:37:52 2002
+++ linux-2.4.20/arch/alpha/kernel/alpha_ksyms.c	Fri May 24 19:48:14 2002
@@ -264,3 +264,6 @@
 EXPORT_SYMBOL(irongate_ioremap);
 EXPORT_SYMBOL(irongate_iounmap);
 #endif
+
+EXPORT_SYMBOL(plat_node_data);
+
--- linux-2.4.20/drivers/char/ipmi/ipmi_watchdog.c.org	Sat Jun  7 14:11:42 2003
+++ linux-2.4.20/drivers/char/ipmi/ipmi_watchdog.c	Sat Jun  7 14:12:42 2003
@@ -31,6 +31,7 @@
  *  675 Mass Ave, Cambridge, MA 02139, USA.
  */
 
+#include <asm/types.h>
 #include <linux/config.h>
 #include <linux/module.h>
 #include <linux/ipmi.h>
--- linux-2.4.20/drivers/i2c/i2c-velleman.c.org	Sat Jun  7 13:14:12 2003
+++ linux-2.4.20/drivers/i2c/i2c-velleman.c	Sat Jun  7 14:21:02 2003
@@ -20,6 +20,7 @@
 
 /* $Id$ */
 
+#include <asm/param.h>
 #include <linux/kernel.h>
 #include <linux/ioport.h>
 #include <linux/module.h>
