diff -Nur linux.old/arch/sparc64/kernel/sparc64_ksyms.c linux/arch/sparc64/kernel/sparc64_ksyms.c
--- linux.old/arch/sparc64/kernel/sparc64_ksyms.c	Fri Dec 21 18:41:53 2001
+++ linux/arch/sparc64/kernel/sparc64_ksyms.c	Mon Jan 28 10:28:22 2002
@@ -305,12 +305,9 @@
 EXPORT_SYMBOL(sunos_sys_table);
 EXPORT_SYMBOL(sys_call_table32);
 EXPORT_SYMBOL(tl0_solaris);
-EXPORT_SYMBOL(sys_sigsuspend);
-EXPORT_SYMBOL(sys_getppid);
 EXPORT_SYMBOL(svr4_getcontext);
 EXPORT_SYMBOL(svr4_setcontext);
 EXPORT_SYMBOL(prom_cpu_nodes);
-EXPORT_SYMBOL(sys_ioctl);
 EXPORT_SYMBOL(sys32_ioctl);
 EXPORT_SYMBOL(sparc32_open);
 #endif
diff -Nur linux.old/kernel/syscall_ksyms.c linux/kernel/syscall_ksyms.c
--- linux.old/kernel/syscall_ksyms.c	Tue Jan 29 12:49:03 2002
+++ linux/kernel/syscall_ksyms.c	Tue Jan 29 12:41:18 2002
@@ -49,9 +49,11 @@
 EXPORT_SYMBOL(sys_statfs);
 EXPORT_SYMBOL(sys_fstatfs);
 EXPORT_SYMBOL(sys_ftruncate);
-EXPORT_SYMBOL(sys_ftruncate64);
 EXPORT_SYMBOL(sys_truncate);
+#if defined(__sparc__) && !defined(__arch64__) 
+EXPORT_SYMBOL(sys_ftruncate64);
 EXPORT_SYMBOL(sys_truncate64);
+#endif
 #if !defined(__alpha__) && !defined(__ia64__)
 EXPORT_SYMBOL(sys_utime);
 #endif
@@ -177,3 +179,4 @@
 EXPORT_SYMBOL(sys_shmat);
 EXPORT_SYMBOL(sys_shmdt);
 EXPORT_SYMBOL(sys_shmctl);
+
diff -Nur linux/include/asm-sparc64/dma.h.org linux/include/asm-sparc64/dma.h
--- linux/include/asm-sparc64/dma.h.org	Fri Feb  8 14:56:56 2002
+++ linux/include/asm-sparc64/dma.h	Sun Feb 10 11:07:51 2002
@@ -224,4 +224,6 @@
  */
 #define DMA_CHUNK_SIZE 8192
 
+#define DMA_AUTOINIT 0x10
+
 #endif /* !(_ASM_SPARC64_DMA_H) */
diff -Nur linux/drivers/usb/serial/ir-usb.c.org	linux/drivers/usb/serial/ir-usb.c
--- linux/drivers/usb/serial/ir-usb.c.org	Sun Feb 10 10:21:23 2002
+++ linux/drivers/usb/serial/ir-usb.c	Sun Feb 10 10:25:33 2002
@@ -435,7 +435,7 @@
 			case B115200:	baud = SPEED_115200;	break;
 			case B576000:	baud = SPEED_576000;	break;
 			case B1152000:	baud = SPEED_1152000;	break;
-			case B4000000:	baud = SPEED_4000000;	break;
+/*			case B4000000:	baud = SPEED_4000000;	break; */
 			default:
 				err ("ir-usb driver does not support the baudrate (%d) requested", tty_get_baud_rate(port->tty));
 				return;
--- linux/include/asm-sparc64/ide.h.org	Sun Feb 10 12:42:17 2002
+++ linux/include/asm-sparc64/ide.h	Sun Feb 10 12:41:50 2002
@@ -87,7 +87,7 @@
 		unsigned int reserved456: 3;
 		unsigned bit3		: 1;	/* ATA-2 thingy */
 		unsigned int SRST	: 1;	/* host soft reset bit */
-		unsigned int nIEN	: 1;	/* device INTRQ to host *
+		unsigned int nIEN	: 1;	/* device INTRQ to host */
 		unsigned int bit0	: 1;
 	} b;
 } control_t;
