diff -urN linux.org/drivers/md/lvm-snap.c linux/drivers/md/lvm-snap.c
--- linux.org/drivers/md/lvm-snap.c	2003-11-21 18:22:24.000000000 +0100
+++ linux/drivers/md/lvm-snap.c	2003-11-21 18:22:53.000000000 +0100
@@ -120,6 +120,9 @@
 	lv_block_exception_t *ret;
 	int i = 0;
 
+	if (!hash_table)
+	   BUG();
+	
 	hash_table =
 	    &hash_table[hashfn(org_dev, org_start, mask, chunk_size)];
 	ret = NULL;
