--- linux/drivers/block/cmd64x.c.old	Thu Feb  8 03:42:17 2001
+++ linux/drivers/block/cmd64x.c	Thu Feb  8 10:02:16 2001
@@ -19,9 +19,9 @@
 #include <linux/delay.h>
 #include <linux/init.h>
 #include <linux/hdreg.h>
-#include <linux/ide.h>
 
 #include <asm/io.h>
+#include <linux/ide.h>
 
 #include "ide_modes.h"
 
--- linux/drivers/block/ide.c.old	Thu Feb  8 03:42:18 2001
+++ linux/drivers/block/ide.c	Thu Feb  8 09:08:19 2001
@@ -138,13 +138,13 @@
 #endif /* MODULE */
 #include <linux/pci.h>
 #include <linux/delay.h>
-#include <linux/ide.h>
 
 #include <asm/byteorder.h>
 #include <asm/irq.h>
 #include <asm/uaccess.h>
 #include <asm/io.h>
 #include <asm/bitops.h>
+#include <linux/ide.h>
 
 #include "ide_modes.h"
 
--- linux/drivers/block/ide-dma.c.old	Thu Feb  8 10:19:26 2001
+++ linux/drivers/block/ide-dma.c	Thu Feb  8 10:19:37 2001
@@ -86,10 +86,10 @@
 #include <linux/interrupt.h>
 #include <linux/pci.h>
 #include <linux/init.h>
-#include <linux/ide.h>
 
 #include <asm/io.h>
 #include <asm/irq.h>
+#include <linux/ide.h>
 
 #undef CONFIG_BLK_DEV_IDEDMA_TIMEOUT
 
--- linux/drivers/block/ide-features.c.old	Thu Feb  8 03:42:18 2001
+++ linux/drivers/block/ide-features.c	Thu Feb  8 09:51:07 2001
@@ -28,13 +28,13 @@
 #include <linux/pci.h>
 #include <linux/delay.h>
 #include <linux/hdreg.h>
-#include <linux/ide.h>
 
 #include <asm/byteorder.h>
 #include <asm/irq.h>
 #include <asm/uaccess.h>
 #include <asm/io.h>
 #include <asm/bitops.h>
+#include <linux/ide.h>
 
 /*
  * A Verbose noise maker for debugging on the attempted transfer rates.
--- linux/drivers/block/ide-pci.c.old	Thu Feb  8 10:22:01 2001
+++ linux/drivers/block/ide-pci.c	Thu Feb  8 10:22:09 2001
@@ -20,10 +20,10 @@
 #include <linux/interrupt.h>
 #include <linux/pci.h>
 #include <linux/init.h>
-#include <linux/ide.h>
 
 #include <asm/io.h>
 #include <asm/irq.h>
+#include <linux/ide.h>
 
 #define DEVID_PIIXa	((ide_pci_devid_t){PCI_VENDOR_ID_INTEL,   PCI_DEVICE_ID_INTEL_82371FB_0})
 #define DEVID_PIIXb	((ide_pci_devid_t){PCI_VENDOR_ID_INTEL,   PCI_DEVICE_ID_INTEL_82371FB_1})
--- linux/drivers/block/ide-probe.c.old	Thu Feb  8 03:42:18 2001
+++ linux/drivers/block/ide-probe.c	Thu Feb  8 11:16:51 2001
@@ -41,12 +41,12 @@
 #include <linux/malloc.h>
 #include <linux/delay.h>
 #include <linux/mc146818rtc.h> /* CMOS defines */
-#include <linux/ide.h>
 
 #include <asm/byteorder.h>
 #include <asm/irq.h>
 #include <asm/uaccess.h>
 #include <asm/io.h>
+#include <linux/ide.h>
 
 static inline void do_identify (ide_drive_t *drive, byte cmd)
 {
--- linux/drivers/block/ide-proc.c.old	Thu Feb  8 03:42:18 2001
+++ linux/drivers/block/ide-proc.c	Thu Feb  8 11:15:07 2001
@@ -65,9 +65,9 @@
 #include <linux/mm.h>
 #include <linux/pci.h>
 #include <linux/ctype.h>
-#include <linux/ide.h>
 
 #include <asm/io.h>
+#include <linux/ide.h>
 
 #ifndef MIN
 #define MIN(a,b) (((a) < (b)) ? (a) : (b))
--- linux/drivers/block/ide-taskfile.c.old	Thu Feb  8 03:42:18 2001
+++ linux/drivers/block/ide-taskfile.c	Thu Feb  8 09:59:21 2001
@@ -25,13 +25,13 @@
 #include <linux/pci.h>
 #include <linux/delay.h>
 #include <linux/hdreg.h>
-#include <linux/ide.h>
 
 #include <asm/byteorder.h>
 #include <asm/irq.h>
 #include <asm/uaccess.h>
 #include <asm/io.h>
 #include <asm/bitops.h>
+#include <linux/ide.h>
 
 #define DEBUG_TASKFILE	0	/* unset when fixed */
 
--- linux/drivers/block/ns87415.c.old	Thu Feb  8 03:42:18 2001
+++ linux/drivers/block/ns87415.c	Thu Feb  8 10:56:30 2001
@@ -17,9 +17,9 @@
 #include <linux/hdreg.h>
 #include <linux/pci.h>
 #include <linux/delay.h>
-#include <linux/ide.h>
 
 #include <asm/io.h>
+#include <linux/ide.h>
 
 static unsigned int ns87415_count = 0, ns87415_control[MAX_HWIFS] = { 0 };
 
--- linux/drivers/block/ide-tape.c.old	Thu Feb  8 10:26:05 2001
+++ linux/drivers/block/ide-tape.c	Thu Feb  8 11:18:45 2001
@@ -400,7 +400,6 @@
 #include <linux/genhd.h>
 #include <linux/malloc.h>
 #include <linux/pci.h>
-#include <linux/ide.h>
 
 #include <asm/byteorder.h>
 #include <asm/irq.h>
@@ -408,6 +407,7 @@
 #include <asm/io.h>
 #include <asm/unaligned.h>
 #include <asm/bitops.h>
+#include <linux/ide.h>
 
 #define NO_LONGER_REQUIRED	(1)
 
--- linux/drivers/block/ide-disk.c.old	Thu Feb  8 03:42:18 2001
+++ linux/drivers/block/ide-disk.c	Thu Feb  8 10:22:43 2001
@@ -48,12 +48,12 @@
 #include <linux/genhd.h>
 #include <linux/malloc.h>
 #include <linux/delay.h>
-#include <linux/ide.h>
 
 #include <asm/byteorder.h>
 #include <asm/irq.h>
 #include <asm/uaccess.h>
 #include <asm/io.h>
+#include <linux/ide.h>
 #include <asm/ide.h>
 
 #ifdef CONFIG_BLK_DEV_PDC4030
--- linux/drivers/block/ide-cd.c.old	Thu Feb  8 03:42:17 2001
+++ linux/drivers/block/ide-cd.c	Thu Feb  8 10:23:07 2001
@@ -299,13 +299,13 @@
 #include <linux/interrupt.h>
 #include <linux/errno.h>
 #include <linux/cdrom.h>
-#include <linux/ide.h>
 
 #include <asm/irq.h>
 #include <asm/io.h>
 #include <asm/byteorder.h>
 #include <asm/uaccess.h>
 #include <asm/unaligned.h>
+#include <linux/ide.h>
 
 #include "ide-cd.h"
 
--- linux/drivers/block/ide-floppy.c.old	Thu Feb  8 03:42:18 2001
+++ linux/drivers/block/ide-floppy.c	Thu Feb  8 10:26:21 2001
@@ -62,7 +62,6 @@
 #include <linux/genhd.h>
 #include <linux/malloc.h>
 #include <linux/cdrom.h>
-#include <linux/ide.h>
 
 #include <asm/byteorder.h>
 #include <asm/irq.h>
@@ -70,6 +69,7 @@
 #include <asm/io.h>
 #include <asm/unaligned.h>
 #include <asm/bitops.h>
+#include <linux/ide.h>
 
 /*
  *	The following are used to debug the driver.
--- linux/arch/sparc64/config.in.old	Thu Feb  8 10:10:53 2001
+++ linux/arch/sparc64/config.in	Thu Feb  8 10:15:42 2001
@@ -138,6 +138,7 @@
 	  define_bool CONFIG_IDEDMA_AUTO y
 	  define_bool CONFIG_BLK_DEV_NS87415 y
 	  define_bool CONFIG_BLK_DEV_CMD64X y
+	  define_bool CONFIG_BLK_DEV_IDE_MODES y
 	fi
 fi
 
