diff -Nur linux-2.4.19.org/drivers/message/fusion/isense.c linux-2.4.19/drivers/message/fusion/isense.c
--- linux-2.4.19.org/drivers/message/fusion/isense.c	Sat Aug  3 02:39:44 2002
+++ linux-2.4.19/drivers/message/fusion/isense.c	Thu Nov  7 21:04:04 2002
@@ -92,6 +92,7 @@
 EXPORT_NO_SYMBOLS;
 MODULE_AUTHOR(MODULEAUTHOR);
 MODULE_DESCRIPTION(my_NAME);
+MODULE_LICENSE("GPL");
 
 /*=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=-=*/
 int __init isense_init(void)
diff -urNp linux-10020/drivers/net/pcmcia/xircom_tulip_cb.c linux-10030/drivers/net/pcmcia/xircom_tulip_cb.c
--- linux-10020/drivers/net/pcmcia/xircom_tulip_cb.c	
+++ linux-10030/drivers/net/pcmcia/xircom_tulip_cb.c	
@@ -115,7 +115,7 @@ KERN_INFO " unofficial 2.4.x kernel port
 
 MODULE_AUTHOR("Donald Becker <becker@scyld.com>");
 MODULE_DESCRIPTION("Xircom CBE-100 ethernet driver");
-MODULE_LICENSE("GPL v2");
+MODULE_LICENSE("GPL");
 
 MODULE_PARM(debug, "i");
 MODULE_PARM(max_interrupt_work, "i");
diff -Nur linux-2.4.19.org/drivers/net/3c523.c linux-2.4.19/drivers/net/3c523.c
--- linux-2.4.19.org/drivers/net/3c523.c	Mon Feb 25 20:37:59 2002
+++ linux-2.4.19/drivers/net/3c523.c	Thu Nov  7 21:04:05 2002
@@ -114,6 +114,10 @@
 
 #include "3c523.h"
 
+MODULE_AUTHOR("");
+MODULE_DESCRIPTION("net-3-driver for the 3c523 Etherlink/MC card (i82586 Ethernet chip)");
+MODULE_LICENSE("GPL");
+
 /*************************************************************************/
 #define DEBUG			/* debug on */
 #define SYSBUSVAL 0		/* 1 = 8 Bit, 0 = 16 bit - 3c523 only does 16 bit */
diff -Nur linux-2.4.19.org/drivers/net/3c527.c linux-2.4.19/drivers/net/3c527.c
--- linux-2.4.19.org/drivers/net/3c527.c	Mon Feb 25 20:37:59 2002
+++ linux-2.4.19/drivers/net/3c527.c	Thu Nov  7 21:04:05 2002
@@ -111,6 +111,10 @@
 
 #include "3c527.h"
 
+MODULE_AUTHOR("Alan Cox");
+MODULE_DESCRIPTION("3Com Etherlink/MC32 driver");
+MODULE_LICENSE("GPL");
+
 /*
  * The name of the card. Is used for messages and in the requests for
  * io regions, irqs and dma channels
diff -Nur linux-2.4.19.org/drivers/net/tc35815.c linux-2.4.19/drivers/net/tc35815.c
--- linux-2.4.19.org/drivers/net/tc35815.c	Sat Aug  3 02:39:44 2002
+++ linux-2.4.19/drivers/net/tc35815.c	Thu Nov  7 21:15:50 2002
@@ -64,6 +64,9 @@
 #include <linux/proc_fs.h>
 #include <asm/byteorder.h>
 
+MODULE_AUTHOR("MontaVista Software, Inc. Based on skelton.c by Donald Becker.");
+MODULE_DESCRIPTION("TOSHIBA TC35815CF PCI 10/100Mbps ethernet driver");
+MODULE_LICENSE("GPL");
 
 /*
  * The name of the card. Is used for messages and in the requests for
diff -Nur linux-2.4.19.org/drivers/pcmcia/i82092.c linux-2.4.19/drivers/pcmcia/i82092.c
--- linux-2.4.19.org/drivers/pcmcia/i82092.c	Fri Dec 21 18:41:55 2001
+++ linux-2.4.19/drivers/pcmcia/i82092.c	Thu Nov  7 21:04:05 2002
@@ -25,6 +25,9 @@
 #include "i82092aa.h"
 #include "i82365.h"
 
+MODULE_AUTHOR("Arjan Van De Ven <arjanv@redhat.com>");
+MODULE_DESCRIPTION("Driver for Intel I82092AA PCI-PCMCIA bridge.");
+
 /* PCI core routines */
 static struct pci_device_id i82092aa_pci_ids[] = {
 	{
diff -Nur linux-2.4.19.org/drivers/scsi/fd_mcs.c linux-2.4.19/drivers/scsi/fd_mcs.c
--- linux-2.4.19.org/drivers/scsi/fd_mcs.c	Sun Sep 30 21:26:07 2001
+++ linux-2.4.19/drivers/scsi/fd_mcs.c	Thu Nov  7 21:18:49 2002
@@ -97,6 +97,9 @@
 
 #define DRIVER_VERSION "v0.2 by ZP Gu<zpg@castle.net>"
   
+MODULE_DESCRIPTION("Future Domain MCS 600/700 (or IBM OEM) driver");
+MODULE_LICENSE("GPL");
+
 /* START OF USER DEFINABLE OPTIONS */
 
 #define DEBUG            0	/* Enable debugging output */
diff -Nur linux-2.4.19.org/drivers/scsi/ibmmca.c linux-2.4.19/drivers/scsi/ibmmca.c
--- linux-2.4.19.org/drivers/scsi/ibmmca.c	Wed May  2 01:05:00 2001
+++ linux-2.4.19/drivers/scsi/ibmmca.c	Thu Nov  7 21:19:45 2002
@@ -38,6 +38,9 @@
 #include "ibmmca.h"
 #include <linux/config.h>
 
+MODULE_DESCRIPTION("Low Level Linux Driver for the IBM Microchannel SCSI Subsystem");
+MODULE_LICENSE("GPL");
+
 /* current version of this driver-source: */
 #define IBMMCA_SCSI_DRIVER_VERSION "4.0b"
 
diff -Nur linux-2.4.19.org/drivers/sound/rme96xx.c linux-2.4.19/drivers/sound/rme96xx.c
--- linux-2.4.19.org/drivers/sound/rme96xx.c	Sat Aug  3 02:39:44 2002
+++ linux-2.4.19/drivers/sound/rme96xx.c	Thu Nov  7 21:04:05 2002
@@ -66,7 +66,7 @@
 
 MODULE_AUTHOR("Guenter Geiger, geiger@debian.org");
 MODULE_DESCRIPTION("RME9652/36 \"Hammerfall\" Driver");
-/*MODULE_LICENSE("GPL");*/
+MODULE_LICENSE("GPL");
 
 
 #ifdef DEBUG
diff -Nur linux-2.4.19.org/drivers/usb/auerswald.c linux-2.4.19/drivers/usb/auerswald.c
--- linux-2.4.19.org/drivers/usb/auerswald.c	Thu Nov  7 20:47:44 2002
+++ linux-2.4.19/drivers/usb/auerswald.c	Thu Nov  7 21:04:05 2002
@@ -2188,6 +2188,7 @@
 
 MODULE_AUTHOR (DRIVER_AUTHOR);
 MODULE_DESCRIPTION (DRIVER_DESC);
+MODULE_LICENSE("GPL");
 
 module_init (auerswald_init);
 module_exit (auerswald_cleanup);
diff -Nur linux-2.4.19.org/drivers/video/pm3fb.c linux-2.4.19/drivers/video/pm3fb.c
--- linux-2.4.19.org/drivers/video/pm3fb.c	Sat Aug  3 02:39:45 2002
+++ linux-2.4.19/drivers/video/pm3fb.c	Thu Nov  7 21:04:05 2002
@@ -3734,6 +3734,7 @@
 #ifdef MODULE
 MODULE_AUTHOR("Romain Dolbeau");
 MODULE_DESCRIPTION("Permedia3 framebuffer device driver");
+MODULE_LICENSE("GPL");
 static char *mode[PM3_MAX_BOARD];
 MODULE_PARM(mode,PM3_MAX_BOARD_MODULE_ARRAY_STRING);
 MODULE_PARM_DESC(mode,"video mode");
diff -Nur linux-2.4.19.org/net/ipv4/netfilter/ip_nat_talk.c linux-2.4.19/net/ipv4/netfilter/ip_nat_talk.c
--- linux-2.4.19.org/net/ipv4/netfilter/ip_nat_talk.c	Thu Nov  7 20:47:13 2002
+++ linux-2.4.19/net/ipv4/netfilter/ip_nat_talk.c	Thu Nov  7 21:04:05 2002
@@ -38,6 +38,7 @@
 static int ntalk2 = 1;
 MODULE_AUTHOR("Jozsef Kadlecsik <kadlec@blackhole.kfki.hu>");
 MODULE_DESCRIPTION("talk network address translation module");
+MODULE_LICENSE("GPL");
 #ifdef MODULE_PARM
 MODULE_PARM(talk, "i");
 MODULE_PARM_DESC(talk, "support (old) talk protocol");
diff -Nur linux-2.4.19.org/net/ipv4/netfilter/ipt_CONNMARK.c linux-2.4.19/net/ipv4/netfilter/ipt_CONNMARK.c
--- linux-2.4.19.org/net/ipv4/netfilter/ipt_CONNMARK.c	Thu Nov  7 20:45:53 2002
+++ linux-2.4.19/net/ipv4/netfilter/ipt_CONNMARK.c	Thu Nov  7 21:24:05 2002
@@ -10,6 +10,8 @@
 #include <linux/netfilter_ipv4/ipt_CONNMARK.h>
 #include <linux/netfilter_ipv4/ip_conntrack.h>
 
+MODULE_LICENSE("GPL");
+
 static unsigned int
 target(struct sk_buff **pskb,
        unsigned int hooknum,
diff -Nur linux-2.4.19.org/net/ipv4/netfilter/ipt_POOL.c linux-2.4.19/net/ipv4/netfilter/ipt_POOL.c
--- linux-2.4.19.org/net/ipv4/netfilter/ipt_POOL.c	Thu Nov  7 20:45:14 2002
+++ linux-2.4.19/net/ipv4/netfilter/ipt_POOL.c	Thu Nov  7 21:24:31 2002
@@ -22,6 +22,8 @@
 #include <linux/netfilter_ipv4/ip_nat_rule.h>
 #include <linux/netfilter_ipv4/ipt_pool.h>
 
+MODULE_LICENSE("GPL");
+
 #if 0
 #define DEBUGP printk
 #else
diff -Nur linux-2.4.19.org/net/ipv4/netfilter/ipt_connmark.c linux-2.4.19/net/ipv4/netfilter/ipt_connmark.c
--- linux-2.4.19.org/net/ipv4/netfilter/ipt_connmark.c	Thu Nov  7 20:45:53 2002
+++ linux-2.4.19/net/ipv4/netfilter/ipt_connmark.c	Thu Nov  7 21:04:05 2002
@@ -6,6 +6,8 @@
 #include <linux/netfilter_ipv4/ipt_connmark.h>
 #include <linux/netfilter_ipv4/ip_conntrack.h>
 
+MODULE_LICENSE("GPL");
+
 static int
 match(const struct sk_buff *skb,
       const struct net_device *in,
diff -Nur linux-2.4.19.org/net/ipv4/netfilter/ipt_pool.c linux-2.4.19/net/ipv4/netfilter/ipt_pool.c
--- linux-2.4.19.org/net/ipv4/netfilter/ipt_pool.c	Thu Nov  7 20:45:14 2002
+++ linux-2.4.19/net/ipv4/netfilter/ipt_pool.c	Thu Nov  7 21:04:05 2002
@@ -8,6 +8,8 @@
 #include <linux/netfilter_ipv4/ip_pool.h>
 #include <linux/netfilter_ipv4/ipt_pool.h>
 
+MODULE_LICENSE("GPL");
+
 static inline int match_pool(
 	ip_pool_t index,
 	__u32 addr,
diff -Nur linux-2.4.19.org/net/ipv4/netfilter/ipt_string.c linux-2.4.19/net/ipv4/netfilter/ipt_string.c
--- linux-2.4.19.org/net/ipv4/netfilter/ipt_string.c	Thu Nov  7 20:47:07 2002
+++ linux-2.4.19/net/ipv4/netfilter/ipt_string.c	Thu Nov  7 21:25:22 2002
@@ -26,6 +26,8 @@
 #include <linux/netfilter_ipv4/ip_tables.h>
 #include <linux/netfilter_ipv4/ipt_string.h>
 
+MODULE_LICENSE("GPL");
+
 struct string_per_cpu {
 	int *skip;
 	int *shift;
diff -Nur linux-2.4.19.org/net/ipv6/netfilter/ip6t_REJECT.c linux-2.4.19/net/ipv6/netfilter/ip6t_REJECT.c
--- linux-2.4.19.org/net/ipv6/netfilter/ip6t_REJECT.c	Thu Nov  7 20:44:03 2002
+++ linux-2.4.19/net/ipv6/netfilter/ip6t_REJECT.c	Thu Nov  7 21:04:05 2002
@@ -18,6 +18,8 @@
 #define DEBUGP(format, args...)
 #endif
 
+MODULE_LICENSE("GPL");
+
 #if 0
 /* Send RST reply */
 static void send_reset(struct sk_buff *oldskb)
diff -Nur linux-2.4.19.org/net/ipv6/netfilter/ip6t_multiport.c linux-2.4.19/net/ipv6/netfilter/ip6t_multiport.c
--- linux-2.4.19.org/net/ipv6/netfilter/ip6t_multiport.c	Wed Oct 31 00:08:12 2001
+++ linux-2.4.19/net/ipv6/netfilter/ip6t_multiport.c	Thu Nov  7 21:04:05 2002
@@ -9,6 +9,8 @@
 #include <linux/netfilter_ipv6/ip6t_multiport.h>
 #include <linux/netfilter_ipv6/ip6_tables.h>
 
+MODULE_LICENSE("GPL");
+
 #if 0
 #define duprintf(format, args...) printk(format , ## args)
 #else
diff -Nur linux-2.4.19.org/net/sched/sch_wrr.c linux-2.4.19/net/sched/sch_wrr.c
--- linux-2.4.19.org/net/sched/sch_wrr.c	Thu Nov  7 20:47:43 2002
+++ linux-2.4.19/net/sched/sch_wrr.c	Thu Nov  7 21:26:32 2002
@@ -113,6 +113,10 @@
 #include <linux/if_arp.h>
 #include <linux/version.h>
 
+MODULE_AUTHOR("Christian Worm Mortensen, cworm@it-c.dk.");
+MODULE_DESCRIPTION("Weighted Round Robin scheduler.");
+MODULE_LICENSE("GPL");
+
 // Kernel depend stuff:
 #if LINUX_VERSION_CODE < KERNEL_VERSION(2,4,0)
   #define KERNEL22
diff -Nur linux-2.4.19.org/net/wanrouter/wanmain.c linux-2.4.19/net/wanrouter/wanmain.c
--- linux-2.4.19.org/net/wanrouter/wanmain.c	Wed Aug 15 10:22:18 2001
+++ linux-2.4.19/net/wanrouter/wanmain.c	Thu Nov  7 21:27:04 2002
@@ -55,6 +55,9 @@
 #include <asm/byteorder.h>	/* htons(), etc. */
 #include <linux/wanrouter.h>	/* WAN router API definitions */
 
+MODULE_AUTHOR("Gideon Hack");
+MODULE_DESCRIPTION("WAN Multiprotocol Router Module. Main code.");
+MODULE_LICENSE("GPL");
 
 #if defined(LINUX_2_4)
  #include <linux/vmalloc.h>	/* vmalloc, vfree */
--- linux-2.4.20/drivers/char/pcmcia/synclink_cs.c.org	Fri Nov 29 00:53:12 2002
+++ linux-2.4.20/drivers/char/pcmcia/synclink_cs.c	Tue Dec 31 10:45:02 2002
@@ -500,6 +500,10 @@
 MODULE_PARM(debug_level,"i");
 MODULE_PARM(maxframe,"1-" __MODULE_STRING(MAX_DEVICE_COUNT) "i");
 
+MODULE_DESCRIPTION("Device driver for Microgate SyncLink PC Card");
+MODULE_AUTHOR("Paul Fulghum <paulkf@microgate.com>");
+MODULE_LICENSE("GPL");
+
 static char *driver_name = "SyncLink PC Card driver";
 static char *driver_version = "$Revision$";
 
--- linux/fs/quota_v1.c.org	Fri Apr 26 11:33:20 2002
+++ linux/fs/quota_v1.c	Fri Apr 26 11:58:33 2002
@@ -234,6 +234,10 @@
 
 EXPORT_NO_SYMBOLS;
 
+MODULE_AUTHOR("");
+MODULE_DESCRIPTION("QUOTA Version 1 support.");
+MODULE_LICENSE("GPL");
+
 module_init(init_v1_quota_format);
 module_exit(exit_v1_quota_format);
 
--- linux/fs/quota_v2.c.org	Fri Apr 26 11:33:20 2002
+++ linux/fs/quota_v2.c	Fri Apr 26 11:58:20 2002
@@ -686,5 +686,9 @@
 
 EXPORT_NO_SYMBOLS;
 
+MODULE_AUTHOR("");
+MODULE_DESCRIPTION("QUOTA Version 2 support.");
+MODULE_LICENSE("GPL");
+
 module_init(init_v2_quota_format);
 module_exit(exit_v2_quota_format);
diff -urN linux.org/net/unix/af_unix.c linux/net/unix/af_unix.c
--- linux.org/net/unix/af_unix.c	Sun May 19 02:49:28 2002
+++ linux/net/unix/af_unix.c	Sun May 19 02:50:14 2002
@@ -1880,6 +1880,9 @@
 	remove_proc_entry("net/unix", 0);
 }
 
+MODULE_AUTHOR("Alan Cox");
+MODULE_DESCRIPTION("Implementation of BSD Unix domain sockets.");
+
 module_init(af_unix_init);
 module_exit(af_unix_exit);
 
--- linux-2.4.20/drivers/char/synclinkmp.c.org	Fri Nov 29 00:53:12 2002
+++ linux-2.4.20/drivers/char/synclinkmp.c	Mon Dec 16 15:00:41 2002
@@ -87,6 +87,10 @@
 
 #include "linux/synclink.h"
 
+MODULE_AUTHOR("Paul Fulghum <paulkf@microgate.com>");
+MODULE_DESCRIPTION("Device driver for Microgate SyncLink Multiport");
+MODULE_LICENSE("GPL");
+
 static MGSL_PARAMS default_params = {
 	MGSL_MODE_HDLC,			/* unsigned long mode */
 	0,				/* unsigned char loopback; */
--- linux-2.4.20/drivers/net/hamradio/soundmodem/sm.c.org	Wed Apr 18 23:40:05 2001
+++ linux-2.4.20/drivers/net/hamradio/soundmodem/sm.c	Mon Dec 16 15:12:44 2002
@@ -634,6 +634,7 @@
 
 MODULE_AUTHOR("Thomas M. Sailer, sailer@ife.ee.ethz.ch, hb9jnx@hb9w.che.eu");
 MODULE_DESCRIPTION("Soundcard amateur radio modem driver");
+MODULE_LICENSE("GPL");
 
 /* --------------------------------------------------------------------- */
 
--- linux-2.4.20/drivers/net/wan/8253x/8253xini.c.org	Sat Aug  3 02:39:44 2002
+++ linux-2.4.20/drivers/net/wan/8253x/8253xini.c	Mon Dec 16 15:36:42 2002
@@ -3041,3 +3041,4 @@
 module_exit(auraXX20_cleanup);
 MODULE_DESCRIPTION("Aurora Multiport Multiprotocol Serial Driver");
 MODULE_AUTHOR("Joachim Martillo <martillo@telfordtools.com>");
+MODULE_LICENSE("GPL");
--- linux-2.4.20/drivers/scsi/mca_53c9x.c.org	Thu Oct 25 22:53:51 2001
+++ linux-2.4.20/drivers/scsi/mca_53c9x.c	Mon Dec 16 15:24:43 2002
@@ -51,6 +51,11 @@
 
 #include <asm/pgtable.h>
 
+#include <linux/module.h>
+MODULE_DESCRIPTION("Driver for the SCSI adapter on NCR 35xx Microchannel machines");
+MODULE_AUTHOR("unknown");
+MODULE_LICENSE("GPL");
+
 static int  dma_bytes_sent(struct NCR_ESP *, int);
 static int  dma_can_transfer(struct NCR_ESP *, Scsi_Cmnd *);
 static void dma_dump_state(struct NCR_ESP *);
--- linux-2.4.20/net/ipv4/netfilter/ipfwadm_core.c.org	Mon Feb 25 20:38:14 2002
+++ linux-2.4.20/net/ipv4/netfilter/ipfwadm_core.c	Mon Dec 16 15:27:52 2002
@@ -130,6 +130,10 @@
 #include <linux/stat.h>
 #include <linux/version.h>
 
+MODULE_AUTHOR("");
+MODULE_DESCRIPTION("IP firewalling code");
+MODULE_LICENSE("Dual BSD/GPL");
+
 /*
  *	Implement IP packet firewall
  */
--- linux-2.4.20/net/ipv6/netfilter/ip6t_length.c.org	Fri Nov 29 00:53:15 2002
+++ linux-2.4.20/net/ipv6/netfilter/ip6t_length.c	Mon Dec 16 15:31:08 2002
@@ -5,6 +5,8 @@
 #include <linux/netfilter_ipv6/ip6t_length.h>
 #include <linux/netfilter_ipv6/ip6_tables.h>
 
+MODULE_LICENSE("GPL");
+
 static int
 match(const struct sk_buff *skb,
       const struct net_device *in,
