diff -Nru a/drivers/video/Kconfig b/drivers/video/Kconfig
--- a/drivers/video/Kconfig	2004-10-22 16:37:37 +08:00
+++ b/drivers/video/Kconfig	2004-10-26 21:53:01 +08:00
@@ -811,7 +811,7 @@
 	  will be called savagefb.
 
 config FB_SAVAGE_I2C
-       bool "Enable DDC2 Support"
+       tristate "Enable DDC2 Support"
        depends on FB_SAVAGE
        help
 	  This enables I2C support for S3 Savage Chipsets.  This is used
@@ -823,7 +823,7 @@
 	  here.
        
 config FB_SAVAGE_ACCEL
-       bool "Enable Console Acceleration"
+       tristate "Enable Console Acceleration"
        depends on FB_SAVAGE
        default n
        help
diff -Nru a/drivers/video/savage/savagefb-i2c.c b/drivers/video/savage/savagefb-i2c.c
--- a/drivers/video/savage/savagefb-i2c.c	2004-10-22 16:37:37 +08:00
+++ b/drivers/video/savage/savagefb-i2c.c	2004-10-26 21:53:01 +08:00
@@ -197,6 +197,7 @@
 
 	savage_setup_i2c_bus(&par->chan, "SAVAGE DDC2");
 }
+EXPORT_SYMBOL(savagefb_create_i2c_busses);
 
 void savagefb_delete_i2c_busses(struct fb_info *info)
 {
@@ -206,6 +207,7 @@
 		i2c_bit_del_bus(&par->chan.adapter);
 	par->chan.par = NULL;
 }
+EXPORT_SYMBOL(savagefb_delete_i2c_busses);
 
 static u8 *savage_do_probe_i2c_edid(struct savagefb_i2c_chan *chan)
 {
@@ -255,3 +257,6 @@
 
 	return 0;
 }
+EXPORT_SYMBOL(savagefb_probe_i2c_connector);
+
+MODULE_LICENSE("GPL");
diff -Nru a/drivers/video/savage/savagefb_accel.c b/drivers/video/savage/savagefb_accel.c
--- a/drivers/video/savage/savagefb_accel.c	2004-10-22 16:37:38 +08:00
+++ b/drivers/video/savage/savagefb_accel.c	2004-10-26 21:53:01 +08:00
@@ -26,6 +26,7 @@
 	par->SavageWaitIdle(par);
 	return 0;
 }
+EXPORT_SYMBOL(savagefb_sync);
 
 void savagefb_copyarea(struct fb_info *info, const struct fb_copyarea *region) 
 {
@@ -60,6 +61,7 @@
 	BCI_SEND(BCI_X_Y(dx, dy));
 	BCI_SEND(BCI_W_H(region->width, region->height));
 }
+EXPORT_SYMBOL(savagefb_copyarea);
 
 void savagefb_fillrect(struct fb_info *info, const struct fb_fillrect *rect)
 {
@@ -87,6 +89,7 @@
 	BCI_SEND( BCI_X_Y(rect->dx, rect->dy) );
 	BCI_SEND( BCI_W_H(rect->width, rect->height) );
 }
+EXPORT_SYMBOL(savagefb_fillrect);
 
 void savagefb_imageblit(struct fb_info *info, const struct fb_image *image)
 {
@@ -132,4 +135,6 @@
 	for (i = 0; i < size; i++)
 		BCI_SEND(src[i]);
 }
+EXPORT_SYMBOL(savagefb_imageblit);
 
+MODULE_LICENSE("GPL");
diff -Nru a/drivers/video/savage/savagefb.c b/drivers/video/savage/savagefb.c
--- a/drivers/video/savage/savagefb.c	2004-10-23 22:48:12 +08:00
+++ b/drivers/video/savage/savagefb.c	2004-10-25 17:01:45 +08:00
@@ -2468,9 +2468,9 @@
 
 /* ************************* init in-kernel code ************************** */
 
-#ifndef MODULE
 int __init savagefb_setup(char *options)
 {
+#ifndef MODULE
 	char *this_opt;
 
 	if (!options || !*options)
@@ -2483,10 +2483,9 @@
 			mode_option = this_opt;
 		}
 	}
-
+#endif /* !MODULE */
 	return 0;
 }
-#endif /* !MODULE */
 
 int __init savagefb_init(void)
 {
diff -Nru a/drivers/video/savage/savagefb.h b/drivers/video/savage/savagefb.h
--- a/drivers/video/savage/savagefb.h	2004-10-23 22:48:13 +08:00
+++ b/drivers/video/savage/savagefb.h	2004-10-25 16:59:51 +08:00
@@ -21,7 +21,7 @@
 # define DBG(x)		printk (KERN_DEBUG "savagefb: %s\n", (x));
 #else
 # define DBG(x)
-# define SavagePrintRegs(x)
+# define SavagePrintRegs(...)
 #endif
 
 
