--- linux-2.5.72/drivers/ide/legacy/pdc4030.c~	2003-06-17 06:20:02.000000000 +0200
+++ linux-2.5.72/drivers/ide/legacy/pdc4030.c	2003-06-17 16:00:05.000000000 +0200
@@ -818,7 +818,8 @@
 	   Feature register.
 	   FIXME: Is promise_selectproc now redundant??
 	*/
-	int drive_number = (HWIF(drive)->channel << 1) + drive->select.b.unit;
+    	ide_hwif_t *hwif	= HWIF(drive);
+	int drive_number	= (hwif->channel << 1) + drive->select.b.unit;
 #ifdef CONFIG_IDE_TASKFILE_IO
 	struct hd_drive_task_hdr taskfile;
 	ide_task_t args;
