--- linux-2.4.21/drivers/net/irda/ma600.c.orig	Fri Nov 29 00:53:13 2002
+++ linux-2.4.21/drivers/net/irda/ma600.c	Thu Jun 26 18:12:16 2003
@@ -32,7 +32,7 @@
  ********************************************************************/
 
 /* define this macro for release version */
-//#define NDEBUG
+#define NDEBUG
 
 #include <linux/module.h>
 #include <linux/delay.h>
--- linux-2.4.21/drivers/net/wan/sdla_chdlc.c.orig	Thu Jun 26 18:18:08 2003
+++ linux-2.4.21/drivers/net/wan/sdla_chdlc.c	Thu Jun 26 18:18:26 2003
@@ -592,8 +592,7 @@
 	
 
 		if (chdlc_set_intr_mode(card, APP_INT_ON_TIMER)){
-			printk (KERN_INFO "%s: 
-				Failed to set interrupt triggers!\n",
+			printk (KERN_INFO "%s: Failed to set interrupt triggers!\n",
 				card->devname);
 			return -EIO;	
         	}
--- linux-2.4.21/drivers/net/wan/sbni.c.orig	Thu Jun 26 18:20:25 2003
+++ linux-2.4.21/drivers/net/wan/sbni.c	Thu Jun 26 18:32:25 2003
@@ -149,11 +149,11 @@
 static int  enslave( struct net_device *, struct net_device * );
 static int  emancipate( struct net_device * );
 #endif
-
+/*
 #ifdef __i386__
 #define ASM_CRC 1
 #endif
-
+*/
 static const char  version[] =
 	"Granch SBNI12 driver ver 5.0.1  Jun 22 2001  Denis I.Timofeev.\n";
 
--- linux-2.4.21/drivers/sound/cs46xx.c.orig	Thu Jun 26 18:52:19 2003
+++ linux-2.4.21/drivers/sound/cs46xx.c	Thu Jun 26 18:53:24 2003
@@ -947,8 +947,8 @@
 
 struct InitStruct
 {
-    u32 long off;
-    u32 long val;
+    u32 off;
+    u32 val;
 } InitArray[] = { {0x00000040, 0x3fc0000f},
                   {0x0000004c, 0x04800000},
 
--- linux-2.4.21/net/decnet/dn_table.c~	Fri Dec 21 18:42:05 2001
+++ linux-2.4.21/net/decnet/dn_table.c	Thu Jun 26 19:27:45 2003
@@ -836,8 +836,7 @@
                 return NULL;
 
         if (in_interrupt() && net_ratelimit()) {
-                printk(KERN_DEBUG "DECnet: BUG! Attempt to create routing table 
-from interrupt\n"); 
+                printk(KERN_DEBUG "DECnet: BUG! Attempt to create routing table from interrupt\n"); 
                 return NULL;
         }
         if ((t = kmalloc(sizeof(struct dn_fib_table), GFP_KERNEL)) == NULL)
--- linux-2.4.21/fs/partitions/ldm.c.orig	2003-08-16 03:27:54.000000000 +0200
+++ linux-2.4.21/fs/partitions/ldm.c	2003-08-16 03:28:09.000000000 +0200
@@ -44,7 +44,7 @@
  * driver was compiled with debug enabled. Otherwise, the call turns into a NOP.
  */
 #ifndef CONFIG_LDM_DEBUG
-#define ldm_debug(...)	do {} while (0)
+#define ldm_debug(f, a...)	do {} while (0)
 #else
 #define ldm_debug(f, a...) _ldm_printk (KERN_DEBUG, __FUNCTION__, f, ##a)
 #endif
--- linux-2.4.21/net/bridge/netfilter/ebt_vlan.c.orig	2003-08-16 23:05:47.000000000 +0200
+++ linux-2.4.21/net/bridge/netfilter/ebt_vlan.c	2003-08-16 23:06:35.000000000 +0200
@@ -35,7 +35,7 @@
 MODULE_LICENSE("GPL");
 
 
-#define DEBUG_MSG(...) if (debug) printk (KERN_DEBUG "ebt_vlan: " __VA_ARGS__)
+#define DEBUG_MSG(f, a...) if (debug) printk (KERN_DEBUG "ebt_vlan: " f, ##a)
 #define INV_FLAG(_inv_flag_) (info->invflags & _inv_flag_) ? "!" : ""
 #define GET_BITMASK(_BIT_MASK_) info->bitmask & _BIT_MASK_
 #define SET_BITMASK(_BIT_MASK_) info->bitmask |= _BIT_MASK_
