--- linux-2.4.19/arch/ppc/kernel/ppc_ksyms.c.org	Thu Sep 19 20:41:52 2002
+++ linux-2.4.19/arch/ppc/kernel/ppc_ksyms.c	Thu Sep 19 20:41:40 2002
@@ -371,3 +371,6 @@
 #endif /* defined(CONFIG_ALL_PPC) */
 
 
+EXPORT_SYMBOL(ioremap_bot);
+
+EXPORT_SYMBOL(pmac_get_partition);
--- linux-2.4.20/arch/ppc/kernel/ppc_ksyms.c.orig	2004-01-11 20:41:02.000000000 +0100
+++ linux-2.4.20/arch/ppc/kernel/ppc_ksyms.c	2004-01-11 21:05:13.000000000 +0100
@@ -354,6 +354,10 @@
 EXPORT_SYMBOL(next_mmu_context);
 EXPORT_SYMBOL(set_context);
 EXPORT_SYMBOL(handle_mm_fault); /* For MOL */
+#ifdef CONFIG_SMP
+extern int *hash_table_lock;
+EXPORT_SYMBOL_NOVERS(hash_table_lock); /* For MOL */
+#endif
 EXPORT_SYMBOL_NOVERS(disarm_decr);
 #ifdef CONFIG_PPC_STD_MMU
 EXPORT_SYMBOL(flush_hash_page); /* For MOL */
