--- linux.org/include/asm-sparc64/ide.h	Wed Apr 10 13:55:19 2002
+++ linux/include/asm-sparc64/ide.h	Thu Apr 11 09:40:23 2002
@@ -87,7 +87,7 @@
 		unsigned int reserved456: 3;
 		unsigned bit3		: 1;	/* ATA-2 thingy */
 		unsigned int SRST	: 1;	/* host soft reset bit */
-		unsigned int nIEN	: 1;	/* device INTRQ to host *
+		unsigned int nIEN	: 1;	/* device INTRQ to host */
 		unsigned int bit0	: 1;
 	} b;
 } control_t;
