--- linux/include/linux/tasks.h	Tue Jan 26 01:04:20 1999
+++ linux.pld/include/linux/tasks.h	Tue Jan 26 13:23:29 1999
@@ -11,10 +11,10 @@
 #define NR_CPUS 1
 #endif
 
-#define NR_TASKS	512	/* On x86 Max 4092, or 4090 w/APM configured. */
+#define NR_TASKS	4080	/* On x86 Max 4092, or 4090 w/APM configured. */
 
-#define MAX_TASKS_PER_USER (NR_TASKS/2)
-#define MIN_TASKS_LEFT_FOR_ROOT 4
+#define MAX_TASKS_PER_USER 256
+#define MIN_TASKS_LEFT_FOR_ROOT 64
 
 
 /*
