diff -urNp linux-5270/arch/alpha/kernel/sys_sable.c linux-10010/arch/alpha/kernel/sys_sable.c
--- linux-5270/arch/alpha/kernel/sys_sable.c	
+++ linux-10010/arch/alpha/kernel/sys_sable.c	
@@ -253,8 +253,9 @@ sable_init_irq(void)
  */
 
 static int __init
-sable_map_irq(struct pci_dev *dev, u8 slot, u8 pin)
+sable_map_irq(struct pci_dev *dev, u8 slota, u8 pin)
 {
+	int slot=slota;
         static char irq_tab[9][5] __initdata = {
 		/*INT    INTA   INTB   INTC   INTD */
 		{ 32+0,  32+0,  32+0,  32+0,  32+0},  /* IdSel 0,  TULIP  */
diff -urNp linux-5270/arch/alpha/kernel/sys_titan.c linux-10010/arch/alpha/kernel/sys_titan.c
--- linux-5270/arch/alpha/kernel/sys_titan.c	2001-10-13 00:35:53.000000000 +0200
+++ linux-10010/arch/alpha/kernel/sys_titan.c	
@@ -83,7 +83,7 @@ titan_update_irq_hw(unsigned long mask)
 	*dim2;
 	*dim3;
 #else
-	volatile unsigned long *dimB;
+	volatile unsigned long *dimB=NULL;
 	if (bcpu == 0) dimB = &cchip->dim0.csr;
 	else if (bcpu == 1) dimB = &cchip->dim1.csr;
 	else if (bcpu == 2) dimB = &cchip->dim2.csr;
diff -urNp linux-5270/arch/alpha/kernel/sys_wildfire.c linux-10010/arch/alpha/kernel/sys_wildfire.c
--- linux-5270/arch/alpha/kernel/sys_wildfire.c	2001-03-02 20:12:07.000000000 +0100
+++ linux-10010/arch/alpha/kernel/sys_wildfire.c	
@@ -300,8 +300,9 @@ wildfire_device_interrupt(unsigned long 
  */
 
 static int __init
-wildfire_map_irq(struct pci_dev *dev, u8 slot, u8 pin)
+wildfire_map_irq(struct pci_dev *dev, u8 slotu, u8 pin)
 {
+	int slot=slotu;
 	static char irq_tab[8][5] __initdata = {
 		/*INT    INTA   INTB   INTC   INTD */
 		{ -1,    -1,    -1,    -1,    -1}, /* IdSel 0 ISA Bridge */
diff -urNp linux-5270/drivers/atm/ambassador.c linux-10010/drivers/atm/ambassador.c
--- linux-5270/drivers/atm/ambassador.c	2001-09-14 00:21:32.000000000 +0200
+++ linux-10010/drivers/atm/ambassador.c	
@@ -290,12 +290,11 @@ static inline void __init show_version (
 /********** microcode **********/
 
 #ifdef AMB_NEW_MICROCODE
-#define UCODE(x) UCODE1(atmsar12.,x)
+#define UCODE(x) UCODE2(atmsar12.x)
 #else
-#define UCODE(x) UCODE1(atmsar11.,x)
+#define UCODE(x) UCODE2(atmsar11.x)
 #endif
 #define UCODE2(x) #x
-#define UCODE1(x,y) UCODE2(x ## y)
 
 static u32 __initdata ucode_start = 
 #include UCODE(start)
diff -urNp linux-5270/drivers/block/blkpg.c linux-10010/drivers/block/blkpg.c
--- linux-5270/drivers/block/blkpg.c	
+++ linux-10010/drivers/block/blkpg.c	
@@ -32,6 +32,7 @@
 #include <linux/fs.h>			/* for BLKRASET, ... */
 #include <linux/sched.h>		/* for capable() */
 #include <linux/blk.h>			/* for set_device_ro() */
+#include <linux/blkdev.h>
 #include <linux/blkpg.h>
 #include <linux/genhd.h>
 #include <linux/swap.h>			/* for is_swap_partition() */
diff -urNp linux-5270/drivers/block/DAC960.c linux-10010/drivers/block/DAC960.c
--- linux-5270/drivers/block/DAC960.c	
+++ linux-10010/drivers/block/DAC960.c	
@@ -5491,11 +5491,7 @@ static int DAC960_IOCTL(Inode_T *Inode, 
 				       .part[MINOR(Inode->i_rdev)]
 				       .nr_sects << 9,
 		      (u64 *) Argument);
-    case BLKRAGET:
-    case BLKRASET:
-    case BLKFLSBUF:
-    case BLKBSZGET:
-    case BLKBSZSET:
+    default:
       return blk_ioctl(Inode->i_rdev, Request, Argument);
     case BLKRRPART:
       /* Re-Read Partition Table. */
diff -urNp linux-5270/drivers/block/nbd.c linux-10010/drivers/block/nbd.c
--- linux-5270/drivers/block/nbd.c	
+++ linux-10010/drivers/block/nbd.c	
@@ -552,7 +552,7 @@ static struct block_device_operations nb
  *  (Just smiley confuses emacs :-)
  */
 
-static int __init nbd_init(void)
+int __init nbd_init(void)
 {
 	int i;
 
diff -urNp linux-5270/drivers/char/Config.in linux-10010/drivers/char/Config.in
--- linux-5270/drivers/char/Config.in	
+++ linux-10010/drivers/char/Config.in	
@@ -159,7 +159,7 @@ if [ "$CONFIG_PARPORT" != "n" ]; then
    dep_tristate 'Texas Instruments parallel link cable support' CONFIG_TIPAR $CONFIG_PARPORT
 fi
 
-if [ "$CONFIG_PPC64" ] ; then 
+if [ "$CONFIG_PPC64" = "y" ] ; then 
    bool 'pSeries Hypervisor Virtual Console support' CONFIG_HVC_CONSOLE
 fi
 
diff -urNp linux-5270/drivers/media/video/msp3400.c linux-10010/drivers/media/video/msp3400.c
--- linux-5270/drivers/media/video/msp3400.c	
+++ linux-10010/drivers/media/video/msp3400.c	
@@ -1234,7 +1234,7 @@ static int msp_attach(struct i2c_adapter
 	DECLARE_MUTEX_LOCKED(sem);
 	struct msp3400c *msp;
         struct i2c_client *c;
-	int              rev1,rev2,i;
+	int              rev1,rev2=0,i;
 
         client_template.adapter = adap;
         client_template.addr = addr;
diff -urNp linux-5270/drivers/media/video/saa7110.c linux-10010/drivers/media/video/saa7110.c
--- linux-5270/drivers/media/video/saa7110.c	2001-11-09 23:01:22.000000000 +0100
+++ linux-10010/drivers/media/video/saa7110.c	
@@ -392,7 +392,7 @@ int saa7110_command(struct i2c_device *d
 		break;
 
 	 default:
-		DEBUG(printk(KERN_INFO "unknown saa7110_command??(%d)\n",cmd));
+		DEBUG(printk(KERN_INFO "unknown saa7110_command?(%d)\n",cmd));
 		return -EINVAL;
 	}
 	return 0;
diff -urNp linux-5270/drivers/net/ns83820.c linux-10010/drivers/net/ns83820.c
--- linux-5270/drivers/net/ns83820.c	
+++ linux-10010/drivers/net/ns83820.c	
@@ -94,6 +94,7 @@
 
 #include <linux/module.h>
 #include <linux/types.h>
+#include <asm/io.h>
 #include <linux/pci.h>
 #include <linux/netdevice.h>
 #include <linux/etherdevice.h>
diff -urNp linux-5270/drivers/net/pcmcia/wavelan_cs.h linux-10010/drivers/net/pcmcia/wavelan_cs.h
--- linux-5270/drivers/net/pcmcia/wavelan_cs.h	
+++ linux-10010/drivers/net/pcmcia/wavelan_cs.h	
@@ -489,7 +489,7 @@
 #undef DEBUG_RX_INFO		/* Header of the transmitted packet */
 #undef DEBUG_RX_FAIL		/* Normal failure conditions */
 #define DEBUG_RX_ERROR		/* Unexpected conditions */
-#undef DEBUG_PACKET_DUMP	32	/* Dump packet on the screen */
+#undef DEBUG_PACKET_DUMP /* 32	 Dump packet on the screen */
 #undef DEBUG_IOCTL_TRACE	/* Misc call by Linux */
 #undef DEBUG_IOCTL_INFO		/* Various debug info */
 #define DEBUG_IOCTL_ERROR	/* What's going wrong */
diff -urNp linux-5270/drivers/net/plip.c linux-10010/drivers/net/plip.c
--- linux-5270/drivers/net/plip.c	2001-09-30 21:26:07.000000000 +0200
+++ linux-10010/drivers/net/plip.c	
@@ -1380,7 +1380,7 @@ static void __exit plip_cleanup_module (
 
 static int parport_ptr;
 
-static int __init plip_setup(char *str)
+int __init plip_setup(char *str)
 {
 	int ints[4];
 
diff -urNp linux-5270/drivers/net/sk98lin/h/skgepnm2.h linux-10010/drivers/net/sk98lin/h/skgepnm2.h
--- linux-5270/drivers/net/sk98lin/h/skgepnm2.h	
+++ linux-10010/drivers/net/sk98lin/h/skgepnm2.h	
@@ -362,8 +362,16 @@ typedef struct s_PnmiStatAddr {
 #if SK_TICKS_PER_SEC == 100
 #define SK_PNMI_HUNDREDS_SEC(t)	(t)
 #else
+#if SK_TICKS_PER_SEC == 800
+#define SK_PNMI_HUNDREDS_SEC(t)	(t>>3)
+#else
+#if SK_TICKS_PER_SEC == 1000
+#define SK_PNMI_HUNDREDS_SEC(t)	((t*1000)>>10)
+#else
 #define SK_PNMI_HUNDREDS_SEC(t)	(((t) * 100) / (SK_TICKS_PER_SEC))
 #endif
+#endif
+#endif
 
 /*
  * Macros to work around alignment problems
diff -urNp linux-5270/drivers/net/starfire.c linux-10010/drivers/net/starfire.c
--- linux-5270/drivers/net/starfire.c	
+++ linux-10010/drivers/net/starfire.c	
@@ -112,6 +112,7 @@ TODO:
 
 #include <linux/version.h>
 #include <linux/module.h>
+#include <asm/io.h>
 #include <linux/kernel.h>
 #include <linux/pci.h>
 #include <linux/netdevice.h>
diff -urNp linux-5270/drivers/net/tokenring/Config.in linux-10010/drivers/net/tokenring/Config.in
--- linux-5270/drivers/net/tokenring/Config.in	
+++ linux-10010/drivers/net/tokenring/Config.in	
@@ -21,10 +21,10 @@ if [ "$CONFIG_TR" != "n" ]; then
    dep_tristate '  3Com 3C359 Token Link Velocity XL adapter support' CONFIG_3C359 $CONFIG_TR $CONFIG_PCI
    tristate '  Generic TMS380 Token Ring ISA/PCI adapter support' CONFIG_TMS380TR
    if [ "$CONFIG_TMS380TR" != "n" ]; then
-      dep_tristate '    Generic TMS380 PCI support' CONFIG_TMSPCI $CONFIG_PCI
-      dep_tristate '    Generic TMS380 ISA support' CONFIG_TMSISA $CONFIG_ISA
-      dep_tristate '    Madge Smart 16/4 PCI Mk2 support' CONFIG_ABYSS $CONFIG_PCI
-      dep_tristate '    Madge Smart 16/4 Ringnode MicroChannel' CONFIG_MADGEMC $CONFIG_MCA
+      dep_tristate '    Generic TMS380 PCI support' CONFIG_TMSPCI $CONFIG_PCI $CONFIG_TMS380TR
+      dep_tristate '    Generic TMS380 ISA support' CONFIG_TMSISA $CONFIG_ISA $CONFIG_TMS380TR
+      dep_tristate '    Madge Smart 16/4 PCI Mk2 support' CONFIG_ABYSS $CONFIG_PCI $CONFIG_TMS380TR
+      dep_tristate '    Madge Smart 16/4 Ringnode MicroChannel' CONFIG_MADGEMC $CONFIG_MCA $CONFIG_TMS380TR
    fi
    if [ "$CONFIG_ISA" = "y" -o "$CONFIG_MCA" = "y" ]; then
       tristate '  SMC ISA/MCA adapter support' CONFIG_SMCTR
diff -urNp linux-5270/drivers/net/wan/sdla_chdlc.c linux-10010/drivers/net/wan/sdla_chdlc.c
--- linux-5270/drivers/net/wan/sdla_chdlc.c	
+++ linux-10010/drivers/net/wan/sdla_chdlc.c	
@@ -50,6 +50,7 @@
 
 #include <linux/module.h>
 #include <linux/version.h>
+#include <linux/module.h>
 #include <linux/kernel.h>	/* printk(), and other useful stuff */
 #include <linux/stddef.h>	/* offsetof(), etc. */
 #include <linux/errno.h>	/* return codes */
diff -urNp linux-5270/drivers/net/wan/sdla_fr.c linux-10010/drivers/net/wan/sdla_fr.c
--- linux-5270/drivers/net/wan/sdla_fr.c	
+++ linux-10010/drivers/net/wan/sdla_fr.c	
@@ -140,6 +140,7 @@
 
 #include <linux/module.h>
 #include <linux/version.h>
+#include <linux/module.h>
 #include <linux/kernel.h>	/* printk(), and other useful stuff */
 #include <linux/stddef.h>	/* offsetof(), etc. */
 #include <linux/errno.h>	/* return codes */
diff -urNp linux-5270/drivers/net/wan/sdla_ppp.c linux-10010/drivers/net/wan/sdla_ppp.c
--- linux-5270/drivers/net/wan/sdla_ppp.c	
+++ linux-10010/drivers/net/wan/sdla_ppp.c	
@@ -92,6 +92,7 @@
 
 #include <linux/module.h>
 #include <linux/version.h>
+#include <linux/module.h>
 #include <linux/kernel.h>	/* printk(), and other useful stuff */
 #include <linux/stddef.h>	/* offsetof(), etc. */
 #include <linux/errno.h>	/* return codes */
diff -urNp linux-5270/drivers/net/wan/sdla_x25.c linux-10010/drivers/net/wan/sdla_x25.c
--- linux-5270/drivers/net/wan/sdla_x25.c	
+++ linux-10010/drivers/net/wan/sdla_x25.c	
@@ -83,6 +83,7 @@
 
 #include <linux/module.h>
 #include <linux/version.h>
+#include <linux/module.h>
 #include <linux/kernel.h>	/* printk(), and other useful stuff */
 #include <linux/stddef.h>	/* offsetof(), etc. */
 #include <linux/errno.h>	/* return codes */
diff -urNp linux-5270/drivers/net/wan/wanpipe_multppp.c linux-10010/drivers/net/wan/wanpipe_multppp.c
--- linux-5270/drivers/net/wan/wanpipe_multppp.c	2001-09-14 01:04:43.000000000 +0200
+++ linux-10010/drivers/net/wan/wanpipe_multppp.c	
@@ -19,6 +19,7 @@
 
 #include <linux/module.h>
 #include <linux/version.h>
+#include <linux/module.h>
 #include <linux/kernel.h>	/* printk(), and other useful stuff */
 #include <linux/stddef.h>	/* offsetof(), etc. */
 #include <linux/errno.h>	/* return codes */
diff -urNp linux-5270/drivers/scsi/advansys.c linux-10010/drivers/scsi/advansys.c
--- linux-5270/drivers/scsi/advansys.c	
+++ linux-10010/drivers/scsi/advansys.c	
@@ -5551,7 +5551,7 @@ advansys_detect(Scsi_Host_Template *tpnt
                 }
             } else {
                 ADV_CARR_T      *carrp;
-                int             req_cnt;
+                int             req_cnt=0;
                 adv_req_t       *reqp = NULL;
                 int             sg_cnt = 0;
 
@@ -9258,7 +9258,6 @@ DvcAdvWritePCIConfigByte(
             ASC_PCI_ID2FUNC(asc_dvc->cfg->pci_slot_info)),
         offset, byte_data);
 #else /* CONFIG_PCI */
-    return 0;
 #endif /* CONFIG_PCI */
 }
 
diff -urNp linux-5270/drivers/scsi/eata_dma.c linux-10010/drivers/scsi/eata_dma.c
--- linux-5270/drivers/scsi/eata_dma.c	2001-09-30 21:26:07.000000000 +0200
+++ linux-10010/drivers/scsi/eata_dma.c	
@@ -1067,7 +1067,7 @@ short register_HBA(u32 base, struct get_
     char *buff = 0;
     unchar bugs = 0;
     struct Scsi_Host *sh;
-    hostdata *hd;
+    hostdata *hd=NULL;
     int x;
     
     
diff -urNp linux-5270/drivers/scsi/imm.c linux-10010/drivers/scsi/imm.c
--- linux-5270/drivers/scsi/imm.c	
+++ linux-10010/drivers/scsi/imm.c	
@@ -322,10 +322,10 @@ static unsigned char imm_wait(int host_n
      * STR      imm     imm
      * ===================================
      * 0x80     S_REQ   S_REQ
-     * 0x40     !S_BSY  (????)
+     * 0x40     !S_BSY  (?)
      * 0x20     !S_CD   !S_CD
      * 0x10     !S_IO   !S_IO
-     * 0x08     (????)  !S_BSY
+     * 0x08     (?)  !S_BSY
      *
      * imm      imm     meaning
      * ==================================
@@ -927,7 +927,7 @@ static void imm_interrupt(void *data)
 	printk("imm: told to abort\n");
 	break;
     case DID_PARITY:
-	printk("imm: parity error (???)\n");
+	printk("imm: parity error (?)\n");
 	break;
     case DID_ERROR:
 	printk("imm: internal driver error\n");
@@ -936,7 +936,7 @@ static void imm_interrupt(void *data)
 	printk("imm: told to reset device\n");
 	break;
     case DID_BAD_INTR:
-	printk("imm: bad interrupt (???)\n");
+	printk("imm: bad interrupt (?)\n");
 	break;
     default:
 	printk("imm: bad return code (%02x)\n", (cmd->result >> 16) & 0xff);
diff -urNp linux-5270/drivers/scsi/ppa.c linux-10010/drivers/scsi/ppa.c
--- linux-5270/drivers/scsi/ppa.c	
+++ linux-10010/drivers/scsi/ppa.c	
@@ -824,7 +824,7 @@ static void ppa_interrupt(void *data)
 	printk("ppa: told to abort\n");
 	break;
     case DID_PARITY:
-	printk("ppa: parity error (???)\n");
+	printk("ppa: parity error (?)\n");
 	break;
     case DID_ERROR:
 	printk("ppa: internal driver error\n");
@@ -833,7 +833,7 @@ static void ppa_interrupt(void *data)
 	printk("ppa: told to reset device\n");
 	break;
     case DID_BAD_INTR:
-	printk("ppa: bad interrupt (???)\n");
+	printk("ppa: bad interrupt (?)\n");
 	break;
     default:
 	printk("ppa: bad return code (%02x)\n", (cmd->result >> 16) & 0xff);
diff -urNp linux-5270/drivers/scsi/qlogicisp.c linux-10010/drivers/scsi/qlogicisp.c
--- linux-5270/drivers/scsi/qlogicisp.c	2001-10-25 22:53:51.000000000 +0200
+++ linux-10010/drivers/scsi/qlogicisp.c	
@@ -1034,6 +1034,9 @@ void isp1020_intr_handler(int irq, void 
 
 		cmd_slot = sts->handle;
 		Cmnd = hostdata->cmd_slots[cmd_slot];
+#ifdef __alpha__
+		if (Cmnd == NULL) mb(); /* Magic, don't move */
+#endif
 		hostdata->cmd_slots[cmd_slot] = NULL;
 
 		TRACE("done", out_ptr, Cmnd);
diff -urNp linux-5270/drivers/scsi/sym53c8xx_2/Makefile linux-10010/drivers/scsi/sym53c8xx_2/Makefile
--- linux-5270/drivers/scsi/sym53c8xx_2/Makefile	2001-11-10 00:22:54.000000000 +0100
+++ linux-10010/drivers/scsi/sym53c8xx_2/Makefile	
@@ -1,14 +1,14 @@
 # File: drivers/sym53c8xx/Makefile
 # Makefile for the NCR/SYMBIOS/LSI 53C8XX PCI SCSI controllers driver.
 
-list-multi := sym53c8xx.o
-sym53c8xx-objs := sym_fw.o sym_glue.o sym_hipd.o sym_malloc.o sym_misc.o sym_nvram.o
-obj-$(CONFIG_SCSI_SYM53C8XX_2) := sym53c8xx.o
+list-multi := sym53c8xx_2.o
+sym53c8xx_2-objs := sym_fw.o sym_glue.o sym_hipd.o sym_malloc.o sym_misc.o sym_nvram.o
+obj-$(CONFIG_SCSI_SYM53C8XX_2) := sym53c8xx_2.o
 
 EXTRA_CFLAGS += -I.
 
-sym53c8xx.o: $(sym53c8xx-objs)
-	$(LD) -r -o $@ $(sym53c8xx-objs)
+sym53c8xx_2.o: $(sym53c8xx_2-objs)
+	$(LD) -r -o $@ $(sym53c8xx_2-objs)
 
 include $(TOPDIR)/Rules.make
 
diff -urNp linux-5270/drivers/scsi/sym53c8xx_2/sym_hipd.c linux-10010/drivers/scsi/sym53c8xx_2/sym_hipd.c
--- linux-5270/drivers/scsi/sym53c8xx_2/sym_hipd.c	
+++ linux-10010/drivers/scsi/sym53c8xx_2/sym_hipd.c	
@@ -221,7 +221,7 @@ static void sym_chip_reset (hcb_p np)
  */
 static void sym_soft_reset (hcb_p np)
 {
-	u_char istat;
+	u_char istat=0;
 	int i;
 
 	if (!(np->features & FE_ISTAT1) || !(INB (nc_istat1) & SCRUN))
diff -urNp linux-5270/drivers/scsi/sym53c8xx.c linux-10010/drivers/scsi/sym53c8xx.c
--- linux-5270/drivers/scsi/sym53c8xx.c	
+++ linux-10010/drivers/scsi/sym53c8xx.c	
@@ -6992,7 +6992,7 @@ static void ncr_chip_reset (ncb_p np)
 
 static void ncr_soft_reset(ncb_p np)
 {
-	u_char istat;
+	u_char istat=0;
 	int i;
 
 	if (!(np->features & FE_ISTAT1) || !(INB (nc_istat1) & SRUN))
diff -urNp linux-5270/drivers/sound/cs4281/cs4281m.c linux-10010/drivers/sound/cs4281/cs4281m.c
--- linux-5270/drivers/sound/cs4281/cs4281m.c	
+++ linux-10010/drivers/sound/cs4281/cs4281m.c	
@@ -4597,8 +4597,8 @@ struct pci_driver cs4281_pci_driver = {
 	id_table:cs4281_pci_tbl,
 	probe:cs4281_probe,
 	remove:cs4281_remove,
-	suspend:CS4281_SUSPEND_TBL,
-	resume:CS4281_RESUME_TBL,
+	suspend:NULL,
+	resume:NULL,
 };
 
 int __init cs4281_init_module(void)
diff -urNp linux-5270/drivers/video/sis/init301.c linux-10010/drivers/video/sis/init301.c
--- linux-5270/drivers/video/sis/init301.c	
+++ linux-10010/drivers/video/sis/init301.c	
@@ -2983,7 +2983,7 @@ SiS_GetCRT2Data301(SiS_Private *SiS_Pr, 
                    USHORT RefreshRateTableIndex,
 		   PSIS_HW_DEVICE_INFO HwDeviceExtension)
 {
-  USHORT tempax,tempbx,modeflag;
+  USHORT tempax=0,tempbx=0,modeflag;
   USHORT resinfo;
   USHORT CRT2Index,ResIndex;
   const SiS_LCDDataStruct *LCDPtr = NULL;
diff -urNp linux-5270/fs/dquot.c linux-10010/fs/dquot.c
--- linux-5270/fs/dquot.c	
+++ linux-10010/fs/dquot.c	
@@ -70,6 +70,7 @@
 #include <linux/proc_fs.h>
 
 #include <asm/uaccess.h>
+#include <asm/cache.h>
 
 static char *quotatypes[] = INITQFNAMES;
 static struct quota_format_type *quota_formats;	/* List of registered formats */
diff -urNp linux-5270/fs/locks.c linux-10010/fs/locks.c
--- linux-5270/fs/locks.c	
+++ linux-10010/fs/locks.c	
@@ -415,9 +415,15 @@ static void locks_insert_block(struct fi
 			       struct file_lock *waiter)
 {
 	if (!list_empty(&waiter->fl_block)) {
+#ifdef __alpha__
+		printk(KERN_ERR "locks_insert_block: removing duplicated lock "
+			"(pid=%d %ld-%ld type=%d)\n", waiter->fl_pid,
+			waiter->fl_start, waiter->fl_end, waiter->fl_type);
+#else
 		printk(KERN_ERR "locks_insert_block: removing duplicated lock "
 			"(pid=%d %Ld-%Ld type=%d)\n", waiter->fl_pid,
 			waiter->fl_start, waiter->fl_end, waiter->fl_type);
+#endif
 		locks_delete_block(waiter);
 	}
 	list_add_tail(&waiter->fl_block, &blocker->fl_block);
@@ -1858,11 +1864,19 @@ static void lock_get_status(char* out, s
 		     fl->fl_pid,
 		     inode ? kdevname(inode->i_dev) : "<none>",
 		     inode ? inode->i_ino : 0);
+#if __alpha__
+	out += sprintf(out, "%ld ", fl->fl_start);
+#else
 	out += sprintf(out, "%Ld ", fl->fl_start);
+#endif
 	if (fl->fl_end == OFFSET_MAX)
 		out += sprintf(out, "EOF ");
 	else
+#if __alpha__
+		out += sprintf(out, "%ld ", fl->fl_end);
+#else
 		out += sprintf(out, "%Ld ", fl->fl_end);
+#endif
 	sprintf(out, "%08lx %08lx %08lx %08lx %08lx\n",
 		(long)fl, (long)fl->fl_link.prev, (long)fl->fl_link.next,
 		(long)fl->fl_next, (long)fl->fl_block.next);
diff -urNp linux-5270/fs/proc/proc_tty.c linux-10010/fs/proc/proc_tty.c
--- linux-5270/fs/proc/proc_tty.c	2000-04-22 00:17:57.000000000 +0200
+++ linux-10010/fs/proc/proc_tty.c	
@@ -161,7 +161,7 @@ void proc_tty_unregister_driver(struct t
 	if (!ent)
 		return;
 		
-	remove_proc_entry(driver->driver_name, proc_tty_driver);
+	remove_proc_entry(ent->name, proc_tty_driver);
 	
 	driver->proc_entry = 0;
 }
diff -urNp linux-5270/fs/reiserfs/ioctl.c linux-10010/fs/reiserfs/ioctl.c
--- linux-5270/fs/reiserfs/ioctl.c	
+++ linux-10010/fs/reiserfs/ioctl.c	
@@ -84,6 +84,7 @@ int reiserfs_ioctl (struct inode * inode
 	default:
 		return -ENOTTY;
 	}
+	return 0;
 }
 
 /*
diff -urNp linux-5270/include/asm-alpha/hardirq.h linux-10010/include/asm-alpha/hardirq.h
--- linux-5270/include/asm-alpha/hardirq.h	2001-07-09 23:47:39.000000000 +0200
+++ linux-10010/include/asm-alpha/hardirq.h	
@@ -41,6 +41,7 @@ extern unsigned long __irq_attempt[];
 
 #define synchronize_irq()	barrier()
 
+#define release_irqlock(cpu) ((void) 0)
 #else
 
 #define irq_attempt(cpu, irq) (cpu_data[cpu].irq_attempt[irq])
diff -urNp linux-5270/include/asm-alpha/smplock.h linux-10010/include/asm-alpha/smplock.h
--- linux-5270/include/asm-alpha/smplock.h	2000-03-23 21:50:09.000000000 +0100
+++ linux-10010/include/asm-alpha/smplock.h	
@@ -7,7 +7,7 @@
 #include <linux/sched.h>
 #include <linux/interrupt.h>
 #include <linux/spinlock.h>
-
+#include <asm/hardirq.h>
 extern spinlock_t kernel_flag;
 
 #define kernel_locked()		spin_is_locked(&kernel_flag)
diff -urNp linux-5270/include/asm-i386/rwsem.h linux-10010/include/asm-i386/rwsem.h
--- linux-5270/include/asm-i386/rwsem.h	
+++ linux-10010/include/asm-i386/rwsem.h	
@@ -42,6 +42,7 @@
 
 #include <linux/list.h>
 #include <linux/spinlock.h>
+#include <asm/system.h>
 
 struct rwsem_waiter;
 
diff -urNp linux-5270/include/asm-i386/smpboot.h linux-10010/include/asm-i386/smpboot.h
--- linux-5270/include/asm-i386/smpboot.h	
+++ linux-10010/include/asm-i386/smpboot.h	
@@ -99,4 +99,12 @@ extern volatile int cpu_2_physical_apici
 #define cpu_to_boot_apicid(cpu) cpu_2_physical_apicid[cpu]
 #endif /* CONFIG_MULTIQUAD */
 
+static inline int cpu_to_apicid (int cpu)
+{
+	if (clustered_apic_mode)
+		return cpu_to_physical_apicid(cpu);
+	else
+		return cpu_to_logical_apicid(cpu);
+}
+
 #endif
diff -urNp linux-5270/include/linux/binfmts.h linux-10010/include/linux/binfmts.h
--- linux-5270/include/linux/binfmts.h	
+++ linux-10010/include/linux/binfmts.h	
@@ -18,6 +18,8 @@
 
 struct file;
 
+struct mm_struct;
+
 /*
  * This structure is used to hold the arguments that are used when loading binaries.
  */
diff -urNp linux-5270/include/linux/list.h linux-10010/include/linux/list.h
--- linux-5270/include/linux/list.h	
+++ linux-10010/include/linux/list.h	
@@ -15,9 +15,9 @@
  * using the generic single-entry routines.
  */
 
-struct list_head {
+typedef struct list_head {
 	struct list_head *next, *prev;
-};
+} list_t;
 
 #define LIST_HEAD_INIT(name) { &(name), &(name) }
 
diff -urNp linux-5270/include/linux/proc_fs.h linux-10010/include/linux/proc_fs.h
--- linux-5270/include/linux/proc_fs.h	
+++ linux-10010/include/linux/proc_fs.h	
@@ -210,6 +210,8 @@ static inline void proc_tty_register_dri
 static inline void proc_tty_unregister_driver(struct tty_driver *driver) {};
 
 extern struct proc_dir_entry proc_root;
+extern struct proc_dir_entry proc_root_driver;
+
 
 #endif /* CONFIG_PROC_FS */
 
diff -urNp linux-5270/include/linux/vt_kern.h linux-10010/include/linux/vt_kern.h
--- linux-5270/include/linux/vt_kern.h	
+++ linux-10010/include/linux/vt_kern.h	
@@ -9,6 +9,7 @@
 #include <linux/config.h>
 #include <linux/vt.h>
 #include <linux/kd.h>
+#include <linux/tty.h>
 
 /*
  * Presently, a lot of graphics programs do not restore the contents of
diff -urNp linux-5270/init/do_mounts.c linux-10010/init/do_mounts.c
--- linux-5270/init/do_mounts.c	
+++ linux-10010/init/do_mounts.c	
@@ -443,6 +443,10 @@ static void __init change_floppy(char *f
 		close(fd);
 	}
 }
+#else
+static void __init change_floppy(char *fmt, ...)
+{
+}
 #endif
 
 #ifdef CONFIG_BLK_DEV_RAM
diff -urNp linux-5270/kernel/ksyms.c linux-10010/kernel/ksyms.c
--- linux-5270/kernel/ksyms.c	
+++ linux-10010/kernel/ksyms.c	
@@ -526,9 +526,6 @@ EXPORT_SYMBOL(simple_strtoul);
 EXPORT_SYMBOL(simple_strtoull);
 EXPORT_SYMBOL(system_utsname);	/* UTS data */
 EXPORT_SYMBOL(uts_sem);		/* UTS semaphore */
-#ifndef __mips__
-EXPORT_SYMBOL(sys_call_table);
-#endif
 EXPORT_SYMBOL(machine_restart);
 EXPORT_SYMBOL(machine_halt);
 EXPORT_SYMBOL(machine_power_off);
diff -urNp linux-5270/lib/rwsem-spinlock.c linux-10010/lib/rwsem-spinlock.c
--- linux-5270/lib/rwsem-spinlock.c	
+++ linux-10010/lib/rwsem-spinlock.c	
@@ -280,6 +280,8 @@ EXPORT_SYMBOL(__down_read);
 EXPORT_SYMBOL(__down_write);
 EXPORT_SYMBOL(__up_read);
 EXPORT_SYMBOL(__up_write);
+EXPORT_SYMBOL(__down_write_trylock);
+EXPORT_SYMBOL(__down_read_trylock);
 #if RWSEM_DEBUG
 EXPORT_SYMBOL(rwsemtrace);
 #endif
diff -urNp linux-5270/Makefile linux-10010/Makefile
--- linux-5270/Makefile	
+++ linux-10010/Makefile	
@@ -340,7 +340,7 @@ xconfig: symlinks
 	$(MAKE) -C scripts kconfig.tk
 	wish -f scripts/kconfig.tk
 
-menuconfig: include/linux/version.h symlinks
+menuconfig: symlinks
 	$(MAKE) -C scripts/lxdialog all
 	$(CONFIG_SHELL) scripts/Menuconfig arch/$(ARCH)/config.in
 
diff -urNp linux-5270/mm/vmalloc.c linux-10010/mm/vmalloc.c
--- linux-5270/mm/vmalloc.c	
+++ linux-10010/mm/vmalloc.c	
@@ -12,6 +12,7 @@
 #include <linux/spinlock.h>
 #include <linux/highmem.h>
 #include <linux/smp_lock.h>
+#include <linux/interrupt.h>
 
 #include <asm/uaccess.h>
 #include <asm/pgalloc.h>
@@ -19,6 +20,9 @@
 rwlock_t vmlist_lock = RW_LOCK_UNLOCKED;
 struct vm_struct * vmlist;
 
+extern void show_stack(unsigned long *);
+
+
 static inline void free_area_pte(pmd_t * pmd, unsigned long address, unsigned long size)
 {
 	pte_t * pte;
@@ -218,6 +222,12 @@ void vfree(void * addr)
 		printk(KERN_ERR "Trying to vfree() bad address (%p)\n", addr);
 		return;
 	}
+	if (in_interrupt()) {
+		printk("vfree(): sleeping in interrupt!! \n");
+#ifdef __i386__	
+		show_stack(NULL);
+#endif	
+	}
 	write_lock(&vmlist_lock);
 	for (p = &vmlist ; (tmp = *p) ; p = &tmp->next) {
 		if (tmp->addr == addr) {
diff -urNp linux-5270/net/sunrpc/auth.c linux-10010/net/sunrpc/auth.c
--- linux-5270/net/sunrpc/auth.c	2001-09-21 20:24:50.000000000 +0200
+++ linux-10010/net/sunrpc/auth.c	
@@ -13,6 +13,7 @@
 #include <linux/socket.h>
 #include <linux/sunrpc/clnt.h>
 #include <linux/spinlock.h>
+#include <linux/module.h>
 
 #ifdef RPC_DEBUG
 # define RPCDBG_FACILITY	RPCDBG_AUTH
diff -urNp linux-5270/net/wanrouter/wanproc.c linux-10010/net/wanrouter/wanproc.c
--- linux-5270/net/wanrouter/wanproc.c	2001-09-10 16:58:35.000000000 +0200
+++ linux-10010/net/wanrouter/wanproc.c	
@@ -1081,6 +1081,13 @@ static int device_write(
 }
 #endif
 
+#ifndef CONFIG_PROC_FS
+	int __init wanrouter_proc_init (void)
+	{
+		return 0;
+	}
+#endif
+
 /*
  *	End
  */
--- linux-2.4.20/include/linux/kbd_kern.h.orig	Thu Mar  6 18:47:09 2003
+++ linux-2.4.20/include/linux/kbd_kern.h	Thu Mar  6 19:04:01 2003
@@ -3,6 +3,8 @@
 
 #include <linux/interrupt.h>
 #include <linux/keyboard.h>
+#include <linux/sched.h>
+#include <linux/tty.h>
 
 extern struct tasklet_struct keyboard_tasklet;
 
--- linux-2.4.20/drivers/parport/parport_cs.c.orig	Fri Dec 21 18:41:55 2001
+++ linux-2.4.20/drivers/parport/parport_cs.c	Thu Mar  6 20:06:35 2003
@@ -45,6 +45,7 @@
 #include <linux/string.h>
 #include <linux/timer.h>
 #include <linux/ioport.h>
+#include <linux/major.h>
 
 #include <linux/parport.h>
 #include <linux/parport_pc.h>
--- linux-2.4.20/fs/intermezzo/dir.c.orig	Fri Nov 29 00:53:15 2002
+++ linux-2.4.20/fs/intermezzo/dir.c	Thu Mar  6 20:45:17 2003
@@ -27,6 +27,8 @@
 #include <asm/bitops.h>
 #include <asm/uaccess.h>
 #include <asm/system.h>
+#include <asm/ioctls.h>
+#include <asm/termios.h>
 #include <linux/smp_lock.h>
 
 #include <linux/errno.h>
--- linux-2.4.20/net/bluetooth/bnep/core.c~	Fri Nov 29 00:53:15 2002
+++ linux-2.4.20/net/bluetooth/bnep/core.c	Thu Mar  6 21:02:13 2003
@@ -458,7 +458,7 @@
         sigfillset(&current->blocked);
 	flush_signals(current);
 
-	current->nice = -15;
+	set_user_nice(current, -15);
 
         set_fs(KERNEL_DS);
 
