diff -up new/fs/davfs/dav_debug.h.orig new/fs/davfs/dav_debug.h
--- new/fs/davfs/dav_debug.h.orig	2002-08-09 22:38:28.000000000 +0200
+++ new/fs/davfs/dav_debug.h	2002-08-09 22:35:32.000000000 +0200
@@ -11,14 +11,14 @@
  * these are normally enabled.
  */
 #ifdef SMBFS_PARANOIA
-#define PARANOIA(x...) printk(KERN_NOTICE __FUNCTION__ ": " x)
+#define PARANOIA(fmt,args...) printk(KERN_NOTICE "%s: " fmt, __FUNCTION__, args)
 #else
 #define PARANOIA(x...) do { ; } while(0)
 #endif
 
 /* lots of debug messages */
 #ifdef SMBFS_DEBUG_VERBOSE
-#define VERBOSE(x...) printk(KERN_DEBUG __FUNCTION__ ": " x)
+#define VERBOSE(fmt, args...) printk(KERN_DEBUG "%s: " fmt, __FUNCTION__, args)
 #else
 #define VERBOSE(x...) do { ; } while(0)
 #endif
@@ -28,19 +28,19 @@
  * too common name.
  */
 #ifdef SMBFS_TRACE
-#define TRACE() printk(KERN_DEBUG __FUNCTION__ "--trace--\n")
+#define TRACE() printk(KERN_DEBUG "%s--trace--\n", __FUNCTION__)
 #else
 #define TRACE() do { ; } while(0)
 #endif
 
 #ifdef SMBFS_DEBUG
-#define DEBUG1(x...) printk(KERN_DEBUG __FUNCTION__ ": " x)
+#define DEBUG1(fmt, args...) printk(KERN_DEBUG "%s: " fmt, __FUNCTION__, args)
 #else
 #define DEBUG1(x...) do { ; } while(0)
 #endif
 
 #ifdef SMBFS_DEBUG2
-#define DEBUG2(x...) printk(KERN_DEBUG __FUNCTION__ ": " x)
+#define DEBUG2(fmt, args...) printk(KERN_DEBUG "%s: " fmt, __FUNCTION__, args)
 #else
 #define DEBUG2(x...) do { ; } while(0)
 #endif
diff -ur linux-2.4.20/fs/davfs/proc.c linux-2.4.20.fpu/fs/davfs/proc.c
--- linux-2.4.20/fs/davfs/proc.c	Sun Mar  9 21:12:08 2003
+++ linux-2.4.20.fpu/fs/davfs/proc.c	Sun Mar  9 21:05:33 2003
@@ -156,7 +156,7 @@
 int dav_atoi(char *digit) {
     int ret=0;
     int i = (strlen(digit)-1);
-    double weight=1;
+    unsigned int weight=1;
     
     for(;i;i--)
         weight*=10;
