diff -urN linux/fs/proc/base.c linux.new/fs/proc/base.c
--- linux/fs/proc/base.c	Mon Nov 26 18:23:38 2001
+++ linux.new/fs/proc/base.c	Mon Nov 26 18:20:19 2001
@@ -704,13 +704,15 @@
 	get_task_struct(task);
 	inode->u.proc_i.task = task;
 	inode->i_uid = 0;
+#ifndef CONFIG_GRKERNSEC_PROC_USERGROUP
 	inode->i_gid = 0;
+#else
+	inode->i_gid = CONFIG_GRKERNSEC_PROC_GID;
+#endif
 	if (ino == PROC_PID_INO || task_dumpable(task)) {
 		inode->i_uid = task->euid;
 #ifndef CONFIG_GRKERNSEC_PROC_USERGROUP
 		inode->i_gid = task->egid;
-#else
-               inode->i_gid = CONFIG_GRKERNSEC_PROC_GID;
 #endif
 	}
 
