diff -Naur -X /home/marcelo/lib/dontdiff linux-2.4.22/include/asm-alpha/kmap_types.h linux-2.4.23-pre3/include/asm-alpha/kmap_types.h
--- linux-2.4.22/include/asm-alpha/kmap_types.h	1970-01-01 00:00:00.000000000 +0000
+++ linux-2.4.23-pre3/include/asm-alpha/kmap_types.h	2003-09-03 20:34:21.000000000 +0000
@@ -0,0 +1,18 @@
+#ifdef __KERNEL__
+#ifndef _ASM_KMAP_TYPES_H
+#define _ASM_KMAP_TYPES_H
+
+enum km_type {
+	KM_BOUNCE_READ,
+	KM_SKB_SUNRPC_DATA,
+	KM_SKB_DATA_SOFTIRQ,
+	KM_USER0,
+	KM_USER1,
+	KM_BH_IRQ,
+	KM_SOFTIRQ0,
+	KM_SOFTIRQ1,
+	KM_TYPE_NR
+};
+
+#endif
+#endif /* __KERNEL__ */
