--- linux-2.6.6/Makefile.orig	2004-05-23 13:18:32.000000000 +0200
+++ linux-2.6.6/Makefile	2004-05-23 13:20:16.759551080 +0200
@@ -397,7 +397,6 @@
 # Build targets only - this includes vmlinux, arch specific targets, clean
 # targets and others. In general all targets except *config targets.
 
-ifeq ($(KBUILD_EXTMOD),)
 # Additional helpers built in scripts/
 # Carefully list dependencies so we do not try to build scripts twice
 # in parrallel
@@ -407,6 +406,7 @@
 
 scripts_basic: include/linux/autoconf.h
 
+ifeq ($(KBUILD_EXTMOD),)
 # Objects we will link into vmlinux / subdirs we need to visit
 init-y		:= init/
 drivers-y	:= drivers/ sound/ cluster/
@@ -957,7 +957,7 @@
 
 module-dirs := $(addprefix _module_,$(KBUILD_EXTMOD))
 .PHONY: $(module-dirs) modules
-$(module-dirs): crmodverdir $(objtree)/Module.symvers
+$(module-dirs): crmodverdir scripts $(objtree)/Module.symvers
 	$(Q)$(MAKE) $(build)=$(patsubst _module_%,%,$@)
 
 modules: $(module-dirs)
