diff -Nur /usr/src/linux/include/linux/netfilter_ipv4/ip_conntrack_h323.h linux-2.4.19/include/linux/netfilter_ipv4/ip_conntrack_h323.h
--- linux-2.4.19.org/include/linux/netfilter_ipv4/ip_conntrack_h323.h	Mon Sep  9 10:12:19 2002
+++ linux-2.4.19/include/linux/netfilter_ipv4/ip_conntrack_h323.h	Mon Sep  9 10:23:44 2002
@@ -3,13 +3,15 @@
 /* H.323 connection tracking. */
 
 #ifdef __KERNEL__
+
 /* Protects H.323 related data */
 DECLARE_LOCK_EXTERN(ip_h323_lock);
-#endif
 
 /* Default H.225 port */
 #define H225_PORT	1720
 
+#endif
+
 /* This structure is per expected connection */
 struct ip_ct_h225_expect {
 	u_int16_t port;			/* Port of the H.225 helper/RTCP/RTP channel */
diff -Nur /usr/src/linux/include/linux/netfilter_ipv4/ip_conntrack_rsh.h linux-2.4.19/include/linux/netfilter_ipv4/ip_conntrack_rsh.h
--- linux-2.4.19.org/include/linux/netfilter_ipv4/ip_conntrack_rsh.h	Mon Sep  9 10:12:59 2002
+++ linux-2.4.19/include/linux/netfilter_ipv4/ip_conntrack_rsh.h	Mon Sep  9 10:24:15 2002
@@ -12,9 +12,7 @@
 #ifndef _IP_CONNTRACK_RSH_H
 #define _IP_CONNTRACK_RSH_H
 
+#ifdef __KERNEL__
-#ifndef __KERNEL__
-#error Only in kernel.
-#endif
 
 #include <linux/netfilter_ipv4/lockhelp.h>
 
@@ -20,6 +22,8 @@
 
 DECLARE_LOCK_EXTERN(ip_rsh_lock);
 
+#endif /* __KERNEL__ */
+
 /* This structure is per expected connection */
 struct ip_ct_rsh_expect
 {
diff -Nur /usr/src/linux/include/linux/netfilter_ipv4/ip_conntrack_talk.h linux-2.4.19/include/linux/netfilter_ipv4/ip_conntrack_talk.h
--- linux-2.4.19.org/include/linux/netfilter_ipv4/ip_conntrack_talk.h	Mon Sep  9 10:13:15 2002
+++ linux-2.4.19/include/linux/netfilter_ipv4/ip_conntrack_talk.h	Mon Sep  9 10:24:29 2002
@@ -2,9 +2,7 @@
 #define _IP_CONNTRACK_TALK_H
 /* TALK tracking. */
 
+#ifdef __KERNEL__
-#ifndef __KERNEL__
-#error Only in kernel.
-#endif
 
 #include <linux/in.h>
 #include <linux/netfilter_ipv4/lockhelp.h>
@@ -13,6 +15,8 @@
 #define TALK_PORT	517
 #define NTALK_PORT	518
 
+#endif /* __KERNEL__ */
+
 /* talk structures and constants from <protocols/talkd.h> */
 
 /*
@@ -106,6 +106,8 @@
 	char	r_name[TALK_NSIZE]; /* callee's name            */
 };
 
+#ifdef __KERNEL__
+
 #define TALK_STR(data, talk_str, member) ((struct talk_str *)data)->member)
 #define TALK_RESP(data, ver, member) (ver ? ((struct ntalk_response *)data)->member : ((struct talk_response *)data)->member)
 #define TALK_MSG(data, ver, member) (ver ? ((struct ntalk_msg *)data)->member : ((struct talk_msg *)data)->member)
@@ -140,6 +138,8 @@
 #define SELECTIVE_REFUSAL 11	/* User Filter refusal.               */
 #define MAX_RESPONSE_TYPE 11	/* Make sure this is updated          */
 
+#endif /* __KERNEL__ */
+
 /* We don't really need much for talk */
 struct ip_ct_talk_expect
 {
