--- a/include/linux/netfilter_ipv4/ip_conntrack_core.h	2007-01-04 02:01:13.000000000 +0100
+++ b/include/linux/netfilter_ipv4/ip_conntrack_core.h	2007-01-04 11:55:16.000000000 +0100
@@ -32,7 +32,7 @@ ip_ct_invert_tuple(struct ip_conntrack_t
 		   const struct ip_conntrack_protocol *protocol);
 
 /* Find a connection corresponding to a tuple. */
-struct ip_conntrack_tuple_hash *
+extern struct ip_conntrack_tuple_hash *
 ip_conntrack_find_get(const struct ip_conntrack_tuple *tuple,
 		      const struct ip_conntrack *ignored_conntrack);
 
--- a/net/ipv4/netfilter/ip_conntrack_core.c	2007-01-04 01:59:01.000000000 +0100
+++ b/net/ipv4/netfilter/ip_conntrack_core.c	2007-01-04 01:59:01.000000000 +0100
@@ -403,6 +403,8 @@ ip_conntrack_find_get(const struct ip_co
 	return h;
 }
 
+EXPORT_SYMBOL_GPL(ip_conntrack_find_get);
+
 static void __ip_conntrack_hash_insert(struct ip_conntrack *ct,
 					unsigned int hash,
 					unsigned int repl_hash) 
