--- linux/kernel/sys.c~	Sun Feb  3 01:38:13 2002
+++ linux/kernel/sys.c	Sun Feb  3 20:19:15 2002
@@ -220,7 +220,7 @@
 		}
 		if (error == -ESRCH)
 			error = 0;
-		if (niceval < task_nice(p) && !capable(CAP_SYS_NICE))
+		if (niceval < p->nice && !capable(CAP_SYS_NICE))
 			error = -EACCES;
 		else
 			set_user_nice(p, niceval);
--- linux/kernel/sched.c~	Mon Feb  4 15:01:07 2002
+++ linux/kernel/sched.c	Mon Feb  4 15:04:10 2002
@@ -19,7 +19,6 @@
 #include <linux/smp_lock.h>
 #include <linux/interrupt.h>
 #include <asm/mmu_context.h>
-#include <linux/kernel_stat.h>
 
 /*
  * Priority of a process goes from 0 to 139. The 0-99
--- linux/kernel/printk.c.orig	Tue Feb  5 13:51:53 2002
+++ linux/kernel/printk.c	Tue Feb  5 13:52:12 2002
@@ -26,7 +26,6 @@
 #include <linux/module.h>
 #include <linux/interrupt.h>			/* For in_interrupt() */
 #include <linux/config.h>
-#include <linux/delay.h>
 
 #include <asm/uaccess.h>
 
