--- linux-2.4.21/drivers/md/dm-io.c.orig	Tue Aug 19 18:03:52 2003
+++ linux-2.4.21/drivers/md/dm-io.c	Tue Aug 19 18:08:06 2003
@@ -177,11 +177,6 @@
 /*
  * Primitives for alignment calculations.
  */
-int fls(unsigned n)
-{
-	return generic_fls32(n);
-}
-
 static inline int log2_floor(unsigned n)
 {
 	return ffs(n) - 1;
@@ -189,7 +184,7 @@
 
 static inline int log2_align(unsigned n)
 {
-	return fls(n) - 1;
+	return generic_fls32(n) - 1;
 }
 
 /*
