--- linux-2.4.21/drivers/net/irda/ma600.c.orig	Fri Nov 29 00:53:13 2002
+++ linux-2.4.21/drivers/net/irda/ma600.c	Thu Jun 26 18:12:16 2003
@@ -32,7 +32,7 @@
  ********************************************************************/
 
 /* define this macro for release version */
-//#define NDEBUG
+#define NDEBUG
 
 #include <linux/module.h>
 #include <linux/delay.h>
--- linux-2.4.21/drivers/net/wan/sbni.c.orig	Thu Jun 26 18:20:25 2003
+++ linux-2.4.21/drivers/net/wan/sbni.c	Thu Jun 26 18:32:25 2003
@@ -149,11 +149,11 @@
 static int  enslave( struct net_device *, struct net_device * );
 static int  emancipate( struct net_device * );
 #endif
-
+/*
 #ifdef __i386__
 #define ASM_CRC 1
 #endif
-
+*/
 static const char  version[] =
 	"Granch SBNI12 driver ver 5.0.1  Jun 22 2001  Denis I.Timofeev.\n";
 
--- linux-2.4.21/fs/partitions/ldm.c.orig	2003-08-16 03:27:54.000000000 +0200
+++ linux-2.4.21/fs/partitions/ldm.c	2003-08-16 03:28:09.000000000 +0200
@@ -44,7 +44,7 @@
  * driver was compiled with debug enabled. Otherwise, the call turns into a NOP.
  */
 #ifndef CONFIG_LDM_DEBUG
-#define ldm_debug(...)	do {} while (0)
+#define ldm_debug(f, a...)	do {} while (0)
 #else
 #define ldm_debug(f, a...) _ldm_printk (KERN_DEBUG, __FUNCTION__, f, ##a)
 #endif
