diff -urN linux.orig/drivers/ieee1394/pcilynx.c linux/drivers/ieee1394/pcilynx.c
--- linux.orig/drivers/ieee1394/pcilynx.c	Sat Mar 16 21:00:04 2002
+++ linux/drivers/ieee1394/pcilynx.c	Sun Mar 17 10:51:19 2002
@@ -1360,10 +1360,6 @@
         lynx->lock = SPIN_LOCK_UNLOCKED;
         lynx->phy_reg_lock = SPIN_LOCK_UNLOCKED;
 
-        if (pci_set_dma_mask(dev, 0xffffffff)) {
-                FAIL("DMA address limits not supported for PCILynx hardware %d",
-                     lynx->id);
-        }
         if (pci_enable_device(dev)) {
                 FAIL("failed to enable PCILynx hardware %d", lynx->id);
         }
