--- a/drivers/char/misc.c.orig	2004-12-24 22:35:28.000000000 +0100
+++ b/drivers/char/misc.c	2005-01-05 11:16:47.000000000 +0100
@@ -220,7 +220,7 @@
 	}
 
 	if (misc->minor == MISC_DYNAMIC_MINOR) {
-		int i = DYNAMIC_MINORS;
+		int i = DYNAMIC_MINORS - 3 /* reserved */;
 		while (--i >= 0)
 			if ( (misc_minors[i>>3] & (1 << (i&7))) == 0)
 				break;
--- a/drivers/md/dm-ioctl.c.orig	2004-12-24 22:34:29.000000000 +0100
+++ b/drivers/md/dm-ioctl.c	2005-01-05 11:11:39.000000000 +0100
@@ -1374,7 +1374,7 @@
 };
 
 static struct miscdevice _dm_misc = {
-	.minor 		= MISC_DYNAMIC_MINOR,
+	.minor 		= 63 /* MISC_DYNAMIC_MINOR */,
 	.name  		= DM_NAME,
 	.fops  		= &_ctl_fops
 };
