--- linux/drivers/video/vesafb.c~	Sun Jan  7 03:43:00 2001
+++ linux/drivers/video/vesafb.c	Sun Jan  7 04:58:01 2001
@@ -637,7 +637,7 @@
 		int temp_size = video_size;
 		/* Find the largest power-of-two */
 		while (temp_size & (temp_size - 1))
-                	temp_sze &= (temp_size - 1);
+                	temp_size &= (temp_size - 1);
                         
                 /* Try and find a power of two to add */
 		while (temp_size && mtrr_add(video_base, temp_size, MTRR_TYPE_WRCOMB, 1)==-EINVAL) {
