--- linux-2.4.20/fs/ntfs/aops.c.org	Thu Dec  5 09:51:21 2002
+++ linux-2.4.20/fs/ntfs/aops.c	Thu Dec  5 09:51:27 2002
@@ -45,6 +45,9 @@
  * record size, and index_block_size_bits, to the log(base 2) of the ntfs
  * record size.
  */
+
+#define KM_BIO_IRQ	KM_BH_IRQ
+
 static void ntfs_end_buffer_async_read(struct buffer_head *bh, int uptodate)
 {
 	static spinlock_t page_uptodate_lock = SPIN_LOCK_UNLOCKED;
