diff -ur linux-2.4.20/drivers/media/video/tda7432.c linux-2.4.20.fpu/drivers/media/video/tda7432.c
--- linux-2.4.20/drivers/media/video/tda7432.c	Fri Nov 29 00:53:13 2002
+++ linux-2.4.20.fpu/drivers/media/video/tda7432.c	Sun Mar  9 20:55:29 2003
@@ -398,7 +398,7 @@
 		if(!maxvol){  /* max +20db */
 			va->volume = ( 0x6f - (t->volume & 0x7F) ) * 630;
 		} else {     /* max 0db   */
-			va->volume = (int )(( 0x6f - (t->volume & 0x7F) ) * 829.557);
+			va->volume = (int )(( 0x6f - (t->volume & 0x7F) ) * 829557/1000);
 		}
 		
 		/* Balance depends on L,R attenuation
@@ -440,7 +440,7 @@
 			if(!maxvol){ /* max +20db */
 				t->volume = 0x6f - ( (va->volume)/630 );
 			} else {    /* max 0db   */
-				t->volume = 0x6f - ((int) (va->volume)/829.557 );
+				t->volume = 0x6f - ((int) (va->volume*1000)/829557 );
 			}
 		
 		if (loudness)		/* Turn on the loudness bit */
