diff -Nur linux-2.4.20.org/net/netsyms.c linux-2.4.20/net/netsyms.c
--- linux-2.4.20.org/net/netsyms.c	Wed Nov  5 11:21:18 2003
+++ linux-2.4.20/net/netsyms.c	Wed Nov  5 10:32:28 2003
@@ -606,10 +606,14 @@
 #if !defined (CONFIG_IPV6_MODULE) && !defined (CONFIG_KHTTPD) && !defined (CONFIG_KHTTPD_MODULE)
 EXPORT_SYMBOL(tcp_v4_lookup_listener);
 #endif
+#endif
+
+#if defined(CONFIG_IP_NF_MATCH_STEALTH_MODULE)
 extern struct sock *udp_v4_lookup(u32 saddr, u16 sport, u32 daddr, u16 dport, int dif);
-#if defined(CONFIG_IP_NF_MATCH_OWNER)||defined(CONFIG_IP_NF_MATCH_OWNER_MODULE)
-EXPORT_SYMBOL(udp_v4_lookup);
 #endif
+
+#if defined(CONFIG_IP_NF_MATCH_STEALTH_MODULE)||defined(CONFIG_IP_NF_MATCH_OWNER)||defined(CONFIG_IP_NF_MATCH_OWNER_MODULE)
+EXPORT_SYMBOL(udp_v4_lookup);
 #endif
 
 #if defined(CONFIG_GRKERNSEC_RANDID)
