diff -urN drm.org/drmP.h drm/drmP.h
--- drm.org/drmP.h	Sun May 26 10:07:41 2002
+++ drm/drmP.h	Sun May 26 10:52:19 2002
@@ -32,6 +32,8 @@
 #ifndef _DRM_P_H_
 #define _DRM_P_H_
 
+#define CONFIG_X86_CMPXCHG 1
+
 #ifdef __KERNEL__
 #ifdef __alpha__
 /* add include of current.h so that "current" is defined
diff -urN drm.org/i810_dma.c drm/i810_dma.c
--- drm.org/i810_dma.c	Sun May 26 10:07:41 2002
+++ drm/i810_dma.c	Sun May 26 23:33:53 2002
@@ -31,6 +31,7 @@
  */
 
 #define __NO_VERSION__
+#include <linux/config.h>
 #include "i810.h"
 #include "drmP.h"
 #include "i810_drv.h"
diff -urN drm.org/i830_dma.c drm/i830_dma.c
--- drm.org/i830_dma.c	Sun May 26 10:07:41 2002
+++ drm/i830_dma.c	Sun May 26 23:34:02 2002
@@ -32,6 +32,7 @@
  */
 
 #define __NO_VERSION__
+#include <linux/config.h>
 #include "i830.h"
 #include "drmP.h"
 #include "i830_drv.h"
