--- linux-2.4.20/net/ipv4/netfilter/ip_conntrack_amanda.c.orig	Wed Apr 21 12:12:39 2004
+++ linux-2.4.20/net/ipv4/netfilter/ip_conntrack_amanda.c	Thu Apr 22 15:48:50 2004
@@ -154,11 +154,11 @@
 						{ { ct->tuplehash[IP_CT_DIR_ORIGINAL].tuple.src.ip,
 						  { 0 } },
 				  		{ ct->tuplehash[IP_CT_DIR_ORIGINAL].tuple.dst.ip,
-						  { htons(exp_amanda_info->port) },
+						  { .tcp = { htons(exp_amanda_info->port) } },
 				    	  IPPROTO_TCP }});
 				expect.mask = ((struct ip_conntrack_tuple)
 							   { { 0, { 0 } },
-				  				 { 0xFFFFFFFF, { 0xFFFF }, 0xFFFF }});
+				  				 { 0xFFFFFFFF, { .tcp = { 0xFFFF } }, 0xFFFF }});
 
 				expect.expectfn = NULL;
 
