diff -urNp --exclude-from=/home/mitica/quintela/config/misc/dontdiff 2.4.19-rc3/drivers/video/radeonfb.c test/drivers/video/radeonfb.c
--- 2.4.19-rc3/drivers/video/radeonfb.c	2002-07-22 18:43:34.000000000 +0200
+++ test/drivers/video/radeonfb.c	2002-07-22 18:48:09.000000000 +0200
@@ -599,11 +599,6 @@ static int panel_yres __initdata = 0;
 static char force_dfp __initdata = 0;
 static struct radeonfb_info *board_list = NULL;
 
-#ifdef FBCON_HAS_CFB8
-static struct display_switch fbcon_radeon8;
-#endif
-
-
 /*
  * prototypes
  */
@@ -2948,89 +2943,3 @@ int radeon_sleep_notify(struct pmu_sleep
 
 #endif /* CONFIG_PMAC_PBOOK */
 
-/*
- * text console acceleration
- */
-
-
-static void fbcon_radeon_bmove(struct display *p, int srcy, int srcx,
-			       int dsty, int dstx, int height, int width)
-{
-	struct radeonfb_info *rinfo = (struct radeonfb_info *)(p->fb_info);
-	u32 dp_cntl = DST_LAST_PEL;
-
-	srcx *= fontwidth(p);
-	srcy *= fontheight(p);
-	dstx *= fontwidth(p);
-	dsty *= fontheight(p);
-	width *= fontwidth(p);
-	height *= fontheight(p);
-
-	if (srcy < dsty) {
-		srcy += height - 1;
-		dsty += height - 1;
-	} else
-		dp_cntl |= DST_Y_TOP_TO_BOTTOM;
-
-	if (srcx < dstx) {
-		srcx += width - 1;
-		dstx += width - 1;
-	} else
-		dp_cntl |= DST_X_LEFT_TO_RIGHT;
-
-	radeon_fifo_wait(6);
-	OUTREG(DP_GUI_MASTER_CNTL, (rinfo->dp_gui_master_cntl |
-				    GMC_BRUSH_NONE |
-				    GMC_SRC_DATATYPE_COLOR |
-				    ROP3_S |
-				    DP_SRC_SOURCE_MEMORY));
-	OUTREG(DP_WRITE_MSK, 0xffffffff);
-	OUTREG(DP_CNTL, dp_cntl);
-	OUTREG(SRC_Y_X, (srcy << 16) | srcx);
-	OUTREG(DST_Y_X, (dsty << 16) | dstx);
-	OUTREG(DST_HEIGHT_WIDTH, (height << 16) | width);
-}
-
-
-
-static void fbcon_radeon_clear(struct vc_data *conp, struct display *p,
-			       int srcy, int srcx, int height, int width)
-{
-	struct radeonfb_info *rinfo = (struct radeonfb_info *)(p->fb_info);
-	u32 clr;
-
-	clr = attr_bgcol_ec(p, conp);
-	clr |= (clr << 8);
-	clr |= (clr << 16);
-
-	srcx *= fontwidth(p);
-	srcy *= fontheight(p);
-	width *= fontwidth(p);
-	height *= fontheight(p);
-
-	radeon_fifo_wait(6);
-	OUTREG(DP_GUI_MASTER_CNTL, (rinfo->dp_gui_master_cntl |
-				    GMC_BRUSH_SOLID_COLOR |
-				    GMC_SRC_DATATYPE_COLOR |
-				    ROP3_P));
-	OUTREG(DP_BRUSH_FRGD_CLR, clr);
-	OUTREG(DP_WRITE_MSK, 0xffffffff);
-	OUTREG(DP_CNTL, (DST_X_LEFT_TO_RIGHT | DST_Y_TOP_TO_BOTTOM));
-	OUTREG(DST_Y_X, (srcy << 16) | srcx);
-	OUTREG(DST_WIDTH_HEIGHT, (width << 16) | height);
-}
-
-
-
-#ifdef FBCON_HAS_CFB8
-static struct display_switch fbcon_radeon8 = {
-	setup:			fbcon_cfb8_setup,
-	bmove:			fbcon_radeon_bmove,
-	clear:			fbcon_radeon_clear,
-	putc:			fbcon_cfb8_putc,
-	putcs:			fbcon_cfb8_putcs,
-	revc:			fbcon_cfb8_revc,
-	clear_margins:		fbcon_cfb8_clear_margins,
-	fontwidthmask:		FONTWIDTH(4)|FONTWIDTH(8)|FONTWIDTH(12)|FONTWIDTH(16)
-};
-#endif
