diff -ur linux-2.4.20/arch/sparc64/kernel/sparc64_ksyms.c linux-2.4.20.xxx/arch/sparc64/kernel/sparc64_ksyms.c
--- linux-2.4.20/arch/sparc64/kernel/sparc64_ksyms.c	Fri Nov 29 00:53:12 2002
+++ linux-2.4.20.xxx/arch/sparc64/kernel/sparc64_ksyms.c	Mon Dec  9 19:34:01 2002
@@ -310,12 +310,12 @@
 EXPORT_SYMBOL(sunos_sys_table);
 EXPORT_SYMBOL(sys_call_table32);
 EXPORT_SYMBOL(tl0_solaris);
-EXPORT_SYMBOL(sys_sigsuspend);
-EXPORT_SYMBOL(sys_getppid);
+//EXPORT_SYMBOL(sys_sigsuspend);
+//EXPORT_SYMBOL(sys_getppid);
 EXPORT_SYMBOL(svr4_getcontext);
 EXPORT_SYMBOL(svr4_setcontext);
 EXPORT_SYMBOL(prom_cpu_nodes);
-EXPORT_SYMBOL(sys_ioctl);
+//EXPORT_SYMBOL(sys_ioctl);
 EXPORT_SYMBOL(sys32_ioctl);
 EXPORT_SYMBOL(sparc32_open);
 #endif
