--- linux-2.4.25.orig/drivers/net/imq.c	Mon Mar 22 01:19:20 2004
+++ linux-2.4.25/drivers/net/imq.c	Mon Mar 22 01:31:26 2004
@@ -38,7 +38,7 @@
 	imq_nf_hook,
 	PF_INET,
 	NF_IP_PRE_ROUTING,
-	NF_IP_PRI_MANGLE + 1
+	NF_IP_PRI_NAT_DST + 1
 };
 
 static struct nf_hook_ops imq_egress_ipv4 = {
@@ -55,7 +55,7 @@
 	imq_nf_hook,
 	PF_INET6,
 	NF_IP6_PRE_ROUTING,
-	NF_IP6_PRI_MANGLE + 1
+	NF_IP_PRI_NAT_DST + 1
 };
 
 static struct nf_hook_ops imq_egress_ipv6 = {
