--- linux-2.6.10-rc1/drivers/input/mousedev.c.btc	2004-11-07 00:14:06.527476789 +0100
+++ linux-2.6.10-rc1/drivers/input/mousedev.c	2004-11-07 00:18:25.423624344 +0100
@@ -173,13 +173,16 @@
 	int index;
 
 	switch (code) {
+		case 0x11F:
 		case BTN_TOUCH:
 		case BTN_0:
 		case BTN_FORWARD:
 		case BTN_LEFT:		index = 0; break;
+		case BTN_TRIGGER:
 		case BTN_STYLUS:
 		case BTN_1:
 		case BTN_RIGHT:		index = 1; break;
+		case BTN_THUMB:
 		case BTN_2:
 		case BTN_STYLUS2:
 		case BTN_MIDDLE:	index = 2; break;
