--- linux-2.4.10-pre7/include/scsi/scsi.h	Fri Apr 27 13:59:19 2001
+++ linux/include/scsi/scsi.h	Mon Sep 10 03:53:58 2001
@@ -214,6 +214,12 @@
 /* Used to get the PCI location of a device */
 #define SCSI_IOCTL_GET_PCI 0x5387
 
+/* Used to invoke Target Defice Reset for Fibre Channel */
+#define SCSI_IOCTL_FC_TDR 0x5388
+
+/* Used to get Fibre Channel WWN and port_id from device */
+#define SCSI_IOCTL_FC_TARGET_ADDRESS 0x5389
+
 /*
  * Overrides for Emacs so that we follow Linus's tabbing style.
  * Emacs will notice this stuff at the end of the file and automatically
--- linux-2.4.10-pre7/drivers/scsi/cpqfcTSinit.c	Sun Aug 12 10:51:41 2001
+++ linux/drivers/scsi/cpqfcTSinit.c	Mon Sep 10 03:54:23 2001
@@ -63,6 +63,7 @@
 
 #include <linux/config.h>  
 #include <linux/module.h>
+#include <linux/init.h>
 #include <linux/version.h> 
 
 /* Embedded module documentation macros - see module.h */
