diff -urN linux-2.4.21.org/drivers/char/pcwd.c linux-2.4.21/drivers/char/pcwd.c
--- linux-2.4.21.org/drivers/char/pcwd.c	Tue Aug 12 22:25:59 2003
+++ linux-2.4.21/drivers/char/pcwd.c	Tue Aug 12 22:28:53 2003
@@ -933,8 +933,7 @@
 
 	release_region (pcwd_info.io_addr, pcwd_info.card_info->io_size);
 
-	if (pcwd_info.flags & PCWD_PCI_REG)
-		pci_unregister_driver (&pcwd_driver);
+	pci_unregister_driver (&pcwd_driver);
 
 	return;
 }
