--- linux-2.4.19/kernel/ksyms.c.org	Tue Sep 24 10:32:40 2002
+++ linux-2.4.19/kernel/ksyms.c	Wed Sep 25 09:21:21 2002
@@ -611,6 +611,9 @@
 EXPORT_SYMBOL(gr_is_capable);
 EXPORT_SYMBOL(gr_pid_is_chrooted);
 EXPORT_SYMBOL(gr_learn_resource);
+EXPORT_SYMBOL(gr_handle_chroot_unix);
+EXPORT_SYMBOL(gr_acl_handle_mknod);
+EXPORT_SYMBOL(gr_handle_create);
  
 EXPORT_SYMBOL(elevator_init);
 EXPORT_SYMBOL(elevator_noop_merge_req);
