--- linux/include/crypto/des.h~	Wed Feb 25 23:59:49 2004
+++ linux/include/crypto/des.h	Thu Feb 26 00:06:03 2004
@@ -136,7 +136,6 @@
 #define ecb_encrypt des_ecb_encrypt
 #define cbc_encrypt des_cbc_encrypt
 #define ncbc_encrypt des_ncbc_encrypt
-#define xcbc_encrypt des_xcbc_encrypt
 #define cbc_cksum des_cbc_cksum
 #define quad_cksum des_quad_cksum
 
@@ -165,9 +164,6 @@
 	des_key_schedule schedule,des_cblock *ivec,int enc);
 void des_ncbc_encrypt(des_cblock *input,des_cblock *output,long length,
 	des_key_schedule schedule,des_cblock *ivec,int enc);
-void des_xcbc_encrypt(des_cblock *input,des_cblock *output,long length,
-	des_key_schedule schedule,des_cblock *ivec,
-	des_cblock *inw,des_cblock *outw,int enc);
 void des_cfb_encrypt(unsigned char *in,unsigned char *out,int numbits,
 	long length,des_key_schedule schedule,des_cblock *ivec,int enc);
 void des_ecb_encrypt(des_cblock *input,des_cblock *output,
@@ -250,7 +246,6 @@
 DES_LONG des_cbc_cksum();
 void des_cbc_encrypt();
 void des_ncbc_encrypt();
-void des_xcbc_encrypt();
 void des_cfb_encrypt();
 void des_ede3_cfb64_encrypt();
 void des_ede3_ofb64_encrypt();
