--- linux-2.6.11.10/drivers/input/serio/i8042.c	2005-05-16 19:50:32.000000000 +0200
+++ linux-2.6.11.10-old/drivers/input/serio/i8042.c	2005-05-20 23:06:04.251079159 +0200
@@ -1015,7 +1015,12 @@
 	int err;
 
 	dbg_init();
-
+#ifdef CONFIG_PPC_MULTIPLATFORM
+        if(_machine == _MACH_Pmac) {
+                printk("%s: nothing to do on PowerMac\n",__FUNCTION__);
+                return -ENODEV;
+        }
+#endif
 	init_timer(&i8042_timer);
 	i8042_timer.function = i8042_timer_func;
 
