--- linux-2.4.20/drivers/usb/storage/unusual_devs.h.org	Tue Feb 25 22:23:56 2003
+++ linux-2.4.20/drivers/usb/storage/unusual_devs.h	Wed Feb 26 10:14:05 2003
@@ -76,6 +76,11 @@
 		US_SC_SCSI, US_PR_BULK, NULL,
 		US_FL_FIX_INQUIRY | US_FL_START_STOP),
 
+UNUSUAL_DEV(  0x0421, 0x0404, 0x0001, 0x0001,
+		"Nokia",
+		"Nokia 5510",
+		US_SC_SCSI, US_PR_BULK, NULL, 0 ),
+
 #ifdef CONFIG_USB_STORAGE_DPCM
 UNUSUAL_DEV(  0x0436, 0x0005, 0x0100, 0x0100,
 		"Microtech",
