diff -Nur include/linux/netfilter_ipv4.org/ipt_DSCP.h include/linux/netfilter_ipv4/ipt_DSCP.h
--- include/linux/netfilter_ipv4.org/ipt_DSCP.h	Thu Jan  1 01:00:00 1970
+++ include/linux/netfilter_ipv4/ipt_DSCP.h	Thu Mar 14 13:14:44 2002
@@ -0,0 +1,20 @@
+/* iptables module for setting the IPv4 DSCP field
+ *
+ * (C) 2002 Harald Welte <laforge@gnumonks.org>
+ * based on ipt_FTOS.c (C) 2000 by Matthew G. Marsh <mgm@paktronix.com>
+ * This software is distributed under GNU GPL v2, 1991
+ * 
+ * See RFC2474 for a description of the DSCP field within the IP Header.
+ *
+ * Id: ipt_DSCP.h,v 1.7 2002/03/14 12:03:13 laforge Exp 
+*/
+#ifndef _IPT_DSCP_TARGET_H
+#define _IPT_DSCP_TARGET_H
+#include <linux/netfilter_ipv4/ipt_dscp.h>
+
+/* target info */
+struct ipt_DSCP_info {
+	u_int8_t dscp;
+};
+
+#endif /* _IPT_DSCP_TARGET_H */
diff -Nur include/linux/netfilter_ipv4.org/ipt_ECN.h include/linux/netfilter_ipv4/ipt_ECN.h
--- include/linux/netfilter_ipv4.org/ipt_ECN.h	Thu Jan  1 01:00:00 1970
+++ include/linux/netfilter_ipv4/ipt_ECN.h	Sun Feb 17 22:26:49 2002
@@ -0,0 +1,25 @@
+/* Header file for iptables ipt_ECN target
+ *
+ * (C) 2002 by Harald Welte <laforge@gnumonks.org>
+ *
+ * This software is distributed under GNU GPL v2, 1991
+ * 
+ * ipt_ECN.h,v 1.1 2002/02/17 21:30:16 laforge Exp
+*/
+#ifndef _IPT_DSCP_H
+#define _IPT_DSCP_H
+#include <linux/netfilter_ipv4/ipt_DSCP.h>
+
+#define IPT_ECN_MASK	(~IPT_DSCP_MASK)
+
+enum ipt_ecn_operation {
+	IPT_ECN_OP_NONE = 0,
+	IPT_ECN_OP_REMOVE,
+};
+#define IPT_ECN_OP_MAX	IPT_ECN_OP_REMOVE
+
+struct ipt_ECN_info {
+	enum ipt_ecn_operation operation;
+};
+
+#endif /* _IPT_ECN_H */
diff -Nur include/linux/netfilter_ipv4.org/ipt_dscp.h include/linux/netfilter_ipv4/ipt_dscp.h
--- include/linux/netfilter_ipv4.org/ipt_dscp.h	Thu Jan  1 01:00:00 1970
+++ include/linux/netfilter_ipv4/ipt_dscp.h	Thu Mar 14 13:14:56 2002
@@ -0,0 +1,23 @@
+/* iptables module for matching the IPv4 DSCP field
+ *
+ * (C) 2002 Harald Welte <laforge@gnumonks.org>
+ * This software is distributed under GNU GPL v2, 1991
+ * 
+ * See RFC2474 for a description of the DSCP field within the IP Header.
+ *
+ * Id: ipt_dscp.h,v 1.1 2002/03/14 12:03:13 laforge Exp 
+*/
+#ifndef _IPT_DSCP_H
+#define _IPT_DSCP_H
+
+#define IPT_DSCP_MASK	0xfc	/* 11111100 */
+#define IPT_DSCP_SHIFT	2	/* shift DSCP two bits for ECN */
+#define IPT_DSCP_MAX	0x3f	/* 00111111 */
+
+/* match info */
+struct ipt_dscp_info {
+	u_int8_t dscp;
+	u_int8_t invert;
+};
+
+#endif /* _IPT_DSCP_H */
