--- kernel/syscall_ksyms.c.org	Tue Feb  5 11:06:06 2002
+++ kernel/syscall_ksyms.c	Wed Feb  6 08:53:07 2002
@@ -49,18 +49,20 @@
 EXPORT_SYMBOL(sys_statfs);
 EXPORT_SYMBOL(sys_fstatfs);
 EXPORT_SYMBOL(sys_ftruncate);
+#if !defined(__alpha__) && !defined(__arch64__)
 EXPORT_SYMBOL(sys_ftruncate64);
+#endif
 EXPORT_SYMBOL(sys_truncate);
-EXPORT_SYMBOL(sys_truncate64);
 #if !defined(__alpha__) && !defined(__ia64__)
+EXPORT_SYMBOL(sys_truncate64);
 EXPORT_SYMBOL(sys_utime);
 #endif
 EXPORT_SYMBOL(sys_utimes);
 EXPORT_SYMBOL(sys_vhangup);
 #if !defined(__alpha__)
-EXPORT_SYMBOL(sys_lseek);
-#endif
 EXPORT_SYMBOL(sys_llseek);
+#endif
+EXPORT_SYMBOL(sys_lseek);
 EXPORT_SYMBOL(sys_read);
 EXPORT_SYMBOL(sys_readv);
 EXPORT_SYMBOL(sys_pread);
@@ -78,9 +80,9 @@
 EXPORT_SYMBOL(sys_setitimer);
 EXPORT_SYMBOL(sys_gettimeofday);
 EXPORT_SYMBOL(sys_settimeofday);
+#if !defined(__alpha__)
 EXPORT_SYMBOL(sys_stime);
 EXPORT_SYMBOL(sys_time);
-#if !defined(__alpha__)
 EXPORT_SYMBOL(sys_nice);
 #endif
 EXPORT_SYMBOL(sys_sched_setscheduler);
@@ -98,8 +100,10 @@
 EXPORT_SYMBOL(sys_rt_sigtimedwait);
 EXPORT_SYMBOL(sys_sigaltstack);
 EXPORT_SYMBOL(sys_sigpending);
+#if !defined(__alpha__)
 EXPORT_SYMBOL(sys_sigprocmask);
 EXPORT_SYMBOL(sys_sigsuspend);
+#endif
 EXPORT_SYMBOL(sys_gethostname);
 EXPORT_SYMBOL(sys_sethostname);
 EXPORT_SYMBOL(sys_setdomainname);
@@ -140,7 +144,9 @@
 EXPORT_SYMBOL(sys_setgroups16);
 #endif
 EXPORT_SYMBOL(sys_wait4);
+#if !defined(__alpha__)
 EXPORT_SYMBOL(sys_waitpid);
+#endif
 EXPORT_SYMBOL(sys_sendfile);
 EXPORT_SYMBOL(sys_brk);
 EXPORT_SYMBOL(sys_msync);
--- arch/alpha/kernel/alpha_ksyms.c.org	Tue Feb  5 11:40:48 2002
+++ arch/alpha/kernel/alpha_ksyms.c	Tue Feb  5 11:41:18 2002
@@ -149,19 +149,6 @@
 EXPORT_SYMBOL(alpha_write_fp_reg);
 EXPORT_SYMBOL(alpha_write_fp_reg_s);
 
-/* In-kernel system calls.  */
-EXPORT_SYMBOL(kernel_thread);
-EXPORT_SYMBOL(sys_open);
-EXPORT_SYMBOL(sys_dup);
-EXPORT_SYMBOL(sys_exit);
-EXPORT_SYMBOL(sys_write);
-EXPORT_SYMBOL(sys_read);
-EXPORT_SYMBOL(sys_lseek);
-EXPORT_SYMBOL(__kernel_execve);
-EXPORT_SYMBOL(sys_setsid);
-EXPORT_SYMBOL(sys_sync);
-EXPORT_SYMBOL(sys_wait4);
-
 /* Networking helper routines. */
 EXPORT_SYMBOL(csum_tcpudp_magic);
 EXPORT_SYMBOL(ip_compute_csum);
