--- drivers/usb/storage/b	Sat Aug  3 02:39:45 2002
+++ drivers/usb/storage/unusual_devs.h	Tue Nov 26 01:21:14 2002
@@ -65,6 +65,11 @@
 		US_SC_8070, US_PR_SCM_ATAPI, init_8200e, 0), 
 #endif
 
+UNUSUAL_DEV(  0x03f0, 0x4002, 0x0100, 0x0100,
+						"HP",
+						"Photosmart 715",
+						US_SC_SCSI, US_PR_BULK, NULL,
+						US_FL_START_STOP ),
 #ifdef CONFIG_USB_STORAGE_DPCM
 UNUSUAL_DEV(  0x0436, 0x0005, 0x0100, 0x0100,
 		"Microtech",
