--- linux-2.6.10-rc1/drivers/scsi/Kconfig.orig	2004-10-22 23:40:27.000000000 +0200
+++ linux-2.6.10-rc1/drivers/scsi/Kconfig	2004-10-31 03:06:57.228702256 +0100
@@ -374,7 +374,7 @@
 
 config SCSI_ADVANSYS
 	tristate "AdvanSys SCSI support"
-	depends on (ISA || EISA || PCI) && SCSI && BROKEN
+	depends on (ISA || EISA || PCI) && SCSI && X86 && !64BIT
 	help
 	  This is a driver for all SCSI host adapters manufactured by
 	  AdvanSys. It is documented in the kernel source in
