--- linux-2.4.20/arch/alpha/kernel/entry.S.org	Mon Dec  9 10:48:36 2002
+++ linux-2.4.20/arch/alpha/kernel/entry.S	Mon Dec  9 10:47:19 2002
@@ -10,7 +10,7 @@
 
 #define SIGCHLD 20
 
-#define NR_SYSCALLS 382
+#define NR_SYSCALLS 394
 
 /*
  * These offsets must match with alpha_mv in <asm/machvec.h>.
--- linux-2.4.20/arch/alpha/kernel/alpha_ksyms.c.org	Fri Dec  6 13:52:55 2002
+++ linux-2.4.20/arch/alpha/kernel/alpha_ksyms.c	Tue Dec 10 08:47:55 2002
@@ -153,16 +153,8 @@
 
 /* In-kernel system calls.  */
 EXPORT_SYMBOL(kernel_thread);
-EXPORT_SYMBOL(sys_open);
-EXPORT_SYMBOL(sys_dup);
-EXPORT_SYMBOL(sys_exit);
-EXPORT_SYMBOL(sys_write);
-EXPORT_SYMBOL(sys_read);
 EXPORT_SYMBOL(sys_lseek);
 EXPORT_SYMBOL(__kernel_execve);
-EXPORT_SYMBOL(sys_setsid);
-EXPORT_SYMBOL(sys_sync);
-EXPORT_SYMBOL(sys_wait4);
 
 /* Networking helper routines. */
 EXPORT_SYMBOL(csum_tcpudp_magic);
@@ -268,3 +260,5 @@
 EXPORT_SYMBOL(irongate_ioremap);
 EXPORT_SYMBOL(irongate_iounmap);
 #endif
+
+EXPORT_SYMBOL(plat_node_data);
--- linux-2.4.19/arch/ppc/kernel/ppc_ksyms.c.org	Thu Sep 19 20:41:52 2002
+++ linux-2.4.19/arch/ppc/kernel/ppc_ksyms.c	Thu Sep 19 20:41:40 2002
@@ -371,3 +371,5 @@
 EXPORT_SYMBOL_NOVERS(agp_special_page);
 #endif /* defined(CONFIG_ALL_PPC) */
 
+EXPORT_SYMBOL(ioremap_bot);
+EXPORT_SYMBOL(pmac_get_partition);
--- linux-2.4.20/arch/sparc/kernel/sparc_ksyms.c.org	Fri Nov 29 00:53:12 2002
+++ linux-2.4.20/arch/sparc/kernel/sparc_ksyms.c	Mon Dec  9 22:16:09 2002
@@ -251,6 +251,7 @@
 EXPORT_SYMBOL_NOVERS(memscan);
 EXPORT_SYMBOL_NOVERS(strlen);
 EXPORT_SYMBOL(strnlen);
+EXPORT_SYMBOL(__strnlen_user);
 EXPORT_SYMBOL(strcpy);
 EXPORT_SYMBOL(strncpy);
 EXPORT_SYMBOL(strcat);
diff -ur linux-2.4.20/arch/sparc64/kernel/sparc64_ksyms.c linux-2.4.20.xxx/arch/sparc64/kernel/sparc64_ksyms.c
--- linux-2.4.20/arch/sparc64/kernel/sparc64_ksyms.c	Fri Nov 29 00:53:12 2002
+++ linux-2.4.20.xxx/arch/sparc64/kernel/sparc64_ksyms.c	Mon Dec  9 19:34:01 2002
@@ -310,12 +310,12 @@
 EXPORT_SYMBOL(sunos_sys_table);
 EXPORT_SYMBOL(sys_call_table32);
 EXPORT_SYMBOL(tl0_solaris);
-EXPORT_SYMBOL(sys_sigsuspend);
-EXPORT_SYMBOL(sys_getppid);
+//EXPORT_SYMBOL(sys_sigsuspend);
+//EXPORT_SYMBOL(sys_getppid);
 EXPORT_SYMBOL(svr4_getcontext);
 EXPORT_SYMBOL(svr4_setcontext);
 EXPORT_SYMBOL(prom_cpu_nodes);
-EXPORT_SYMBOL(sys_ioctl);
+//EXPORT_SYMBOL(sys_ioctl);
 EXPORT_SYMBOL(sys32_ioctl);
 EXPORT_SYMBOL(sparc32_open);
 #endif
--- linux-2.4.19/kernel/ksyms.c.org	Tue Aug 20 12:02:25 2002
+++ linux-2.4.19/kernel/ksyms.c	Tue Aug 20 12:15:31 2002
@@ -384,6 +384,7 @@
 EXPORT_SYMBOL(proc_dointvec_minmax);
 EXPORT_SYMBOL(proc_doulongvec_ms_jiffies_minmax);
 EXPORT_SYMBOL(proc_doulongvec_minmax);
+EXPORT_SYMBOL(proc_get_inode);
 
 /* interrupt handling */
 EXPORT_SYMBOL(add_timer);
@@ -555,6 +555,10 @@
 #include <linux/kdb.h>
 EXPORT_SYMBOL(kdb_usb_infos);
 #endif
+ 
+EXPORT_SYMBOL(elevator_init);
+EXPORT_SYMBOL(elevator_noop_merge_req);
+EXPORT_SYMBOL(elevator_noop_merge);
 
 /* debug */
 EXPORT_SYMBOL(dump_stack);
--- linux-2.4.20/kernel/ksyms.c~	Fri Mar  7 01:58:56 2003
+++ linux-2.4.20/kernel/ksyms.c	Fri Mar  7 14:10:58 2003
@@ -192,6 +192,7 @@
 EXPORT_SYMBOL(__d_path);
 EXPORT_SYMBOL(mark_buffer_dirty);
 EXPORT_SYMBOL(set_buffer_async_io); /* for reiserfs_writepage */
+EXPORT_SYMBOL(end_buffer_io_sync);
 EXPORT_SYMBOL(balance_dirty);
 EXPORT_SYMBOL(__mark_dirty);
 EXPORT_SYMBOL(__mark_buffer_dirty);
--- linux-2.4.20/net/netsyms.c.orig	Fri Mar  7 01:58:53 2003
+++ linux-2.4.20/net/netsyms.c	Fri Mar  7 14:20:19 2003
@@ -86,6 +86,8 @@
 #include <net/sock.h>
 #endif
 
+extern struct socket *sockfd_lookup(int fd, int *err);
+
 #ifdef CONFIG_SYSCTL
 extern int sysctl_max_syn_backlog;
 #endif
@@ -159,6 +161,7 @@
 EXPORT_SYMBOL(put_cmsg);
 EXPORT_SYMBOL(sock_kmalloc);
 EXPORT_SYMBOL(sock_kfree_s);
+EXPORT_SYMBOL(sockfd_lookup);
 
 #ifdef CONFIG_FILTER
 EXPORT_SYMBOL(sk_run_filter);
--- linux-2.4.20/drivers/atm/idt77105.c.org	Fri Mar 21 08:27:41 2003
+++ linux-2.4.20/drivers/atm/idt77105.c	Fri Mar 21 08:36:05 2003
@@ -370,6 +370,7 @@
 }
 
 EXPORT_SYMBOL(idt77105_init);
+EXPORT_SYMBOL(idt77105_stop);
 
 static void __exit idt77105_exit(void)
 {
