diff -Nru linux-2.6.16.21.orig/sound/pci/hda/patch_realtek.c linux-2.6.16.21/sound/pci/hda/patch_realtek.c
--- linux-2.6.16.21.orig/sound/pci/hda/patch_realtek.c	2006-06-22 11:37:47.806532750 +0000
+++ linux-2.6.16.21/sound/pci/hda/patch_realtek.c	2006-06-22 11:38:20.088550250 +0000
@@ -2953,7 +2953,7 @@
 	{ .modelname = "hp", .config = ALC260_HP },
 	{ .pci_subvendor = 0x103c, .pci_subdevice = 0x3010, .config = ALC260_HP },
 	{ .pci_subvendor = 0x103c, .pci_subdevice = 0x3011, .config = ALC260_HP },
-	{ .pci_subvendor = 0x103c, .pci_subdevice = 0x3012, .config = ALC260_HP },
+	{ .pci_subvendor = 0x103c, .pci_subdevice = 0x3012, .config = ALC260_HP_3013 },
 	{ .pci_subvendor = 0x103c, .pci_subdevice = 0x3013, .config = ALC260_HP_3013 },
 	{ .pci_subvendor = 0x103c, .pci_subdevice = 0x3014, .config = ALC260_HP },
 	{ .pci_subvendor = 0x103c, .pci_subdevice = 0x3015, .config = ALC260_HP },
