diff -urN linux-2.4.22.org/drivers/ide/ide.c linux-2.4.22/drivers/ide/ide.c
--- linux-2.4.22.org/drivers/ide/ide.c	2003-11-22 22:11:12.000000000 +0100
+++ linux-2.4.22/drivers/ide/ide.c	2003-11-23 00:13:00.000000000 +0100
@@ -3040,6 +3040,8 @@
 	return 0;
 }
 
+#include "ide-probe.c"
+
 #ifdef MODULE
 char *options = NULL;
 MODULE_PARM(options,"s");
@@ -3061,6 +3063,13 @@
 
 int init_module (void)
 {
+        unsigned int index;
+
+        for (index = 0; index < MAX_HWIFS; ++index)
+                ide_unregister(index);
+        ideprobe_init();
+        create_proc_ide_interfaces();
+        ide_xlate_1024_hook = ide_xlate_1024;
 	parse_options(options);
 	return ide_init();
 }
@@ -3069,6 +3078,9 @@
 {
 	int index;
 
+	ide_probe = NULL;
+	ide_xlate_1024_hook = 0;
+
 	unregister_reboot_notifier(&ide_notifier);
 	for (index = 0; index < MAX_HWIFS; ++index) {
 		ide_unregister(index);
diff -urN linux-2.4.22.org/drivers/ide/ide-probe.c linux-2.4.22/drivers/ide/ide-probe.c
--- linux-2.4.22.org/drivers/ide/ide-probe.c	2003-11-22 22:08:09.000000000 +0100
+++ linux-2.4.22/drivers/ide/ide-probe.c	2003-11-22 23:37:35.000000000 +0100
@@ -1423,23 +1423,4 @@
 
 #ifdef MODULE
 extern int (*ide_xlate_1024_hook)(kdev_t, int, int, const char *);
-
-int init_module (void)
-{
-	unsigned int index;
-	
-	for (index = 0; index < MAX_HWIFS; ++index)
-		ide_unregister(index);
-	ideprobe_init();
-	create_proc_ide_interfaces();
-	ide_xlate_1024_hook = ide_xlate_1024;
-	return 0;
-}
-
-void cleanup_module (void)
-{
-	ide_probe = NULL;
-	ide_xlate_1024_hook = 0;
-}
-MODULE_LICENSE("GPL");
 #endif /* MODULE */
diff -urN linux-2.4.22.org/drivers/ide/Makefile linux-2.4.22/drivers/ide/Makefile
--- linux-2.4.22.org/drivers/ide/Makefile	2003-11-22 22:11:09.000000000 +0100
+++ linux-2.4.22/drivers/ide/Makefile	2003-11-23 00:13:44.000000000 +0100
@@ -9,7 +9,7 @@
 #
 
 
-export-objs := ide-iops.o ide-taskfile.o ide-proc.o ide.o ide-probe.o ide-dma.o ide-lib.o setup-pci.o ide-io.o ide-disk.o
+export-objs := ide-iops.o ide-taskfile.o ide-proc.o ide.o ide-dma.o ide-lib.o setup-pci.o ide-io.o ide-disk.o
 
 all-subdirs	:= arm legacy pci ppc raid
 mod-subdirs	:= arm legacy pci ppc raid
@@ -28,8 +28,7 @@
 
 # Core IDE code - must come before legacy
 
-ide-core-objs	:= ide-iops.o ide-taskfile.o ide.o ide-lib.o ide-io.o ide-default.o ide-proc.o
-ide-detect-objs	:= ide-probe.o ide-geometry.o
+ide-core-objs	:= ide-iops.o ide-taskfile.o ide.o ide-lib.o ide-io.o ide-default.o ide-proc.o ide-geometry.o
 
 
 ifeq ($(CONFIG_BLK_DEV_IDEPCI),y)
@@ -66,8 +65,6 @@
 obj-$(CONFIG_BLK_DEV_IDETAPE)		+= ide-tape.o
 obj-$(CONFIG_BLK_DEV_IDEFLOPPY)		+= ide-floppy.o
 
-obj-$(CONFIG_BLK_DEV_IDE) += ide-detect.o
-
 ifeq ($(CONFIG_BLK_DEV_IDE),y)
 # RAID must be last of all
   obj-y		+= raid/idedriver-raid.o
@@ -81,6 +78,3 @@
 ide-core.o:	$(ide-core-objs)
 	$(LD) -r -o $@ $(ide-core-objs)
 
-ide-detect.o:	$(ide-detect-objs)
-	$(LD) -r -o $@ $(ide-detect-objs)
-
