--- linux-wlan-ng-0.2.7/src/Makefile.orig	2006-07-22 16:49:36.000000000 +0200
+++ linux-wlan-ng-0.2.7/src/Makefile	2007-01-19 01:04:13.000000000 +0100
@@ -44,7 +44,7 @@
 
 -include ../config.mk
 
-DIRS=mkmeta shared wlanctl nwepgen wlancfg p80211 prism2
+DIRS=mkmeta shared wlanctl nwepgen wlancfg prism2
 
 ifneq ($(wildcard *.addon),) 
 DIRS+=`cat *.addon`
--- linux-wlan-ng-0.2.7/src/prism2/Makefile.orig	2005-11-02 15:13:30.000000000 +0100
+++ linux-wlan-ng-0.2.7/src/prism2/Makefile	2007-01-19 01:08:11.000000000 +0100
@@ -44,7 +44,7 @@
 
 -include ../../config.mk
 
-DIRS=driver ridlist download
+DIRS=ridlist download
 
 ifneq ($(wildcard *.addon),) 
 DIRS+=`cat *.addon`
--- linux-wlan-ng-0.2.7/man/Makefile.orig	2005-01-26 23:01:07.000000000 +0100
+++ linux-wlan-ng-0.2.7/man/Makefile	2007-01-19 01:15:33.000000000 +0100
@@ -48,8 +48,8 @@
 	@echo "Nothing to do..."
 
 install:
-	mkdir -p $(TARGET_ROOT_ON_HOST)/usr/local/man/man1
-	cp *.1 	$(TARGET_ROOT_ON_HOST)/usr/local/man/man1
+	mkdir -p $(TARGET_ROOT_ON_HOST)/usr/share/man/man1
+	cp *.1 	$(TARGET_ROOT_ON_HOST)/usr/share/man/man1
 
 clean:
 	rm -f core core.* *.o .*.o *.s *.a .depend tmp_make *~ tags
