--- linux.org/kernel/power/suspend.c.org	2004-09-26 01:20:35.304106773 +0200
+++ linux/kernel/power/suspend.c	2004-09-26 01:23:29.340823488 +0200
@@ -66,6 +67,8 @@
 #include <asm/i387.h> /* for kernel_fpu_end */
 #endif
 
+#include <asm/uaccess.h>
+
 static unsigned long avenrun_save[3];
 static u32 pm_disk_mode_save;
 struct partial_device_tree * suspend_device_tree;
--- linux.org/kernel/power/suspend_gzip.c.org	2004-09-26 01:26:01.468233758 +0200
+++ linux/kernel/power/suspend_gzip.c	2004-09-26 01:27:01.486376279 +0200
@@ -417,7 +417,7 @@
 
 static unsigned long gzip_memory_needed(void)
 {
-	return PAGE_SIZE + MAX( zlib_deflate_workspacesize(),
+	return PAGE_SIZE + max( zlib_deflate_workspacesize(),
 				zlib_inflate_workspacesize());
 }
 
--- linux.org/arch/i386/kernel/apm.c.org	2004-09-26 01:28:46.748826268 +0200
+++ linux/arch/i386/kernel/apm.c	2004-09-26 01:32:46.482468957 +0200
@@ -230,6 +230,7 @@
 #include <asm/desc.h>
 
 #include "io_ports.h"
+#include "../../../kernel/power/suspend.h"
 
 extern spinlock_t i8253_lock;
 extern unsigned long get_cmos_time(void);
--- linux.org/kernel/power/suspend_builtin.c.org	2004-09-26 02:08:00.658556314 +0200
+++ linux/kernel/power/suspend_builtin.c	2004-09-26 02:08:31.620923353 +0200
@@ -13,6 +13,8 @@
 #include <linux/module.h>
 #include <linux/reboot.h>
 
+#include <asm/uaccess.h>
+
 #include "suspend.h"
 /* 
  *---------------------  Variables ---------------------------
--- linux.org/kernel/power/process.c.org	2004-09-26 15:20:19.791412023 +0200
+++ linux/kernel/power/process.c	2004-09-26 15:20:38.477408986 +0200
@@ -159,12 +159,12 @@
 	}
 }
 
-extern void __smp_suspend_lowlevel(void * info);
+extern void smp_suspend_lowlevel(void * info);
 
 void smp_suspend(void)
 {
 	set_suspend_state(SUSPEND_FREEZE_SMP);
-	smp_call_function(__smp_suspend_lowlevel, NULL, 0, 0);
+	smp_call_function(smp_suspend_lowlevel, NULL, 0, 0);
 
 	while (atomic_read(&suspend_cpu_counter) < (num_online_cpus() - 1)) {
 		cpu_relax();

