--- linux-2.4.21/drivers/ide/ide-disk.c~	Fri Jun 13 16:51:33 2003
+++ linux-2.4.21/drivers/ide/ide-disk.c	Sun Jun 29 18:48:33 2003
@@ -698,12 +698,9 @@
 {
 	ide_hwif_t *hwif	= HWIF(drive);
 	if (hwif->rw_disk)
-		return hwif->rw_disk(drive, rq, block);
-	else 
-		return __ide_do_rw_disk(drive, rq, block);
+		hwif->rw_disk(drive, rq, block);
+	return __ide_do_rw_disk(drive, rq, block);
 }
-
-EXPORT_SYMBOL_GPL(__ide_do_rw_disk);
 
 static int idedisk_open (struct inode *inode, struct file *filp, ide_drive_t *drive)
 {
--- linux-2.4.21/drivers/ide/pci/hpt366.c~	Sun Jun 29 15:09:43 2003
+++ linux-2.4.21/drivers/ide/pci/hpt366.c	Sun Jun 29 18:50:33 2003
@@ -742,7 +742,7 @@
 		hpt372n_set_clock(drive, wantclock);
 		HWIF(drive)->config_data = wantclock;
 	}
-	return __ide_do_rw_disk(drive, rq, block);
+	return ide_started;
 }
 
 /*
