--- linux-2.4.19/drivers/video/riva/accel.c.orig	Sat Aug  3 02:39:45 2002
+++ linux-2.4.19/drivers/video/riva/accel.c	Thu Nov 28 12:08:49 2002
@@ -293,8 +293,8 @@
 
 static inline void convert_bgcolor_16(u32 *col)
 {
-	*col = ((*col & 0x00007C00) << 9)
-             | ((*col & 0x000003E0) << 6)
+	*col = ((*col & 0x0000F800) << 8)
+             | ((*col & 0x000007E0) << 5)
              | ((*col & 0x0000001F) << 3)
              |          0xFF000000;
 }
