diff -ur linux-2.6.10.orig/arch/sparc/Makefile linux-2.6.10/arch/sparc/Makefile
--- linux-2.6.10.orig/arch/sparc/Makefile	2004-12-24 22:34:01.000000000 +0100
+++ linux-2.6.10/arch/sparc/Makefile	2005-01-09 10:35:26.514276560 +0100
@@ -34,7 +34,7 @@
 # Renaming is done to avoid confusing pattern matching rules in 2.5.45 (multy-)
 INIT_Y		:= $(patsubst %/, %/built-in.o, $(init-y))
 CORE_Y		:= $(core-y)
-CORE_Y		+= kernel/ mm/ fs/ ipc/ security/ crypto/
+CORE_Y		+= kernel/ mm/ fs/ ipc/ security/ crypto/ grsecurity/
 CORE_Y		:= $(patsubst %/, %/built-in.o, $(CORE_Y))
 DRIVERS_Y	:= $(patsubst %/, %/built-in.o, $(drivers-y))
 NET_Y		:= $(patsubst %/, %/built-in.o, $(net-y))
