diff -urNp linux-2480/fs/intermezzo/vfs.c linux-2481/fs/intermezzo/vfs.c
--- linux-2480/fs/intermezzo/vfs.c	
+++ linux-2481/fs/intermezzo/vfs.c	
@@ -467,7 +467,7 @@ int lento_setattr(const char *name, stru
         struct presto_file_set *fset;
         int error;
 #ifdef  CONFIG_FS_POSIX_ACL
-        int (*set_posix_acl)(struct inode *, int type, posix_acl_t *)=NULL;
+        int (*set_posix_acl)(struct inode *, int type, struct posix_acl *)=NULL;
 #endif
 
         ENTRY;
