--- linux-2.4.21.orig/kernel/panic.c	2004-04-16 14:19:50.000000000 +0200
+++ linux-2.4.21/kernel/panic.c	2004-04-16 14:20:23.000000000 +0200
@@ -69,7 +69,7 @@ NORET_TYPE void panic(const char * fmt, 
 	machine_paniced = 1; 
 	bust_spinlocks(1);
 	va_start(args, fmt);
-	vsprintf(buf, fmt, args);
+	vsnprintf(buf, sizeof(buf), fmt, args);
 	va_end(args);
 	printk(KERN_EMERG "Kernel panic: %s\n",buf);
 	if (in_interrupt())
