--- linux-2.4.21/drivers/i2c/i2c-keywest.c.orig	Tue Aug 19 17:00:14 2003
+++ linux-2.4.21/drivers/i2c/i2c-keywest.c	Tue Aug 19 17:00:37 2003
@@ -515,7 +515,7 @@
 		struct keywest_chan* chan = &iface->channels[i];
 		u8 addr;
 		
-		chan->adapter.name = THIS_MODULE;
+		chan->adapter.owner = THIS_MODULE;
 		sprintf(chan->adapter.name, "%s %d", np->parent->name, i);
 		chan->iface = iface;
 		chan->chan_no = i;
