--- linux-2.6.4/drivers/video/riva/fbdev.c.orig	2004-03-11 03:55:36.000000000 +0100
+++ linux-2.6.4/drivers/video/riva/fbdev.c	2004-04-22 18:42:33.063742432 +0200
@@ -1368,8 +1368,8 @@
 
 static inline void convert_bgcolor_16(u32 *col)
 {
-	*col = ((*col & 0x00007C00) << 9)
-		| ((*col & 0x000003E0) << 6)
+	*col = ((*col & 0x0000F800) << 8)
+		| ((*col & 0x000007E0) << 5)
 		| ((*col & 0x0000001F) << 3)
 		|	   0xFF000000;
 }
