--- linux-2.4.19/arch/ppc/kernel/ppc_ksyms.c.org	Thu Sep 19 20:41:52 2002
+++ linux-2.4.19/arch/ppc/kernel/ppc_ksyms.c	Thu Sep 19 20:41:40 2002
@@ -371,3 +371,5 @@
 #endif /* defined(CONFIG_ALL_PPC) */
 
 
+EXPORT_SYMBOL(ioremap_bot);
+
--- linux-2.4.20/arch/ppc/kernel/ppc_ksyms.c.org	Fri Feb 21 12:24:39 2003
+++ linux-2.4.20/arch/ppc/kernel/ppc_ksyms.c	Fri Feb 21 13:29:56 2003
@@ -370,3 +370,4 @@
 
 EXPORT_SYMBOL(ioremap_bot);
 
+EXPORT_SYMBOL(pmac_get_partition);
