--- drivers/char/genrtc.c.orig	2003-04-08 21:15:52.000000000 +0200
+++ drivers/char/genrtc.c	2003-04-08 21:43:37.000000000 +0200
@@ -486,16 +486,21 @@
 		     "update_IRQ\t: %s\n"
 		     "periodic_IRQ\t: %s\n"
 		     "periodic_freq\t: %ld\n"
-		     "batt_status\t: %s\n",
-		     (flags & RTC_DST_EN) ? "yes" : "no",
+#ifdef RTC_BATT_BAD
+		     "batt_status\t: %s\n"
+#endif
+		     ,(flags & RTC_DST_EN) ? "yes" : "no",
 		     (flags & RTC_DM_BINARY) ? "no" : "yes",
 		     (flags & RTC_24H) ? "yes" : "no",
 		     (flags & RTC_SQWE) ? "yes" : "no",
 		     (flags & RTC_AIE) ? "yes" : "no",
 		     irq_active ? "yes" : "no",
 		     (flags & RTC_PIE) ? "yes" : "no",
-		     0L /* freq */,
-		     (flags & RTC_BATT_BAD) ? "bad" : "okay");
+		     0L /* freq */
+#ifdef RTC_BATT_BAD
+		     ,(flags & RTC_BATT_BAD) ? "bad" : "okay"
+#endif
+		     );
 	if (!get_rtc_pll(&pll))
 	    p += sprintf(p,
 			 "PLL adjustment\t: %d\n"
