diff -urNp 2.4.21pre5/fs/seq_file.c 2.4.21pre6/fs/seq_file.c
--- 2.4.21pre5/fs/seq_file.c	2003-03-15 03:25:08.000000000 +0100
+++ 2.4.21pre6/fs/seq_file.c	2003-04-01 10:13:19.000000000 +0200
@@ -94,8 +94,10 @@ ssize_t seq_read(struct file *file, char
 		m->buf = kmalloc(m->size <<= 1, GFP_KERNEL);
 		if (!m->buf)
 			goto Enomem;
+		m->count = 0;
 	}
 	m->op->stop(m, p);
+	m->count = 0;
 	goto Done;
 Fill:
 	/* they want more? let's try to get some more */
