--- linux-5.16/drivers/pci/controller/pci-mvebu.c.orig	2022-02-04 16:52:46.089869659 +0100
+++ linux-5.16/drivers/pci/controller/pci-mvebu.c	2022-02-04 17:00:31.993825135 +0100
@@ -1288,7 +1288,8 @@
 		 * indirectly via kernel emulated PCI bridge driver.
 		 */
 		mvebu_pcie_setup_hw(port);
-		mvebu_pcie_set_local_dev_nr(port, 0);
+		mvebu_pcie_set_local_dev_nr(port, 1);
+		mvebu_pcie_set_local_bus_nr(port, 0);
 	}
 
 	pcie->nports = i;
