diff -urN linux.org/net/ipv4/netfilter/ip_conntrack_irc.c linux/net/ipv4/netfilter/ip_conntrack_irc.c
--- linux.org/net/ipv4/netfilter/ip_conntrack_irc.c	Thu Feb 28 14:20:29 2002
+++ linux/net/ipv4/netfilter/ip_conntrack_irc.c	Thu Feb 28 14:22:58 2002
@@ -111,9 +112,9 @@
 
 	struct ip_ct_irc *info = &ct->help.ct_irc_info;
 
-	memset(&mask, 0, sizeof(struct ip_conntrack_tuple));
-	mask.dst.u.tcp.port = 0xFFFF;
-	mask.dst.protonum = 0xFFFF;
+	mask = ((struct ip_conntrack_tuple)
+		 { { 0, { 0 } },
+		 { 0xFFFFFFFF, { 0xFFFF }, 0xFFFF }});
 
 	DEBUGP("entered\n");
 	/* Can't track connections formed before we registered */
