diff -Nur /usr/src/linux/include/linux/netfilter_ipv4/ipt_DSCP.h rpm/netfilter/userspace/include/linux/netfilter_ipv4/ipt_DSCP.h
--- linux/include/linux/netfilter_ipv4/ipt_DSCP.h	Sat Jun  1 10:07:21 2002
+++ linux/include/linux/netfilter_ipv4/ipt_DSCP.h	Thu Mar 14 13:22:06 2002
@@ -6,17 +6,15 @@
  * 
  * See RFC2474 for a description of the DSCP field within the IP Header.
  *
- * ipt_DSCP.h,v 1.6 2002/02/25 09:41:23 laforge Exp
+ * Id: ipt_DSCP.h,v 1.7 2002/03/14 12:03:13 laforge Exp 
 */
-#ifndef _IPT_DSCP_H
-#define _IPT_DSCP_H
-
-#define IPT_DSCP_MASK	0xfc	/* 11111100 */
-#define IPT_DSCP_SHIFT	2	/* shift DSCP two bits for ECN */
-#define IPT_DSCP_MAX	0x3f	/* 00111111 */
+#ifndef _IPT_DSCP_TARGET_H
+#define _IPT_DSCP_TARGET_H
+#include <linux/netfilter_ipv4/ipt_dscp.h>
 
+/* target info */
 struct ipt_DSCP_info {
 	u_int8_t dscp;
 };
 
-#endif /* _IPT_DSCP_H */
+#endif /* _IPT_DSCP_TARGET_H */
diff -Nur /usr/src/linux/include/linux/netfilter_ipv4/ipt_ECN.h rpm/netfilter/userspace/include/linux/netfilter_ipv4/ipt_ECN.h
--- linux/include/linux/netfilter_ipv4/ipt_ECN.h	Sat Jun  1 10:07:29 2002
+++ linux/include/linux/netfilter_ipv4/ipt_ECN.h	Wed May 29 17:12:30 2002
@@ -4,7 +4,7 @@
  *
  * This software is distributed under GNU GPL v2, 1991
  * 
- * $Id$
+ * $Id$
 */
 #ifndef _IPT_ECN_TARGET_H
 #define _IPT_ECN_TARGET_H
diff -Nur /usr/src/linux/include/linux/netfilter_ipv4/ipt_FTOS.h rpm/netfilter/userspace/include/linux/netfilter_ipv4/ipt_FTOS.h
--- linux/include/linux/netfilter_ipv4/ipt_FTOS.h	Thu Jan  1 01:00:00 1970
+++ linux/include/linux/netfilter_ipv4/ipt_FTOS.h	Wed Feb 28 12:21:54 2001
@@ -0,0 +1,16 @@
+/* Set TOS field in header to any value
+ *
+ * (C) 2000 by Matthew G. Marsh <mgm@paktronix.com>
+ *
+ * This software is distributed under GNU GPL v2, 1991
+ * 
+ * ipt_FTOS.h borrowed heavily from ipt_TOS.h  11/09/2000
+*/
+#ifndef _IPT_FTOS_H
+#define _IPT_FTOS_H
+
+struct ipt_FTOS_info {
+	u_int8_t ftos;
+};
+
+#endif /*_IPT_FTOS_H*/
diff -Nur /usr/src/linux/include/linux/netfilter_ipv4/ipt_TTL.h rpm/netfilter/userspace/include/linux/netfilter_ipv4/ipt_TTL.h
--- linux/include/linux/netfilter_ipv4/ipt_TTL.h	Sat Jun  1 10:07:37 2002
+++ linux/include/linux/netfilter_ipv4/ipt_TTL.h	Wed Feb 28 12:21:54 2001
@@ -16,6 +16,4 @@
 	u_int8_t	mode;
 	u_int8_t	ttl;
 };
-
-
 #endif
diff -Nur /usr/src/linux/include/linux/netfilter_ipv4/ipt_ecn.h rpm/netfilter/userspace/include/linux/netfilter_ipv4/ipt_ecn.h
--- linux/include/linux/netfilter_ipv4/ipt_ecn.h	Thu Jan  1 01:00:00 1970
+++ linux/include/linux/netfilter_ipv4/ipt_ecn.h	Wed May 29 17:12:30 2002
@@ -0,0 +1,34 @@
+/* iptables module for matching the ECN header in IPv4 and TCP header
+ *
+ * (C) 2002 Harald Welte <laforge@gnumonks.org>
+ * This software is distributed under GNU GPL v2, 1991
+ * 
+ * See 
+ *
+ * $Id$
+*/
+#ifndef _IPT_ECN_H
+#define _IPT_ECN_H
+#include <linux/netfilter_ipv4/ipt_dcsp.h>
+
+#define IPT_ECN_IP_MASK	(~IPT_DSCP_MASK)
+
+#define IPT_ECN_OP_MATCH_IP	0x01
+#define IPT_ECN_OP_MATCH_ECE	0x10
+#define IPT_ECN_OP_MATCH_CWR	0x20
+
+#define IPT_ECN_OP_MATCH_MASK	0xce
+
+/* match info */
+struct ipt_ecn_info {
+	u_int8_t operation;
+	u_int8_t invert;
+	u_int8_t ip_ect;
+	union {
+		struct {
+			u_int8_t ect;
+		} tcp;
+	} proto;
+};
+
+#endif /* _IPT_ECN_H */
