--- net/ipv4/netfilter/ip_nat_pptp.c.org	Thu Apr 18 13:22:26 2002
+++ net/ipv4/netfilter/ip_nat_pptp.c	Thu Apr 18 13:21:38 2002
@@ -424,3 +424,4 @@
 
 module_init(init);
 module_exit(fini);
+EXPORT_SYMBOL(ip_pptp_lock);
--- net/ipv4/netfilter/Makefile.org	Thu Apr 18 13:34:38 2002
+++ net/ipv4/netfilter/Makefile	Fri Apr 19 10:15:18 2002
@@ -9,7 +9,7 @@
 
 O_TARGET := netfilter.o
 
-export-objs = ip_conntrack_standalone.o ip_fw_compat.o ip_nat_standalone.o ip_tables.o arp_tables.o
+export-objs = ip_conntrack_standalone.o ip_fw_compat.o ip_nat_standalone.o ip_tables.o arp_tables.o ip_nat_pptp.o
 
 # Multipart objects.
 list-multi		:= ip_conntrack.o iptable_nat.o ipfwadm.o ipchains.o
