diff -Nur linux-2.4.20.org/Documentation/Configure.help linux-2.4.20/Documentation/Configure.help
--- linux-2.4.20.org/Documentation/Configure.help	Thu Mar 13 09:16:32 2003
+++ linux-2.4.20/Documentation/Configure.help	Thu Mar 13 09:20:23 2003
@@ -583,6 +583,35 @@
   if you want to do performance tuning, by tweaking the elevator, e.g.
 
   If unsure, say N.
+  
+Network disk device support
+CONFIG_BLK_DEV_NWD 
+  If you say Y here, your kernel will be able to act as network disk 
+  client and provide you with network disk block devices. This is 
+  similar to NBD in that the driver forwards block device requests 
+  over network to server, which performs the operations on files on 
+  its local filesystem and returns results back to the client. The 
+  driver will enlarge your kernel by about 8 KiB.
+  
+  Unlike NBD though, NWD it uses RPC to transfer block device requests 
+  to the server. The network disk devices can be also used for root 
+  filesystems, making it easy to setup diskless environment -- all you 
+  need is kernel image with network drivers. The RPC transport protocol
+  can be either TCP/IP or UDP/IP, the latter being the default one and 
+  probably the one more suitable for reliable, local area networks. The
+  driver utilizes built-in kernel SunRPC infrastructure, thus saying Y 
+  here will force SunRPC into your kernel, which will enlarge it by 
+  about 65 KiB.
+  
+  Consult <file:Documentation/nwd.txt> for more information concerning 
+  the kernel client driver and user space server.
+  
+  If you want to compile this driver as a module (i.e. code which can be
+  inserted in and removed from the running kernel whenever you want),
+  say M here and read <file:Documentation/modules.txt>. The module will be
+  called nwd.o.
+
+  If unsure, say N.
 
 ATA/IDE/MFM/RLL support
 CONFIG_IDE
diff -Nur linux-2.4.20.org/Makefile linux-2.4.20/Makefile
--- linux-2.4.20.org/Makefile	Thu Mar 13 09:16:33 2003
+++ linux-2.4.20/Makefile	Thu Mar 13 09:20:23 2003
@@ -187,6 +187,7 @@
 DRIVERS-$(CONFIG_SGI_IP22) += drivers/sgi/sgi.a
 DRIVERS-$(CONFIG_VT) += drivers/video/video.o
 DRIVERS-$(CONFIG_PARIDE) += drivers/block/paride/paride.a
+DRIVERS-$(CONFIG_BLK_DEV_NWD) += drivers/block/nwd/nwd.o
 DRIVERS-$(CONFIG_HAMRADIO) += drivers/net/hamradio/hamradio.o
 DRIVERS-$(CONFIG_TC) += drivers/tc/tc.a
 DRIVERS-$(CONFIG_USB) += drivers/usb/usbdrv.o
diff -Nur linux-2.4.20.org/drivers/block/Config.in linux-2.4.20/drivers/block/Config.in
--- linux-2.4.20.org/drivers/block/Config.in	Thu Mar 13 09:16:27 2003
+++ linux-2.4.20/drivers/block/Config.in	Thu Mar 13 09:20:23 2003
@@ -46,6 +46,7 @@
    
 tristate 'Loopback device support' CONFIG_BLK_DEV_LOOP
 dep_tristate 'Network block device support' CONFIG_BLK_DEV_NBD $CONFIG_NET
+dep_tristate 'Network disk device support' CONFIG_BLK_DEV_NWD $CONFIG_NET
 
 tristate 'RAM disk support' CONFIG_BLK_DEV_RAM
 if [ "$CONFIG_BLK_DEV_RAM" = "y" -o "$CONFIG_BLK_DEV_RAM" = "m" ]; then
diff -Nur linux-2.4.20.org/drivers/block/Makefile linux-2.4.20/drivers/block/Makefile
--- linux-2.4.20.org/drivers/block/Makefile	Thu Mar 13 09:16:27 2003
+++ linux-2.4.20/drivers/block/Makefile	Thu Mar 13 09:20:23 2003
@@ -34,5 +34,6 @@
 obj-$(CONFIG_CDROM_PKTCDVD)	+= pktcdvd.o
 
 subdir-$(CONFIG_PARIDE) += paride
+subdir-$(CONFIG_BLK_DEV_NWD)	+= nwd
 
 include $(TOPDIR)/Rules.make
diff -Nur linux-2.4.20.org/fs/Config.in linux-2.4.20/fs/Config.in
--- linux-2.4.20.org/fs/Config.in	Thu Mar 13 09:16:25 2003
+++ linux-2.4.20/fs/Config.in	Thu Mar 13 09:20:23 2003
@@ -150,17 +150,25 @@
    dep_mbool '  Provide NFS server over TCP support (EXPERIMENTAL)' CONFIG_NFSD_TCP $CONFIG_NFSD $CONFIG_EXPERIMENTAL
 
    if [ "$CONFIG_NFS_FS" = "y" -o "$CONFIG_NFSD" = "y" ]; then
-      define_tristate CONFIG_SUNRPC y
       define_tristate CONFIG_LOCKD y
    else
       if [ "$CONFIG_NFS_FS" = "m" -o "$CONFIG_NFSD" = "m" ]; then
-	 define_tristate CONFIG_SUNRPC m
 	 define_tristate CONFIG_LOCKD m
-   else
-	 define_tristate CONFIG_SUNRPC n
-	 define_tristate CONFIG_LOCKD n
+      else
+         define_tristate CONFIG_LOCKD n
       fi
    fi
+   
+   if [ "$CONFIG_NFS_FS" = "y" -o "$CONFIG_NFSD" = "y" -o "$CONFIG_BLK_DEV_NWD" = "y" ]; then
+     define_tristate CONFIG_SUNRPC y
+   else
+     if [ "$CONFIG_NFS_FS" = "m" -o "$CONFIG_NFSD" = "m" -o "$CONFIG_BLK_DEV_NWD" = "m" ]; then
+       define_tristate CONFIG_SUNRPC m
+     else
+       define_tristate CONFIG_SUNRPC n
+     fi
+   fi
+   
    if [ "$CONFIG_NFSD_V3" = "y" -o "$CONFIG_NFS_V3" = "y" ]; then
      define_bool CONFIG_LOCKD_V4 y
    fi
diff -Nur linux-2.4.20.org/include/linux/blk.h linux-2.4.20/include/linux/blk.h
--- linux-2.4.20.org/include/linux/blk.h	Sat Aug  3 02:39:45 2002
+++ linux-2.4.20/include/linux/blk.h	Thu Mar 13 09:20:23 2003
@@ -295,6 +295,12 @@
 #define DEVICE_REQUEST do_nbd_request
 #define DEVICE_NR(device) (MINOR(device))
 
+#elif (MAJOR_NR == NWD_MAJOR)
+
+#define DEVICE_NAME "nwd"
+#define DEVICE_REQUEST nwd_request
+#define DEVICE_NR(device) (MINOR(device))
+
 #elif (MAJOR_NR == MDISK_MAJOR)
 
 #define DEVICE_NAME "mdisk"
diff -Nur linux-2.4.20.org/include/linux/major.h linux-2.4.20/include/linux/major.h
--- linux-2.4.20.org/include/linux/major.h	Thu Mar 13 09:16:27 2003
+++ linux-2.4.20/include/linux/major.h	Thu Mar 13 09:20:23 2003
@@ -87,6 +87,8 @@
 
 #define LVM_BLK_MAJOR	58	/* Logical Volume Manager */
 
+#define NWD_MAJOR	62	/* Network disk driver	*/
+
 #define SCSI_DISK1_MAJOR	65
 #define SCSI_DISK2_MAJOR	66
 #define SCSI_DISK3_MAJOR	67
diff -Nur linux-2.4.20.org/init/do_mounts.c linux-2.4.20/init/do_mounts.c
--- linux-2.4.20.org/init/do_mounts.c	Thu Mar 13 09:16:31 2003
+++ linux-2.4.20/init/do_mounts.c	Thu Mar 13 09:20:23 2003
@@ -240,6 +240,9 @@
 	{ "ftlc", 0x2c10 },
 	{ "ftld", 0x2c18 },
 	{ "mtdblock", 0x1f00 },
+#ifdef CONFIG_BLK_DEV_NWD
+        { "nwd", 0x3E00 },
+#endif
 	{ NULL, 0 }
 };
 
diff -Nur linux-2.4.20.org/init/main.c linux-2.4.20/init/main.c
--- linux-2.4.20.org/init/main.c	Thu Mar 13 09:16:32 2003
+++ linux-2.4.20/init/main.c	Thu Mar 13 09:20:23 2003
@@ -103,6 +103,7 @@
 #endif
 
 extern void ecard_init(void);
+extern int nwd_boot_init(void);
 
 #if defined(CONFIG_SYSVIPC)
 extern void ipc_init(void);
@@ -536,6 +537,12 @@
 	start_context_thread();
 	do_initcalls();
 
+#ifdef CONFIG_BLK_DEV_NWD
+	/* Must be done after NET/RPC initialization */
+	if (nwd_boot_init())
+		panic ("Unable to setup network disk root device.");
+#endif
+
 #ifdef CONFIG_IRDA
 	irda_proto_init();
 	irda_device_init(); /* Must be done after protocol initialization */
