--- linux-2.4.19/include/asm-ppc/pgtable.h.org	Thu Aug 29 12:01:52 2002
+++ linux-2.4.19/include/asm-ppc/pgtable.h	Thu Aug 29 12:11:13 2002
@@ -286,6 +286,16 @@
 #define PAGE_COPY	__pgprot(_PAGE_BASE | _PAGE_USER)
 #define PAGE_COPY_X	__pgprot(_PAGE_BASE | _PAGE_USER | _PAGE_EXEC)
 
+#ifdef CONFIG_GRKERNSEC_PAX
+#define PAGE_COPY_NOEXEC	__pgproto(_PAGE_BASE)
+#define PAGE_SHARED_NOEXEC	__pgproto(_PAGE_BASE | _PAGE_RW)
+#define	PAGE_READONLY_NOEXEC	__pgproto(_PAGE_BASE)
+#else
+#define PAGE_COPY_NOEXEC	PAGE_COPY
+#define PAGE_SHARED_NOEXEC	PAGE_SHARED
+#define	PAGE_READONLY_NOEXEC	PAGE_READONLY
+#endif
+
 #define PAGE_KERNEL	__pgprot(_PAGE_KERNEL)
 #define PAGE_KERNEL_RO	__pgprot(_PAGE_BASE | _PAGE_SHARED)
 #define PAGE_KERNEL_CI	__pgprot(_PAGE_IO)
