diff -Nur drivers/net/smc-mca.c.org drivers/net/smc-mca.c
--- drivers/net/smc-mca.c.org	2001/03/04 02:33:56	1.1.1.3.10.1
+++ drivers/net/smc-mca.c	2001/03/04 03:20:30
@@ -213,7 +213,7 @@
 	dev->mem_start = 0;
 	num_pages      = 40;
 
-	switch (j) {	/* 'j' = card-# in const array above [hs] */
+	switch (adapter) {	/* card-# in const array above [hs] */
 		case _61c8_SMC_Ethercard_PLUS_Elite_A_BNC_AUI_WD8013EP_A:
 		case _61c9_SMC_Ethercard_PLUS_Elite_A_UTP_AUI_WD8013EP_A:
 		{
