--- squashfs1.0c/squashfs-patch.org	Wed Oct 30 01:34:58 2002
+++ squashfs1.0c/squashfs-patch	Thu Oct 31 08:58:31 2002
@@ -26,13 +26,13 @@
 --- linux-2.4.19/fs/Makefile	Mon Feb 25 19:38:07 2002
 +++ linux/fs/Makefile	Wed Oct 30 00:26:09 2002
 @@ -67,6 +67,7 @@
- subdir-$(CONFIG_REISERFS_FS)	+= reiserfs
- subdir-$(CONFIG_DEVPTS_FS)	+= devpts
  subdir-$(CONFIG_SUN_OPENPROMFS)	+= openpromfs
+ subdir-$(CONFIG_XFS_FS)		+= xfs
+ subdir-$(CONFIG_JFS_FS)		+= jfs
 +subdir-$(CONFIG_SQUASHFS)	+= squashfs
  
- 
  obj-$(CONFIG_BINFMT_AOUT)	+= binfmt_aout.o
+ obj-$(CONFIG_BINFMT_XOUT)	+= binfmt_xout.o
 diff --new-file -ur linux-2.4.19/fs/squashfs/Makefile linux/fs/squashfs/Makefile
 --- linux-2.4.19/fs/squashfs/Makefile	Thu Jan  1 01:00:00 1970
 +++ linux/fs/squashfs/Makefile	Wed Oct 30 00:26:09 2002
--- squashfs1.0c/squashfs-patch.org	Wed Dec  4 09:35:09 2002
+++ squashfs1.0c/squashfs-patch	Wed Dec  4 09:43:47 2002
@@ -9,19 +9,6 @@
  bool 'Virtual memory file system support (former shm fs)' CONFIG_TMPFS
  define_bool CONFIG_RAMFS y
  
-@@ -146,10 +147,10 @@
- else
-    define_tristate CONFIG_ZISOFS_FS n
- fi
--if [ "$CONFIG_CRAMFS" = "y" -o "$CONFIG_ZISOFS_FS" = "y" ]; then
-+if [ "$CONFIG_CRAMFS" = "y" -o "$CONFIG_ZISOFS_FS" = "y" -o $CONFIG_SQUASHFS = "y" ]; then
-    define_tristate CONFIG_ZLIB_FS_INFLATE y
- else
--  if [ "$CONFIG_CRAMFS" = "m" -o "$CONFIG_ZISOFS_FS" = "m" ]; then
-+  if [ "$CONFIG_CRAMFS" = "m" -o "$CONFIG_ZISOFS_FS" = "m" -o $CONFIG_SQUASHFS = "m" ]; then
-      define_tristate CONFIG_ZLIB_FS_INFLATE m
-   else
-      define_tristate CONFIG_ZLIB_FS_INFLATE n
 diff --new-file -ur linux-2.4.19/fs/Makefile linux/fs/Makefile
 --- linux-2.4.19/fs/Makefile	Mon Feb 25 19:38:07 2002
 +++ linux/fs/Makefile	Wed Oct 30 00:26:09 2002
